<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>alignments | Integrated Bioanalytics</title>
<meta name="author" content="Lucas Busta and members of the Busta lab">
<meta name="description" content="Multiple sequence alignments form the bridge between raw sequences and downstream phylogenetic analysis. In this chapter we first use alignSequences() to assemble consistent nucleotide,...">
<meta name="generator" content="bookdown 0.44 with bs4_book()">
<meta property="og:title" content="alignments | Integrated Bioanalytics">
<meta property="og:type" content="book">
<meta property="og:description" content="Multiple sequence alignments form the bridge between raw sequences and downstream phylogenetic analysis. In this chapter we first use alignSequences() to assemble consistent nucleotide,...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="alignments | Integrated Bioanalytics">
<meta name="twitter:description" content="Multiple sequence alignments form the bridge between raw sequences and downstream phylogenetic analysis. In this chapter we first use alignSequences() to assemble consistent nucleotide,...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><link href="libs/Lato-0.4.10/font.css" rel="stylesheet">
<link href="libs/Roboto_Mono-0.4.10/font.css" rel="stylesheet">
<link href="libs/Montserrat-0.4.10/font.css" rel="stylesheet">
<script src="libs/bs3compat-0.9.0/transition.js"></script><script src="libs/bs3compat-0.9.0/tabs.js"></script><script src="libs/bs3compat-0.9.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><link rel="icon" type="image/png" sizes="32x32" href="images/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="images/favicon-16x16.png">
<script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-99618359-1', 'auto');
      ga('send', 'pageview');

    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="style/style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h2>
        <a href="index.html" title="">Integrated Bioanalytics</a>
      </h2>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">WELCOME</a></li>
<li class="book-part">GETTING STARTED</li>
<li><a class="" href="overview.html">overview</a></li>
<li><a class="" href="installation.html">installation</a></li>
<li class="book-part">DATA VISUALIZATION</li>
<li><a class="" href="data-visualization-i.html">data visualization I</a></li>
<li><a class="" href="data-visualization-ii.html">data visualization II</a></li>
<li><a class="" href="data-visualization-iii.html">data visualization III</a></li>
<li class="book-part">STATISTICAL METHODS</li>
<li><a class="" href="wrangling-and-summaries.html">wrangling and summaries</a></li>
<li><a class="" href="hierarchical-clustering.html">hierarchical clustering</a></li>
<li><a class="" href="dimensional-reduction.html">dimensional reduction</a></li>
<li><a class="" href="flat-clustering.html">flat clustering</a></li>
<li><a class="" href="comparing-means.html">comparing means</a></li>
<li class="book-part">MODELS</li>
<li><a class="" href="numerical-models.html">numerical models</a></li>
<li><a class="" href="embedding-models.html">embedding models</a></li>
<li class="book-part">SEQUENCE ANALYSIS</li>
<li><a class="" href="homology.html">homology</a></li>
<li><a class="active" href="alignments.html">alignments</a></li>
<li><a class="" href="phylogenies.html">phylogenies</a></li>
<li><a class="" href="phylogenetic-analyses.html">phylogenetic analyses</a></li>
<li class="book-part">ANALYTICAL REPORTS</li>
<li><a class="" href="figures-captions.html">figures &amp; captions</a></li>
<li><a class="" href="question-driven-report.html">question-driven report</a></li>
<li class="book-part">APPENDIX</li>
<li><a class="" href="links.html">links</a></li>
<li><a class="" href="datasets.html">datasets</a></li>
<li><a class="" href="r-faq.html">r faq</a></li>
<li><a class="" href="templates.html">templates</a></li>
<li><a class="" href="loading-analyzegcmsdata.html">loading analyzeGCMSdata</a></li>
<li><a class="" href="using-analyzegcmsdata.html">using analyzeGCMSdata</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="alignments" class="section level1 unnumbered">
<h1>alignments<a class="anchor" aria-label="anchor" href="#alignments"><i class="fas fa-link"></i></a>
</h1>
<div class="inline-figure"><img src="https://thebustalab.github.io/integrated_bioanalytics/images/alignments.jpeg" width="100%" style="display: block; margin: auto;"></div>
<p>Multiple sequence alignments form the bridge between raw sequences and downstream phylogenetic analysis. In this chapter we first use <code>alignSequences()</code> to assemble consistent nucleotide, amino-acid, or codon alignments. This is easily accomplished from the monolist generated by <code>polyBlast()</code>, if your sequences were generated that way. Then we turn to <code>analyzeAlignment()</code> to diagnose gap-heavy or poorly conserved sites, interactively trim them away, and export a cleaned alignment for tree building.</p>
<div id="alignsequences" class="section level2 unnumbered">
<h2>alignSequences<a class="anchor" aria-label="anchor" href="#alignsequences"><i class="fas fa-link"></i></a>
</h2>
<p>There are, of course, many tools for aligning sequences. <code>alignSequences()</code>, from the phylochemistry toolkit, is designed to be flexible: it will align nucleotide, amino-acid, or codon sequences, and it can restrict the alignment to any subset of records in your BLAST monolist. If you already ran <code>polyBlast()</code>, most of the inputs outlined below should be ready to go. The function writes the aligned FASTA to <code>alignment_out_path</code> and returns nothing (invisibly).</p>
<ul>
<li>
<code>"monolist"</code>: a data frame where each row represents a hit of interest. It must contain an <code>accession</code> column whose values match individual FASTA files in <code>sequences_of_interest_directory_path</code>. The helper <code>readMonolist()</code> returns this object from the CSV generated by <code>polyBlast()</code>.</li>
<li>
<code>"subset"</code>: the name of a logical column in the monolist (for example <code>subset_all</code>). Only rows where that column is <code>TRUE</code> are aligned. Create additional <code>subset_*</code> columns if you want multiple alignment configurations.</li>
<li>
<code>"alignment_out_path"</code>: full file path (including filename) for the alignment that will be written, e.g.¬†<code>/tmp/my_alignment.fa</code>. Existing files at this path are overwritten.</li>
<li>
<code>"sequences_of_interest_directory_path"</code>: directory that stores one FASTA per accession, as produced by <code>polyBlast()</code>. The function normalises the path internally, so either absolute or relative locations are fine.</li>
<li>
<code>"input_sequence_type"</code>: set to <code>"nucl"</code> when the subject FASTA files contain nucleotide sequences and <code>"amin"</code> for amino-acid sequences. Codon alignments start from nucleotides, while amino-acid alignments can start from either input type (translation happens automatically when needed).</li>
<li>
<code>"mode"</code>: chooses the alignment strategy. <code>"nucl_align"</code> performs a straight nucleotide multiple sequence alignment. <code>"amin_align"</code> aligns amino-acid sequences (translated from nucleotides if required). <code>"codon_align"</code> translates to protein, aligns there, and then projects the alignment back to nucleotides. <code>"fragment_align"</code> is reserved for fragment-to-fragment workflows and currently expects a base fragment defined via <code>base_fragment</code>.</li>
<li>
<code>"base_fragment"</code>: optional path to a FASTA file containing the fragment that other sequences should be aligned against. Only used when <code>mode = "fragment_align"</code>.</li>
</ul>
<div class="sourceCode" id="cb164"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">alignSequences</span><span class="op">(</span></span>
<span>  monolist <span class="op">=</span> <span class="fu">readMonolist</span><span class="op">(</span><span class="st">"/path_to/a_csv_file_that_will_list_all_blast_hits.csv"</span><span class="op">)</span>,</span>
<span>  subset <span class="op">=</span> <span class="st">"subset_all"</span>,</span>
<span>  alignment_out_path <span class="op">=</span> <span class="st">"/path_to/a_folder_for_alignments/subset_all_amin_seqs_aligned.fa"</span>,</span>
<span>  sequences_of_interest_directory_path <span class="op">=</span> <span class="st">"/path_to/a_folder_for_hit_sequences/"</span>,</span>
<span>  input_sequence_type <span class="op">=</span> <span class="st">"amin"</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"amin_align"</span>,</span>
<span>  base_fragment <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div id="analyzealignment" class="section level2 unnumbered">
<h2>analyzeAlignment<a class="anchor" aria-label="anchor" href="#analyzealignment"><i class="fas fa-link"></i></a>
</h2>
<p>Once you have an alignment on disk, <code>analyzeAlignment()</code> helps you decide which positions to keep before phylogeny building. It reads a FASTA alignment, profiles each column for gap content and conservation, and launches a Shiny interface that lets you tune thresholds while inspecting the original and trimmed alignments alongside neighbour-joining trees. The function returns the trimmed alignment (as a <code>DNAStringSet</code> or <code>AAStringSet</code>) invisibly and also writes it to <code>&lt;alignment_in_path&gt;_trimmed</code>.</p>
<ul>
<li>
<code>"alignment_in_path"</code>: path to the FASTA alignment to review. This is typically the file produced by <code>alignSequences()</code>, e.g.¬†<code>/tmp/subset_all_amin_seqs_aligned.fa</code>.</li>
<li>
<code>"type"</code>: <code>"DNA"</code> for nucleotide alignments or <code>"AA"</code> for amino acids. This determines how the alignment is read, which tree scaffold is built, and which Biostrings container is used when writing the trimmed alignment.</li>
<li>
<code>"jupyter"</code>: set to <code>TRUE</code> when launching from a Jupyter environment so the app claims an available internal port and prints the connection URL. Leave at the default <code>FALSE</code> for regular RStudio or terminal sessions.</li>
</ul>
<p>The interface plots gap percentages and conservation across the alignment, highlights the positions that survive the current filters, and previews the trimmed tree. When you click <strong>Return Trimmed Alignment &amp; Close App</strong>, the filtered alignment is written to <code>&lt;alignment_in_path&gt;_trimmed</code> (with <code>_trimmed</code> appended) and returned to the caller.</p>
<div class="sourceCode" id="cb165"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">analyzeAlignment</span><span class="op">(</span></span>
<span>  alignment_in_path <span class="op">=</span> <span class="st">"/path_to/a_folder_for_alignments/subset_all_amin_seqs_aligned.fa"</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"AA"</span>,</span>
<span>  jupyter <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="homology.html">homology</a></div>
<div class="next"><a href="phylogenies.html">phylogenies</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <h2>Note: Second Edition is under construction üèó</h2>
    <p>Now is a great time to provide feedback</p>
        <ul class="list-unstyled">
<li><a href="https://forms.gle/SZmB2Ct2exE2dBwv9">Provide feedback (5 min)</a></li>
          <!-- <li><a href="https://geocompr.robinlovelace.net/#reproducibility">Install updated packages</a></li> -->
          <!-- <li><a href="https://github.com/Robinlovelace/geocompr/issues">Open an issue <i class="fas fa-question"></i></a></li> -->
          <!-- <li><a href="https://discord.gg/Te3gWeDwmf">Chat on Discord <i class="fab fa-discord"></i></a></li> -->
        </ul>
<hr>
<nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#alignments">alignments</a></li>
<li><a class="nav-link" href="#alignsequences">alignSequences</a></li>
<li><a class="nav-link" href="#analyzealignment">analyzeAlignment</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/blob/main/index.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/edit/main/index.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Integrated Bioanalytics</strong>" was written by Lucas Busta and members of the Busta lab. It was last built on 2025-12-16.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
