<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

  <title>blast | Integrated Bioanalytics</title>

    <meta name="author" content="Lucas Busta and members of the Busta lab" />
  
   <meta name="description" content="blast | Integrated Bioanalytics is a book describing how to perform chemical, phylogenetic, and genomic analyses." />
   <meta name="generator" content="placeholder" />
  <meta property="og:title" content="blast | Integrated Bioanalytics" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="blast | Integrated Bioanalytics is a book describing how to perform chemical, phylogenetic, and genomic analyses." />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="blast | Integrated Bioanalytics" />
  
  <meta name="twitter:description" content="blast | Integrated Bioanalytics is a book describing how to perform chemical, phylogenetic, and genomic analyses." />
  
  <!-- JS -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script>
  <script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script>
    <script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no" />
    <link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet" />
    <script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script>
    <link href="libs/Lato-0.4.6/font.css" rel="stylesheet" />
    <link href="libs/Roboto_Mono-0.4.6/font.css" rel="stylesheet" />
    <link href="libs/Montserrat-0.4.6/font.css" rel="stylesheet" />
    <script src="libs/bs3compat-0.4.2/transition.js"></script>
    <script src="libs/bs3compat-0.4.2/tabs.js"></script>
    <script src="libs/bs3compat-0.4.2/bs3compat.js"></script>
    <link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet" />
    <script src="libs/bs4_book-1.0.0/bs4_book.js"></script>
    <link rel="icon" type="image/png" sizes="32x32" href="images/favicon-32x32.png">
    <link rel="icon" type="image/png" sizes="16x16" href="images/favicon-16x16.png">

    <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-99618359-1', 'auto');
      ga('send', 'pageview');

    </script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script>

  <!-- CSS -->
    <link rel="stylesheet" href="style/style.css" />
  
</head>

<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book">
    <a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h2>
        <a href="index.html" title="">Integrated Bioanalytics</a>
      </h2>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
      </form>

      <nav aria-label="Table of contents">
        <h2>Table of contents</h2>
        <div id="book-toc"></div>

        <div class="book-extra">
          <p><a id="book-repo" href="#">View book source <i class="fab fa-github"></i></a></li></p>
        </div>
      </nav>
    </div>
  </header>

  <main class="col-sm-12 col-md-9 col-lg-7" id="content">
<div id="blast" class="section level1 unnumbered">
<h1>blast</h1>
<p><img src="https://thebustalab.github.io/integrated_bioanalytics/images/homology.png" width="100%" style="display: block; margin: auto;" /></p>
<div id="polyblast" class="section level2" number="1.11">
<h2><span class="header-section-number">1.11</span> polyBlast</h2>
<div id="setup-1" class="section level3" number="1.11.1">
<h3><span class="header-section-number">1.11.1</span> setup</h3>
<p>On NCBI, you can search various sequence collections with one or more queries. However, often we want to search a custom library, or multiple libraries. For example, maybe we have downloaded some genomes of interest and want to run blast searches on them. That is what polyBlast() is designed to do. polyBlast() relies on the BLAST+ program available from <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">NCBI BLAST+</a>. Download the program and then point this function to the executable via the <code>blast_module_directory_path</code> argument. You can search multiple sequence libraries at once using multiple queries, and all the usual blast configurations (blastp, blastn, tblastn, etc.) are available. Please note that searches with protein sequences or translated DNA sequences are 5–10-fold more sensitive than DNA:DNA sequence comparison.</p>
<p>Let’s check out <code>polyBlast()</code> by looking at an example. For this example, we need to set up a few things:</p>
<ul>
<li>“named_subjects_list”: A named list of sequence collections (often transcriptomes) to search (one fasta for each collection, often one collection for each species or accession).</li>
<li>“query_in_path”: One or more queries, all listed in a single fasta file.</li>
<li>“sequences_of_interest_directory_path”: The path to a directory where the BLAST hits will be written as individual files (this will be useful later on).</li>
<li>“blast_module_directory_path”: The path to the folder of BLAST+ executable.</li>
<li>“blast_mode”: The format is XYblastZ where X is subject type (the transcriptome or proteome, use “n” for nucleotide, “p” for amino acids). Y is whether the subjects should be translated (“t” for translate, “n” for no translation, if you choose “t” your subjects will not be searched for ORFs, every three base pairs are just translated verbatim). Z is the format of the query/queries (“n” for nucleotide, “p” for amino acids). Allowed formats: nnblastn, ntblastp, pnblastp.</li>
<li>“e_value_cutoff”: Hits with a e-value below this cutoff will not be returned. Default = 1.</li>
<li>“queries_in_outout”: TRUE/FALSE, should the queries be included in the output? If you want to build a tree of BLAST hits and want the queries in the tree, then set this to be TRUE.</li>
<li>“monolist_out_path”: The path to where we want a summary file of the BLAST hits to be written.</li>
</ul>
<p>Once we have those things, we can set up the search (see below). There are two main outputs from the search: a list of the hits (“monolist_out”, which is written to “monolist_out_path”), and the hits themselves, written as individual files to “sequences_of_interest_directory_path”. These two things can be used in downstream analyses, such as alignments. The function does not return an object.</p>
<div class="sourceCode" id="cb145"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb145-1"><a href="blast.html#cb145-1" aria-hidden="true" tabindex="-1"></a>the_transcriptomes <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb145-2"><a href="blast.html#cb145-2" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;/path_to/the_transcriptomes_or_proteomes/Nicotiana_glauca.fa&quot;</span>,</span>
<span id="cb145-3"><a href="blast.html#cb145-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;/path_to/the_transcriptomes_or_proteomes/Nicotiana_tabacum.fa&quot;</span>,</span>
<span id="cb145-4"><a href="blast.html#cb145-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;/path_to/the_transcriptomes_or_proteomes/Nicotiana_benthamiana.fa&quot;</span></span>
<span id="cb145-5"><a href="blast.html#cb145-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb145-6"><a href="blast.html#cb145-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb145-7"><a href="blast.html#cb145-7" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(the_transcriptomes) <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb145-8"><a href="blast.html#cb145-8" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;Nicotiana_glauca.fa&quot;</span>,</span>
<span id="cb145-9"><a href="blast.html#cb145-9" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;Nicotiana_tabacum.fa&quot;</span>,</span>
<span id="cb145-10"><a href="blast.html#cb145-10" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;Nicotiana_benthamiana.fa&quot;</span></span>
<span id="cb145-11"><a href="blast.html#cb145-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb145-12"><a href="blast.html#cb145-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb145-13"><a href="blast.html#cb145-13" aria-hidden="true" tabindex="-1"></a><span class="fu">polyBlast</span>(</span>
<span id="cb145-14"><a href="blast.html#cb145-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">named_subjects_list =</span> the_transcriptomes,</span>
<span id="cb145-15"><a href="blast.html#cb145-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">query_in_path =</span> <span class="st">&quot;/path_to/sequences_you_want_to_find_in_the_transcriptomes.fa&quot;</span>,</span>
<span id="cb145-16"><a href="blast.html#cb145-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">sequences_of_interest_directory_path =</span> <span class="st">&quot;/path_to/a_folder_for_hit_sequences/&quot;</span>,</span>
<span id="cb145-17"><a href="blast.html#cb145-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">blast_module_directory_path =</span> <span class="st">&quot;/path_to/the_blast_module/&quot;</span>,</span>
<span id="cb145-18"><a href="blast.html#cb145-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">blast_mode =</span> <span class="fu">c</span>(<span class="st">&quot;nnblastn&quot;</span>, <span class="st">&quot;ntblastp&quot;</span>, <span class="st">&quot;pnblastp&quot;</span>, <span class="st">&quot;dc-megablast&quot;</span>), </span>
<span id="cb145-19"><a href="blast.html#cb145-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">e_value_cutoff =</span> <span class="dv">1</span>,</span>
<span id="cb145-20"><a href="blast.html#cb145-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">queries_in_output =</span> <span class="cn">TRUE</span>,</span>
<span id="cb145-21"><a href="blast.html#cb145-21" aria-hidden="true" tabindex="-1"></a>  <span class="at">monolist_out_path =</span> <span class="st">&quot;/path_to/a_csv_file_that_will_list_all_blast_hits.csv&quot;</span></span>
<span id="cb145-22"><a href="blast.html#cb145-22" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="interpretation" class="section level3" number="1.11.2">
<h3><span class="header-section-number">1.11.2</span> interpretation</h3>
<p>The “30% identity rule-of-thumb” is too conservative. Statistically significant (E &lt; 10−6 – 10−3) protein homologs can share less than 20% identity. E-values and bit scores (bits &gt; 50) are far more sensitive and reliable than percent identity for inferring homology.</p>
<p>The expect value (E-value) can be changed in order to limit the number of hits to the most significant ones. The lower the E-value, the better the hit. The E-value is dependent on the length of the query sequence and the size of the database. For example, an alignment obtaining an E-value of 0.05 means that there is a 5 in 100 chance of occurring by chance alone. E-values are very dependent on the query sequence length and the database size. Short identical sequence may have a high E-value and may be regarded as “false positive” hits. This is often seen if one searches for short primer regions, small domain regions etc. The default threshold for the E-value on the BLAST web page is 10, the default for polyBlast is 1. Increasing this value will most likely generate more hits. Below are some rules of thumb which can be used as loose guidelines:</p>
<ul>
<li>E-value &lt; 10e-100 Identical sequences. You will get long alignments across the entire query and hit sequence.</li>
<li>10e-100 &lt; E-value &lt; 10e-50 Almost identical sequences. A long stretch of the query protein is matched to the database.</li>
<li>10e-50 &lt; E-value &lt; 10e-10 Closely related sequences, could be a domain match or similar.</li>
<li>10e-10 &lt; E-value &lt; 1 Could be a true homologue but it is a gray area.</li>
<li>E-value &gt; 1 Proteins are most likely not related</li>
<li>E-value &gt; 10 Hits are most likely junk unless the query sequence is very short.</li>
</ul>
<p>reference: <a href="https://resources.qiagenbioinformatics.com/manuals/clcgenomicsworkbench/650/_E_value.html" class="uri">https://resources.qiagenbioinformatics.com/manuals/clcgenomicsworkbench/650/_E_value.html</a></p>
<p>reference: Pearson W. R. (2013). An introduction to sequence similarity (“homology”) searching. Current protocols in bioinformatics, Chapter 3, Unit3.1. <a href="https://doi.org/10.1002/0471250953.bi0301s42" class="uri">https://doi.org/10.1002/0471250953.bi0301s42</a>.</p>
<!-- * bit-scores Taylor to write something? -->
</div>
</div>
</div>
  </main>

  <div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <h2>Note: Second Edition is under construction 🏗</h2>
    <p>Now is a great time to provide feedback</p>
        <ul class="list-unstyled">
          <li><a href="https://forms.gle/SZmB2Ct2exE2dBwv9">Provide feedback (5 min)</a></li>
          <!-- <li><a href="https://geocompr.robinlovelace.net/#reproducibility">Install updated packages</a></li> -->
          <!-- <li><a href="https://github.com/Robinlovelace/geocompr/issues">Open an issue <i class="fas fa-question"></i></a></li> -->
          <!-- <li><a href="https://discord.gg/Te3gWeDwmf">Chat on Discord <i class="fab fa-discord"></i></a></li> -->
        </ul>
        <hr>
    <nav id="toc" data-toggle="toc" aria-label="On this page">
      <h2>On this page</h2>
      <div id="book-on-this-page"></div>

      <div class="book-extra">
        <ul class="list-unstyled">
          <li><a id="book-source" href="#">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="#">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
      </div>
    </nav>
  </div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5">
  <div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Integrated Bioanalytics</strong>" was written by Lucas Busta and members of the Busta lab. It was last built on 2023-06-16.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>


</body>

</html>
