<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>clustering | Integrated Bioanalytics</title>
<meta name="author" content="Lucas Busta and members of the Busta lab">
<meta name="description" content="heirarchical clustering  theory ‚ÄúWhich of my samples are most closely related?‚Äù So far we have been looking at how to plot raw data, summarize data, and reduce a data set‚Äôs dimensionality. It‚Äôs...">
<meta name="generator" content="bookdown 0.25 with bs4_book()">
<meta property="og:title" content="clustering | Integrated Bioanalytics">
<meta property="og:type" content="book">
<meta property="og:description" content="heirarchical clustering  theory ‚ÄúWhich of my samples are most closely related?‚Äù So far we have been looking at how to plot raw data, summarize data, and reduce a data set‚Äôs dimensionality. It‚Äôs...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="clustering | Integrated Bioanalytics">
<meta name="twitter:description" content="heirarchical clustering  theory ‚ÄúWhich of my samples are most closely related?‚Äù So far we have been looking at how to plot raw data, summarize data, and reduce a data set‚Äôs dimensionality. It‚Äôs...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><link href="libs/Lato-0.4.1/font.css" rel="stylesheet">
<link href="libs/Roboto%20Mono-0.4.1/font.css" rel="stylesheet">
<link href="libs/Montserrat-0.4.1/font.css" rel="stylesheet">
<script src="libs/bs3compat-0.3.1.9000/transition.js"></script><script src="libs/bs3compat-0.3.1.9000/tabs.js"></script><script src="libs/bs3compat-0.3.1.9000/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><link rel="icon" type="image/png" sizes="32x32" href="images/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="images/favicon-16x16.png">
<script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-99618359-1', 'auto');
      ga('send', 'pageview');

    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="style/style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h2>
        <a href="index.html" title="">Integrated Bioanalytics</a>
      </h2>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">WELCOME</a></li>
<li class="book-part">DATA ANALYSIS IN R</li>
<li><a class="" href="overview.html">overview</a></li>
<li><a class="" href="installation.html">installation</a></li>
<li><a class="" href="data-visualization.html">data visualization</a></li>
<li><a class="" href="data-wrangling.html">data wrangling</a></li>
<li><a class="" href="dimensionality-reduction.html">dimensionality reduction</a></li>
<li><a class="active" href="clustering.html">clustering</a></li>
<li><a class="" href="models.html">models</a></li>
<li><a class="" href="comparing-means.html">comparing means</a></li>
<li class="book-part">GC-MS DATA</li>
<li><a class="" href="mass-spectrometric-analysis.html">mass spectrometric analysis</a></li>
<li class="book-part">DNA AND RNA SEQUENCE DATA</li>
<li><a class="" href="data-acquisition.html">data acquisition</a></li>
<li><a class="" href="data-transfer.html">data transfer</a></li>
<li><a class="" href="sequence-assessment.html">sequence assessment</a></li>
<li><a class="" href="illumina-read-assessment.html">illumina read assessment</a></li>
<li class="book-part">TRANSCRIPTOME ANALYSIS</li>
<li><a class="" href="transcriptomic-analyses.html">transcriptomic analyses</a></li>
<li class="book-part">GENOMIC ANALYSIS</li>
<li><a class="" href="setup.html">setup</a></li>
<li><a class="" href="genome-assembly.html">genome assembly</a></li>
<li><a class="" href="comparative-genomics.html">comparative genomics</a></li>
<li class="book-part">EVOLUTIONARY ANALYSIS</li>
<li><a class="" href="blast-searches.html">blast searches</a></li>
<li><a class="" href="alignments.html">alignments</a></li>
<li><a class="" href="phylogenies.html">phylogenies</a></li>
<li><a class="" href="phylogenetic-analyses.html">phylogenetic analyses</a></li>
<li class="book-part">SCIENTIFIC WRITING</li>
<li><a class="" href="overview-1.html">overview</a></li>
<li><a class="" href="a-mini-manuscript.html">a mini manuscript</a></li>
<li><a class="" href="figures-captions.html">figures &amp; captions</a></li>
<li><a class="" href="results-and-discussion.html">results and discussion</a></li>
<li><a class="" href="conclusion-and-introduction.html">conclusion and introduction</a></li>
<li><a class="" href="abstract-and-title.html">abstract and title</a></li>
<li><a class="" href="proposals.html">proposals</a></li>
<li class="book-part">IMAGE ANALYSIS</li>
<li><a class="" href="image-color-analysis.html">image color analysis</a></li>
<li><a class="" href="images-of-mass-spectra.html">images of mass spectra</a></li>
<li class="book-part">APPENDIX</li>
<li><a class="" href="links.html">links</a></li>
<li><a class="" href="r-faq.html">r faq</a></li>
<li><a class="" href="templates.html">templates</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="clustering" class="section level1 unnumbered">
<h1>clustering<a class="anchor" aria-label="anchor" href="#clustering"><i class="fas fa-link"></i></a>
</h1>
<div class="inline-figure"><img src="https://thebustalab.github.io/integrated_bioanalytics/images/art_tree.png" width="100%" style="display: block; margin: auto;"></div>
<div id="heirarchical-clustering" class="section level2 unnumbered">
<h2>heirarchical clustering<a class="anchor" aria-label="anchor" href="#heirarchical-clustering"><i class="fas fa-link"></i></a>
</h2>
<div id="theory-1" class="section level3 unnumbered">
<h3>theory<a class="anchor" aria-label="anchor" href="#theory-1"><i class="fas fa-link"></i></a>
</h3>
<p>‚ÄúWhich of my samples are most closely related?‚Äù</p>
<p>So far we have been looking at how to plot raw data, summarize data, and reduce a data set‚Äôs dimensionality. It‚Äôs time to look at how to identify relationships between the samples in our data sets. For example: in the Alaska lakes dataset, which lake is most similar, chemically speaking, to Lake Narvakrak? Answering this requires calculating numeric distances between samples based on their chemical properties. For this, the first thing we need is a distance matrix:</p>
<div class="inline-figure"><img src="https://thebustalab.github.io/integrated_bioanalytics/images/dist_matrix.jpg" width="100%" style="display: block; margin: auto;"></div>
<p>Please note that we can get distance matrices directly from <code>runMatrixAnalysis</code> by specifying <code>analysis = "dist"</code>:</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dist</span> <span class="op">&lt;-</span> <span class="fu">runMatrixAnalysis</span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">alaska_lake_data</span>,
    analysis <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"dist"</span><span class="op">)</span>,
    column_w_names_of_multiple_analytes <span class="op">=</span> <span class="st">"element"</span>,
    column_w_values_for_multiple_analytes <span class="op">=</span> <span class="st">"mg_per_L"</span>,
    columns_w_values_for_single_analyte <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"water_temp"</span>, <span class="st">"pH"</span><span class="op">)</span>,
    columns_w_additional_analyte_info <span class="op">=</span> <span class="st">"element_type"</span>,
    columns_w_sample_ID_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lake"</span>, <span class="st">"park"</span><span class="op">)</span>
<span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">dist</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>
<span class="co">##                          Devil_Mountain_Lake_BELA</span>
<span class="co">## Devil_Mountain_Lake_BELA                 0.000000</span>
<span class="co">## Imuruk_Lake_BELA                         3.667875</span>
<span class="co">## Kuzitrin_Lake_BELA                       1.642500</span>
<span class="co">##                          Imuruk_Lake_BELA</span>
<span class="co">## Devil_Mountain_Lake_BELA         3.667875</span>
<span class="co">## Imuruk_Lake_BELA                 0.000000</span>
<span class="co">## Kuzitrin_Lake_BELA               3.061149</span>
<span class="co">##                          Kuzitrin_Lake_BELA</span>
<span class="co">## Devil_Mountain_Lake_BELA           1.642500</span>
<span class="co">## Imuruk_Lake_BELA                   3.061149</span>
<span class="co">## Kuzitrin_Lake_BELA                 0.000000</span></code></pre></div>
<p>There is more that we can do with distance matrices though, lots more. Let‚Äôs start by looking at an example of hierarchical clustering. For this, we just need to tell <code>runMatrixAnalysis()</code> to use <code>analysis = "hclust"</code>:</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">AK_lakes_clustered</span> <span class="op">&lt;-</span> <span class="fu">runMatrixAnalysis</span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">alaska_lake_data</span>,
    analysis <span class="op">=</span> <span class="st">"hclust"</span>,
    column_w_names_of_multiple_analytes <span class="op">=</span> <span class="st">"element"</span>,
    column_w_values_for_multiple_analytes <span class="op">=</span> <span class="st">"mg_per_L"</span>,
    columns_w_values_for_single_analyte <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"water_temp"</span>, <span class="st">"pH"</span><span class="op">)</span>,
    columns_w_additional_analyte_info <span class="op">=</span> <span class="st">"element_type"</span>,
    columns_w_sample_ID_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lake"</span>, <span class="st">"park"</span><span class="op">)</span>,
    na_replacement <span class="op">=</span> <span class="st">"mean"</span>
<span class="op">)</span>
<span class="co">## Replacing NAs in your data with mean</span>
<span class="va">AK_lakes_clustered</span>
<span class="co">## # A tibble: 39 √ó 25</span>
<span class="co">##    sample_unique_ID   lake  park  parent  node branch.length</span>
<span class="co">##    &lt;chr&gt;              &lt;chr&gt; &lt;chr&gt;  &lt;int&gt; &lt;int&gt;         &lt;dbl&gt;</span>
<span class="co">##  1 Devil_Mountain_La‚Ä¶ Devi‚Ä¶ BELA      33     1         0.987</span>
<span class="co">##  2 Imuruk_Lake_BELA   Imur‚Ä¶ BELA      39     2         0.820</span>
<span class="co">##  3 Kuzitrin_Lake_BELA Kuzi‚Ä¶ BELA      38     3         0.703</span>
<span class="co">##  4 Lava_Lake_BELA     Lava‚Ä¶ BELA      31     4         0.743</span>
<span class="co">##  5 North_Killeak_Lak‚Ä¶ Nort‚Ä¶ BELA      21     5         5.62 </span>
<span class="co">##  6 White_Fish_Lake_B‚Ä¶ Whit‚Ä¶ BELA      22     6         3.89 </span>
<span class="co">##  7 Iniakuk_Lake_GAAR  Inia‚Ä¶ GAAR      28     7         1.25 </span>
<span class="co">##  8 Kurupa_Lake_GAAR   Kuru‚Ä¶ GAAR      35     8         0.954</span>
<span class="co">##  9 Lake_Matcharak_GA‚Ä¶ Lake‚Ä¶ GAAR      35     9         0.954</span>
<span class="co">## 10 Lake_Selby_GAAR    Lake‚Ä¶ GAAR      36    10         1.12 </span>
<span class="co">## # ‚Ä¶ with 29 more rows, and 19 more variables: label &lt;chr&gt;,</span>
<span class="co">## #   isTip &lt;lgl&gt;, x &lt;dbl&gt;, y &lt;dbl&gt;, branch &lt;dbl&gt;,</span>
<span class="co">## #   angle &lt;dbl&gt;, water_temp &lt;dbl&gt;, pH &lt;dbl&gt;, C &lt;dbl&gt;,</span>
<span class="co">## #   N &lt;dbl&gt;, P &lt;dbl&gt;, Cl &lt;dbl&gt;, S &lt;dbl&gt;, F &lt;dbl&gt;, Br &lt;dbl&gt;,</span>
<span class="co">## #   Na &lt;dbl&gt;, K &lt;dbl&gt;, Ca &lt;dbl&gt;, Mg &lt;dbl&gt;</span></code></pre></div>
<p>It works! Now we can plot our cluster diagram with a ggplot add-on called ggtree. We‚Äôve seen that ggplot takes a ‚Äúdata‚Äù argument (i.e.¬†<code>ggplot(data = &lt;some_data&gt;) + geom_*()</code> etc.). In contrast, ggtree takes an argument called <code>tr</code>, though if you‚Äôre using the <code>runMatrixAnalysis()</code> function, you can treat these two (<code>data</code> and <code>tr</code>) the same, so, use: <code>ggtree(tr = &lt;output_from_runMatrixAnalysis&gt;) + geom_*()</code> etc.</p>
<p>Note that <code>ggtree</code> also comes with several great new geoms: <code><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab()</a></code> and <code><a href="https://rdrr.io/pkg/ggtree/man/geom_tippoint.html">geom_tippoint()</a></code>. Let‚Äôs try those out:</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/treedata-book/">ggtree</a></span><span class="op">)</span>
<span class="va">AK_lakes_clustered</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tippoint.html">geom_tippoint</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-115-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Cool! Though that plot could use some tweaking‚Ä¶ let‚Äôs try:</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">AK_lakes_clustered</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">lake</span><span class="op">)</span>, offset <span class="op">=</span> <span class="fl">1</span>, align <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tippoint.html">geom_tippoint</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">21</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">park</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_x_continuous</span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_fill_brewer</span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="op">+</span>
    <span class="co"># theme_classic() +</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>
      legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.2</span>,<span class="fl">0.8</span><span class="op">)</span>
    <span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-116-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Very nice!</p>
</div>
<div id="further-reading-4" class="section level3 unnumbered">
<h3>further reading<a class="anchor" aria-label="anchor" href="#further-reading-4"><i class="fas fa-link"></i></a>
</h3>
<p>For more information on plotting annotated trees, see: <a href="https://yulab-smu.top/treedata-book/chapter10.html" class="uri">https://yulab-smu.top/treedata-book/chapter10.html</a>.</p>
<p>For more on clustering, see: <a href="https://ryanwingate.com/intro-to-machine-learning/unsupervised/hierarchical-and-density-based-clustering/" class="uri">https://ryanwingate.com/intro-to-machine-learning/unsupervised/hierarchical-and-density-based-clustering/</a>.</p>
</div>
<div id="exercises-2" class="section level3 unnumbered">
<h3>exercises<a class="anchor" aria-label="anchor" href="#exercises-2"><i class="fas fa-link"></i></a>
</h3>
<p>For this set of exercises, please use <code>runMatrixAnalysis()</code> to run and visualize a hierarchical cluster analysis with each of the main datasets that we have worked with so far, except for NY_trees. This means: <code>algae_data</code> (which algae strains are most similar to each other?), <code>alaska_lake_data</code> (which lakes are most similar to each other?). and <code>solvents</code> (which solvents are most similar to each other?). It also means you should use the periodic table (which elements are most similar to each other?), though please don‚Äôt use the whole periodic table, rather, use <code>periodic_table_subset</code>. Please also conduct a heirarchical clustering analysis for a dataset of your own choice that is not provided by the <code><a href="https://rdrr.io/r/base/source.html">source()</a></code> code. For each of these, create (i) a tree diagram that shows how the ‚Äúsamples‚Äù in each data set are related to each other based on the numerical data associated with them, (ii) a caption for each diagram, and (iii) describe, in two or so sentences, an interesting trend you see in the diagram. You can ignore columns that contain categorical data, or you can list those columns as ‚Äúadditional_analyte_info‚Äù.</p>
<p>For this assignment, you may again find the <code><a href="https://rdrr.io/r/base/colnames.html">colnames()</a></code> function and square bracket-subsetting useful. It will list all or a subset of the column names in a dataset for you. For example:</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">solvents</span><span class="op">)</span>
<span class="co">##  [1] "solvent"             "formula"            </span>
<span class="co">##  [3] "boiling_point"       "melting_point"      </span>
<span class="co">##  [5] "density"             "miscible_with_water"</span>
<span class="co">##  [7] "solubility_in_water" "relative_polarity"  </span>
<span class="co">##  [9] "vapor_pressure"      "CAS_number"         </span>
<span class="co">## [11] "formula_weight"      "refractive_index"   </span>
<span class="co">## [13] "specific_gravity"    "category"</span>

<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">solvents</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>
<span class="co">## [1] "solvent"       "formula"       "boiling_point"</span>

<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">solvents</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">5</span>,<span class="fl">7</span><span class="op">)</span><span class="op">]</span>
<span class="co">## [1] "solvent"             "density"            </span>
<span class="co">## [3] "solubility_in_water"</span></code></pre></div>
</div>
</div>
<div id="k-means-clustering" class="section level2 unnumbered">
<h2>k-means clustering<a class="anchor" aria-label="anchor" href="#k-means-clustering"><i class="fas fa-link"></i></a>
</h2>
<p>‚ÄúDo my samples fall into definable clusters?‚Äù</p>
<div id="theory-2" class="section level3 unnumbered">
<h3>theory<a class="anchor" aria-label="anchor" href="#theory-2"><i class="fas fa-link"></i></a>
</h3>
<p>One of the questions we‚Äôve been asking is ‚Äúwhich of my samples are most closely related?‚Äù. We‚Äôve been answering that question using clustering. However, now that we know how to run principal components analyses, we can use another approach. This alternative approach is called k-means, and can help us decide how to assign our data into clusters. It is generally desirable to have a small number of clusters, however, this must be balanced by not having the variance within each cluster be too big. To strike this balance point, the elbow method is used. For it, we must first determine the maximum within-group variance at each possible number of clusters. An illustration of this is shown in <strong>A</strong> below:</p>
<div class="inline-figure"><img src="https://thebustalab.github.io/integrated_bioanalytics/images/kmeans.png" width="100%" style="display: block; margin: auto;"></div>
<p>One we know within-group variances, we find the ‚Äúelbow‚Äù point - the point with minimum angle theta - thus picking the outcome with a good balance of cluster number and within-cluster variance (illustrated above in <strong>B</strong> and <strong>C</strong>.)</p>
<p>Let‚Äôs try k-means using <code>runMatrixAnalysis</code>. We can use it in conjunction with <code>analysis = "pca"</code> or <code>analysis = "hclust"</code>. Let‚Äôs do PCA first. To include k-means, we can just set <code>kmeans = "auto"</code>. It‚Äôs important to note that kmeans cannot handle NAs. We must set something for the <code>na_replacement</code> argument. One solution is to ignore variables that have NAs for some values, which can be done by setting <code>na_replacement = "drop"</code>.</p>
<p>With <code>kmeans = "auto"</code> and <code>na_replacement = "drop"</code>, we can now run our analyssis. The output now has an additional column called <code>kmeans_cluster</code>, which indicates what cluster each sample is in:</p>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">solvents_pca_kmeans</span> <span class="op">&lt;-</span> <span class="fu">runMatrixAnalysis</span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">solvents</span>,
  analysis <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"pca"</span><span class="op">)</span>,
  column_w_names_of_multiple_analytes <span class="op">=</span> <span class="cn">NULL</span>,
  column_w_values_for_multiple_analytes <span class="op">=</span> <span class="cn">NULL</span>,
  columns_w_values_for_single_analyte <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">solvents</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">7</span><span class="op">:</span><span class="fl">9</span>, <span class="fl">11</span><span class="op">:</span><span class="fl">12</span><span class="op">)</span><span class="op">]</span>,
  columns_w_additional_analyte_info <span class="op">=</span> <span class="cn">NULL</span>,
  columns_w_sample_ID_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"solvent"</span>, <span class="st">"formula"</span>, <span class="st">"miscible_with_water"</span>, <span class="st">"CAS_number"</span>, <span class="st">"category"</span><span class="op">)</span>,
  transpose <span class="op">=</span> <span class="cn">FALSE</span>,
  kmeans <span class="op">=</span> <span class="st">"auto"</span>,
  na_replacement <span class="op">=</span> <span class="st">"drop"</span>
<span class="op">)</span>
<span class="co">## Dropping any variables in your dataset that have NA as a value.</span>
<span class="co">## Variables dropped:</span>
<span class="co">## solubility_in_water vapor_pressure</span>

<span class="va">solvents_pca_kmeans</span>
<span class="co">## # A tibble: 32 √ó 14</span>
<span class="co">##    sample_unique_ID         solvent formula miscible_with_w‚Ä¶</span>
<span class="co">##    &lt;chr&gt;                    &lt;chr&gt;   &lt;chr&gt;   &lt;lgl&gt;           </span>
<span class="co">##  1 acetic_acid_C2H4O2_TRUE‚Ä¶ acetic‚Ä¶ C2H4O2  TRUE            </span>
<span class="co">##  2 acetone_C3H6O_TRUE_67-6‚Ä¶ acetone C3H6O   TRUE            </span>
<span class="co">##  3 acetonitrile_C2H3N_TRUE‚Ä¶ aceton‚Ä¶ C2H3N   TRUE            </span>
<span class="co">##  4 benzene_C6H6_FALSE_71-4‚Ä¶ benzene C6H6    FALSE           </span>
<span class="co">##  5 benzonitrile_C7H5N_FALS‚Ä¶ benzon‚Ä¶ C7H5N   FALSE           </span>
<span class="co">##  6 1-butanol_C4H10O_FALSE_‚Ä¶ 1-buta‚Ä¶ C4H10O  FALSE           </span>
<span class="co">##  7 2-butanone_C4H8O_FALSE_‚Ä¶ 2-buta‚Ä¶ C4H8O   FALSE           </span>
<span class="co">##  8 carbon_disulfide_CS2_FA‚Ä¶ carbon‚Ä¶ CS2     FALSE           </span>
<span class="co">##  9 carbon_tetrachloride_CC‚Ä¶ carbon‚Ä¶ CCl4    FALSE           </span>
<span class="co">## 10 chlorobenzene_C6H5Cl_FA‚Ä¶ chloro‚Ä¶ C6H5Cl  FALSE           </span>
<span class="co">## # ‚Ä¶ with 22 more rows, and 10 more variables:</span>
<span class="co">## #   CAS_number &lt;chr&gt;, category &lt;chr&gt;, Dim.1 &lt;dbl&gt;,</span>
<span class="co">## #   Dim.2 &lt;dbl&gt;, boiling_point &lt;dbl&gt;, melting_point &lt;dbl&gt;,</span>
<span class="co">## #   density &lt;dbl&gt;, relative_polarity &lt;dbl&gt;,</span>
<span class="co">## #   formula_weight &lt;dbl&gt;, refractive_index &lt;dbl&gt;</span></code></pre></div>
<p>We can plot the results and color them according to the group that kmeans suggested:</p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># ggplot(solvents_pca_kmeans) +</span>
<span class="co">#   geom_point(aes(x = Dim.1, y = Dim.2, fill = kmeans_cluster), shape = 21, size = 5, alpha = 0.6)</span></code></pre></div>
<p>Hmmm, it looks like the elbow algorithm is suggesting lots of clusters. Why is this? Let‚Äôs look at the elbow plot itself. For this, we can just set <code>kmeans = "elbow"</code>:</p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># solvents_pca_kmeans_elbow &lt;- runMatrixAnalysis(</span>
<span class="co">#   data = solvents,</span>
<span class="co">#   analysis = c("pca"),</span>
<span class="co">#   column_w_names_of_multiple_analytes = NULL,</span>
<span class="co">#   column_w_values_for_multiple_analytes = NULL,</span>
<span class="co">#   columns_w_values_for_single_analyte = colnames(solvents)[c(3:5, 7:9, 11:12)],</span>
<span class="co">#   columns_w_additional_analyte_info = NULL,</span>
<span class="co">#   columns_w_sample_ID_info = c("solvent", "formula", "miscible_with_water", "CAS_number", "category"),</span>
<span class="co">#   transpose = FALSE,</span>
<span class="co">#   kmeans = "elbow",</span>
<span class="co">#   na_replacement = "drop"</span>
<span class="co"># )</span>
<span class="co"># </span>
<span class="co"># solvents_pca_kmeans_elbow</span></code></pre></div>
<p>This gives us the maximum variance within a cluster for each number of clusters. Let‚Äôs plot that:</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># ggplot(</span>
<span class="co">#   solvents_pca_kmeans_elbow,</span>
<span class="co">#   aes(x = cluster_number, y = variance_within_cluster)</span>
<span class="co"># ) +</span>
<span class="co">#   geom_col() +</span>
<span class="co">#   geom_point() +</span>
<span class="co">#   geom_line()</span></code></pre></div>
<p>Hmm, it looks like there aren‚Äôt any strong elbows in this plot - probably the reason that the elbow method chooses such a high number of clusters. Suppose we want to manually set the number of clusters? We can set <code>kmeans = 3</code> if we want three clusters in the output. Below, let‚Äôs do just that. Let‚Äôs also plot the results and use <code>geom_mark_ellipse</code>.</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># runMatrixAnalysis(</span>
<span class="co">#   data = solvents,</span>
<span class="co">#   analysis = c("pca"),</span>
<span class="co">#   column_w_names_of_multiple_analytes = NULL,</span>
<span class="co">#   column_w_values_for_multiple_analytes = NULL,</span>
<span class="co">#   columns_w_values_for_single_analyte = colnames(solvents)[c(3:5, 7:9, 11:12)],</span>
<span class="co">#   columns_w_additional_analyte_info = NULL,</span>
<span class="co">#   columns_w_sample_ID_info = c("solvent", "formula", "miscible_with_water", "CAS_number", "category"),</span>
<span class="co">#   transpose = FALSE,</span>
<span class="co">#   kmeans = 3,</span>
<span class="co">#   na_replacement = "drop"</span>
<span class="co"># ) %&gt;%</span>
<span class="co"># </span>
<span class="co"># ggplot(aes(x = Dim.1, y = Dim.2, fill = kmeans_cluster)) +</span>
<span class="co">#   geom_point(shape = 21, size = 5) +</span>
<span class="co">#   geom_mark_ellipse(aes(label = kmeans_cluster), alpha = 0.2) +</span>
<span class="co">#   theme_classic() +</span>
<span class="co">#   coord_cartesian(xlim = c(-4,4), ylim = c(-4,4))</span></code></pre></div>
<p>Cool!</p>
<p>One more important point: when using kmeans, the output of runMatrixAnalysis (specifically the kmeans_cluster column) can be used to create groupings for summary statistics. For example, suppose we want two groups of solvents and we want to calculate the mean and standard deviation in boiling points for each of those groups:</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># solvents_clustered &lt;- runMatrixAnalysis(</span>
<span class="co">#   data = solvents,</span>
<span class="co">#   analysis = c("pca"),</span>
<span class="co">#   column_w_names_of_multiple_analytes = NULL,</span>
<span class="co">#   column_w_values_for_multiple_analytes = NULL,</span>
<span class="co">#   columns_w_values_for_single_analyte = colnames(solvents)[c(3:5, 7:9, 11:12)],</span>
<span class="co">#   columns_w_additional_analyte_info = NULL,</span>
<span class="co">#   columns_w_sample_ID_info = c("solvent", "formula", "miscible_with_water", "CAS_number", "category"),</span>
<span class="co">#   transpose = FALSE,</span>
<span class="co">#   kmeans = 2,</span>
<span class="co">#   na_replacement = "drop"</span>
<span class="co"># )</span>
<span class="co"># </span>
<span class="co"># solvents_clustered_summary &lt;- solvents_clustered %&gt;%</span>
<span class="co">#   group_by(kmeans_cluster) %&gt;%</span>
<span class="co">#   summarize(mean_bp = mean(boiling_point))</span>
<span class="co"># </span>
<span class="co"># ggplot() + </span>
<span class="co">#   geom_col(</span>
<span class="co">#     data = solvents_clustered_summary,</span>
<span class="co">#     aes(x = kmeans_cluster, y = mean_bp),</span>
<span class="co">#     color = "black", fill = "white"</span>
<span class="co">#   ) +</span>
<span class="co">#   geom_point(</span>
<span class="co">#     data = solvents_clustered,</span>
<span class="co">#     aes(x = kmeans_cluster, y = boiling_point)</span>
<span class="co">#   )</span></code></pre></div>
<p>Very good! Since we can use the outputs of our k-means analyses to run and visualize summary statistics, it‚Äôs possible that we‚Äôll want to see the cluster plot (dendrogram or pca plot) alongside the summary stats plot. For this we can use the <code>plot_grid</code> function from the <code>cowplot</code> package. Let‚Äôs check it out:</p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># solvents_clustered &lt;- runMatrixAnalysis(</span>
<span class="co">#   data = solvents,</span>
<span class="co">#   analysis = c("pca"),</span>
<span class="co">#   column_w_names_of_multiple_analytes = NULL,</span>
<span class="co">#   column_w_values_for_multiple_analytes = NULL,</span>
<span class="co">#   columns_w_values_for_single_analyte = colnames(solvents)[c(3:5, 7:9, 11:12)],</span>
<span class="co">#   columns_w_additional_analyte_info = NULL,</span>
<span class="co">#   columns_w_sample_ID_info = c("solvent", "formula", "miscible_with_water", "CAS_number", "category"),</span>
<span class="co">#   transpose = FALSE,</span>
<span class="co">#   kmeans = 4,</span>
<span class="co">#   na_replacement = "drop"</span>
<span class="co"># )</span>
<span class="co"># </span>
<span class="co"># colors &lt;- c("maroon", "gold", "grey", "white")</span>
<span class="co"># </span>
<span class="co"># pca_plot &lt;- ggplot( data = solvents_clustered, aes(x = Dim.1, y = Dim.2, fill = kmeans_cluster) ) +</span>
<span class="co">#   geom_mark_ellipse(</span>
<span class="co">#     aes(label = kmeans_cluster), </span>
<span class="co">#     alpha = 0.5, label.lineheight = 0.2, size = 0.5) +</span>
<span class="co">#   geom_point(shape = 21, size = 2) +</span>
<span class="co">#   theme_classic() +</span>
<span class="co">#   guides(fill = "none") +</span>
<span class="co">#   scale_x_continuous(name = "PCA dimension 1", breaks = seq(-8,8,1)) +</span>
<span class="co">#   scale_y_continuous(name = "PCA dimension 2", breaks = seq(-7,7,1)) +</span>
<span class="co">#   scale_fill_manual(values = colors) +</span>
<span class="co">#   coord_cartesian(xlim = c(-8,8), ylim = c(-7,7))</span>
<span class="co"># </span>
<span class="co"># solvents_clustered_summary &lt;- solvents_clustered %&gt;%</span>
<span class="co">#   group_by(kmeans_cluster) %&gt;%</span>
<span class="co">#   summarize(mean_bp = mean(boiling_point))</span>
<span class="co"># </span>
<span class="co"># bar_plot &lt;- ggplot() + </span>
<span class="co">#   geom_violin(</span>
<span class="co">#     data = solvents_clustered,</span>
<span class="co">#     aes(x = kmeans_cluster, y = boiling_point, fill = kmeans_cluster),</span>
<span class="co">#     size = 0.5, color = "black", alpha = 0.6, width = 0.5</span>
<span class="co">#   ) +</span>
<span class="co">#   geom_crossbar(</span>
<span class="co">#     data = solvents_clustered_summary,</span>
<span class="co">#     aes(x = kmeans_cluster, y = mean_bp, ymin = mean_bp, ymax = mean_bp),</span>
<span class="co">#     color = "black", width = 0.5</span>
<span class="co">#   ) +</span>
<span class="co">#   geom_point(</span>
<span class="co">#     data = solvents_clustered,</span>
<span class="co">#     aes(x = kmeans_cluster, y = boiling_point),</span>
<span class="co">#     size = 2, color = "black", alpha = 0.6</span>
<span class="co">#   ) +</span>
<span class="co">#   scale_y_continuous(name = "Boiling point", breaks = seq(0,250,20)) +</span>
<span class="co">#   scale_x_discrete(name = "Cluster") +</span>
<span class="co">#   scale_fill_manual(values = colors) +</span>
<span class="co">#   theme_classic() +</span>
<span class="co">#   coord_flip() +</span>
<span class="co">#   guides(fill = "none") +</span>
<span class="co">#   theme(legend.position = "bottom")</span>
<span class="co"># </span>
<span class="co"># cowplot::plot_grid(pca_plot, bar_plot, align = "h", axis = "b", labels = "AUTO")</span></code></pre></div>
<p>Now we are really rockin!!</p>
</div>
<div id="exercises-3" class="section level3 unnumbered">
<h3>exercises<a class="anchor" aria-label="anchor" href="#exercises-3"><i class="fas fa-link"></i></a>
</h3>
<p>Use the wine grapes dataset (it‚Äôs stored as <code>wine_grape_data</code> after you run the <code>source(...)</code> command).</p>
<ul>
<li>Question 1</li>
</ul>
<p>Run a principal components analysis on the dataset. Use <code>na_replacement = "drop"</code> (so that variables with NA values are not included in the analysis) and generate clusters automatically using kmeans by setting <code>kmeans = "auto"</code>. Make scatter plot of the results. How many clusters does kmeans recommend?</p>
<ul>
<li>Question 2</li>
</ul>
<p>Modify your code from Question 1 so that only two clusters are generated. Plot the results. Use <code>geom_mark_ellipse</code> to highlight each cluster in your plot (note that the <code>fill</code> aesthetic is required to mark groups). Which varieties are put into each of the two clusters?</p>
<ul>
<li>Question 3</li>
</ul>
<p>Use an ordination plot to determine what chemicals makes Chardonnay so different from the other varieties. To what class of compounds do these chemical belong?</p>
<ul>
<li>Question 4</li>
</ul>
<p>Modify your code from Question 2 so that five clusters are generated. Plot the results. Use <code>geom_mark_ellipse</code> to highlight each cluster in your plot (note that the <code>fill</code> aesthetic is required to mark groups). Based on this plot, which grape variety undergoes the least amount of change, chemically speaking, between dry and well-watered conditions?</p>
<ul>
<li>Question 5</li>
</ul>
<p>Run a heirarchical clustering analysis on the wine grapes data set, using kmeans to create five groups, and also continue using <code>na_replacement = "drop"</code>. Plot the results. Which grape variety undergoes the most change in terms of its chemistry between well-watered and dry conditions? (hint: remember that the x-axis shows the distances between nodes and tips, the y-axis is meaningless). Compare the method you used to compare sample shifts in question 4 (i.e.¬†pca+kmeans) versus the method you used in this question (i.e.¬†hclust+kmeans). Which do you is better? Would this change depending on the circumstances?</p>
<ul>
<li>Question 6</li>
</ul>
<p>Google ‚ÄúQuercetin‚Äù. What kind of compound is it? Use the clusters created by the heirarchical clustering analysis in question 5 as groups for which to calculate summary statistics. Calculate the mean and standard deviation of the concentration of Quercetin in each group. Plot the result using <code>geom_pointrange</code> and adjust axis font sizes so that they are in good proportion with the size of the plot. Also specify a theme (for example, <code>theme_classic()</code>).</p>
<p>Does one cluster have a large amount of variation in Quercetin abundance? Why do you think this might be?</p>
<ul>
<li>Question 7</li>
</ul>
<p>Use <code><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">cowplot::plot_grid</a></code> to display your plots from questions 4 and 5 next to each other.</p>
<ul>
<li>Challenge (optional)</li>
</ul>
<p>Use cowplot to display your plots from questions 4, 5, and 6 alongside each other. <strong>Make your combined plot as attractive as possible!</strong> Use each of the following:</p>
<p><code>align = TRUE</code> inside <code><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab()</a></code></p>
<p><code>nrow = 1</code> inside <code>plot_grid()</code></p>
<p><code>rel_widths = &lt;your_choice&gt;</code> inside <code>plot_grid()</code></p>
<p><code>name = &lt;your_choice&gt;</code> inside <code>scale_*_*</code></p>
<p><code>label = kmeans_cluster</code> inside <code>geom_mark_ellipse()</code></p>
<p><code>breaks = &lt;your_choice&gt;</code> inside <code>scale_x_continuous()</code> or <code>scale_y_continuous()</code> (as an example, <code>breaks = seq(0,10,1)</code>)</p>
<p>Also, consider using:</p>
<p><code>guides(fill = "none", color = "none")</code></p>
<p>Install the RColorBrewer package, and use one of its color schemes. As an example with the color scheme <code>Set1</code>:</p>
<p><code>scale_fill_brewer(palette = "Set1", na.value = "grey")</code></p>
<p><code>scale_color_brewer(palette = "Set1", na.value = "grey")</code></p>
<p>Save your plot as a png using <code><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave()</a></code>.</p>
<p>Maybe something like this:</p>
<div class="inline-figure"><img src="https://thebustalab.github.io/integrated_bioanalytics/images/Ex6Challenge.png" width="100%" style="display: block; margin: auto;"></div>
</div>
</div>
<div id="dbscan" class="section level2 unnumbered">
<h2>dbscan<a class="anchor" aria-label="anchor" href="#dbscan"><i class="fas fa-link"></i></a>
</h2>
<p>Density based clustering</p>
<div id="further-reading-5" class="section level3 unnumbered">
<h3>further reading<a class="anchor" aria-label="anchor" href="#further-reading-5"><i class="fas fa-link"></i></a>
</h3>
<p><a href="http://www.sthda.com/english/wiki/wiki.php?id_contents=7940" class="uri">http://www.sthda.com/english/wiki/wiki.php?id_contents=7940</a></p>
<p><a href="https://ryanwingate.com/intro-to-machine-learning/unsupervised/hierarchical-and-density-based-clustering/" class="uri">https://ryanwingate.com/intro-to-machine-learning/unsupervised/hierarchical-and-density-based-clustering/</a></p>
<p><a href="https://ryanwingate.com/intro-to-machine-learning/unsupervised/hierarchical-and-density-based-clustering/hierarchical-4.png" class="uri">https://ryanwingate.com/intro-to-machine-learning/unsupervised/hierarchical-and-density-based-clustering/hierarchical-4.png</a></p>
<p><a href="https://www.geeksforgeeks.org/dbscan-clustering-in-r-programming/" class="uri">https://www.geeksforgeeks.org/dbscan-clustering-in-r-programming/</a></p>
<!-- end -->
<!-- start models -->
</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="dimensionality-reduction.html">dimensionality reduction</a></div>
<div class="next"><a href="models.html">models</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <h2>Note: Second Edition is under construction üèó</h2>
    <p>Now is a great time to provide feedback</p>
        <ul class="list-unstyled">
<li><a href="https://forms.gle/SZmB2Ct2exE2dBwv9">Provide feedback (5 min)</a></li>
          <!-- <li><a href="https://geocompr.robinlovelace.net/#reproducibility">Install updated packages</a></li> -->
          <!-- <li><a href="https://github.com/Robinlovelace/geocompr/issues">Open an issue <i class="fas fa-question"></i></a></li> -->
          <!-- <li><a href="https://discord.gg/Te3gWeDwmf">Chat on Discord <i class="fab fa-discord"></i></a></li> -->
        </ul>
<hr>
<nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#clustering">clustering</a></li>
<li>
<a class="nav-link" href="#heirarchical-clustering">heirarchical clustering</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#theory-1">theory</a></li>
<li><a class="nav-link" href="#further-reading-4">further reading</a></li>
<li><a class="nav-link" href="#exercises-2">exercises</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#k-means-clustering">k-means clustering</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#theory-2">theory</a></li>
<li><a class="nav-link" href="#exercises-3">exercises</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#dbscan">dbscan</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#further-reading-5">further reading</a></li></ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/blob/main/index.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/edit/main/index.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Integrated Bioanalytics</strong>" was written by Lucas Busta and members of the Busta lab. It was last built on 2022-10-02.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
