<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>clustering | Integrated Bioanalytics</title>
<meta name="author" content="Lucas Busta and members of the Busta lab">
<meta name="description" content="1.5 theory “Which of my samples are most closely related?” So far we have been looking at how to plot raw data, as well as data that has been summarize across samples. This is important stuff and...">
<meta name="generator" content="bookdown 0.25 with bs4_book()">
<meta property="og:title" content="clustering | Integrated Bioanalytics">
<meta property="og:type" content="book">
<meta property="og:description" content="1.5 theory “Which of my samples are most closely related?” So far we have been looking at how to plot raw data, as well as data that has been summarize across samples. This is important stuff and...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="clustering | Integrated Bioanalytics">
<meta name="twitter:description" content="1.5 theory “Which of my samples are most closely related?” So far we have been looking at how to plot raw data, as well as data that has been summarize across samples. This is important stuff and...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><link href="libs/Lato-0.4.1/font.css" rel="stylesheet">
<link href="libs/Roboto%20Mono-0.4.1/font.css" rel="stylesheet">
<link href="libs/Montserrat-0.4.1/font.css" rel="stylesheet">
<script src="libs/bs3compat-0.3.1.9000/transition.js"></script><script src="libs/bs3compat-0.3.1.9000/tabs.js"></script><script src="libs/bs3compat-0.3.1.9000/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><link rel="icon" type="image/png" sizes="32x32" href="images/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="images/favicon-16x16.png">
<script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-99618359-1', 'auto');
      ga('send', 'pageview');

    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="style/style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h2>
        <a href="index.html" title="">Integrated Bioanalytics</a>
      </h2>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">WELCOME</a></li>
<li class="book-part">DATA ANALYSIS IN R</li>
<li><a class="" href="overview.html">overview</a></li>
<li><a class="" href="installation.html">installation</a></li>
<li><a class="" href="data-visualization.html">data visualization</a></li>
<li><a class="" href="data-wranging.html">data wranging</a></li>
<li><a class="active" href="clustering.html">clustering</a></li>
<li><a class="" href="pca.html">pca</a></li>
<li><a class="" href="k-means.html">k-means</a></li>
<li><a class="" href="models.html">models</a></li>
<li><a class="" href="comparing-means.html">comparing means</a></li>
<li class="book-part">GC-MS DATA</li>
<li><a class="" href="mass-spectrometric-analysis.html">mass spectrometric analysis</a></li>
<li class="book-part">DNA AND RNA SEQUENCE DATA</li>
<li><a class="" href="data-acquisition.html">data acquisition</a></li>
<li><a class="" href="data-transfer.html">data transfer</a></li>
<li><a class="" href="sequence-assessment.html">sequence assessment</a></li>
<li><a class="" href="illumina-read-assessment.html">illumina read assessment</a></li>
<li class="book-part">TRANSCRIPTOME ANALYSIS</li>
<li><a class="" href="transcriptomic-analyses.html">transcriptomic analyses</a></li>
<li class="book-part">GENOMIC ANALYSIS</li>
<li><a class="" href="setup.html">setup</a></li>
<li><a class="" href="genome-assembly.html">genome assembly</a></li>
<li><a class="" href="comparative-genomics.html">comparative genomics</a></li>
<li class="book-part">EVOLUTIONARY ANALYSIS</li>
<li><a class="" href="blast-searches.html">blast searches</a></li>
<li><a class="" href="alignments.html">alignments</a></li>
<li><a class="" href="phylogenies.html">phylogenies</a></li>
<li><a class="" href="phylogenetic-analyses.html">phylogenetic analyses</a></li>
<li class="book-part">SCIENTIFIC WRITING</li>
<li><a class="" href="overview-1.html">overview</a></li>
<li><a class="" href="a-mini-manuscript.html">a mini manuscript</a></li>
<li><a class="" href="figures-captions.html">figures &amp; captions</a></li>
<li><a class="" href="results-and-discussion.html">results and discussion</a></li>
<li><a class="" href="conclusion-and-introduction.html">conclusion and introduction</a></li>
<li><a class="" href="abstract-and-title.html">abstract and title</a></li>
<li><a class="" href="proposals.html">proposals</a></li>
<li class="book-part">IMAGE ANALYSIS</li>
<li><a class="" href="image-color-analysis.html">image color analysis</a></li>
<li><a class="" href="images-of-mass-spectra.html">images of mass spectra</a></li>
<li class="book-part">APPENDIX</li>
<li><a class="" href="links.html">links</a></li>
<li><a class="" href="r-faq.html">r faq</a></li>
<li><a class="" href="templates.html">templates</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="clustering" class="section level1 unnumbered">
<h1>clustering<a class="anchor" aria-label="anchor" href="#clustering"><i class="fas fa-link"></i></a>
</h1>
<div id="theory" class="section level2" number="1.5">
<h2>
<span class="header-section-number">1.5</span> theory<a class="anchor" aria-label="anchor" href="#theory"><i class="fas fa-link"></i></a>
</h2>
<p>“Which of my samples are most closely related?”</p>
<p>So far we have been looking at how to plot raw data, as well as data that has been summarize across samples. This is important stuff and very useful. However, we often have questions about how samples in our datasets relate to one another. For example: in the Alaska lakes dataset, which lake is most similar, chemically speaking, to Lake Narvakrak? Answering this requires calculating numeric distances between samples based on their chemical properties. For this, and other questions, we need to use matrix analyses. For this, we will use <code>runMatrixAnalysis()</code>, a function that is loaded into your R Session when you run the source() command.</p>
<p>Matrix analyses can be a bit difficult to set up. There are two things that we can do to help us with this: (i) we will use a template for <code>runMatrixAnalysis()</code> (see below) and (ii) it is <em>critical</em> that we think about our data in terms of <strong>samples</strong> and <strong>analytes</strong>. Let’s consider our Alaska lakes data set:</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">alaska_lake_data</span>
<span class="co">## # A tibble: 220 × 7</span>
<span class="co">##    lake              park  water_temp    pH element mg_per_L</span>
<span class="co">##    &lt;chr&gt;             &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;</span>
<span class="co">##  1 Devil_Mountain_L… BELA        6.46  7.69 C          3.4  </span>
<span class="co">##  2 Devil_Mountain_L… BELA        6.46  7.69 N          0.028</span>
<span class="co">##  3 Devil_Mountain_L… BELA        6.46  7.69 P          0    </span>
<span class="co">##  4 Devil_Mountain_L… BELA        6.46  7.69 Cl        10.4  </span>
<span class="co">##  5 Devil_Mountain_L… BELA        6.46  7.69 S          0.62 </span>
<span class="co">##  6 Devil_Mountain_L… BELA        6.46  7.69 F          0.04 </span>
<span class="co">##  7 Devil_Mountain_L… BELA        6.46  7.69 Br         0.02 </span>
<span class="co">##  8 Devil_Mountain_L… BELA        6.46  7.69 Na         8.92 </span>
<span class="co">##  9 Devil_Mountain_L… BELA        6.46  7.69 K          1.2  </span>
<span class="co">## 10 Devil_Mountain_L… BELA        6.46  7.69 Ca         5.73 </span>
<span class="co">## # … with 210 more rows, and 1 more variable:</span>
<span class="co">## #   element_type &lt;chr&gt;</span></code></pre></div>
<p>We can see that this dataset is comprised of measurements of various <em>analytes</em> (i.e. several chemical elements, as well as water_temp, and pH), in different <em>samples</em> (i.e. lakes). We need to tell the <code>runMatrixAnalysis()</code> function how each column relates to this samples and analytes structure. See the image below for an explanation.</p>
<div class="inline-figure"><img src="https://thebustalab.github.io/integrated_bioanalytics/images/runMatrixAnalysis1.png" width="100%" style="display: block; margin: auto;"></div>
<p>With this in mind, let’s try out our template:</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">AK_lakes_clustered</span> <span class="op">&lt;-</span> <span class="fu">runMatrixAnalysis</span><span class="op">(</span>
                                
    data <span class="op">=</span> <span class="va">alaska_lake_data</span>,

    analysis <span class="op">=</span> <span class="st">"hclust"</span>,

    column_w_names_of_multiple_analytes <span class="op">=</span> <span class="st">"element"</span>,
    column_w_values_for_multiple_analytes <span class="op">=</span> <span class="st">"mg_per_L"</span>,
    
    columns_w_values_for_single_analyte <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"water_temp"</span>, <span class="st">"pH"</span><span class="op">)</span>,
    
    columns_w_additional_analyte_info <span class="op">=</span> <span class="st">"element_type"</span>,

    columns_w_sample_ID_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lake"</span>, <span class="st">"park"</span><span class="op">)</span>

<span class="op">)</span>
<span class="va">AK_lakes_clustered</span>
<span class="co">## # A tibble: 39 × 25</span>
<span class="co">##    sample_unique_ID   lake  park  parent  node branch.length</span>
<span class="co">##    &lt;chr&gt;              &lt;chr&gt; &lt;chr&gt;  &lt;int&gt; &lt;int&gt;         &lt;dbl&gt;</span>
<span class="co">##  1 Devil_Mountain_La… Devi… BELA      33     1         0.979</span>
<span class="co">##  2 Imuruk_Lake_BELA   Imur… BELA      39     2         0.816</span>
<span class="co">##  3 Kuzitrin_Lake_BELA Kuzi… BELA      38     3         0.703</span>
<span class="co">##  4 Lava_Lake_BELA     Lava… BELA      27     4         0.738</span>
<span class="co">##  5 North_Killeak_Lak… Nort… BELA      21     5         5.58 </span>
<span class="co">##  6 White_Fish_Lake_B… Whit… BELA      22     6         3.89 </span>
<span class="co">##  7 Iniakuk_Lake_GAAR  Inia… GAAR      29     7         1.25 </span>
<span class="co">##  8 Kurupa_Lake_GAAR   Kuru… GAAR      35     8         0.954</span>
<span class="co">##  9 Lake_Matcharak_GA… Lake… GAAR      35     9         0.954</span>
<span class="co">## 10 Lake_Selby_GAAR    Lake… GAAR      36    10         1.12 </span>
<span class="co">## # … with 29 more rows, and 19 more variables: label &lt;chr&gt;,</span>
<span class="co">## #   isTip &lt;lgl&gt;, x &lt;dbl&gt;, y &lt;dbl&gt;, branch &lt;dbl&gt;,</span>
<span class="co">## #   angle &lt;dbl&gt;, water_temp &lt;dbl&gt;, pH &lt;dbl&gt;, C &lt;dbl&gt;,</span>
<span class="co">## #   N &lt;dbl&gt;, P &lt;dbl&gt;, Cl &lt;dbl&gt;, S &lt;dbl&gt;, F &lt;dbl&gt;, Br &lt;dbl&gt;,</span>
<span class="co">## #   Na &lt;dbl&gt;, K &lt;dbl&gt;, Ca &lt;dbl&gt;, Mg &lt;dbl&gt;</span></code></pre></div>
<p>It works! Now we can plot our cluster diagram with a ggplot add-on called ggtree. We’ve seen that ggplot takes a “data” argument (i.e. <code>ggplot(data = &lt;some_data&gt;) + geom_*()</code> etc.). In contrast, ggtree takes an argument called <code>tr</code>, though if you’re using the <code>runMatrixAnalysis()</code> function, you can treat these two (<code>data</code> and <code>tr</code>) the same, so, use: <code>ggtree(tr = &lt;output_from_runMatrixAnalysis&gt;) + geom_*()</code> etc.</p>
<p>Note that <code>ggtree</code> also comes with several great new geoms: <code><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab()</a></code> and <code><a href="https://rdrr.io/pkg/ggtree/man/geom_tippoint.html">geom_tippoint()</a></code>. Let’s try those out:</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/treedata-book/">ggtree</a></span><span class="op">)</span>
<span class="va">AK_lakes_clustered</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tippoint.html">geom_tippoint</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-93-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Cool! Though that plot could use some tweaking… let’s try:</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">AK_lakes_clustered</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">lake</span><span class="op">)</span>, offset <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tippoint.html">geom_tippoint</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">21</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">park</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_x_continuous</span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">15</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-94-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Very nice!</p>
</div>
<div id="further-reading-1" class="section level2" number="1.6">
<h2>
<span class="header-section-number">1.6</span> further reading<a class="anchor" aria-label="anchor" href="#further-reading-1"><i class="fas fa-link"></i></a>
</h2>
<p>For more information on plotting annotated trees, see: <a href="https://yulab-smu.top/treedata-book/chapter10.html" class="uri">https://yulab-smu.top/treedata-book/chapter10.html</a>.</p>
</div>
<div id="exercises-1" class="section level2" number="1.7">
<h2>
<span class="header-section-number">1.7</span> exercises<a class="anchor" aria-label="anchor" href="#exercises-1"><i class="fas fa-link"></i></a>
</h2>
<p>For this set of exercises, please use <code>runMatrixAnalysis()</code> to run and visualize a hierarchical cluster analysis with each of the main datasets that we have worked with so far, except for NY_trees. This means: <code>algae_data</code> (which algae strains are most similar to each other?), <code>alaska_lake_data</code> (which lakes are most similar to each other?). and <code>solvents</code> (which solvents are most similar to each other?). It also means you should use the periodic table (which elements are most similar to each other?), though please don’t use the whole periodic table, rather, use <code>periodic_table_subset</code>. For each of these, create (i) a tree diagram that shows how the “samples” in each data set are related to each other based on the numerical data associated with them, (ii) a caption for each diagram, and (iii) describe, in two or so sentences, an interesting trend you see in the diagram. You can ignore columns that contain categorical data, or you can list those columns as “additional_analyte_info”.</p>
<p>For this assignment, you may find the <code><a href="https://rdrr.io/r/base/colnames.html">colnames()</a></code> function and square bracket-subsetting useful. It will list all or a subset of the column names in a dataset for you. For example:</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">solvents</span><span class="op">)</span>
<span class="co">##  [1] "solvent"             "formula"            </span>
<span class="co">##  [3] "boiling_point"       "melting_point"      </span>
<span class="co">##  [5] "density"             "miscible_with_water"</span>
<span class="co">##  [7] "solubility_in_water" "relative_polarity"  </span>
<span class="co">##  [9] "vapor_pressure"      "CAS_number"         </span>
<span class="co">## [11] "formula_weight"      "refractive_index"   </span>
<span class="co">## [13] "specific_gravity"    "category"</span>

<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">solvents</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>
<span class="co">## [1] "solvent"       "formula"       "boiling_point"</span>

<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">solvents</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">5</span>,<span class="fl">7</span><span class="op">)</span><span class="op">]</span>
<span class="co">## [1] "solvent"             "density"            </span>
<span class="co">## [3] "solubility_in_water"</span></code></pre></div>
<!-- end -->
<!-- start principal components analysis -->
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="data-wranging.html">data wranging</a></div>
<div class="next"><a href="pca.html">pca</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <h2>Note: Second Edition is under construction 🏗</h2>
    <p>Now is a great time to provide feedback</p>
        <ul class="list-unstyled">
<li><a href="https://forms.gle/SZmB2Ct2exE2dBwv9">Provide feedback (5 min)</a></li>
          <!-- <li><a href="https://geocompr.robinlovelace.net/#reproducibility">Install updated packages</a></li> -->
          <!-- <li><a href="https://github.com/Robinlovelace/geocompr/issues">Open an issue <i class="fas fa-question"></i></a></li> -->
          <!-- <li><a href="https://discord.gg/Te3gWeDwmf">Chat on Discord <i class="fab fa-discord"></i></a></li> -->
        </ul>
<hr>
<nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#clustering">clustering</a></li>
<li><a class="nav-link" href="#theory"><span class="header-section-number">1.5</span> theory</a></li>
<li><a class="nav-link" href="#further-reading-1"><span class="header-section-number">1.6</span> further reading</a></li>
<li><a class="nav-link" href="#exercises-1"><span class="header-section-number">1.7</span> exercises</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/blob/main/index.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/edit/main/index.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Integrated Bioanalytics</strong>" was written by Lucas Busta and members of the Busta lab. It was last built on 2022-08-24.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
