<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>data wrangling and summaries | Integrated Bioanalytics</title>
<meta name="author" content="Lucas Busta and members of the Busta lab">
<meta name="description" content="Data wrangling refers to the process of organizing, cleaning up, and making a “raw” data set more ready for downstream analysis. It is a key piece of any data analysis process. Here we will look...">
<meta name="generator" content="bookdown 0.35 with bs4_book()">
<meta property="og:title" content="data wrangling and summaries | Integrated Bioanalytics">
<meta property="og:type" content="book">
<meta property="og:description" content="Data wrangling refers to the process of organizing, cleaning up, and making a “raw” data set more ready for downstream analysis. It is a key piece of any data analysis process. Here we will look...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="data wrangling and summaries | Integrated Bioanalytics">
<meta name="twitter:description" content="Data wrangling refers to the process of organizing, cleaning up, and making a “raw” data set more ready for downstream analysis. It is a key piece of any data analysis process. Here we will look...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><link href="libs/Lato-0.4.6/font.css" rel="stylesheet">
<link href="libs/Roboto_Mono-0.4.6/font.css" rel="stylesheet">
<link href="libs/Montserrat-0.4.6/font.css" rel="stylesheet">
<script src="libs/bs3compat-0.4.2/transition.js"></script><script src="libs/bs3compat-0.4.2/tabs.js"></script><script src="libs/bs3compat-0.4.2/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><link rel="icon" type="image/png" sizes="32x32" href="images/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="images/favicon-16x16.png">
<script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-99618359-1', 'auto');
      ga('send', 'pageview');

    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="style/style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h2>
        <a href="index.html" title="">Integrated Bioanalytics</a>
      </h2>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">WELCOME</a></li>
<li class="book-part">DATA ANALYSIS IN R</li>
<li><a class="" href="overview.html">overview</a></li>
<li><a class="" href="installation.html">installation</a></li>
<li><a class="" href="data-visualization-i.html">data visualization I</a></li>
<li><a class="" href="data-visualization-ii.html">data visualization II</a></li>
<li><a class="" href="data-visualization-iii.html">data visualization III</a></li>
<li><a class="active" href="data-wrangling-and-summaries.html">data wrangling and summaries</a></li>
<li><a class="" href="dimensionality-reduction.html">dimensionality reduction</a></li>
<li><a class="" href="clustering.html">clustering</a></li>
<li><a class="" href="models.html">models</a></li>
<li><a class="" href="comparing-means.html">comparing means</a></li>
<li><a class="" href="special-topics.html">special topics</a></li>
<li class="book-part">GC-MS DATA</li>
<li><a class="" href="mass-spectrometric-analysis.html">mass spectrometric analysis</a></li>
<li class="book-part">TRANSCRIPTOME ANALYSIS</li>
<li><a class="" href="transcriptome-assembly.html">transcriptome assembly</a></li>
<li class="book-part">EVOLUTIONARY ANALYSIS</li>
<li><a class="" href="similarity-searching.html">similarity searching</a></li>
<li><a class="" href="alignments.html">alignments</a></li>
<li><a class="" href="phylogenies.html">phylogenies</a></li>
<li><a class="" href="phylogenetic-analyses.html">phylogenetic analyses</a></li>
<li><a class="" href="comparative-genomics.html">comparative genomics</a></li>
<li class="book-part">LANGUAGE MODELS</li>
<li><a class="" href="completiongpt.html">completionGPT</a></li>
<li><a class="" href="analyzeliterature.html">analyzeLiterature</a></li>
<li class="book-part">SCIENTIFIC WRITING</li>
<li><a class="" href="overview-1.html">overview</a></li>
<li><a class="" href="figures-captions.html">figures &amp; captions</a></li>
<li><a class="" href="results-and-discussion.html">results and discussion</a></li>
<li><a class="" href="conclusion-and-introduction.html">conclusion and introduction</a></li>
<li><a class="" href="abstract-and-title.html">abstract and title</a></li>
<li class="book-part">APPENDIX</li>
<li><a class="" href="links.html">links</a></li>
<li><a class="" href="r-faq.html">r faq</a></li>
<li><a class="" href="templates.html">templates</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="data-wrangling-and-summaries" class="section level1 unnumbered">
<h1>data wrangling and summaries<a class="anchor" aria-label="anchor" href="#data-wrangling-and-summaries"><i class="fas fa-link"></i></a>
</h1>
<div class="inline-figure"><img src="https://thebustalab.github.io/integrated_bioanalytics/images/wrangling.png" width="100%" style="display: block; margin: auto;"></div>
<p>Data wrangling refers to the process of organizing, cleaning up, and making a “raw” data set more ready for downstream analysis. It is a key piece of any data analysis process. Here we will look at a few different aspects of wrangling, including data import, subsetting, pivoting, and summarizing data.</p>
<div id="data-import" class="section level2 unnumbered">
<h2>data import<a class="anchor" aria-label="anchor" href="#data-import"><i class="fas fa-link"></i></a>
</h2>
<p>To analyze data that is stored on your own computer you can indeed import it into RStudio.</p>
<p>The easiest way to do this is to use the interactive command <code>readCSV()</code>, a function that comes with the phylochemistry source command. You run <code>readCSV()</code> in your console, then navigate to the data on your hard drive.</p>
<p>Another option is to read the data in from a path. For this, you will need to know the “path” to your data file. This is essentially the street address of your data on your computer’s hard drive. Paths look different on Mac and PC.</p>
<ul>
<li>On Mac: <code>/Users/lucasbusta/Documents/sample_data_set.csv</code> (note the forward slashes!)</li>
<li>On PC: <code>C:\\My Computer\\Documents\\sample_data_set.csv</code> (note double backward slashes!)</li>
</ul>
<p>You can quickly find paths to files via the following:</p>
<ul>
<li>On Mac: Locate the file in Finder. Right-click on the file, hold the Option key, then click “Copy <file> as Pathname”</file>
</li>
<li>On PC: Locate the file in Windows Explorer. Hold down the Shift key then right-click on the file. Click “Copy As Path”</li>
</ul>
<p>With these paths, we can read in data using the <code>read_csv</code> command. We’ll run <code>read_csv("&lt;path_to_your_data&gt;")</code>. Note the use of QUOTES <code>""</code>! Those are necessary. Also make sure your path uses the appropriate direction of slashes for your operating system.</p>
</div>
<div id="subsetting" class="section level2 unnumbered">
<h2>subsetting<a class="anchor" aria-label="anchor" href="#subsetting"><i class="fas fa-link"></i></a>
</h2>
<p>So far, we have always been passing whole data sets to ggplot to do our plotting. However, suppose we wanted to get at just certain portions of our dataset, say, specific columns, or specific rows? Here are a few ways to do that:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># To look at a single column (the third column)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">alaska_lake_data</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 6 × 1</span></span>
<span><span class="co">##   water_temp</span></span>
<span><span class="co">##        &lt;dbl&gt;</span></span>
<span><span class="co">## 1       6.46</span></span>
<span><span class="co">## 2       6.46</span></span>
<span><span class="co">## 3       6.46</span></span>
<span><span class="co">## 4       6.46</span></span>
<span><span class="co">## 5       6.46</span></span>
<span><span class="co">## 6       6.46</span></span>
<span></span>
<span><span class="co"># To look at select columns:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">alaska_lake_data</span><span class="op">[</span>,<span class="fl">2</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 6 × 4</span></span>
<span><span class="co">##   park  water_temp    pH element</span></span>
<span><span class="co">##   &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;  </span></span>
<span><span class="co">## 1 BELA        6.46  7.69 C      </span></span>
<span><span class="co">## 2 BELA        6.46  7.69 N      </span></span>
<span><span class="co">## 3 BELA        6.46  7.69 P      </span></span>
<span><span class="co">## 4 BELA        6.46  7.69 Cl     </span></span>
<span><span class="co">## 5 BELA        6.46  7.69 S      </span></span>
<span><span class="co">## 6 BELA        6.46  7.69 F</span></span>
<span></span>
<span><span class="co"># To look at a single row (the second row)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">alaska_lake_data</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 1 × 7</span></span>
<span><span class="co">##   lake  park  water_temp    pH element mg_per_L element_type</span></span>
<span><span class="co">##   &lt;chr&gt; &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt; &lt;chr&gt;       </span></span>
<span><span class="co">## 1 Devi… BELA        6.46  7.69 N          0.028 bound</span></span>
<span></span>
<span><span class="co"># To look at select rows:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">alaska_lake_data</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 4 × 7</span></span>
<span><span class="co">##   lake  park  water_temp    pH element mg_per_L element_type</span></span>
<span><span class="co">##   &lt;chr&gt; &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt; &lt;chr&gt;       </span></span>
<span><span class="co">## 1 Devi… BELA        6.46  7.69 N          0.028 bound       </span></span>
<span><span class="co">## 2 Devi… BELA        6.46  7.69 P          0     bound       </span></span>
<span><span class="co">## 3 Devi… BELA        6.46  7.69 Cl        10.4   free        </span></span>
<span><span class="co">## 4 Devi… BELA        6.46  7.69 S          0.62  free</span></span>
<span></span>
<span><span class="co"># To look at just a single column, by name</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">alaska_lake_data</span><span class="op">$</span><span class="va">pH</span><span class="op">)</span></span>
<span><span class="co">## [1] 7.69 7.69 7.69 7.69 7.69 7.69</span></span>
<span></span>
<span><span class="co"># To look at select columns by name</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">select</span><span class="op">(</span><span class="va">alaska_lake_data</span>, <span class="va">park</span>, <span class="va">water_temp</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 6 × 2</span></span>
<span><span class="co">##   park  water_temp</span></span>
<span><span class="co">##   &lt;chr&gt;      &lt;dbl&gt;</span></span>
<span><span class="co">## 1 BELA        6.46</span></span>
<span><span class="co">## 2 BELA        6.46</span></span>
<span><span class="co">## 3 BELA        6.46</span></span>
<span><span class="co">## 4 BELA        6.46</span></span>
<span><span class="co">## 5 BELA        6.46</span></span>
<span><span class="co">## 6 BELA        6.46</span></span></code></pre></div>
</div>
<div id="wide-and-long-data" class="section level2 unnumbered">
<h2>wide and long data<a class="anchor" aria-label="anchor" href="#wide-and-long-data"><i class="fas fa-link"></i></a>
</h2>
<p>When we make data tables by hand, it’s often easy to make a <strong>wide-style table</strong> like the following. In it, the abundances of 7 different fatty acids in 10 different species are tabulated. Each fatty acid gets its own row, each species, its own column.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">fadb_sample</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 6 × 11</span></span>
<span><span class="co">##   fatty_acid      Agonandra_brasiliensis Agonandra_silvatica</span></span>
<span><span class="co">##   &lt;chr&gt;                            &lt;dbl&gt;               &lt;dbl&gt;</span></span>
<span><span class="co">## 1 Hexadecanoic a…                    3.4                 1  </span></span>
<span><span class="co">## 2 Octadecanoic a…                    6.2                 0.1</span></span>
<span><span class="co">## 3 Eicosanoic acid                    4.7                 3.5</span></span>
<span><span class="co">## 4 Docosanoic acid                   77.4                 0.4</span></span>
<span><span class="co">## 5 Tetracosanoic …                    1.4                 1  </span></span>
<span><span class="co">## 6 Hexacosanoic a…                    1.9                12.6</span></span>
<span><span class="co">## # ℹ 8 more variables: Agonandra_excelsa &lt;dbl&gt;,</span></span>
<span><span class="co">## #   Heisteria_silvianii &lt;dbl&gt;, Malania_oleifera &lt;dbl&gt;,</span></span>
<span><span class="co">## #   Ximenia_americana &lt;dbl&gt;, Ongokea_gore &lt;dbl&gt;,</span></span>
<span><span class="co">## #   Comandra_pallida &lt;dbl&gt;, Buckleya_distichophylla &lt;dbl&gt;,</span></span>
<span><span class="co">## #   Nuytsia_floribunda &lt;dbl&gt;</span></span></code></pre></div>
<p>While this format is very nice for filling in my hand (such as in a lab notebook or similar), it does not groove with ggplot and other <code>tidyverse</code> functions very well. We need to convert it into a <strong>long-style table</strong>. This is done using <code>pivot_longer()</code>. You can think of this function as transforming both your data’s column names (or some of the column names) and your data matrix’s values (in this case, the measurements) each into their own variables (i.e. columns). We can do this for our fatty acid dataset using the command below. In it, we specify what data we want to transform (<code>data = fadb_sample</code>), we need to tell it what columns we want to transform (<code>cols = 2:11</code>), what we want the new variable that contains column names to be called (<code>names_to = "plant_species"</code>) and what we want the new variable that contains matrix values to be called (<code>values_to = "relative_abundance"</code>). All together now:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pivot_longer</span><span class="op">(</span>data <span class="op">=</span> <span class="va">fadb_sample</span>, cols <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">11</span>, names_to <span class="op">=</span> <span class="st">"plant_species"</span>, values_to <span class="op">=</span> <span class="st">"relative_abundance"</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 70 × 3</span></span>
<span><span class="co">##    fatty_acid        plant_species        relative_abundance</span></span>
<span><span class="co">##    &lt;chr&gt;             &lt;chr&gt;                             &lt;dbl&gt;</span></span>
<span><span class="co">##  1 Hexadecanoic acid Agonandra_brasilien…                3.4</span></span>
<span><span class="co">##  2 Hexadecanoic acid Agonandra_silvatica                 1  </span></span>
<span><span class="co">##  3 Hexadecanoic acid Agonandra_excelsa                   1.2</span></span>
<span><span class="co">##  4 Hexadecanoic acid Heisteria_silvianii                 2.9</span></span>
<span><span class="co">##  5 Hexadecanoic acid Malania_oleifera                    0.7</span></span>
<span><span class="co">##  6 Hexadecanoic acid Ximenia_americana                   3.3</span></span>
<span><span class="co">##  7 Hexadecanoic acid Ongokea_gore                        1  </span></span>
<span><span class="co">##  8 Hexadecanoic acid Comandra_pallida                    2.3</span></span>
<span><span class="co">##  9 Hexadecanoic acid Buckleya_distichoph…                1.6</span></span>
<span><span class="co">## 10 Hexadecanoic acid Nuytsia_floribunda                  3.8</span></span>
<span><span class="co">## # ℹ 60 more rows</span></span></code></pre></div>
<p>Brilliant! Now we have a long-style table that can be used with ggplot.</p>
</div>
<div id="the-pipe" class="section level2 unnumbered">
<h2>the pipe (%&gt;%)<a class="anchor" aria-label="anchor" href="#the-pipe"><i class="fas fa-link"></i></a>
</h2>
<p>We have seen how to create new objects using <code>&lt;-</code>, and we have been filtering and plotting data using, for example:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">alaska_lake_data</span>, <span class="va">park</span> <span class="op">==</span> <span class="st">"BELA"</span><span class="op">)</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">pH</span>, y <span class="op">=</span> <span class="va">lake</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_col</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-79-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>However, as our analyses get more complex, the code can get long and hard to read. We’re going to use the pipe <code>%&gt;%</code> to help us with this. Check it out:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alaska_lake_data</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">park</span> <span class="op">==</span> <span class="st">"BELA"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">pH</span>, y <span class="op">=</span> <span class="va">lake</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_col</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-80-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Neat! Another way to think about the pipe:</p>
<div class="inline-figure"><img src="https://thebustalab.github.io/integrated_bioanalytics/images/pipe.jpg" width="100%" style="display: block; margin: auto;"></div>
<p>The pipe will become more important as our analyses become more sophisticated, which happens very quickly when we start working with summary statistics, as we shall now see…</p>
</div>
<div id="summary-statistics" class="section level2 unnumbered">
<h2>summary statistics<a class="anchor" aria-label="anchor" href="#summary-statistics"><i class="fas fa-link"></i></a>
</h2>
<p>So far, we have been plotting raw data. This is well and good, but it is not always suitable. Often we have scientific questions that cannot be answered by looking at raw data alone, or sometimes there is too much raw data to plot. For this, we need summary statistics - things like averages, standard deviations, and so on. While these metrics can be computed in Excel, programming such can be time consuming, especially for group statistics. Consider the example below, which uses the <code>ny_trees</code> dataset. The NY Trees dataset contains information on nearly half a million trees in New York City (this is after considerable filtering and simplification):</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ny_trees</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 6 × 14</span></span>
<span><span class="co">##   tree_height tree_diameter address        tree_loc pit_type</span></span>
<span><span class="co">##         &lt;dbl&gt;         &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;    &lt;chr&gt;   </span></span>
<span><span class="co">## 1        21.1             6 1139 57 STREET Front    Sidewal…</span></span>
<span><span class="co">## 2        59.0             6 2220 BERGEN A… Across   Sidewal…</span></span>
<span><span class="co">## 3        92.4            13 2254 BERGEN A… Across   Sidewal…</span></span>
<span><span class="co">## 4        50.2            15 2332 BERGEN A… Across   Sidewal…</span></span>
<span><span class="co">## 5        95.0            21 2361 EAST   7… Front    Sidewal…</span></span>
<span><span class="co">## 6        67.5            19 2409 EAST   7… Front    Continu…</span></span>
<span><span class="co">## # ℹ 9 more variables: soil_lvl &lt;chr&gt;, status &lt;chr&gt;,</span></span>
<span><span class="co">## #   spc_latin &lt;chr&gt;, spc_common &lt;chr&gt;, trunk_dmg &lt;chr&gt;,</span></span>
<span><span class="co">## #   zipcode &lt;dbl&gt;, boroname &lt;chr&gt;, latitude &lt;dbl&gt;,</span></span>
<span><span class="co">## #   longitude &lt;dbl&gt;</span></span></code></pre></div>
<p>More than 300,000 observations of 14 variables! That’s 4.2M data points! Now, what is the average and standard deviation of the height and diameter of each tree species within each NY borough? Do those values change for trees that are in parks versus sidewalk pits?? I don’t even know how one would begin to approach such questions using traditional spreadsheets. Here, we will answer these questions with ease using two new commands: <code>group_by()</code> and <code>summarize()</code>. Let’s get to it.</p>
<p>Say that we want to know (and of course, visualize) the mean and standard deviation of the heights of each tree species in NYC. We can see that data in first few columns of the NY trees dataset above, but how to calculate these statistics? In R, mean can be computed with <code><a href="https://rdrr.io/r/base/mean.html">mean()</a></code> and standard deviation can be calculated with <code><a href="https://rdrr.io/r/stats/sd.html">sd()</a></code>. We will use the function <code>summarize()</code> to calculate summary statistics. So, we can calculate the average and standard deviation of all the trees in the data set as follows:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ny_trees</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarize</span><span class="op">(</span>mean_height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">tree_height</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 1 × 1</span></span>
<span><span class="co">##   mean_height</span></span>
<span><span class="co">##         &lt;dbl&gt;</span></span>
<span><span class="co">## 1        72.6</span></span>
<span></span>
<span><span class="va">ny_trees</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarize</span><span class="op">(</span>stdev_height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span><span class="op">(</span><span class="va">tree_height</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 1 × 1</span></span>
<span><span class="co">##   stdev_height</span></span>
<span><span class="co">##          &lt;dbl&gt;</span></span>
<span><span class="co">## 1         28.7</span></span></code></pre></div>
<p>Great! But how to do this for each species? We need to subdivide the data by species, then compute the mean and standard deviation, then recombine the results into a new table. First, we use <code>group_by()</code>. Note that in ny_trees, species are indicated in the column called <code>spc_latin</code>. Once the data is grouped, we can use <code>summarize()</code> to compute statistics.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ny_trees</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">spc_latin</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarize</span><span class="op">(</span>mean_height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">tree_height</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 12 × 2</span></span>
<span><span class="co">##    spc_latin              mean_height</span></span>
<span><span class="co">##    &lt;chr&gt;                        &lt;dbl&gt;</span></span>
<span><span class="co">##  1 ACER PLATANOIDES              82.6</span></span>
<span><span class="co">##  2 ACER RUBRUM                  106. </span></span>
<span><span class="co">##  3 ACER SACCHARINUM              65.6</span></span>
<span><span class="co">##  4 FRAXINUS PENNSYLVANICA        60.6</span></span>
<span><span class="co">##  5 GINKGO BILOBA                 90.4</span></span>
<span><span class="co">##  6 GLEDITSIA TRIACANTHOS         53.0</span></span>
<span><span class="co">##  7 PLATANUS ACERIFOLIA           82.0</span></span>
<span><span class="co">##  8 PYRUS CALLERYANA              21.0</span></span>
<span><span class="co">##  9 QUERCUS PALUSTRIS             65.5</span></span>
<span><span class="co">## 10 QUERCUS RUBRA                111. </span></span>
<span><span class="co">## 11 TILIA CORDATA                 98.8</span></span>
<span><span class="co">## 12 ZELKOVA SERRATA              101.</span></span></code></pre></div>
<p>Bam. Mean height of each tree species. We can also count the number of observations using <code>n()</code>:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ny_trees</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">spc_latin</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarize</span><span class="op">(</span>number_of_individuals <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 12 × 2</span></span>
<span><span class="co">##    spc_latin              number_of_individuals</span></span>
<span><span class="co">##    &lt;chr&gt;                                  &lt;int&gt;</span></span>
<span><span class="co">##  1 ACER PLATANOIDES                       67260</span></span>
<span><span class="co">##  2 ACER RUBRUM                            11506</span></span>
<span><span class="co">##  3 ACER SACCHARINUM                       13161</span></span>
<span><span class="co">##  4 FRAXINUS PENNSYLVANICA                 16987</span></span>
<span><span class="co">##  5 GINKGO BILOBA                          15672</span></span>
<span><span class="co">##  6 GLEDITSIA TRIACANTHOS                  48707</span></span>
<span><span class="co">##  7 PLATANUS ACERIFOLIA                    80075</span></span>
<span><span class="co">##  8 PYRUS CALLERYANA                       39125</span></span>
<span><span class="co">##  9 QUERCUS PALUSTRIS                      37058</span></span>
<span><span class="co">## 10 QUERCUS RUBRA                          10020</span></span>
<span><span class="co">## 11 TILIA CORDATA                          25970</span></span>
<span><span class="co">## 12 ZELKOVA SERRATA                        13221</span></span></code></pre></div>
<p>Cool! <code>summarize()</code> is more powerful though, we can do many summary statistics at once:</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ny_trees</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">spc_latin</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarize</span><span class="op">(</span></span>
<span>    mean_height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">tree_height</span><span class="op">)</span>,</span>
<span>    stdev_height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span><span class="op">(</span><span class="va">tree_height</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">-&gt;</span> <span class="va">ny_trees_by_spc_summ</span></span>
<span><span class="va">ny_trees_by_spc_summ</span></span>
<span><span class="co">## # A tibble: 12 × 3</span></span>
<span><span class="co">##    spc_latin              mean_height stdev_height</span></span>
<span><span class="co">##    &lt;chr&gt;                        &lt;dbl&gt;        &lt;dbl&gt;</span></span>
<span><span class="co">##  1 ACER PLATANOIDES              82.6        17.6 </span></span>
<span><span class="co">##  2 ACER RUBRUM                  106.         15.7 </span></span>
<span><span class="co">##  3 ACER SACCHARINUM              65.6        16.6 </span></span>
<span><span class="co">##  4 FRAXINUS PENNSYLVANICA        60.6        21.3 </span></span>
<span><span class="co">##  5 GINKGO BILOBA                 90.4        24.5 </span></span>
<span><span class="co">##  6 GLEDITSIA TRIACANTHOS         53.0        13.0 </span></span>
<span><span class="co">##  7 PLATANUS ACERIFOLIA           82.0        16.0 </span></span>
<span><span class="co">##  8 PYRUS CALLERYANA              21.0         5.00</span></span>
<span><span class="co">##  9 QUERCUS PALUSTRIS             65.5         6.48</span></span>
<span><span class="co">## 10 QUERCUS RUBRA                111.         20.7 </span></span>
<span><span class="co">## 11 TILIA CORDATA                 98.8        32.6 </span></span>
<span><span class="co">## 12 ZELKOVA SERRATA              101.         10.7</span></span></code></pre></div>
<p>Now we can use this data in plotting. For this, we will use a new geom, <code>geom_pointrange</code>, which takes <code>x</code> and <code>y</code> aesthetics, as usual, but also requires two additional y-ish aesthetics <code>ymin</code> and <code>ymax</code> (or <code>xmin</code> and <code>xmax</code> if you want them to vary along x). Also, note that in the aesthetic mappings for <code>xmin</code> and <code>xmax</code>, we can use a mathematical expression: <code>mean-stdev</code> and <code>mean+stdev</code>, respectivey. In our case, these are <code>mean_height - stdev_height</code> and <code>mean_height + stdev_height</code>. Let’s see it in action:</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ny_trees_by_spc_summ</span> <span class="op">%&gt;%</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_pointrange</span><span class="op">(</span></span>
<span>      <span class="fu">aes</span><span class="op">(</span></span>
<span>        y <span class="op">=</span> <span class="va">spc_latin</span>,</span>
<span>        x <span class="op">=</span> <span class="va">mean_height</span>,</span>
<span>        xmin <span class="op">=</span> <span class="va">mean_height</span> <span class="op">-</span> <span class="va">stdev_height</span>,</span>
<span>        xmax <span class="op">=</span> <span class="va">mean_height</span> <span class="op">+</span> <span class="va">stdev_height</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-87-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Cool! Just like that, we’ve found (and visualized) the average and standard deviation of tree heights, by species, in NYC. But it doesn’t stop there. We can use <code>group_by()</code> and <code>summarize()</code> on multiple variables (i.e. more groups). We can do this to examine the properties of each tree species in each NYC borough. Let’s check it out:</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ny_trees</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">spc_latin</span>, <span class="va">boroname</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarize</span><span class="op">(</span></span>
<span>    mean_diam <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">tree_diameter</span><span class="op">)</span>,</span>
<span>    stdev_diam <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span><span class="op">(</span><span class="va">tree_diameter</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">-&gt;</span> <span class="va">ny_trees_by_spc_boro_summ</span></span>
<span><span class="va">ny_trees_by_spc_boro_summ</span></span>
<span><span class="co">## # A tibble: 48 × 4</span></span>
<span><span class="co">## # Groups:   spc_latin [12]</span></span>
<span><span class="co">##    spc_latin        boroname  mean_diam stdev_diam</span></span>
<span><span class="co">##    &lt;chr&gt;            &lt;chr&gt;         &lt;dbl&gt;      &lt;dbl&gt;</span></span>
<span><span class="co">##  1 ACER PLATANOIDES Bronx         13.9        6.74</span></span>
<span><span class="co">##  2 ACER PLATANOIDES Brooklyn      15.4       14.9 </span></span>
<span><span class="co">##  3 ACER PLATANOIDES Manhattan     11.6        8.45</span></span>
<span><span class="co">##  4 ACER PLATANOIDES Queens        15.1       12.9 </span></span>
<span><span class="co">##  5 ACER RUBRUM      Bronx         11.4        7.88</span></span>
<span><span class="co">##  6 ACER RUBRUM      Brooklyn      10.5        7.41</span></span>
<span><span class="co">##  7 ACER RUBRUM      Manhattan      6.63       4.23</span></span>
<span><span class="co">##  8 ACER RUBRUM      Queens        14.1        8.36</span></span>
<span><span class="co">##  9 ACER SACCHARINUM Bronx         19.7       10.5 </span></span>
<span><span class="co">## 10 ACER SACCHARINUM Brooklyn      22.2       10.1 </span></span>
<span><span class="co">## # ℹ 38 more rows</span></span></code></pre></div>
<p>Now we have summary statistics for each tree species within each borough. This is different from the previous plot in that we now have an additional variable (boroname) in our summarized dataset. This additional variable needs to be encoded in our plot. Let’s map boroname to x and facet over tree species, which used to be on x. We’ll also manually modify the theme element <code>strip.text.y</code> to get the species names in a readable position.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ny_trees_by_spc_boro_summ</span> <span class="op">%&gt;%</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_pointrange</span><span class="op">(</span></span>
<span>    <span class="fu">aes</span><span class="op">(</span></span>
<span>      y <span class="op">=</span> <span class="va">boroname</span>,</span>
<span>      x <span class="op">=</span> <span class="va">mean_diam</span>,</span>
<span>      xmin <span class="op">=</span> <span class="va">mean_diam</span><span class="op">-</span><span class="va">stdev_diam</span>,</span>
<span>      xmax <span class="op">=</span> <span class="va">mean_diam</span><span class="op">+</span><span class="va">stdev_diam</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_grid</span><span class="op">(</span><span class="va">spc_latin</span><span class="op">~</span><span class="va">.</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span></span>
<span>    strip.text.y <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-89-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Excellent! And if we really want to go for something pretty:</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ny_trees_by_spc_boro_summ</span> <span class="op">%&gt;%</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_pointrange</span><span class="op">(</span></span>
<span>    <span class="fu">aes</span><span class="op">(</span></span>
<span>      y <span class="op">=</span> <span class="va">boroname</span>,</span>
<span>      x <span class="op">=</span> <span class="va">mean_diam</span>,</span>
<span>      xmin <span class="op">=</span> <span class="va">mean_diam</span><span class="op">-</span><span class="va">stdev_diam</span>,</span>
<span>      xmax <span class="op">=</span> <span class="va">mean_diam</span><span class="op">+</span><span class="va">stdev_diam</span>,</span>
<span>      fill <span class="op">=</span> <span class="va">spc_latin</span></span>
<span>    <span class="op">)</span>, color <span class="op">=</span> <span class="st">"black"</span>, shape <span class="op">=</span> <span class="fl">21</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span></span>
<span>    y <span class="op">=</span> <span class="st">"Borough"</span>, </span>
<span>    x <span class="op">=</span> <span class="st">"Trunk diameter"</span>,</span>
<span>    caption <span class="op">=</span> <span class="fu">str_wrap</span><span class="op">(</span><span class="st">"Figure 1: Diameters of trees in New York City. Points correspond to average diameters of each tree species in each borough. Horizontal lines indicate the standard deviation of tree diameters. Points are colored according to tree species."</span>, width <span class="op">=</span> <span class="fl">80</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_grid</span><span class="op">(</span><span class="va">spc_latin</span><span class="op">~</span><span class="va">.</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">guides</span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_fill_brewer</span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Paired"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span></span>
<span>    strip.text.y <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>    plot.caption <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-90-1.png" width="100%" style="display: block; margin: auto;"></div>
<p><em>Now</em> we are getting somewhere. It looks like there are some really big maple trees (Acer) in Queens.</p>
</div>
<div id="ordering" class="section level2 unnumbered">
<h2>ordering<a class="anchor" aria-label="anchor" href="#ordering"><i class="fas fa-link"></i></a>
</h2>
<p>We can also sort or order a data frame based on a specific column with the command <code>arrange()</code>. Let’s have a quick look. Suppose we wanted to know which lake was the coldest:</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">arrange</span><span class="op">(</span><span class="va">alaska_lake_data</span>, <span class="va">water_temp</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 220 × 7</span></span>
<span><span class="co">##    lake             park  water_temp    pH element mg_per_L</span></span>
<span><span class="co">##    &lt;chr&gt;            &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;</span></span>
<span><span class="co">##  1 Desperation_Lake NOAT        2.95  6.34 C          2.1  </span></span>
<span><span class="co">##  2 Desperation_Lake NOAT        2.95  6.34 N          0.005</span></span>
<span><span class="co">##  3 Desperation_Lake NOAT        2.95  6.34 P          0    </span></span>
<span><span class="co">##  4 Desperation_Lake NOAT        2.95  6.34 Cl         0.2  </span></span>
<span><span class="co">##  5 Desperation_Lake NOAT        2.95  6.34 S          2.73 </span></span>
<span><span class="co">##  6 Desperation_Lake NOAT        2.95  6.34 F          0.01 </span></span>
<span><span class="co">##  7 Desperation_Lake NOAT        2.95  6.34 Br         0    </span></span>
<span><span class="co">##  8 Desperation_Lake NOAT        2.95  6.34 Na         1.11 </span></span>
<span><span class="co">##  9 Desperation_Lake NOAT        2.95  6.34 K          0.16 </span></span>
<span><span class="co">## 10 Desperation_Lake NOAT        2.95  6.34 Ca         5.87 </span></span>
<span><span class="co">## # ℹ 210 more rows</span></span>
<span><span class="co">## # ℹ 1 more variable: element_type &lt;chr&gt;</span></span></code></pre></div>
<p>Or suppose we wanted to know which was the warmest?</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">arrange</span><span class="op">(</span><span class="va">alaska_lake_data</span>, <span class="fu">desc</span><span class="op">(</span><span class="va">water_temp</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 220 × 7</span></span>
<span><span class="co">##    lake      park  water_temp    pH element mg_per_L</span></span>
<span><span class="co">##    &lt;chr&gt;     &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;</span></span>
<span><span class="co">##  1 Lava_Lake BELA        20.2  7.42 C          8.3  </span></span>
<span><span class="co">##  2 Lava_Lake BELA        20.2  7.42 N          0.017</span></span>
<span><span class="co">##  3 Lava_Lake BELA        20.2  7.42 P          0.001</span></span>
<span><span class="co">##  4 Lava_Lake BELA        20.2  7.42 Cl         2.53 </span></span>
<span><span class="co">##  5 Lava_Lake BELA        20.2  7.42 S          0.59 </span></span>
<span><span class="co">##  6 Lava_Lake BELA        20.2  7.42 F          0.04 </span></span>
<span><span class="co">##  7 Lava_Lake BELA        20.2  7.42 Br         0.01 </span></span>
<span><span class="co">##  8 Lava_Lake BELA        20.2  7.42 Na         2.93 </span></span>
<span><span class="co">##  9 Lava_Lake BELA        20.2  7.42 K          0.57 </span></span>
<span><span class="co">## 10 Lava_Lake BELA        20.2  7.42 Ca        11.8  </span></span>
<span><span class="co">## # ℹ 210 more rows</span></span>
<span><span class="co">## # ℹ 1 more variable: element_type &lt;chr&gt;</span></span></code></pre></div>
<p><code>arrange()</code> will work on grouped data, which is particularly useful in combination with <code>slice()</code>, which can show us the first n elements in each group:</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alaska_lake_data</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">park</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="va">water_temp</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">slice</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 3 × 7</span></span>
<span><span class="co">## # Groups:   park [3]</span></span>
<span><span class="co">##   lake  park  water_temp    pH element mg_per_L element_type</span></span>
<span><span class="co">##   &lt;chr&gt; &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt; &lt;chr&gt;       </span></span>
<span><span class="co">## 1 Devi… BELA        6.46  7.69 C            3.4 bound       </span></span>
<span><span class="co">## 2 Wild… GAAR        5.5   6.98 C            6.5 bound       </span></span>
<span><span class="co">## 3 Desp… NOAT        2.95  6.34 C            2.1 bound</span></span></code></pre></div>
<p>It looks like the coldest lakes in the three parks are Devil Mountain Lake, Wild Lake, and Desperation Lake!</p>
</div>
<div id="mutate" class="section level2 unnumbered">
<h2>mutate<a class="anchor" aria-label="anchor" href="#mutate"><i class="fas fa-link"></i></a>
</h2>
<p>One last thing before our exercises… there is another command called <code>mutate()</code>. It is like summarize it calculates user-defined statistics, but it creates output on a per-observation level instead of for each group. This means that it doesn’t make the data set smaller, in fact it makes it bigger, by creating a new row for the new variables defined inside <code>mutate()</code>. It can also take grouped data. This is really useful for calculating percentages within groups. For example: within each park, what percent of the park’s total dissolved sulfur does each lake have?</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alaska_lake_data</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">element</span> <span class="op">==</span> <span class="st">"S"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">park</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">lake</span>, <span class="va">park</span>, <span class="va">element</span>, <span class="va">mg_per_L</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>percent_S <span class="op">=</span> <span class="va">mg_per_L</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">mg_per_L</span><span class="op">)</span><span class="op">*</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 20 × 5</span></span>
<span><span class="co">## # Groups:   park [3]</span></span>
<span><span class="co">##    lake                park  element mg_per_L percent_S</span></span>
<span><span class="co">##    &lt;chr&gt;               &lt;chr&gt; &lt;chr&gt;      &lt;dbl&gt;     &lt;dbl&gt;</span></span>
<span><span class="co">##  1 Devil_Mountain_Lake BELA  S           0.62     32.3 </span></span>
<span><span class="co">##  2 Imuruk_Lake         BELA  S           0.2      10.4 </span></span>
<span><span class="co">##  3 Kuzitrin_Lake       BELA  S           0.29     15.1 </span></span>
<span><span class="co">##  4 Lava_Lake           BELA  S           0.59     30.7 </span></span>
<span><span class="co">##  5 North_Killeak_Lake  BELA  S           0.04      2.08</span></span>
<span><span class="co">##  6 White_Fish_Lake     BELA  S           0.18      9.38</span></span>
<span><span class="co">##  7 Iniakuk_Lake        GAAR  S          12.1      13.2 </span></span>
<span><span class="co">##  8 Kurupa_Lake         GAAR  S          12.4      13.6 </span></span>
<span><span class="co">##  9 Lake_Matcharak      GAAR  S          13.3      14.5 </span></span>
<span><span class="co">## 10 Lake_Selby          GAAR  S           7.92      8.64</span></span>
<span><span class="co">## 11 Nutavukti_Lake      GAAR  S           2.72      2.97</span></span>
<span><span class="co">## 12 Summit_Lake         GAAR  S           3.21      3.50</span></span>
<span><span class="co">## 13 Takahula_Lake       GAAR  S           5.53      6.03</span></span>
<span><span class="co">## 14 Walker_Lake         GAAR  S           5.77      6.30</span></span>
<span><span class="co">## 15 Wild_Lake           GAAR  S          28.7      31.3 </span></span>
<span><span class="co">## 16 Desperation_Lake    NOAT  S           2.73     25.8 </span></span>
<span><span class="co">## 17 Feniak_Lake         NOAT  S           4.93     46.5 </span></span>
<span><span class="co">## 18 Lake_Kangilipak     NOAT  S           0.55      5.19</span></span>
<span><span class="co">## 19 Lake_Narvakrak      NOAT  S           1.38     13.0 </span></span>
<span><span class="co">## 20 Okoklik_Lake        NOAT  S           1.01      9.53</span></span></code></pre></div>
<p>The percent columns for each park add to 100%, so, for example, Devil Mountain Lake has 32.3% of BELA’s dissolved sulfur.</p>
</div>
<div id="exercises-2" class="section level2 unnumbered">
<h2>exercises<a class="anchor" aria-label="anchor" href="#exercises-2"><i class="fas fa-link"></i></a>
</h2>
<p>Isn’t seven the most powerfully magical number? <em>Isn’t seven the most powerfully magical number?</em> Yes… I think the idea of a seven-part assignment would greatly appeal to an alchemist.</p>
<p>In this set of exercises we are going to use the periodic table. After you run source() you can load that data set using <code>periodic_table</code>. Please use that dataset to run analyses and answer the following questions/prompts. Compile the answers in an R Markdown document, compile it as a pdf, and upload it to the Canvas assignment. Please let me know if you have any questions. Good luck, and have fun!</p>
<p>Some pointers:</p>
<ul>
<li><p>If your code goes off the page, please wrap it across multiple lines, as shown in some of the examples in the previous set of exercises.</p></li>
<li><p>Don’t be afraid to put the variable with the long elements / long text on the y-axis and the continuous variable on the x-axis.</p></li>
<li><p>If your axis tick labels are overlapping or not visible, do something to fix that. Some solutions could be: move the legend to the top of the plot (<code>theme(legend.position = "top")</code>), rotate the text (<code>theme(axis.text.x = element_text(angle = 90)</code>), or make the text smaller (<code>theme(axis.text.x = element_text(size = 8)</code>).</p></li>
</ul>
<ol style="list-style-type: decimal">
<li><p>Make a plot using <code>geom_point()</code> that shows the average atomic weight of the elements discovered in each year spanned by the dataset (i.e. what was the average weight of the elements discovered in 1900? 1901? 1902? etc.). You should see a trend, particularly after 1950. What do you think has caused this trend?</p></li>
<li><p>The column <code>state_at_RT</code> indicates the state of each element at room temperate. Make a plot that shows the average first ionization potential of all the elements belonging to each state group indicated in <code>state_at_RT</code> (i.e. what is the average 1st ionization potential of all elements that are solid at room temp? liquid? etc.). Which is the highest?</p></li>
<li><p>Filter the dataset so that only elements with atomic number less than 85 are included. Considering only these elements, what is the average and standard deviation of boiling points for each type of <code>crystal_structure</code>? Make a plot using <code>geom_pointrange()</code> that shows the mean and standard deviation of each of these groups. What’s up with elements that have a cubic crystal structure?</p></li>
<li><p>Now filter the original dataset so that only elements with atomic number less than 37 are considered. The elements in this dataset belong to the first four periods. What is the average abundance of each of these four <em>periods</em> in seawater? i.e. what is the average abundance of all elements from period 1? period 2? etc. Which period is the most abundant? In this context what does “CHON” mean? (not the rock band, though they are also excellent, especially that song that features GoYama)</p></li>
<li><p>Now filter the original dataset so that only elements with atomic number less than 103 are considered. Filter it further so that elements from group number 18 are excluded. Using this twice-filtered dataset, compute the average, minimum, and maximum values for electronegativiy for each <code>group_number</code>. Use <code>geom_point()</code> and <code>geom_errorbar()</code> to illustrate the average, minimum, and maximum values for each group number.</p></li>
<li><p>Filter the dataset so that only elements with atomic number less than 85 are considered. Group these by <code>color</code>. Now filter out those that have <code>color == "colorless"</code>. Of the remaining elements, which has the widest range of specific heats? Use <code>geom_point()</code> and <code>geom_errorbar()</code> to illustrate the mean and standard deviation of each color’s specific heats.</p></li>
<li><p>You have learned many things in this course so far. <code>read_csv()</code>, <code><a href="https://rdrr.io/r/stats/filter.html">filter()</a></code>, <code>ggplot()</code>, and now <code>group_by()</code>, <code>summarize()</code>, <code>mutate()</code>, <code>arrange()</code>, and <code>slice()</code>. Using <strong>all</strong> these commands, create one or more graphics to illustrate what you consider to be one or more interesting trends in a data set of your own choosing. Use theme elements and scales to enhance your plot. Give your plot a nice caption based on the caption guide in this book.</p></li>
</ol>
</div>
<div id="further-reading-2" class="section level2 unnumbered">
<h2>further reading<a class="anchor" aria-label="anchor" href="#further-reading-2"><i class="fas fa-link"></i></a>
</h2>
<p>Be sure to check out the Tidy Data Tutor: <a href="https://tidydatatutor.com/vis.html" class="uri">https://tidydatatutor.com/vis.html</a>. An easy way to visualize what’s going on during data wrangling!</p>
<!-- end -->
<!-- start dimensionality reduction -->
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="data-visualization-iii.html">data visualization III</a></div>
<div class="next"><a href="dimensionality-reduction.html">dimensionality reduction</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <h2>Note: Second Edition is under construction 🏗</h2>
    <p>Now is a great time to provide feedback</p>
        <ul class="list-unstyled">
<li><a href="https://forms.gle/SZmB2Ct2exE2dBwv9">Provide feedback (5 min)</a></li>
          <!-- <li><a href="https://geocompr.robinlovelace.net/#reproducibility">Install updated packages</a></li> -->
          <!-- <li><a href="https://github.com/Robinlovelace/geocompr/issues">Open an issue <i class="fas fa-question"></i></a></li> -->
          <!-- <li><a href="https://discord.gg/Te3gWeDwmf">Chat on Discord <i class="fab fa-discord"></i></a></li> -->
        </ul>
<hr>
<nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#data-wrangling-and-summaries">data wrangling and summaries</a></li>
<li><a class="nav-link" href="#data-import">data import</a></li>
<li><a class="nav-link" href="#subsetting">subsetting</a></li>
<li><a class="nav-link" href="#wide-and-long-data">wide and long data</a></li>
<li><a class="nav-link" href="#the-pipe">the pipe (%&gt;%)</a></li>
<li><a class="nav-link" href="#summary-statistics">summary statistics</a></li>
<li><a class="nav-link" href="#ordering">ordering</a></li>
<li><a class="nav-link" href="#mutate">mutate</a></li>
<li><a class="nav-link" href="#exercises-2">exercises</a></li>
<li><a class="nav-link" href="#further-reading-2">further reading</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/blob/main/index.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/edit/main/index.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Integrated Bioanalytics</strong>" was written by Lucas Busta and members of the Busta lab. It was last built on 2023-09-28.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
