<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>dimensionality reduction | Integrated Bioanalytics</title>
<meta name="author" content="Lucas Busta and members of the Busta lab">
<meta name="description" content="In the previous chapters, we looked at how to explore our data sets by visualizing many variables and manually identifying trends. Sometimes, we encounter data sets with so many variables, that it...">
<meta name="generator" content="bookdown 0.25 with bs4_book()">
<meta property="og:title" content="dimensionality reduction | Integrated Bioanalytics">
<meta property="og:type" content="book">
<meta property="og:description" content="In the previous chapters, we looked at how to explore our data sets by visualizing many variables and manually identifying trends. Sometimes, we encounter data sets with so many variables, that it...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="dimensionality reduction | Integrated Bioanalytics">
<meta name="twitter:description" content="In the previous chapters, we looked at how to explore our data sets by visualizing many variables and manually identifying trends. Sometimes, we encounter data sets with so many variables, that it...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><link href="libs/_Lato-0.4.1/font.css" rel="stylesheet">
<link href="libs/_Roboto%20Mono-0.4.1/font.css" rel="stylesheet">
<link href="libs/_Montserrat-0.4.1/font.css" rel="stylesheet">
<script src="libs/bs3compat-0.3.1.9000/transition.js"></script><script src="libs/bs3compat-0.3.1.9000/tabs.js"></script><script src="libs/bs3compat-0.3.1.9000/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><link rel="icon" type="image/png" sizes="32x32" href="images/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="images/favicon-16x16.png">
<script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-99618359-1', 'auto');
      ga('send', 'pageview');

    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="style/style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h2>
        <a href="index.html" title="">Integrated Bioanalytics</a>
      </h2>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">WELCOME</a></li>
<li class="book-part">DATA ANALYSIS IN R</li>
<li><a class="" href="overview.html">overview</a></li>
<li><a class="" href="installation.html">installation</a></li>
<li><a class="" href="data-visualization.html">data visualization</a></li>
<li><a class="" href="data-wrangling.html">data wrangling</a></li>
<li><a class="active" href="dimensionality-reduction.html">dimensionality reduction</a></li>
<li><a class="" href="clustering.html">clustering</a></li>
<li><a class="" href="models.html">models</a></li>
<li><a class="" href="comparing-means.html">comparing means</a></li>
<li class="book-part">GC-MS DATA</li>
<li><a class="" href="mass-spectrometric-analysis.html">mass spectrometric analysis</a></li>
<li class="book-part">DNA AND RNA SEQUENCE DATA</li>
<li><a class="" href="data-acquisition.html">data acquisition</a></li>
<li><a class="" href="data-transfer.html">data transfer</a></li>
<li><a class="" href="sequence-assessment.html">sequence assessment</a></li>
<li><a class="" href="illumina-read-assessment.html">illumina read assessment</a></li>
<li class="book-part">TRANSCRIPTOME ANALYSIS</li>
<li><a class="" href="transcriptomic-analyses.html">transcriptomic analyses</a></li>
<li class="book-part">GENOMIC ANALYSIS</li>
<li><a class="" href="setup.html">setup</a></li>
<li><a class="" href="genome-assembly.html">genome assembly</a></li>
<li><a class="" href="comparative-genomics.html">comparative genomics</a></li>
<li class="book-part">EVOLUTIONARY ANALYSIS</li>
<li><a class="" href="blast-searches.html">blast searches</a></li>
<li><a class="" href="alignments.html">alignments</a></li>
<li><a class="" href="phylogenies.html">phylogenies</a></li>
<li><a class="" href="phylogenetic-analyses.html">phylogenetic analyses</a></li>
<li class="book-part">SCIENTIFIC WRITING</li>
<li><a class="" href="overview-1.html">overview</a></li>
<li><a class="" href="a-mini-manuscript.html">a mini manuscript</a></li>
<li><a class="" href="figures-captions.html">figures &amp; captions</a></li>
<li><a class="" href="results-and-discussion.html">results and discussion</a></li>
<li><a class="" href="conclusion-and-introduction.html">conclusion and introduction</a></li>
<li><a class="" href="abstract-and-title.html">abstract and title</a></li>
<li><a class="" href="proposals.html">proposals</a></li>
<li class="book-part">IMAGE ANALYSIS</li>
<li><a class="" href="image-color-analysis.html">image color analysis</a></li>
<li><a class="" href="images-of-mass-spectra.html">images of mass spectra</a></li>
<li class="book-part">APPENDIX</li>
<li><a class="" href="links.html">links</a></li>
<li><a class="" href="r-faq.html">r faq</a></li>
<li><a class="" href="templates.html">templates</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="dimensionality-reduction" class="section level1 unnumbered">
<h1>dimensionality reduction<a class="anchor" aria-label="anchor" href="#dimensionality-reduction"><i class="fas fa-link"></i></a>
</h1>
<div class="inline-figure"><img src="https://thebustalab.github.io/integrated_bioanalytics/images/dimensionality.png" width="100%" style="display: block; margin: auto;"></div>
<p>In the previous chapters, we looked at how to explore our data sets by visualizing many variables and manually identifying trends. Sometimes, we encounter data sets with so many variables, that it is not reasonable to manually select certain variables with which to create plots and manually search for trends. In these cases, we need dimensionality reduction - a set of techniques that helps us identify which variables are driving differences among our samples. In this course, we will conduct dimensionality reduction useing <code>runMatrixAnalysis()</code>, a function that is loaded into your R Session when you run the source() command.</p>
<p>Matrix analyses can be a bit tricky to set up. There are two things that we can do to help us with this: (i) we will use a template for <code>runMatrixAnalysis()</code> (see below) and (ii) it is <em>critical</em> that we think about our data in terms of <strong>samples</strong> and <strong>analytes</strong>. Let‚Äôs consider our Alaska lakes data set:</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">alaska_lake_data</span>
<span class="co">## # A tibble: 220 √ó 7</span>
<span class="co">##    lake              park  water_temp    pH element mg_per_L</span>
<span class="co">##    &lt;chr&gt;             &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;</span>
<span class="co">##  1 Devil_Mountain_L‚Ä¶ BELA        6.46  7.69 C          3.4  </span>
<span class="co">##  2 Devil_Mountain_L‚Ä¶ BELA        6.46  7.69 N          0.028</span>
<span class="co">##  3 Devil_Mountain_L‚Ä¶ BELA        6.46  7.69 P          0    </span>
<span class="co">##  4 Devil_Mountain_L‚Ä¶ BELA        6.46  7.69 Cl        10.4  </span>
<span class="co">##  5 Devil_Mountain_L‚Ä¶ BELA        6.46  7.69 S          0.62 </span>
<span class="co">##  6 Devil_Mountain_L‚Ä¶ BELA        6.46  7.69 F          0.04 </span>
<span class="co">##  7 Devil_Mountain_L‚Ä¶ BELA        6.46  7.69 Br         0.02 </span>
<span class="co">##  8 Devil_Mountain_L‚Ä¶ BELA        6.46  7.69 Na         8.92 </span>
<span class="co">##  9 Devil_Mountain_L‚Ä¶ BELA        6.46  7.69 K          1.2  </span>
<span class="co">## 10 Devil_Mountain_L‚Ä¶ BELA        6.46  7.69 Ca         5.73 </span>
<span class="co">## # ‚Ä¶ with 210 more rows, and 1 more variable:</span>
<span class="co">## #   element_type &lt;chr&gt;</span></code></pre></div>
<p>We can see that this dataset is comprised of measurements of various <em>analytes</em> (i.e.¬†several chemical elements, as well as water_temp, and pH), in different <em>samples</em> (i.e.¬†lakes). We need to tell the <code>runMatrixAnalysis()</code> function how each column relates to this samples and analytes structure. See the image below for an explanation.</p>
<div class="inline-figure"><img src="https://thebustalab.github.io/integrated_bioanalytics/images/runMatrixAnalysis1.png" width="100%" style="display: block; margin: auto;"></div>
<div id="pca" class="section level2 unnumbered">
<h2>pca<a class="anchor" aria-label="anchor" href="#pca"><i class="fas fa-link"></i></a>
</h2>
<p>‚ÄúWhich analytes are driving differences among my samples?‚Äù
‚ÄúWhich analytes in my data set are correlated?‚Äù</p>
<div id="theory" class="section level3 unnumbered">
<h3>theory<a class="anchor" aria-label="anchor" href="#theory"><i class="fas fa-link"></i></a>
</h3>
<p>PCA looks at all the variance in a high dimensional data set and chooses new axes within that data set that align with the directions containing highest variance. These new axes are called principal components. Let‚Äôs look at an example:</p>
<div class="inline-figure"><img src="https://thebustalab.github.io/integrated_bioanalytics/images/PCA.png" width="100%" style="display: block; margin: auto;"></div>
<p>In the example above, the three dimensional space can be reduced to a two dimensional space with the principal components analysis. New axes (principal components) are selected (bold arrows on left) that become the x and y axes in the principal components space (right).</p>
<p>We can run and visualize principal components analyses using the <code>runMatrixAnalysis()</code> function as in the example below:</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">AK_lakes_pca</span> <span class="op">&lt;-</span> <span class="fu">runMatrixAnalysis</span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">alaska_lake_data</span>,
  analysis <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"pca"</span><span class="op">)</span>,
  column_w_names_of_multiple_analytes <span class="op">=</span> <span class="st">"element"</span>,
  column_w_values_for_multiple_analytes <span class="op">=</span> <span class="st">"mg_per_L"</span>,
  columns_w_values_for_single_analyte <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"water_temp"</span>, <span class="st">"pH"</span><span class="op">)</span>,
  columns_w_additional_analyte_info <span class="op">=</span> <span class="st">"element_type"</span>,
  columns_w_sample_ID_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lake"</span>, <span class="st">"park"</span><span class="op">)</span>
<span class="op">)</span>

<span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">AK_lakes_pca</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Dim.1</span>, y <span class="op">=</span> <span class="va">Dim.2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>fill <span class="op">=</span> <span class="va">park</span><span class="op">)</span>, shape <span class="op">=</span> <span class="fl">21</span>, size <span class="op">=</span> <span class="fl">4</span>, alpha <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_label_repel</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>label <span class="op">=</span> <span class="va">lake</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-101-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Great! In this plot we can see that White Fish Lake and North Killeak Lake, both in BELA park, are quite different from the other parks (they are separated from the others along dimension 1, i.e.¬†the first principal component). At the same time, Wild Lake, Iniakuk Lake, Walker Lake, and several other lakes in GAAR park are different from all the others (they are separated from the others along dimension 2, i.e.¬†the second principal component).</p>
<p>Important question: what makes the lakes listed above different from the others? Certainly some aspect of their chemistry, since that‚Äôs the data that this analysis is built upon, but how do we determine which analyte(s) are driving the differences among the lakes that we see in the PCA plot?</p>
</div>
<div id="ordination-plots" class="section level3 unnumbered">
<h3>ordination plots<a class="anchor" aria-label="anchor" href="#ordination-plots"><i class="fas fa-link"></i></a>
</h3>
<p>Let‚Äôs look at how to access the information about which analytes are major contributors to each principal component. This is important because it will tell you which analytes are associated with particular dimensions, and by extension, which analytes are associated with (and are markers for) particular groups in the PCA plot. This can be determined using an ordination plot. Let‚Äôs look at an example. We can obtain the ordination plot information using <code>runMatrixAnalysis()</code> with <code>analysis = "pca_ord"</code>:</p>
<pre><code>## # A tibble: 6 √ó 3
##   analyte      Dim.1   Dim.2
##   &lt;chr&gt;        &lt;dbl&gt;   &lt;dbl&gt;
## 1 water_temp 0.0749  -0.262 
## 2 pH         0.688    0.0156
## 3 C          0.291   -0.243 
## 4 N          0.00739  0.714 
## 5 P          0.475   -0.0828
## 6 Cl         0.953    0.0114</code></pre>
<p>We can now visualize the ordination plot using our standard ggplot plotting techniques. Note the use of <code>geom_label_repel()</code> and <code><a href="https://rdrr.io/r/stats/filter.html">filter()</a></code> to label certain segments in the ordination plot. You do not need to use <code>geom_label_repel()</code>, you could use the built in <code>geom_label()</code>, but <code>geom_label_repel()</code> can make labelling your segments easier.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">AK_lakes_pca_ord</span> <span class="op">&lt;-</span> <span class="fu">runMatrixAnalysis</span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">alaska_lake_data</span>,
  analysis <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"pca_ord"</span><span class="op">)</span>,
  column_w_names_of_multiple_analytes <span class="op">=</span> <span class="st">"element"</span>,
  column_w_values_for_multiple_analytes <span class="op">=</span> <span class="st">"mg_per_L"</span>,
  columns_w_values_for_single_analyte <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"water_temp"</span>, <span class="st">"pH"</span><span class="op">)</span>,
  columns_w_additional_analyte_info <span class="op">=</span> <span class="st">"element_type"</span>,
  columns_w_sample_ID_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lake"</span>, <span class="st">"park"</span><span class="op">)</span>
<span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">AK_lakes_pca_ord</span><span class="op">)</span>
<span class="co">## # A tibble: 6 √ó 3</span>
<span class="co">##   analyte      Dim.1   Dim.2</span>
<span class="co">##   &lt;chr&gt;        &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">## 1 water_temp 0.0749  -0.262 </span>
<span class="co">## 2 pH         0.688    0.0156</span>
<span class="co">## 3 C          0.291   -0.243 </span>
<span class="co">## 4 N          0.00739  0.714 </span>
<span class="co">## 5 P          0.475   -0.0828</span>
<span class="co">## 6 Cl         0.953    0.0114</span>

<span class="fu">ggplot</span><span class="op">(</span><span class="va">AK_lakes_pca_ord</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_segment</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="fl">0</span>, y <span class="op">=</span> <span class="fl">0</span>, xend <span class="op">=</span> <span class="va">Dim.1</span>, yend <span class="op">=</span> <span class="va">Dim.2</span>, color <span class="op">=</span> <span class="va">analyte</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_circle</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x0 <span class="op">=</span> <span class="fl">0</span>, y0 <span class="op">=</span> <span class="fl">0</span>, r <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_label_repel</span><span class="op">(</span>
    data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">AK_lakes_pca_ord</span>, <span class="va">Dim.1</span> <span class="op">&gt;</span> <span class="fl">0.9</span>, <span class="va">Dim.2</span> <span class="op">&lt;</span> <span class="fl">0.1</span>, <span class="va">Dim.2</span> <span class="op">&gt;</span> <span class="op">-</span><span class="fl">0.1</span><span class="op">)</span>,
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Dim.1</span>, y <span class="op">=</span> <span class="va">Dim.2</span>, label <span class="op">=</span> <span class="va">analyte</span><span class="op">)</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1.5</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_label_repel</span><span class="op">(</span>
    data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">AK_lakes_pca_ord</span>, <span class="va">Dim.2</span> <span class="op">&gt;</span> <span class="fl">0.5</span><span class="op">)</span>,
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Dim.1</span>, y <span class="op">=</span> <span class="va">Dim.2</span>, label <span class="op">=</span> <span class="va">analyte</span><span class="op">)</span>, direction <span class="op">=</span> <span class="st">"y"</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1.5</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu">coord_cartesian</span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fl">1.5</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fl">1.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-103-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Great! Here is how to read the ordination plot:</p>
<ol style="list-style-type: decimal">
<li><p>When considering one analyte‚Äôs vector: the vector‚Äôs projected value on an axis shows how much its variance is aligned with that principal component.</p></li>
<li><p>When considering two analyte vectors: the angle between two vectors indicates how correlated those two variables are. If they point in the same direction, they are highly correlated. If they meet each other at 90 degrees, they are not very correlated. If they meet at ~180 degrees, they are negatively correlated. If say that one analyte is ‚Äú1.9‚Äù with respect to dimension 2 and another is ‚Äú-1.9‚Äù with respect to dimension 2. Let‚Äôs also say that these vectors are ~‚Äú0‚Äù with respect to dimension 1.</p></li>
</ol>
<p>With the ordination plot above, we can now see that the abundances of K, Cl, Br, and Na are the major contributors of variance to the first principal component (or the first dimension). The abundances of these elements are what make White Fish Lake and North Killeak Lake different from the other lakes. We can also see that the abundances of N, S, and Ca are the major contributors to variance in teh second dimension, whic means that these elements ar what set Wild Lake, Iniakuk Lake, Walker Lake, and several other lakes in GAAR park apart from the rest of the lakes in the data set.</p>
</div>
<div id="principal-components" class="section level3 unnumbered">
<h3>principal components<a class="anchor" aria-label="anchor" href="#principal-components"><i class="fas fa-link"></i></a>
</h3>
<p>We also can access information about the how much of the variance in the data set is explained by each principal component, and we can plot that using ggplot:</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">AK_lakes_pca_dim</span> <span class="op">&lt;-</span> <span class="fu">runMatrixAnalysis</span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">alaska_lake_data</span>,
  analysis <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"pca_dim"</span><span class="op">)</span>,
  column_w_names_of_multiple_analytes <span class="op">=</span> <span class="st">"element"</span>,
  column_w_values_for_multiple_analytes <span class="op">=</span> <span class="st">"mg_per_L"</span>,
  columns_w_values_for_single_analyte <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"water_temp"</span>, <span class="st">"pH"</span><span class="op">)</span>,
  columns_w_additional_analyte_info <span class="op">=</span> <span class="st">"element_type"</span>,
  columns_w_sample_ID_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lake"</span>, <span class="st">"park"</span><span class="op">)</span>
<span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">AK_lakes_pca_dim</span><span class="op">)</span>
<span class="co">## # A tibble: 6 √ó 2</span>
<span class="co">##   principal_component percent_variance_explained</span>
<span class="co">##                 &lt;dbl&gt;                      &lt;dbl&gt;</span>
<span class="co">## 1                   1                      48.5 </span>
<span class="co">## 2                   2                      18.7 </span>
<span class="co">## 3                   3                      11.7 </span>
<span class="co">## 4                   4                       7.95</span>
<span class="co">## 5                   5                       4.72</span>
<span class="co">## 6                   6                       3.36</span>

<span class="fu">ggplot</span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">AK_lakes_pca_dim</span>, 
  <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">principal_component</span>, y <span class="op">=</span> <span class="va">percent_variance_explained</span><span class="op">)</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-104-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Cool! We can see that the first principal component retains nearly 50% of the variance in the original dataset, while the second dimension contains only about 20%.</p>
</div>
<div id="exercises-1" class="section level3 unnumbered">
<h3>exercises<a class="anchor" aria-label="anchor" href="#exercises-1"><i class="fas fa-link"></i></a>
</h3>
<p>In this set of exercises you will choose to complete ONE of the options below. For either option please refer to Chapter 12 for help with principal component and ordination plots. Also, when you are filling out the <code>runMatrixAnalysis()</code> template, you can use the <code><a href="https://rdrr.io/r/base/colnames.html">colnames()</a></code> function to help you specify a long list of column names rather than typing them out by hand. For example, in the periodic table data set, we can refer to a set of columns (columns 10 through 20) with the following command:</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">periodic_table_subset</span><span class="op">)</span><span class="op">[</span><span class="fl">10</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span>
<span class="co">##  [1] "electronegativity_pauling"         </span>
<span class="co">##  [2] "first_ionization_poten_eV"         </span>
<span class="co">##  [3] "second_ionization_poten_eV"        </span>
<span class="co">##  [4] "third_ionization_poten_eV"         </span>
<span class="co">##  [5] "electron_affinity_eV"              </span>
<span class="co">##  [6] "atomic_radius_ang"                 </span>
<span class="co">##  [7] "ionic_radius_ang"                  </span>
<span class="co">##  [8] "covalent_radius_ang"               </span>
<span class="co">##  [9] "atomic_volume_cm3_per_mol"         </span>
<span class="co">## [10] "electrical_conductivity_mho_per_cm"</span>
<span class="co">## [11] "specific_heat_J_per_g_K"</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">periodic_table_subset</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">18</span><span class="op">:</span><span class="fl">20</span>, <span class="fl">23</span><span class="op">:</span><span class="fl">25</span><span class="op">)</span><span class="op">]</span>
<span class="co">## [1] "atomic_volume_cm3_per_mol"         </span>
<span class="co">## [2] "electrical_conductivity_mho_per_cm"</span>
<span class="co">## [3] "specific_heat_J_per_g_K"           </span>
<span class="co">## [4] "thermal_conductivity_W_per_m_K"    </span>
<span class="co">## [5] "polarizability_A_cubed"            </span>
<span class="co">## [6] "heat_atomization_kJ_per_mol"</span></code></pre></div>
<p>We can use that command in the template, as in the example below. With the notation <code>colnames(periodic_table_subset)[c(5:7,9:25)]</code>, we can mark columns 5 - 7 and 9 - 25 as columns_w_values_for_single_analyte (note what happens when you run <code>c(5:7,9:25)</code> in the console, and what happens when you run <code>colnames(periodic_table_subset)[c(5:7,9:25)]</code> in the console). With the notation <code>colnames(periodic_table_subset)[c(1:4, 8)]</code> we can mark columns 1 - 4 and column 8 as columns_w_sample_ID_info (note what happens when you run <code>c(1:4, 8)</code> in the console, and what happens when you run <code>colnames(periodic_table_subset)[c(1:4, 8)]</code> in the console).</p>
<div id="human-metabolomics" class="section level4 unnumbered">
<h4>human metabolomics<a class="anchor" aria-label="anchor" href="#human-metabolomics"><i class="fas fa-link"></i></a>
</h4>
<p>This first option is to work with a dataset describing metabolomics data (i.e.¬†abundances of &gt; 100 different biochemicals) from each of 93 human patients, some of which have Chronic Kidney Disease. If you choose this option, your task is to discover a biomarker for Chronic Kidney Disease. This means that you will need to determine a metabolite whose abundance is strongly associated with the disease. To do this you should complete the following:</p>
<ol style="list-style-type: decimal">
<li>Run a PCA analysis on <code>metabolomics_data</code> (i.e.¬†<code>runMatrixAnalysis()</code> with <code>analysis = "pca"</code>)</li>
<li>Plot the results of the analysis to determine which principal component (i.e.¬†dimension) separates the healthy and kidney_disease samples.</li>
<li>Obtain the ordination plot coordinates for the analytes in the PCA analysis (i.e.¬†<code>runMatrixAnalysis()</code> with <code>analysis = "pca_ord"</code>).</li>
<li>Visualize the ordination plot and determine which of the analytes are strongly associated with the principal component (i.e.¬†dimension) separates the healthy and kidney_disease samples.</li>
<li>Bingo! These analytes are associated with Chronic Kidney Disease and could be biomarkers for such.</li>
</ol>
</div>
<div id="grape-vine-chemistry" class="section level4 unnumbered">
<h4>grape vine chemistry<a class="anchor" aria-label="anchor" href="#grape-vine-chemistry"><i class="fas fa-link"></i></a>
</h4>
<p>This second option is to work with a dataset describing metabolomics data (i.e.¬†abundances of &gt; 100 different biochemicals) from 5 different wine grape varieties. If you choose this option, your task is to discover a biomarker for Chardonnay and a biomarker for Cabernet Sauvignon. This means that you will need to identify two metabolites, each of which are associated with one of those two grape varieties. To do this you should complete the following:</p>
<ol style="list-style-type: decimal">
<li>Run a PCA analysis on <code>wine_grape_data</code> (i.e.¬†<code>runMatrixAnalysis()</code> with <code>analysis = "pca"</code>)</li>
<li>Plot the results of the analysis to determine which principal component (i.e.¬†dimension) separates the Chardonnay samples from the other varieties and the Cabernet Sauvignon samples from the other varieties.</li>
<li>Obtain the ordination plot coordinates for the analytes in the PCA analysis (i.e.¬†<code>runMatrixAnalysis()</code> with <code>analysis = "pca_ord"</code>).</li>
<li>Visualize the ordination plot and determine which of the analytes are strongly associated with the principal component (i.e.¬†dimension) separates the Chardonnay samples from the other varieties and the Cabernet Sauvignon samples from the other varieties.</li>
<li>Bingo! These analytes are associated with those varieites and could be biomarkers for such.</li>
</ol>
</div>
</div>
<div id="further-reading-2" class="section level3 unnumbered">
<h3>further reading<a class="anchor" aria-label="anchor" href="#further-reading-2"><i class="fas fa-link"></i></a>
</h3>
</div>
</div>
<div id="umap" class="section level2 unnumbered">
<h2>UMAP<a class="anchor" aria-label="anchor" href="#umap"><i class="fas fa-link"></i></a>
</h2>
</div>
<div id="tsne" class="section level2 unnumbered">
<h2>tSNE<a class="anchor" aria-label="anchor" href="#tsne"><i class="fas fa-link"></i></a>
</h2>
<!-- end -->
<!-- start clustering -->
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="data-wrangling.html">data wrangling</a></div>
<div class="next"><a href="clustering.html">clustering</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <h2>Note: Second Edition is under construction üèó</h2>
    <p>Now is a great time to provide feedback</p>
        <ul class="list-unstyled">
<li><a href="https://forms.gle/SZmB2Ct2exE2dBwv9">Provide feedback (5 min)</a></li>
          <!-- <li><a href="https://geocompr.robinlovelace.net/#reproducibility">Install updated packages</a></li> -->
          <!-- <li><a href="https://github.com/Robinlovelace/geocompr/issues">Open an issue <i class="fas fa-question"></i></a></li> -->
          <!-- <li><a href="https://discord.gg/Te3gWeDwmf">Chat on Discord <i class="fab fa-discord"></i></a></li> -->
        </ul>
<hr>
<nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#dimensionality-reduction">dimensionality reduction</a></li>
<li>
<a class="nav-link" href="#pca">pca</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#theory">theory</a></li>
<li><a class="nav-link" href="#ordination-plots">ordination plots</a></li>
<li><a class="nav-link" href="#principal-components">principal components</a></li>
<li><a class="nav-link" href="#exercises-1">exercises</a></li>
<li><a class="nav-link" href="#further-reading-2">further reading</a></li>
</ul>
</li>
<li><a class="nav-link" href="#umap">UMAP</a></li>
<li><a class="nav-link" href="#tsne">tSNE</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/blob/main/index.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/edit/main/index.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Integrated Bioanalytics</strong>" was written by Lucas Busta and members of the Busta lab. It was last built on 2022-09-18.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
