<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>genome assembly | Integrated Bioanalytics</title>
<meta name="author" content="Lucas Busta and members of the Busta lab">
<meta name="description" content="To some degree, refer to: https://github.com/dithiii/ant-pipeline/blob/main/README.md.  1.4 assembly  1.4.1 equipment Genome assembly requires computing resources - and since not all genomes are...">
<meta name="generator" content="bookdown 0.25 with bs4_book()">
<meta property="og:title" content="genome assembly | Integrated Bioanalytics">
<meta property="og:type" content="book">
<meta property="og:description" content="To some degree, refer to: https://github.com/dithiii/ant-pipeline/blob/main/README.md.  1.4 assembly  1.4.1 equipment Genome assembly requires computing resources - and since not all genomes are...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="genome assembly | Integrated Bioanalytics">
<meta name="twitter:description" content="To some degree, refer to: https://github.com/dithiii/ant-pipeline/blob/main/README.md.  1.4 assembly  1.4.1 equipment Genome assembly requires computing resources - and since not all genomes are...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><link href="libs/Lato-0.4.1/font.css" rel="stylesheet">
<link href="libs/Roboto%20Mono-0.4.1/font.css" rel="stylesheet">
<link href="libs/Montserrat-0.4.1/font.css" rel="stylesheet">
<script src="libs/bs3compat-0.3.1.9000/transition.js"></script><script src="libs/bs3compat-0.3.1.9000/tabs.js"></script><script src="libs/bs3compat-0.3.1.9000/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><link rel="icon" type="image/png" sizes="32x32" href="images/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="images/favicon-16x16.png">
<script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-99618359-1', 'auto');
      ga('send', 'pageview');

    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="style/style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h2>
        <a href="index.html" title="">Integrated Bioanalytics</a>
      </h2>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">WELCOME</a></li>
<li class="book-part">DATA ANALYSIS IN R</li>
<li><a class="" href="overview.html">overview</a></li>
<li><a class="" href="installation.html">installation</a></li>
<li><a class="" href="data-visualization.html">data visualization</a></li>
<li><a class="" href="data-wrangling.html">data wrangling</a></li>
<li><a class="" href="dimensionality-reduction.html">dimensionality reduction</a></li>
<li><a class="" href="clustering.html">clustering</a></li>
<li><a class="" href="models.html">models</a></li>
<li><a class="" href="comparing-means.html">comparing means</a></li>
<li><a class="" href="midterm-exam.html">midterm exam</a></li>
<li class="book-part">GC-MS DATA</li>
<li><a class="" href="mass-spectrometric-analysis.html">mass spectrometric analysis</a></li>
<li class="book-part">DNA AND RNA SEQUENCE DATA</li>
<li><a class="" href="data-acquisition.html">data acquisition</a></li>
<li><a class="" href="data-transfer.html">data transfer</a></li>
<li><a class="" href="sequence-assessment.html">sequence assessment</a></li>
<li><a class="" href="illumina-read-assessment.html">illumina read assessment</a></li>
<li class="book-part">TRANSCRIPTOME ANALYSIS</li>
<li><a class="" href="transcriptomic-analyses.html">transcriptomic analyses</a></li>
<li class="book-part">GENOMIC ANALYSIS</li>
<li><a class="" href="setup.html">setup</a></li>
<li><a class="active" href="genome-assembly.html">genome assembly</a></li>
<li><a class="" href="comparative-genomics.html">comparative genomics</a></li>
<li class="book-part">EVOLUTIONARY ANALYSIS</li>
<li><a class="" href="blast-searches.html">blast searches</a></li>
<li><a class="" href="alignments.html">alignments</a></li>
<li><a class="" href="phylogenies.html">phylogenies</a></li>
<li><a class="" href="phylogenetic-analyses.html">phylogenetic analyses</a></li>
<li class="book-part">SCIENTIFIC WRITING</li>
<li><a class="" href="overview-1.html">overview</a></li>
<li><a class="" href="figures-captions.html">figures &amp; captions</a></li>
<li><a class="" href="results-and-discussion.html">results and discussion</a></li>
<li><a class="" href="conclusion-and-introduction.html">conclusion and introduction</a></li>
<li><a class="" href="abstract-and-title.html">abstract and title</a></li>
<li><a class="" href="proposals.html">proposals</a></li>
<li class="book-part">IMAGE ANALYSIS</li>
<li><a class="" href="image-color-analysis.html">image color analysis</a></li>
<li><a class="" href="images-of-mass-spectra.html">images of mass spectra</a></li>
<li class="book-part">APPENDIX</li>
<li><a class="" href="links.html">links</a></li>
<li><a class="" href="r-faq.html">r faq</a></li>
<li><a class="" href="templates.html">templates</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="genome-assembly" class="section level1 unnumbered">
<h1>genome assembly<a class="anchor" aria-label="anchor" href="#genome-assembly"><i class="fas fa-link"></i></a>
</h1>
<p>To some degree, refer to: <a href="https://github.com/dithiii/ant-pipeline/blob/main/README.md" class="uri">https://github.com/dithiii/ant-pipeline/blob/main/README.md</a>.</p>
<div id="assembly" class="section level2" number="1.4">
<h2>
<span class="header-section-number">1.4</span> assembly<a class="anchor" aria-label="anchor" href="#assembly"><i class="fas fa-link"></i></a>
</h2>
<div id="equipment" class="section level3" number="1.4.1">
<h3>
<span class="header-section-number">1.4.1</span> equipment<a class="anchor" aria-label="anchor" href="#equipment"><i class="fas fa-link"></i></a>
</h3>
<p>Genome assembly requires computing resources - and since not all genomes are of equal size, the computing resources required for different assemblies may differ. To run a genome assembly, start by determining what computing resources are available. Some helpful commands when investigating these resources on Linux machines:</p>
<ul>
<li>Assessing RAM (It is recommended to assign about 75% of available RAM to the assembly process):</li>
</ul>
<p><code>grep MeMTotal /proc/meminfo</code></p>
<ul>
<li>Assessing CPU resources (note that “threads per CPU” can denote the availability of hyperthreading):</li>
</ul>
<p><code>lscpu</code></p>
<ul>
<li>Assessing disk/storage space (Make sure you are running your assembly on a disk with lots of open space. Ideally &gt; 2TB):</li>
</ul>
<p><code>df -h</code></p>
</div>
<div id="assembly-software" class="section level3" number="1.4.2">
<h3>
<span class="header-section-number">1.4.2</span> assembly software<a class="anchor" aria-label="anchor" href="#assembly-software"><i class="fas fa-link"></i></a>
</h3>
<div id="canu" class="section level4" number="1.4.2.1">
<h4>
<span class="header-section-number">1.4.2.1</span> canu<a class="anchor" aria-label="anchor" href="#canu"><i class="fas fa-link"></i></a>
</h4>
<p><code>docker pull staphb/canu-racon</code></p>
<p>For assembly on the BustaLab storage box, navigate to the directory that contains your reads. Merge all reads into one file using:</p>
<p><code>cat *.fastq &gt; all_reads.fastq</code></p>
<p>Then use Canu to assemble, we suggest creating a file that contains the Canu call. You can create the file using <code>nano</code>. In it, try something like:</p>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="genome-assembly.html#cb128-1" aria-hidden="true" tabindex="-1"></a>sudo docker run <span class="sc">-</span>u <span class="sc">$</span>(id <span class="sc">-</span>u) <span class="sc">-</span>v <span class="sc">/</span>data<span class="sc">/</span>ben_diatom2<span class="sc">/</span>basecalled_reads<span class="sc">/</span><span class="er">:/</span>canu<span class="sc">-</span>racon_wd staphb<span class="sc">/</span>canu<span class="sc">-</span>racon canu <span class="sc">-</span>p n_frust2assembly <span class="sc">-</span>d <span class="sc">/</span>canu<span class="sc">-</span>racon_wd<span class="sc">/</span> <span class="sc">-</span>genomeSize<span class="ot">=</span>150m <span class="sc">-</span>nanopore <span class="sc">/</span>canu<span class="sc">-</span>racon_wd<span class="sc">/</span>all_reads2.fastq <span class="sc">-</span>minReadLength<span class="ot">=</span><span class="dv">1000</span> <span class="sc">-</span>correctedErrorRate<span class="ot">=</span><span class="fl">0.12</span> <span class="sc">-</span>minOverlapLength<span class="ot">=</span><span class="dv">500</span> <span class="sc">-</span>useGrid<span class="ot">=</span>false <span class="sc">-</span>minInputCoverage<span class="ot">=</span><span class="fl">0.5</span> <span class="sc">-</span>maxInputCoverage<span class="ot">=</span><span class="dv">100</span> <span class="sc">-</span>stopOnLowCoverage<span class="ot">=</span><span class="fl">0.5</span> <span class="sc">-</span>corMemory<span class="ot">=</span><span class="dv">48</span> <span class="sc">-</span>corThreads<span class="ot">=</span><span class="dv">4</span> <span class="sc">-</span>hapMemory<span class="ot">=</span><span class="dv">48</span> <span class="sc">-</span>hapThreads<span class="ot">=</span><span class="dv">4</span> <span class="sc">-</span>merylMemory<span class="ot">=</span><span class="dv">48</span> <span class="sc">-</span>merylthreads<span class="ot">=</span><span class="dv">4</span> <span class="sc">-</span>batMemory<span class="ot">=</span><span class="dv">48</span> <span class="sc">-</span>batThreads<span class="ot">=</span><span class="dv">4</span></span></code></pre></div>
<p>Notes on Canu options:</p>
<p>Defaults:</p>
<ul>
<li>minReadLength=1000</li>
<li>minOverlapLength=500bp</li>
<li>correctedErrorRate=0.114</li>
<li>stopOnLowCoverage &lt;integer=10&gt;</li>
</ul>
<p>Essentially only speed optimization:</p>
<ul>
<li>For over 30X coverage:</li>
<li>Nanopore flip-flop R9.4 or R10.3: try: <code>corMhapOptions=--threshold 0.8 –ordered-sketch-size 1000 –ordered-kmer-size 14’ correctedErrorRate=0.105</code>
</li>
<li>For over 60X coverage: 2 recommendations were made, one said to decrease slightly (~1%). Another suggested using 12%
correctedErrorRate=0.12</li>
<li>Increasing minReadLength increases run time, increasing minOverlapLength improves assembly quality but increasing too much quickly degrades assemblies.</li>
</ul>
</div>
<div id="flye" class="section level4" number="1.4.2.2">
<h4>
<span class="header-section-number">1.4.2.2</span> flye<a class="anchor" aria-label="anchor" href="#flye"><i class="fas fa-link"></i></a>
</h4>
<p><code>docker pull staphb/flye</code></p>
</div>
</div>
<div id="kmer-based-metrics" class="section level3" number="1.4.3">
<h3>
<span class="header-section-number">1.4.3</span> kmer-based metrics<a class="anchor" aria-label="anchor" href="#kmer-based-metrics"><i class="fas fa-link"></i></a>
</h3>
<p>Canu will take some time to run. As it goes along, you can both check on its progress and learn about the genome you are assembling from some intermediate results. Take the k-mer data found in the .histogram files (i.e. in correction/0-mercounts/x.histogram, trimming/0-mercounts/x.histogram, unitigging/0-mercounts/x.histogram) and process them with <code>canuHistogramToKmerTable()</code>, as shown below. You can upload the output to : <a href="http://qb.cshl.edu/genomescope/genomescope2.0/" class="uri">http://qb.cshl.edu/genomescope/genomescope2.0/</a>. This will give you approximate genome size, ploidy, heterozygosity, repeat content, and read error rate. All good stuff to know!</p>
<div class="sourceCode" id="cb129"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">canuHistogramToKmerTable</span><span class="op">(</span>
  file_in_path <span class="op">=</span> <span class="st">"/Users/bust0037/Desktop/n_frust3assemblyB.ms22.histogram"</span>,
  file_out_path <span class="op">=</span> <span class="st">"/Users/bust0037/Desktop/n_frust3assemblyB.ms22.histogram_table"</span>
<span class="op">)</span></code></pre></div>
<div id="merqury" class="section level4" number="1.4.3.1">
<h4>
<span class="header-section-number">1.4.3.1</span> merqury<a class="anchor" aria-label="anchor" href="#merqury"><i class="fas fa-link"></i></a>
</h4>
<p><code>docker pull quay.io/chai/merqury</code></p>
<p><code>sudo docker run -u $(id -u) -v /home/bust0037/data1/Kalanchoe_DNASeq/round2_pass_reads_assembly/:/merqury/ quay.io/chai/merqury:latest quast.py -h</code></p>
<p>References:
<a href="https://www.biorxiv.org/content/10.1101/2020.03.15.992941v1.abstract" class="uri">https://www.biorxiv.org/content/10.1101/2020.03.15.992941v1.abstract</a>
Merqury: reference-free quality, completeness, and phasing assessment for genome assemblies</p>
</div>
</div>
</div>
<div id="evaluating-contigs" class="section level2" number="1.5">
<h2>
<span class="header-section-number">1.5</span> evaluating contigs<a class="anchor" aria-label="anchor" href="#evaluating-contigs"><i class="fas fa-link"></i></a>
</h2>
<p>Can these be merged into a single wrapper that can be run after each step in assembly/polishing/scaffolding etc?</p>
<div id="busco" class="section level3" number="1.5.1">
<h3>
<span class="header-section-number">1.5.1</span> BUSCO<a class="anchor" aria-label="anchor" href="#busco"><i class="fas fa-link"></i></a>
</h3>
<ul>
<li>Real BUSCO input will be /home/bust0037/data1/comparative_genomics/Kfedtschenkoi_382_v1.0.fa</li>
<li>Note BUSCO uses current working directory for input and output</li>
</ul>
<p><code>docker pull ezlabgva/busco:v5.3.2_cv1</code></p>
<p><code>sudo docker run -u $(id -u) -v /home/bust0037/data1/comparative_genomics/:/busco_wd ezlabgva/busco:v5.3.2_cv1 busco -i k_fed.contigs.fa -o busco_our_kfed/ -m genome -l eudicots_odb10</code></p>
</div>
<div id="quast" class="section level3" number="1.5.2">
<h3>
<span class="header-section-number">1.5.2</span> quast<a class="anchor" aria-label="anchor" href="#quast"><i class="fas fa-link"></i></a>
</h3>
<p>Quast provides a score called ALE: alignment liklihood estimate.</p>
<p><code>docker pull longas/quast</code></p>
<p><code>sudo docker run -u $(id -u) -v /home/bust0037/:/tmp/work/quast_results longas/quast:latest quast.py -h</code></p>
<p><code>sudo docker run -u $(id -u) -v /home/bust0037/ben_test/_ben_genomes/:/tmp/work/quast_results longas/quast:latest quast.py --fragmented /tmp/work/quast_results/ben_pre_n_frust.contigs.fasta --nanopore /tmp/work/quast_results/all_pass_reads.fastq --space-efficient --memory-efficient --fast</code></p>
</div>
</div>
<div id="polishing-contigs" class="section level2" number="1.6">
<h2>
<span class="header-section-number">1.6</span> polishing contigs<a class="anchor" aria-label="anchor" href="#polishing-contigs"><i class="fas fa-link"></i></a>
</h2>
<div id="medaka" class="section level3" number="1.6.1">
<h3>
<span class="header-section-number">1.6.1</span> medaka<a class="anchor" aria-label="anchor" href="#medaka"><i class="fas fa-link"></i></a>
</h3>
<p>Pull the docker image:</p>
<p><code>docker pull ontresearch/medaka</code></p>
<p>Run medaka:</p>
<p><code>sudo docker run -u $(id -u) -v /home/bust0037/data1/Kalanchoe_DNASeq/round2_pass_reads_assembly/:/medaka/ ontresearch/medaka:latest medaka_consensus -i medaka/all_reads.fastq -d medaka/k_fed.contigs.fasta -b 50</code></p>
<p>Good documentation here: <a href="https://labs.epi2me.io/notebooks/Introduction_to_how_ONT's_medaka_works.html" class="uri">https://labs.epi2me.io/notebooks/Introduction_to_how_ONT's_medaka_works.html</a></p>
<p>Medaka will:
* Map all your raw reads. Look for feedback like: <code>[M::worker_pipeline::2924.325*0.25] mapped 75823 sequences</code>.
* Do something else, updates in the form of: <code>21.7% Done (88.2/406.1 Mbases) in 6815.1s</code>.</p>
</div>
</div>
<div id="methylation-with-remora" class="section level2" number="1.7">
<h2>
<span class="header-section-number">1.7</span> methylation with remora<a class="anchor" aria-label="anchor" href="#methylation-with-remora"><i class="fas fa-link"></i></a>
</h2>
<p>can we call methylation status on our Kalanchoe genomes? -&gt; yes, we can use Remora -&gt; watch for new guppy release, MinKNOW integration -&gt; currently in bonito</p>
<pre><code>bonito basecaller dna_r10.4_e8.1_sup@v3.4 /data/reads --modified-bases 5mC --reference ref.mmi &gt; basecalls_with_mods.bam
bonito basecaller dna_r10.4_e8.1_sup@v3.4 --reference consensus.fasta --modified-bases 5mC</code></pre>
</div>
<div id="scaffolding-assembly" class="section level2" number="1.8">
<h2>
<span class="header-section-number">1.8</span> scaffolding assembly<a class="anchor" aria-label="anchor" href="#scaffolding-assembly"><i class="fas fa-link"></i></a>
</h2>
<div id="ragtag" class="section level3" number="1.8.1">
<h3>
<span class="header-section-number">1.8.1</span> RagTag<a class="anchor" aria-label="anchor" href="#ragtag"><i class="fas fa-link"></i></a>
</h3>
<p>INSTALL PYTHON (or upgrade python) &lt;- can this be done using docker?</p>
<p>INSTALL BIOCONDA
1. install miniconda:
curl -O <a href="https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh" class="uri">https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh</a>
sh …</p>
</div>
</div>
<div id="annotation" class="section level2" number="1.9">
<h2>
<span class="header-section-number">1.9</span> annotation<a class="anchor" aria-label="anchor" href="#annotation"><i class="fas fa-link"></i></a>
</h2>
<div id="augustus" class="section level3" number="1.9.1">
<h3>
<span class="header-section-number">1.9.1</span> augustus<a class="anchor" aria-label="anchor" href="#augustus"><i class="fas fa-link"></i></a>
</h3>
<p>see: <a href="https://hub.docker.com/r/pegi3s/autoaugustus/" class="uri">https://hub.docker.com/r/pegi3s/autoaugustus/</a></p>
<pre><code>docker pull pegi3s/autoaugustus</code></pre>
<pre><code>sudo docker run --rm -v $(pwd):$(pwd) pegi3s/autoaugustus augustus --species=help
sudo docker run --rm -v $(pwd):/augustus/ pegi3s/autoaugustus augustus --species=tomato /augustus/consensus.fasta &gt; consensus-preductions.gff --progress=TRUE</code></pre>
</div>
</div>
<div id="final-assessment" class="section level2" number="1.10">
<h2>
<span class="header-section-number">1.10</span> final assessment<a class="anchor" aria-label="anchor" href="#final-assessment"><i class="fas fa-link"></i></a>
</h2>
<div id="mosdepth" class="section level3" number="1.10.1">
<h3>
<span class="header-section-number">1.10.1</span> mosdepth<a class="anchor" aria-label="anchor" href="#mosdepth"><i class="fas fa-link"></i></a>
</h3>
<p><code>sudo docker pull quay.io/biocontainers/mosdepth:0.2.4--he527e40_0</code></p>
<p><code>sudo docker run -u $(id -u) -v /home/bust0037/data1/Kalanchoe_DNASeq/rounds_1and2_pass_assembly/:/mosdepth_wd quay.io/biocontainers/mosdepth:0.2.4--he527e40_0 mosdepth -h</code></p>
<p><code>sudo docker run -u $(id -u) -v /home/bust0037/data1/Kalanchoe_DNASeq/rounds_1and2_pass_assembly/:/mosdepth_wd quay.io/biocontainers/mosdepth:0.2.4--he527e40_0 mosdepth -n --fast-mode --by 1000 mosdepth_wd/mosdepth_out /mosdepth_wd/calls_to_draft.bam</code></p>
<p><code>sudo docker run -u $(id -u) -v /home/bust0037/data1/Kalanchoe_DNASeq/rounds_1and2_pass_assembly/:/mosdepth_wd gfanz/mosdepth -n --fast-mode --by 1000 mosdepth_wd/mosdepth_out /mosdepth_wd/calls_to_draft.bam</code></p>
<p><code>sudo docker run -u $(id -u) -v /home/bust0037/data1/Kalanchoe_DNASeq/rounds_1and2_pass_assembly/:/mosdepth_wd quay.io/biocontainers/mosdepth:0.2.4--he527e40_0 mosdepth --quantize 0:1:4:100:200: --fast-mode --by 1000 mosdepth_wd/mosdepth_out /mosdepth_wd/calls_to_draft.bam</code></p>
<ol start="2" style="list-style-type: decimal">
<li>set up channels</li>
</ol>
<p>/home/bust0037/miniconda3/bin/conda config –add channels defaults
/home/bust0037/miniconda3/bin/conda config –add channels bioconda
/home/bust0037/miniconda3/bin/conda config –add channels conda-forge</p>
<ol start="3" style="list-style-type: decimal">
<li>install packages</li>
</ol>
<p>/home/bust0037/miniconda3/bin/conda install sibeliaz
/home/bust0037/miniconda3/bin/conda install -c bioconda ragtag</p>
<ol start="4" style="list-style-type: decimal">
<li>run your stuff</li>
</ol>
<p>/home/bust0037/miniconda3/bin/conda run sibeliaz -n k_fed.contigs.scaffolded.fasta KlaxifloraFTBG2000359A_699_v3.0.fa</p>
<p>/home/bust0037/miniconda3/bin/conda run ragtag</p>
<p>/home/bust0037/miniconda3/bin/conda run ragtag</p>
<p>ragtag.py</p>
<p>/home/bust0037/meryl-1.3/bin/meryl
/home/bust0037/merqury-1.3/merqury.sh</p>
<p>sh $MERQURY/best_k.sh <genome_size></genome_size></p>
<p>Let’s look at some examples. For these example, we will use some fasta files stored in a Google Drive folder:</p>
<div class="sourceCode" id="cb133"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># reads &lt;- readFasta("https://drive.google.com/file/d/1r6E0U5LyYwjWenxy9yqh5QQ2mq1umWOW/view?usp=sharing")</span>

<span class="co"># # post &lt;- readFasta("/Users/bust0037/Desktop/ragtag.scaffold.fasta")</span>
<span class="co"># n_chroms &lt;- 18</span>

<span class="co"># pb &lt;- progress::progress_bar$new(total = n_chroms)</span>

<span class="co"># out &lt;- list()</span>

<span class="co"># for (i in 1:n_chroms) {</span>

<span class="co">#   pb$tick()</span>

<span class="co">#   dat &lt;- strsplit(substr(as.character(post[i]), 1, 50000000), "")[[1]]</span>
  
<span class="co">#   b &lt;- rle(dat)</span>

<span class="co">#   # Create a data frame</span>
<span class="co">#   dt &lt;- data.frame(number = b$values, lengths = b$lengths, scaff = i)</span>
<span class="co">#   # Get the end</span>
<span class="co">#   dt$end &lt;- cumsum(dt$lengths)</span>
<span class="co">#   # Get the start</span>
<span class="co">#   dt$start &lt;- dt$end - dt$lengths + 1</span>

<span class="co">#   # Select columns</span>
<span class="co">#   dt &lt;- dt[, c("number", "start", "end", "scaff")]</span>
<span class="co">#   # Sort rows</span>
<span class="co">#   dt &lt;- dt[order(dt$number), ]</span>

<span class="co">#   dt %&gt;%</span>
<span class="co">#     filter(number == "N") -&gt; N_dat</span>

<span class="co">#   out[[i]] &lt;- N_dat</span>

<span class="co"># }</span>

<span class="co"># out &lt;- do.call(rbind, out)</span>


<span class="co"># chroms &lt;- data.frame(</span>
<span class="co">#   lengths = post@ranges@width[1:n_chroms],</span>
<span class="co">#   scaff = seq(1,n_chroms,1)</span>
<span class="co"># )</span>

<span class="co"># ggplot() +</span>
<span class="co">#   statebins:::geom_rrect(data = chroms, aes(xmin = 0, xmax = lengths, ymin = -1, ymax = 1, fill = scaff), color = "black") +</span>
<span class="co">#   geom_rect(data = out, aes(xmin = start, xmax = end, ymin = -0.95, ymax = 0.95), color = "white", fill = "white", size = 0.08) +</span>
<span class="co">#   facet_grid(scaff~.) +</span>
<span class="co">#   scale_fill_viridis(end = 0.8) +</span>
<span class="co">#   theme_classic()</span>

<span class="co"># ggplot() +</span>
<span class="co">#   geom_rect(data = filter(chroms, scaff == 1 | scaff == 2), aes(xmin = 0, xmax = lengths, ymin = -1, ymax = 1, fill = scaff), color = "black") +</span>
<span class="co">#   geom_rect(data = filter(out, scaff == 1 | scaff == 2), aes(xmin = start, xmax = end, ymin = -0.95, ymax = 0.95), color = "white", fill = "white", size = 0.08) +</span>
<span class="co">#   facet_grid(scaff~.) +</span>
<span class="co">#   scale_y_continuous(limits = c(-2,2)) +</span>
<span class="co">#   scale_fill_viridis(end = 0.8) +</span>
<span class="co">#   theme_classic() +</span>
<span class="co">#   coord_polar()</span></code></pre></div>
</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="setup.html">setup</a></div>
<div class="next"><a href="comparative-genomics.html">comparative genomics</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <h2>Note: Second Edition is under construction 🏗</h2>
    <p>Now is a great time to provide feedback</p>
        <ul class="list-unstyled">
<li><a href="https://forms.gle/SZmB2Ct2exE2dBwv9">Provide feedback (5 min)</a></li>
          <!-- <li><a href="https://geocompr.robinlovelace.net/#reproducibility">Install updated packages</a></li> -->
          <!-- <li><a href="https://github.com/Robinlovelace/geocompr/issues">Open an issue <i class="fas fa-question"></i></a></li> -->
          <!-- <li><a href="https://discord.gg/Te3gWeDwmf">Chat on Discord <i class="fab fa-discord"></i></a></li> -->
        </ul>
<hr>
<nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#genome-assembly">genome assembly</a></li>
<li>
<a class="nav-link" href="#assembly"><span class="header-section-number">1.4</span> assembly</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#equipment"><span class="header-section-number">1.4.1</span> equipment</a></li>
<li><a class="nav-link" href="#assembly-software"><span class="header-section-number">1.4.2</span> assembly software</a></li>
<li><a class="nav-link" href="#kmer-based-metrics"><span class="header-section-number">1.4.3</span> kmer-based metrics</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#evaluating-contigs"><span class="header-section-number">1.5</span> evaluating contigs</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#busco"><span class="header-section-number">1.5.1</span> BUSCO</a></li>
<li><a class="nav-link" href="#quast"><span class="header-section-number">1.5.2</span> quast</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#polishing-contigs"><span class="header-section-number">1.6</span> polishing contigs</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#medaka"><span class="header-section-number">1.6.1</span> medaka</a></li></ul>
</li>
<li><a class="nav-link" href="#methylation-with-remora"><span class="header-section-number">1.7</span> methylation with remora</a></li>
<li>
<a class="nav-link" href="#scaffolding-assembly"><span class="header-section-number">1.8</span> scaffolding assembly</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#ragtag"><span class="header-section-number">1.8.1</span> RagTag</a></li></ul>
</li>
<li>
<a class="nav-link" href="#annotation"><span class="header-section-number">1.9</span> annotation</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#augustus"><span class="header-section-number">1.9.1</span> augustus</a></li></ul>
</li>
<li>
<a class="nav-link" href="#final-assessment"><span class="header-section-number">1.10</span> final assessment</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#mosdepth"><span class="header-section-number">1.10.1</span> mosdepth</a></li></ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/blob/main/index.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/edit/main/index.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Integrated Bioanalytics</strong>" was written by Lucas Busta and members of the Busta lab. It was last built on 2022-10-30.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
