<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

  <title>illumina read assessment | Integrated Bioanalytics</title>

    <meta name="author" content="Lucas Busta and members of the Busta lab" />
  
   <meta name="description" content="illumina read assessment | Integrated Bioanalytics is a book describing how to perform chemical, phylogenetic, and genomic analyses." />
   <meta name="generator" content="placeholder" />
  <meta property="og:title" content="illumina read assessment | Integrated Bioanalytics" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="illumina read assessment | Integrated Bioanalytics is a book describing how to perform chemical, phylogenetic, and genomic analyses." />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="illumina read assessment | Integrated Bioanalytics" />
  
  <meta name="twitter:description" content="illumina read assessment | Integrated Bioanalytics is a book describing how to perform chemical, phylogenetic, and genomic analyses." />
  
  <!-- JS -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script>
  <script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script>
    <script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no" />
    <link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet" />
    <script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script>
    <link href="libs/Lato-0.4.6/font.css" rel="stylesheet" />
    <link href="libs/Roboto_Mono-0.4.6/font.css" rel="stylesheet" />
    <link href="libs/Montserrat-0.4.6/font.css" rel="stylesheet" />
    <script src="libs/bs3compat-0.4.2/transition.js"></script>
    <script src="libs/bs3compat-0.4.2/tabs.js"></script>
    <script src="libs/bs3compat-0.4.2/bs3compat.js"></script>
    <link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet" />
    <script src="libs/bs4_book-1.0.0/bs4_book.js"></script>
    <link rel="icon" type="image/png" sizes="32x32" href="images/favicon-32x32.png">
    <link rel="icon" type="image/png" sizes="16x16" href="images/favicon-16x16.png">

    <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-99618359-1', 'auto');
      ga('send', 'pageview');

    </script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script>

  <!-- CSS -->
    <link rel="stylesheet" href="style/style.css" />
  
</head>

<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book">
    <a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h2>
        <a href="index.html" title="">Integrated Bioanalytics</a>
      </h2>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
      </form>

      <nav aria-label="Table of contents">
        <h2>Table of contents</h2>
        <div id="book-toc"></div>

        <div class="book-extra">
          <p><a id="book-repo" href="#">View book source <i class="fab fa-github"></i></a></li></p>
        </div>
      </nav>
    </div>
  </header>

  <main class="col-sm-12 col-md-9 col-lg-7" id="content">
<div id="illumina-read-assessment" class="section level1 unnumbered">
<h1>illumina read assessment</h1>
<p>Check out: fastqcr: An R Package Facilitating Quality Controls of Sequencing Data for Large Numbers of Samples.</p>
<div id="fastqc" class="section level2" number="2.3">
<h2><span class="header-section-number">2.3</span> fastqc</h2>
<p>See transXpress.</p>
<!-- end -->
<hr />
<hr />
<hr />
</div>
</div>



<!-- # nonmodel species

For nonmodel species transcriptome analysis, `transXpress` is recommended. `transXpress` uses the Trinity assembler by default, which can require at least 500GB of free disk space to run. Depending on your machine, you may also need to make some modifications to `transXpress` for it to run properly.

1. First, make sure to add conda-forge::ncurses as a dependency in the trinity_utils.yaml file to avoid errors with the script not being able to detect the samtools installation. The trinity_utils.yaml file in the emv directory should look like:


```r
channels:
  - conda-forge
  - bioconda
  - r
  - defaults
dependencies:
  - conda-forge::ncurses
  - trinity=2.13.2=hea94271_3
  - bioconductor-edger=3.36.0=r41hc247a5b_2
  - kallisto=0.48.0=h15996b6_2
  - r=4.1=r41hd8ed1ab_1006
  - r-tidyverse=1.3.2=r41hc72bb7e_1
  - bowtie2=2.5.0=py310h8d7afc0_0
  - samtools=1.16.1=h6899075_1
  - rsem=1.3.3=pl5321ha04fe3b_5 
```

2. Also add an explicit pip dependencey to the tmhmm.yaml file (it's also in the env directory):


```r
channels:
  - conda-forge
  - bioconda
  - r
  - defaults
dependencies:
  - pip
  - pip:
    - tmhmm-py
```

3. Also, you may also need to scale your memory usage in trinity by editing the file called `Snakefile`:


```r
rule trinity_inchworm_chrysalis:
...
params:
    memory="50"
```

4. It's possible that you will need to downgrade numpy for tmhmm to work (make sure you do this in the transxpress env):
`pip install "numpy<1.24.0"`

5. You also need to add targetp to the PATH variable in the transexpress environment:
`export PATH=$PATH:/project_data/shared/general_lab_resources/targetp-2.0/bin/`

6. Important! Remember that there must be an empty line on the end of the samples.tex file.

Once transXpress is complete, you may wish to move its output files to a long-term storage device. You may wish to keep the following files handy for downstream analysis though:

* all the sample name folders (ex. "fed_epi_hi_rep1")
* samples.txt
* samples_trimmed.txt
* busco_report.txt
* /transdecoder/longest_orfs.pep -->
<!-- # model species -->
<!-- # transcriptomic analyses {-} -->
<!-- `conda create --name foldseek`
`conda activate foldseek`
`conda install -c conda-forge -c bioconda foldseek`
 -->
<!-- end -->
<hr />
<hr />
<hr />
<!-- # (PART) PROTEOME ANALYSIS {-} -->
<!-- start proteome analyses -->
<!-- ## structure similarity

`conda create --name foldseek`
`conda activate foldseek`
`conda install -c conda-forge -c bioconda foldseek`
 -->
<!-- # proteome analyses {-} -->
<!-- end -->
<hr />
<hr />
<hr />
</div>



</div>
  </main>

  <div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <h2>Note: Second Edition is under construction üèó</h2>
    <p>Now is a great time to provide feedback</p>
        <ul class="list-unstyled">
          <li><a href="https://forms.gle/SZmB2Ct2exE2dBwv9">Provide feedback (5 min)</a></li>
          <!-- <li><a href="https://geocompr.robinlovelace.net/#reproducibility">Install updated packages</a></li> -->
          <!-- <li><a href="https://github.com/Robinlovelace/geocompr/issues">Open an issue <i class="fas fa-question"></i></a></li> -->
          <!-- <li><a href="https://discord.gg/Te3gWeDwmf">Chat on Discord <i class="fab fa-discord"></i></a></li> -->
        </ul>
        <hr>
    <nav id="toc" data-toggle="toc" aria-label="On this page">
      <h2>On this page</h2>
      <div id="book-on-this-page"></div>

      <div class="book-extra">
        <ul class="list-unstyled">
          <li><a id="book-source" href="#">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="#">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
      </div>
    </nav>
  </div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5">
  <div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Integrated Bioanalytics</strong>" was written by Lucas Busta and members of the Busta lab. It was last built on 2023-08-21.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>


</body>

</html>
