<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>models | Integrated Bioanalytics</title>
<meta name="author" content="Lucas Busta and members of the Busta lab">
<meta name="description" content="univariate Next on our quest to develop our abilities in analytical data exploration is modeling. We will start with some of the simplest models - linear models. There are a variety of ways to...">
<meta name="generator" content="bookdown 0.35 with bs4_book()">
<meta property="og:title" content="models | Integrated Bioanalytics">
<meta property="og:type" content="book">
<meta property="og:description" content="univariate Next on our quest to develop our abilities in analytical data exploration is modeling. We will start with some of the simplest models - linear models. There are a variety of ways to...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="models | Integrated Bioanalytics">
<meta name="twitter:description" content="univariate Next on our quest to develop our abilities in analytical data exploration is modeling. We will start with some of the simplest models - linear models. There are a variety of ways to...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><link href="libs/Lato-0.4.6/font.css" rel="stylesheet">
<link href="libs/Roboto_Mono-0.4.6/font.css" rel="stylesheet">
<link href="libs/Montserrat-0.4.6/font.css" rel="stylesheet">
<script src="libs/bs3compat-0.4.2/transition.js"></script><script src="libs/bs3compat-0.4.2/tabs.js"></script><script src="libs/bs3compat-0.4.2/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><link rel="icon" type="image/png" sizes="32x32" href="images/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="images/favicon-16x16.png">
<script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-99618359-1', 'auto');
      ga('send', 'pageview');

    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="style/style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h2>
        <a href="index.html" title="">Integrated Bioanalytics</a>
      </h2>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">WELCOME</a></li>
<li class="book-part">DATA ANALYSIS IN R</li>
<li><a class="" href="overview.html">overview</a></li>
<li><a class="" href="installation.html">installation</a></li>
<li><a class="" href="data-visualization-i.html">data visualization I</a></li>
<li><a class="" href="data-visualization-ii.html">data visualization II</a></li>
<li><a class="" href="data-visualization-iii.html">data visualization III</a></li>
<li><a class="" href="data-wrangling-and-summaries.html">data wrangling and summaries</a></li>
<li><a class="" href="dimensionality-reduction.html">dimensionality reduction</a></li>
<li><a class="" href="clustering.html">clustering</a></li>
<li><a class="" href="comparing-means.html">comparing means</a></li>
<li><a class="active" href="models.html">models</a></li>
<li class="book-part">GC-MS DATA</li>
<li><a class="" href="loading-analyzegcmsdata.html">loading analyzeGCMSdata</a></li>
<li><a class="" href="using-analyzegcmsdata.html">using analyzeGCMSdata</a></li>
<li><a class="" href="cdf-export.html">CDF export</a></li>
<li class="book-part">TRANSCRIPTOME ANALYSIS</li>
<li><a class="" href="transcriptome-assembly.html">transcriptome assembly</a></li>
<li class="book-part">EVOLUTIONARY ANALYSIS</li>
<li><a class="" href="similarity-searching.html">similarity searching</a></li>
<li><a class="" href="alignments.html">alignments</a></li>
<li><a class="" href="phylogenies.html">phylogenies</a></li>
<li><a class="" href="phylogenetic-analyses.html">phylogenetic analyses</a></li>
<li><a class="" href="comparative-genomics.html">comparative genomics</a></li>
<li class="book-part">LANGUAGE MODELS</li>
<li><a class="" href="completiongpt.html">completionGPT</a></li>
<li><a class="" href="analyzeliterature.html">analyzeLiterature</a></li>
<li class="book-part">SCIENTIFIC WRITING</li>
<li><a class="" href="overview-1.html">overview</a></li>
<li><a class="" href="figures-captions.html">figures &amp; captions</a></li>
<li><a class="" href="results-and-discussion.html">results and discussion</a></li>
<li><a class="" href="conclusion-and-introduction.html">conclusion and introduction</a></li>
<li><a class="" href="abstract-and-title.html">abstract and title</a></li>
<li class="book-part">APPENDIX</li>
<li><a class="" href="links.html">links</a></li>
<li><a class="" href="r-faq.html">r faq</a></li>
<li><a class="" href="templates.html">templates</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="models" class="section level1 unnumbered">
<h1>models<a class="anchor" aria-label="anchor" href="#models"><i class="fas fa-link"></i></a>
</h1>
<div id="univariate" class="section level2 unnumbered">
<h2>univariate<a class="anchor" aria-label="anchor" href="#univariate"><i class="fas fa-link"></i></a>
</h2>
<p>Next on our quest to develop our abilities in analytical data exploration is modeling. We will start with some of the simplest models - linear models. There are a variety of ways to build linear models in R, but we will use a function called <code>buildModel</code>. To use it, we simply give it our data, and tell it which to sets of values we want to compare. To tell it what we want to compare, we tell it what variable it should try to predict (outcome variable) and what other variables it should use as inputs for the preduction (predictor variables).</p>
<p>Before we begin, we need to talk about handling missing data. If we have missing data in our data set, oen way forward is to impute it. This means that we need to fill in the missing values with something. There are many ways to do this, but we will use the median value of each column. We can do this using the <code>impute</code> function from the <code>rstatix</code> package. Let’s do that now:</p>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/any.html">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">metabolomics_data</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## [1] TRUE</span></span>
<span><span class="va">metabolomics_data</span><span class="op">[</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">metabolomics_data</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">Hmisc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Hmisc/man/impute.html">impute</a></span><span class="op">(</span><span class="va">x</span>, <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">x</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/any.html">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">metabolomics_data</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## [1] FALSE</span></span></code></pre></div>
<p>Let’s look at an example of model building. Suppose we want to know if the abundances of ADP and AMP are related in our metabolomics dataset (this is the same as trying to predict the abundance of ADP from AMP):</p>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">metabolomics_data</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ADP</span>, y <span class="op">=</span> <span class="va">AMP</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## Don't know how to automatically pick scale for object of</span></span>
<span><span class="co">## type &lt;impute&gt;. Defaulting to continuous.</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-166-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>It looks like there might be a relationship! Let’s build a linear model for that relationship:</p>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">basic_regression_model</span> <span class="op">&lt;-</span> <span class="fu">buildModel</span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">metabolomics_data</span>,</span>
<span>  model_type <span class="op">=</span> <span class="st">"linear_regression"</span>,</span>
<span>  predictor_variables <span class="op">=</span> <span class="st">"ADP"</span>,</span>
<span>  outcome_variable <span class="op">=</span> <span class="st">"AMP"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">basic_regression_model</span></span>
<span><span class="co">## $model_type</span></span>
<span><span class="co">## [1] "linear_regression"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $model</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Call:</span></span>
<span><span class="co">## lm(formula = formula, data = training_data, model = TRUE, x = TRUE, </span></span>
<span><span class="co">##     y = TRUE, qr = TRUE)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Coefficients:</span></span>
<span><span class="co">## (Intercept)          ADP  </span></span>
<span><span class="co">##      4.5021       0.6763  </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $metrics</span></span>
<span><span class="co">##               variable   value std_err        type  p_value</span></span>
<span><span class="co">## 1            r_squared  0.4774      NA   statistic       NA</span></span>
<span><span class="co">## 2    total_sum_squares 38.6346      NA   statistic       NA</span></span>
<span><span class="co">## 3 residual_sum_squares 20.1904      NA   statistic       NA</span></span>
<span><span class="co">## 4          (Intercept)  4.5021  0.9593 coefficient 9.46e-06</span></span>
<span><span class="co">## 5                  ADP  0.6763  0.0742 coefficient 0.00e+00</span></span>
<span><span class="co">##   p_value_adj</span></span>
<span><span class="co">## 1          NA</span></span>
<span><span class="co">## 2          NA</span></span>
<span><span class="co">## 3          NA</span></span>
<span><span class="co">## 4    9.46e-06</span></span>
<span><span class="co">## 5    0.00e+00</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $rmse</span></span>
<span><span class="co">## [1] 0.4050733</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $fold_cross_validation</span></span>
<span><span class="co">## [1] 10</span></span></code></pre></div>
<p>The output of <code>buildModel</code> consists of two thigs: the model, and the metrics. Let’s look at the model:</p>
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">basic_regression_model</span><span class="op">$</span><span class="va">model</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Call:</span></span>
<span><span class="co">## lm(formula = formula, data = training_data, model = TRUE, x = TRUE, </span></span>
<span><span class="co">##     y = TRUE, qr = TRUE)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Coefficients:</span></span>
<span><span class="co">## (Intercept)          ADP  </span></span>
<span><span class="co">##      4.5021       0.6763</span></span></code></pre></div>
<p>This is a linear model stored inside a special object type inside R called an <code>lm</code>. They can be a bit tricky to work with, but we have a way to make it easier - we’ll look at that in a second. Before that, let’s look at the metrics.</p>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">basic_regression_model</span><span class="op">$</span><span class="va">metrics</span></span>
<span><span class="co">##               variable   value std_err        type  p_value</span></span>
<span><span class="co">## 1            r_squared  0.4774      NA   statistic       NA</span></span>
<span><span class="co">## 2    total_sum_squares 38.6346      NA   statistic       NA</span></span>
<span><span class="co">## 3 residual_sum_squares 20.1904      NA   statistic       NA</span></span>
<span><span class="co">## 4          (Intercept)  4.5021  0.9593 coefficient 9.46e-06</span></span>
<span><span class="co">## 5                  ADP  0.6763  0.0742 coefficient 0.00e+00</span></span>
<span><span class="co">##   p_value_adj</span></span>
<span><span class="co">## 1          NA</span></span>
<span><span class="co">## 2          NA</span></span>
<span><span class="co">## 3          NA</span></span>
<span><span class="co">## 4    9.46e-06</span></span>
<span><span class="co">## 5    0.00e+00</span></span></code></pre></div>
<p>It shows us the r-squared, the total and residual sum of squared, the intercept (b in y = m*x + b), and the coefficient for AMP (i.e. the slope, m), as well some other things (we will talk about them in a second).</p>
<p>We can also use a function called <code>predictWithModel</code> to make some predictions using the model. Let’s try that for ADP and AMP. What we do is give it the model, and then tell it what values we want to predict for. In this case, we want to predict the abundance of ADP for each value of AMP in our data set. We can do that like this:</p>
<div class="sourceCode" id="cb115"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">AMP_values_predicted_from_ADP_values</span> <span class="op">&lt;-</span> <span class="fu">predictWithModel</span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">metabolomics_data</span>,</span>
<span>  model_type <span class="op">=</span> <span class="st">"linear_regression"</span>,</span>
<span>  model <span class="op">=</span> <span class="va">basic_regression_model</span><span class="op">$</span><span class="va">model</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">AMP_values_predicted_from_ADP_values</span><span class="op">)</span></span>
<span><span class="co">##        1        2        3        4        5        6 </span></span>
<span><span class="co">## 13.18473 13.35509 13.48334 13.42981 12.57305 13.18377</span></span></code></pre></div>
<p>So, predictWithModel is using the model to predict AMP values from ADP. However, note that we have the measured AMP values in our data set. We can compare the predicted values to the measured values to see how well our model is doing. We can do that like this:</p>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ADP_values</span> <span class="op">&lt;-</span> <span class="va">metabolomics_data</span><span class="op">$</span><span class="va">ADP</span></span>
<span></span>
<span><span class="va">predictions_from_basic_linear_model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>    ADP_values <span class="op">=</span> <span class="va">ADP_values</span>,</span>
<span>    AMP_values_predicted_from_ADP_values <span class="op">=</span> <span class="va">AMP_values_predicted_from_ADP_values</span>,</span>
<span>    AMP_values_measured <span class="op">=</span> <span class="va">metabolomics_data</span><span class="op">$</span><span class="va">AMP</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_line</span><span class="op">(</span></span>
<span>        data <span class="op">=</span> <span class="va">predictions_from_basic_linear_model</span>,</span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ADP_values</span>, y <span class="op">=</span> <span class="va">AMP_values_predicted_from_ADP_values</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"red"</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span></span>
<span>        data <span class="op">=</span> <span class="va">predictions_from_basic_linear_model</span>,</span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ADP_values</span>, y <span class="op">=</span> <span class="va">AMP_values_predicted_from_ADP_values</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"red"</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span></span>
<span>        data <span class="op">=</span> <span class="va">metabolomics_data</span>,</span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ADP</span>, y <span class="op">=</span> <span class="va">AMP</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"blue"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="va">plot1</span></span>
<span><span class="co">## Don't know how to automatically pick scale for object of</span></span>
<span><span class="co">## type &lt;impute&gt;. Defaulting to continuous.</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-171-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Very good. Now let’s talk about evaluating the quality of our model. For this we need some means of assessing how well our line fits our data. We will use residuals - the distance between each of our points and our line.</p>
<div class="sourceCode" id="cb117"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">predictions_from_basic_linear_model</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ADP_values</span>, y <span class="op">=</span> <span class="va">AMP_values_measured</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ADP_values</span>, y <span class="op">=</span> <span class="va">AMP_values_predicted_from_ADP_values</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_segment</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ADP_values</span>, y <span class="op">=</span> <span class="va">AMP_values_measured</span>, xend <span class="op">=</span> <span class="va">ADP_values</span>, yend <span class="op">=</span> <span class="va">AMP_values_predicted_from_ADP_values</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## Don't know how to automatically pick scale for object of</span></span>
<span><span class="co">## type &lt;impute&gt;. Defaulting to continuous.</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-172-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>We can calculate the sum of the squared residuals:</p>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span></span>
<span>  <span class="op">(</span><span class="va">predictions_from_basic_linear_model</span><span class="op">$</span><span class="va">AMP_values_measured</span> <span class="op">-</span> <span class="va">predictions_from_basic_linear_model</span><span class="op">$</span><span class="va">AMP_values_predicted_from_ADP_values</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span></span>
<span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">## [1] 20.1904</span></span></code></pre></div>
<p>Cool! Let’s call that the “residual sum of the squares”. So… does that mean our model is good? I don’t know. We have to compare that number to something. Let’s compare it to a super simple model that is just defined by the mean y value of the input data.</p>
<div class="sourceCode" id="cb119"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">metabolomics_data</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ADP</span>, y <span class="op">=</span> <span class="va">AMP</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_hline</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">AMP</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## Don't know how to automatically pick scale for object of</span></span>
<span><span class="co">## type &lt;impute&gt;. Defaulting to continuous.</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-174-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>A pretty bad model, I agree. How much better is our linear model that the flat line model? Let’s create a measure of the distance between each point and the point predicted for that same x value on the model:</p>
<div class="sourceCode" id="cb120"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">metabolomics_data</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ADP</span>, y <span class="op">=</span> <span class="va">AMP</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_hline</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">ADP</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_segment</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ADP</span>, y <span class="op">=</span> <span class="va">AMP</span>, xend <span class="op">=</span> <span class="va">ADP</span>, yend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">ADP</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## Don't know how to automatically pick scale for object of</span></span>
<span><span class="co">## type &lt;impute&gt;. Defaulting to continuous.</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-175-1.png" width="100%" style="display: block; margin: auto;"></div>
<div class="sourceCode" id="cb121"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span></span>
<span>  <span class="op">(</span><span class="va">metabolomics_data</span><span class="op">$</span><span class="va">AMP</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">metabolomics_data</span><span class="op">$</span><span class="va">AMP</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span></span>
<span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">## [1] 38.63462</span></span></code></pre></div>
<p>38.63! Wow. Let’s call that the “total sum of the squares”, and now we can compare that to our “residual sum of the squares”:</p>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fl">1</span><span class="op">-</span><span class="op">(</span><span class="fl">20.1904</span><span class="op">/</span><span class="fl">38.63</span><span class="op">)</span></span>
<span><span class="co">## [1] 0.4773389</span></span></code></pre></div>
<p>0.47! Alright. That is our R squared value. It is equal to 1 minus the ratio of the “residual sum of the squares” to the “total sum of the squares”. You can think of the R squared value as:
- The amount of variance in the response explained by the dependent variable.
- How much better the line of best fit describes the data than the flat line.
Now, let’s put it all together and make it pretty:</p>
<div class="sourceCode" id="cb123"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_line</span><span class="op">(</span></span>
<span>        data <span class="op">=</span> <span class="va">predictions_from_basic_linear_model</span>,</span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ADP_values</span>, y <span class="op">=</span> <span class="va">AMP_values_predicted_from_ADP_values</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"red"</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span></span>
<span>        data <span class="op">=</span> <span class="va">predictions_from_basic_linear_model</span>,</span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ADP_values</span>, y <span class="op">=</span> <span class="va">AMP_values_predicted_from_ADP_values</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"red"</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span></span>
<span>        data <span class="op">=</span> <span class="va">metabolomics_data</span>,</span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ADP</span>, y <span class="op">=</span> <span class="va">AMP</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"blue"</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">annotate</span><span class="op">(</span>geom <span class="op">=</span> <span class="st">"table"</span>,</span>
<span>      x <span class="op">=</span> <span class="fl">10</span>,</span>
<span>      y <span class="op">=</span> <span class="fl">15</span>,</span>
<span>      label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="fu">select</span><span class="op">(</span><span class="va">basic_regression_model</span><span class="op">$</span><span class="va">metrics</span>, <span class="va">variable</span>, <span class="va">value</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">coord_cartesian</span><span class="op">(</span>ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">16</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bottom</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">predictions_from_basic_linear_model</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_col</span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ADP_values</span>, y <span class="op">=</span> <span class="va">AMP_values_measured</span><span class="op">-</span><span class="va">AMP_values_predicted_from_ADP_values</span><span class="op">)</span>,</span>
<span>    width <span class="op">=</span> <span class="fl">0.03</span>, color <span class="op">=</span> <span class="st">"black"</span>, position <span class="op">=</span> <span class="st">"dodge"</span>, alpha <span class="op">=</span> <span class="fl">0.5</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">top</span>, <span class="va">bottom</span>, ncol <span class="op">=</span> <span class="fl">1</span>, labels <span class="op">=</span> <span class="st">"AUTO"</span>, rel_heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## Don't know how to automatically pick scale for object of</span></span>
<span><span class="co">## type &lt;impute&gt;. Defaulting to continuous.</span></span>
<span><span class="co">## Don't know how to automatically pick scale for object of</span></span>
<span><span class="co">## type &lt;impute&gt;. Defaulting to continuous.</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-177-1.png" width="100%" style="display: block; margin: auto;"></div>
</div>
<div id="multivariate" class="section level2 unnumbered">
<h2>multivariate<a class="anchor" aria-label="anchor" href="#multivariate"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">multiple_regression_model</span> <span class="op">&lt;-</span> <span class="fu">buildModel</span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">metabolomics_data</span>,</span>
<span>    model_type <span class="op">=</span> <span class="st">"linear_regression"</span>,</span>
<span>    predictor_variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">metabolomics_data</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fl">50</span><span class="op">]</span>,</span>
<span>    outcome_variable <span class="op">=</span> <span class="st">"AMP"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">ADP_values</span> <span class="op">&lt;-</span> <span class="va">metabolomics_data</span><span class="op">$</span><span class="va">ADP</span></span>
<span></span>
<span><span class="va">AMP_values_predicted_from_ADP_values</span> <span class="op">&lt;-</span> <span class="fu">predictWithModel</span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">metabolomics_data</span>,</span>
<span>    model_type <span class="op">=</span> <span class="st">"linear_regression"</span>,</span>
<span>    model <span class="op">=</span> <span class="va">multiple_regression_model</span><span class="op">$</span><span class="va">model</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">predictions_from_multiple_linear_model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">ADP_values</span>,</span>
<span>    y <span class="op">=</span> <span class="va">AMP_values_predicted_from_ADP_values</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_line</span><span class="op">(</span></span>
<span>        data <span class="op">=</span> <span class="va">predictions_from_multiple_linear_model</span>,</span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"red"</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span></span>
<span>        data <span class="op">=</span> <span class="va">predictions_from_multiple_linear_model</span>,</span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"red"</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span></span>
<span>        data <span class="op">=</span> <span class="va">metabolomics_data</span>,</span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ADP</span>, y <span class="op">=</span> <span class="va">AMP</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"blue"</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span>    <span class="va">plot1</span> <span class="op">/</span> <span class="va">plot2</span></span>
<span><span class="co">## Don't know how to automatically pick scale for object of</span></span>
<span><span class="co">## type &lt;impute&gt;. Defaulting to continuous.</span></span>
<span><span class="co">## Don't know how to automatically pick scale for object of</span></span>
<span><span class="co">## type &lt;impute&gt;. Defaulting to continuous.</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-178-1.png" width="100%" style="display: block; margin: auto;"></div>
<div class="sourceCode" id="cb125"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">multiple_regression_model</span><span class="op">$</span><span class="va">metrics</span><span class="op">)</span></span>
<span><span class="co">##               variable   value std_err        type</span></span>
<span><span class="co">## 1            r_squared  0.8716      NA   statistic</span></span>
<span><span class="co">## 2    total_sum_squares 38.6346      NA   statistic</span></span>
<span><span class="co">## 3 residual_sum_squares  4.9605      NA   statistic</span></span>
<span><span class="co">## 4          (Intercept)  7.5995 20.0559 coefficient</span></span>
<span><span class="co">## 5             Pyruvate  0.1880  0.3149 coefficient</span></span>
<span><span class="co">## 6      MethylSuccinate -0.8486  0.3767 coefficient</span></span>
<span><span class="co">##      p_value p_value_adj</span></span>
<span><span class="co">## 1         NA          NA</span></span>
<span><span class="co">## 2         NA          NA</span></span>
<span><span class="co">## 3         NA          NA</span></span>
<span><span class="co">## 4 0.70657037           1</span></span>
<span><span class="co">## 5 0.55361280           1</span></span>
<span><span class="co">## 6 0.02929848           1</span></span></code></pre></div>
</div>
<div id="classification" class="section level2 unnumbered">
<h2>classification<a class="anchor" aria-label="anchor" href="#classification"><i class="fas fa-link"></i></a>
</h2>
<p>We will use random forests, a type of machine learning model, to make classification models. Start by watching these videos:
1. Decision trees: <a href="https://www.youtube.com/watch?v=_L39rN6gz7Y" class="uri">https://www.youtube.com/watch?v=_L39rN6gz7Y</a>
2. Random forests: <a href="https://www.youtube.com/watch?v=J4Wdy0Wc_xQ" class="uri">https://www.youtube.com/watch?v=J4Wdy0Wc_xQ</a></p>
<p>```{r]}
random_forest_model &lt;- buildModel(
data = metabolomics_data,
model_type = “random_forest_regression”,
predictor_variables = colnames(metabolomics_data)[1:30],
outcome_variable = “AMP”,
train_test_ratio = 0.8,
fold_cross_validation = 3,
optimization_parameters = list(mtry = seq(5,6,1), trees = seq(100,200,50))
)</p>
<p>ADP_values &lt;- metabolomics_data$ADP</p>
<p>AMP_values_predicted_from_ADP_values &lt;- predictWithModel(
data = metabolomics_data,
model_type = “random_forest_regression”,
model = random_forest_model$model
)</p>
<p>predictions_from_random_forest_model &lt;- data.frame(
x = ADP_values,
y = AMP_values_predicted_from_ADP_values
)</p>
<p>plot3 &lt;- ggplot() +
geom_line(
data = predictions_from_random_forest_model,
aes(x = x, y = y), color = “red”
) +
geom_point(
data = predictions_from_random_forest_model,
aes(x = x, y = y), color = “red”
) +
geom_point(
data = metabolomics_data,
aes(x = ADP, y = AMP), color = “blue”
)</p>
<p>plot1 / plot2 / plot3</p>
<p>basic_regression_model<span class="math inline">\(metrics[1,] multiple_regression_model\)</span>metrics[1,]
random_forest_model$rmse</p>
<pre><code>
## exercises {-}

To practice creating models, try the following:

1. Choose one of the datasets we have used so far, and run a principal components analysis on it. Note that the output of the analysis when you run "pca_ord" contains the Dimension 1 coordinate "Dim.1" for each sample, as well as the abundance of each analyte in that sample.

2. Using the information from the ordination plot, identify two analytes: one that has a variance that is strongly and positively correlated with the first principal component (i.e. dimension 1), and one that has a variance that is slightly less strongly, but still positively correlated with the first principal component. Using `buildLinearModel`, create and plot two linear regression models, one that regresses each of those analytes against dimension 1 (in other words, the x-axis should be the Dim.1 coordinate for each sample, and the y-axis should be the values for one of the two selected analytes). Which has the greater r-squared value? Based on what you know about PCA, does that make sense?

3. Choose two analytes: one should be one of the analytes from question 2 above, the other should be an analyte that, according to your PCA ordination analysis, is negatively correlated with the first principal component. Using `buildModel` and `predictWithModel` create plots showing how those two analytes are correlated with dimension 1. One should be positively correlated, and the other negatively correlated.

4. (algae bloom question)




## further reading {-}

https://github.com/easystats/performance

Here is more information on common machine learning tasks:
https://pythonprogramminglanguage.com/machine-learning-tasks/

Some major tasks include:
  Classification
  Regression
  Clustering

Classification with random forests:
1. http://www.rebeccabarter.com/blog/2020-03-25_machine_learning/
2. https://hansjoerg.me/2020/02/09/tidymodels-for-machine-learning/
3. https://towardsdatascience.com/dials-tune-and-parsnip-tidymodels-way-to-create-and-tune-model-parameters-c97ba31d6173

&lt;!-- end --&gt;

&lt;!-- start map dataspecial topics --&gt;

# special topics {-}

## map data {-}

&lt;img src="https://thebustalab.github.io/integrated_bioanalytics/images/hawaii_aquifers.jpeg" width="100%" style="display: block; margin: auto;" /&gt;

### plotting boundaries {-}

There is a simple way to plot maps with ggplot. The map data comes with `ggplot2`! Let's have a look. See below some of the data sets included. Options included with ggplot are: `world`, `world2`, `usa`, `state` (US), `county` (US), `nz`, `italy`, and `france`. `geom_polygon()` is useful for plotting these, at (at least to me) seems more intuitive than `geom_map()`.


```r
head(map_data("world"))
##        long      lat group order region subregion
## 1 -69.89912 12.45200     1     1  Aruba      &lt;NA&gt;
## 2 -69.89571 12.42300     1     2  Aruba      &lt;NA&gt;
## 3 -69.94219 12.43853     1     3  Aruba      &lt;NA&gt;
## 4 -70.00415 12.50049     1     4  Aruba      &lt;NA&gt;
## 5 -70.06612 12.54697     1     5  Aruba      &lt;NA&gt;
## 6 -70.05088 12.59707     1     6  Aruba      &lt;NA&gt;</code></pre>
<div class="sourceCode" id="cb127"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">map_data</span><span class="op">(</span><span class="st">"state"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##        long      lat group order  region subregion</span></span>
<span><span class="co">## 1 -87.46201 30.38968     1     1 alabama      &lt;NA&gt;</span></span>
<span><span class="co">## 2 -87.48493 30.37249     1     2 alabama      &lt;NA&gt;</span></span>
<span><span class="co">## 3 -87.52503 30.37249     1     3 alabama      &lt;NA&gt;</span></span>
<span><span class="co">## 4 -87.53076 30.33239     1     4 alabama      &lt;NA&gt;</span></span>
<span><span class="co">## 5 -87.57087 30.32665     1     5 alabama      &lt;NA&gt;</span></span>
<span><span class="co">## 6 -87.58806 30.32665     1     6 alabama      &lt;NA&gt;</span></span></code></pre></div>
<div class="sourceCode" id="cb128"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">map_data</span><span class="op">(</span><span class="st">"county"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##        long      lat group order  region subregion</span></span>
<span><span class="co">## 1 -86.50517 32.34920     1     1 alabama   autauga</span></span>
<span><span class="co">## 2 -86.53382 32.35493     1     2 alabama   autauga</span></span>
<span><span class="co">## 3 -86.54527 32.36639     1     3 alabama   autauga</span></span>
<span><span class="co">## 4 -86.55673 32.37785     1     4 alabama   autauga</span></span>
<span><span class="co">## 5 -86.57966 32.38357     1     5 alabama   autauga</span></span>
<span><span class="co">## 6 -86.59111 32.37785     1     6 alabama   autauga</span></span></code></pre></div>
<div class="sourceCode" id="cb129"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">map_data</span><span class="op">(</span><span class="st">"france"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##       long      lat group order region subregion</span></span>
<span><span class="co">## 1 2.557093 51.09752     1     1   Nord      &lt;NA&gt;</span></span>
<span><span class="co">## 2 2.579995 51.00298     1     2   Nord      &lt;NA&gt;</span></span>
<span><span class="co">## 3 2.609101 50.98545     1     3   Nord      &lt;NA&gt;</span></span>
<span><span class="co">## 4 2.630782 50.95073     1     4   Nord      &lt;NA&gt;</span></span>
<span><span class="co">## 5 2.625894 50.94116     1     5   Nord      &lt;NA&gt;</span></span>
<span><span class="co">## 6 2.597699 50.91967     1     6   Nord      &lt;NA&gt;</span></span></code></pre></div>
<p>Cool! We can see that lat, lon, group, order, region, and subregion are included. That makes plotting easy. Note that <code>coord_map()</code> can help preserve aspect ratios:</p>
<div class="sourceCode" id="cb130"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu">map_data</span><span class="op">(</span><span class="st">"world"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">long</span>, y <span class="op">=</span> <span class="va">lat</span>, color <span class="op">=</span> <span class="va">group</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_void</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">coord_map</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-185-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Note that we can use <code>coord_map()</code> to do some pretty cool things!</p>
<div class="sourceCode" id="cb131"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu">map_data</span><span class="op">(</span><span class="st">"world"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">long</span>, y <span class="op">=</span> <span class="va">lat</span>, color <span class="op">=</span> <span class="va">group</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_void</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">coord_map</span><span class="op">(</span>projection <span class="op">=</span> <span class="st">"albers"</span>, lat0 <span class="op">=</span> <span class="fl">39</span>, lat1 <span class="op">=</span> <span class="fl">45</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-186-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>We can use filtering to produce maps of specific regions.</p>
<div class="sourceCode" id="cb132"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_polygon</span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="fu">map_data</span><span class="op">(</span><span class="st">"county"</span><span class="op">)</span>, <span class="va">region</span> <span class="op">==</span> <span class="st">"minnesota"</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">long</span>, y <span class="op">=</span> <span class="va">lat</span>, group <span class="op">=</span> <span class="va">subregion</span>, fill <span class="op">=</span> <span class="va">subregion</span><span class="op">)</span>,</span>
<span>    color <span class="op">=</span> <span class="st">"black"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_void</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">coord_map</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-187-1.png" width="100%" style="display: block; margin: auto;"></div>
<div id="further-reading-8" class="section level3 unnumbered">
<h3>further reading<a class="anchor" aria-label="anchor" href="#further-reading-8"><i class="fas fa-link"></i></a>
</h3>
<p>For more on plotting maps in R: <a href="https://datavizpyr.com/how-to-make-us-state-and-county-level-maps-in-r/">datavizplyr</a></p>
<p>For more advanced map plotting: <a href="https://r-spatial.org/r/2018/10/25/ggplot2-sf.html">R Spatial</a></p>
<!-- ## exercises {-} -->
<!-- Using the `hawaii_aquifers` data set, please complete the following:

1. Choose one analyte and filter the data so only the rows for that analyte are shown.

2. Choose two of the aquifers. Are the mean abundances for your chosen analyte different in these two aquifers? Don't forget to test your data for normality and homogeneity of variance before selecting a statistical test. Use a plot to illustrate whether the means are similar or different.

3. Choose a second analyte, different from the first one you chose. Considering all the aquifers in the dataset, do any of them have the same abundance of this analyte? Again, don't forget about normality and homogeneity of variance tests. Use a plot to illustrate your answer.

4. Repeat #3 above, but switch the type of test used (i.e. use non-parametric if you used parametric for #3 and vice-versa). Compare the *p* values and *p* groups obtained by the two methods. Use a graphic to illustrate this. Why are they different? -->
<!-- end -->
<hr>
<hr>
<hr>
</div>
</div>
</div>



<div class="chapter-nav">
<div class="prev"><a href="comparing-means.html">comparing means</a></div>
<div class="next"><a href="loading-analyzegcmsdata.html">loading analyzeGCMSdata</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <h2>Note: Second Edition is under construction 🏗</h2>
    <p>Now is a great time to provide feedback</p>
        <ul class="list-unstyled">
<li><a href="https://forms.gle/SZmB2Ct2exE2dBwv9">Provide feedback (5 min)</a></li>
          <!-- <li><a href="https://geocompr.robinlovelace.net/#reproducibility">Install updated packages</a></li> -->
          <!-- <li><a href="https://github.com/Robinlovelace/geocompr/issues">Open an issue <i class="fas fa-question"></i></a></li> -->
          <!-- <li><a href="https://discord.gg/Te3gWeDwmf">Chat on Discord <i class="fab fa-discord"></i></a></li> -->
        </ul>
<hr>
<nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#models">models</a></li>
<li><a class="nav-link" href="#univariate">univariate</a></li>
<li><a class="nav-link" href="#multivariate">multivariate</a></li>
<li>
<a class="nav-link" href="#classification">classification</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#further-reading-8">further reading</a></li></ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/blob/main/index.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/edit/main/index.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>
</div>
  

  

</div>
 <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Integrated Bioanalytics</strong>" was written by Lucas Busta and members of the Busta lab. It was last built on 2023-10-09.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
