<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>pca | Integrated Bioanalytics</title>
<meta name="author" content="Lucas Busta">
<meta name="description" content="“Which analytes are driving differences among my samples?” In addition to heirarchical clustering, there is another way to look at our data in a cluster context - i.e. another way to identify...">
<meta name="generator" content="bookdown 0.24.4 with bs4_book()">
<meta property="og:title" content="pca | Integrated Bioanalytics">
<meta property="og:type" content="book">
<meta property="og:description" content="“Which analytes are driving differences among my samples?” In addition to heirarchical clustering, there is another way to look at our data in a cluster context - i.e. another way to identify...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="pca | Integrated Bioanalytics">
<meta name="twitter:description" content="“Which analytes are driving differences among my samples?” In addition to heirarchical clustering, there is another way to look at our data in a cluster context - i.e. another way to identify...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/header-attrs-2.11/header-attrs.js"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><link href="libs/_Lato-0.4.0/font.css" rel="stylesheet">
<link href="libs/_Roboto%20Mono-0.4.0/font.css" rel="stylesheet">
<link href="libs/_Montserrat-0.4.0/font.css" rel="stylesheet">
<script src="libs/bs3compat-0.3.1.9000/transition.js"></script><script src="libs/bs3compat-0.3.1.9000/tabs.js"></script><script src="libs/bs3compat-0.3.1.9000/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><link rel="icon" type="image/png" sizes="32x32" href="images/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="images/favicon-16x16.png">
<script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-99618359-1', 'auto');
      ga('send', 'pageview');

    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="style/style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h2>
        <a href="index.html" title="">Integrated Bioanalytics</a>
      </h2>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">integrated bioanalytics</a></li>
<li class="book-part">chemical instrumentation</li>
<li><a class="" href="overview.html">overview</a></li>
<li><a class="" href="separations.html">separations</a></li>
<li><a class="" href="mass-spectrometry.html">mass spectrometry</a></li>
<li><a class="" href="spectroscopy.html">spectroscopy</a></li>
<li><a class="" href="electrochemistry.html">electrochemistry</a></li>
<li class="book-part">basic data analysis in r</li>
<li><a class="" href="installation.html">installation</a></li>
<li><a class="" href="ggplot-and-markdown.html">ggplot and markdown</a></li>
<li><a class="" href="data-visualization.html">data visualization</a></li>
<li><a class="" href="ggplot-ii.html">ggplot II</a></li>
<li><a class="" href="import-and-tidying.html">import and tidying</a></li>
<li><a class="" href="the-pipe-and-summaries.html">the pipe and summaries</a></li>
<li class="book-part">advanced data analysis in r</li>
<li><a class="" href="wrangling.html">wrangling</a></li>
<li class="book-part">chemometrics</li>
<li><a class="" href="overview-1.html">overview</a></li>
<li><a class="" href="clustering.html">clustering</a></li>
<li><a class="active" href="pca.html">pca</a></li>
<li><a class="" href="k-means.html">k-means</a></li>
<li><a class="" href="models.html">models</a></li>
<li><a class="" href="comparing-means.html">comparing means</a></li>
<li class="book-part">mass spectrometric analysis</li>
<li><a class="" href="mass-spectrometric-analysis.html">mass spectrometric analysis</a></li>
<li class="book-part">transcriptome analysis</li>
<li><a class="" href="transcriptomic-analyses.html">transcriptomic analyses</a></li>
<li class="book-part">genomic analysis</li>
<li><a class="" href="genomic-analyses.html">genomic analyses</a></li>
<li class="book-part">evolutionary analysis</li>
<li><a class="" href="evolutionary-analyses.html">evolutionary analyses</a></li>
<li class="book-part">scientific writing</li>
<li><a class="" href="overview-2.html">overview</a></li>
<li><a class="" href="a-mini-manuscript.html">a mini manuscript</a></li>
<li><a class="" href="figures-captions.html"><span class="header-section-number">5</span> figures &amp; captions</a></li>
<li><a class="" href="results-and-discussion.html"><span class="header-section-number">6</span> results and discussion</a></li>
<li><a class="" href="conclusion-and-introduction.html">conclusion and introduction</a></li>
<li><a class="" href="abstract-and-title.html">abstract and title</a></li>
<li><a class="" href="proposals.html">proposals</a></li>
<li class="book-part">image analysis</li>
<li><a class="" href="section.html"></a></li>
<li><a class="" href="image-analysis.html">IMAGE ANALYSIS</a></li>
<li><a class="" href="image-color-analysis.html">image color analysis</a></li>
<li><a class="" href="images-of-mass-spectra.html"><span class="header-section-number">7</span> images of mass spectra</a></li>
<li class="book-part">appendix</li>
<li><a class="" href="links.html"><span class="header-section-number">8</span> links</a></li>
<li><a class="" href="r-faq.html"><span class="header-section-number">9</span> r faq</a></li>
<li><a class="" href="templates.html"><span class="header-section-number">10</span> templates</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="pca" class="section level1 unnumbered">
<h1>pca<a class="anchor" aria-label="anchor" href="#pca"><i class="fas fa-link"></i></a>
</h1>
<p>“Which analytes are driving differences among my samples?”</p>
<p>In addition to heirarchical clustering, there is another way to look at our data in a cluster context - i.e. another way to identify clusters of samples that have similar properties based on the analytes in the data set. This method is called k-means, which we will look at later, because for it we first need to have a look at dimensionality reduction techniques, particularly principal components analysis (PCA).</p>
<p>explain pca as drawing a new set of axes
pca also answers questions about whether variables are related
hclust is “who is most closely related to whom?” PCA is about clustering</p>
<div id="pca-1" class="section level2" number="4.3">
<h2>
<span class="header-section-number">4.3</span> pca<a class="anchor" aria-label="anchor" href="#pca-1"><i class="fas fa-link"></i></a>
</h2>
<p>PCA looks at all the variance in a high dimensional data set and chooses new axes within that data set that align with the directions containing highest variance. These new axes are called principal components. Let’s look at an example:</p>
<div class="inline-figure"><img src="https://thebustalab.github.io/integrated_bioanalytics/images/PCA.png" width="100%" style="display: block; margin: auto;"></div>
<p>In the example above, the three dimensional space can be reduced to a two dimensional space with the principal components analysis. New axes (principal components) are selected (bold arrows on left) that become the x and y axes in the principal components space (right).</p>
<p>We can run and visualize principal components analyses using the <code>runMatrixAnalysis()</code> function as in the example below:</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">AK_lakes_pca</span> <span class="op">&lt;-</span> <span class="fu">runMatrixAnalysis</span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">alaska_lake_data</span>,
  analysis <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"pca"</span><span class="op">)</span>,
  column_w_names_of_multiple_analytes <span class="op">=</span> <span class="st">"element"</span>,
  column_w_values_for_multiple_analytes <span class="op">=</span> <span class="st">"mg_per_L"</span>,
  columns_w_values_for_single_analyte <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"water_temp"</span>, <span class="st">"pH"</span><span class="op">)</span>,
  columns_w_additional_analyte_info <span class="op">=</span> <span class="st">"element_type"</span>,
  columns_w_sample_ID_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lake"</span>, <span class="st">"park"</span><span class="op">)</span>
<span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">AK_lakes_pca</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Dim.1</span>, y <span class="op">=</span> <span class="va">Dim.2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">park</span><span class="op">)</span>, shape <span class="op">=</span> <span class="fl">21</span>, size <span class="op">=</span> <span class="fl">4</span>, alpha <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_label_repel</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">lake</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-99-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Great! In this plot we can see that White Fish Lake and North Killeak Lake, both in BELA park, are quite different from the other parks (they are separated from the others along dimension 1, i.e. the first principal component). At the same time, Wild Lake, Iniakuk Lake, Walker Lake, and several other lakes in GAAR park are different from all the others (they are separated from the others along dimension 2, i.e. the second principal component).</p>
<p>Important question: what makes the lakes listed above different from the others? Certainly some aspect of their chemistry, since that’s the data that this analysis is built upon, but how do we determine which analyte(s) are driving the differences among the lakes that we see in the PCA plot?</p>
</div>
<div id="ordination-plots" class="section level2" number="4.4">
<h2>
<span class="header-section-number">4.4</span> ordination plots<a class="anchor" aria-label="anchor" href="#ordination-plots"><i class="fas fa-link"></i></a>
</h2>
<p>Let’s look at how to access the information about which analytes are major contributors to each principal component. This is important because it will tell you which analytes are associated with particular dimensions, and by extension, which analytes are associated with (and are markers for) particular groups in the PCA plot. This can be determined using an ordination plot. Let’s look at an example. We can obtain the ordination plot information using <code>runMatrixAnalysis()</code> with <code>analysis = "pca_ord"</code>:</p>
<pre><code>## # A tibble: 6 × 3
##   analyte      Dim.1   Dim.2
##   &lt;chr&gt;        &lt;dbl&gt;   &lt;dbl&gt;
## 1 water_temp 0.0769  -0.267 
## 2 pH         0.704    0.0190
## 3 C          0.297   -0.248 
## 4 N          0.00446  0.732 
## 5 P          0.485   -0.0817
## 6 Cl         0.978    0.0152</code></pre>
<p>We can now visualize the ordination plot using our standard ggplot plotting techniques. Note the use of <code>geom_label_repel()</code> and <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> to label certain segments in the ordination plot. You do not need to use <code>geom_label_repel()</code>, you could use the built in <code><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_label()</a></code>, but <code>geom_label_repel()</code> can make labelling your segments easier.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">AK_lakes_pca_ord</span> <span class="op">&lt;-</span> <span class="fu">runMatrixAnalysis</span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">alaska_lake_data</span>,
  analysis <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"pca_ord"</span><span class="op">)</span>,
  column_w_names_of_multiple_analytes <span class="op">=</span> <span class="st">"element"</span>,
  column_w_values_for_multiple_analytes <span class="op">=</span> <span class="st">"mg_per_L"</span>,
  columns_w_values_for_single_analyte <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"water_temp"</span>, <span class="st">"pH"</span><span class="op">)</span>,
  columns_w_additional_analyte_info <span class="op">=</span> <span class="st">"element_type"</span>,
  columns_w_sample_ID_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lake"</span>, <span class="st">"park"</span><span class="op">)</span>
<span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">AK_lakes_pca_ord</span><span class="op">)</span>
<span class="co">## # A tibble: 6 × 3</span>
<span class="co">##   analyte      Dim.1   Dim.2</span>
<span class="co">##   &lt;chr&gt;        &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">## 1 water_temp 0.0769  -0.267 </span>
<span class="co">## 2 pH         0.704    0.0190</span>
<span class="co">## 3 C          0.297   -0.248 </span>
<span class="co">## 4 N          0.00446  0.732 </span>
<span class="co">## 5 P          0.485   -0.0817</span>
<span class="co">## 6 Cl         0.978    0.0152</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">AK_lakes_pca_ord</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html">geom_segment</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">0</span>, y <span class="op">=</span> <span class="fl">0</span>, xend <span class="op">=</span> <span class="va">Dim.1</span>, yend <span class="op">=</span> <span class="va">Dim.2</span>, color <span class="op">=</span> <span class="va">analyte</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_circle</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x0 <span class="op">=</span> <span class="fl">0</span>, y0 <span class="op">=</span> <span class="fl">0</span>, r <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_label_repel</span><span class="op">(</span>
    data <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">AK_lakes_pca_ord</span>, <span class="va">Dim.1</span> <span class="op">&gt;</span> <span class="fl">0.9</span>, <span class="va">Dim.2</span> <span class="op">&lt;</span> <span class="fl">0.1</span>, <span class="va">Dim.2</span> <span class="op">&gt;</span> <span class="op">-</span><span class="fl">0.1</span><span class="op">)</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Dim.1</span>, y <span class="op">=</span> <span class="va">Dim.2</span>, label <span class="op">=</span> <span class="va">analyte</span><span class="op">)</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1.5</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_label_repel</span><span class="op">(</span>
    data <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">AK_lakes_pca_ord</span>, <span class="va">Dim.2</span> <span class="op">&gt;</span> <span class="fl">0.5</span><span class="op">)</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Dim.1</span>, y <span class="op">=</span> <span class="va">Dim.2</span>, label <span class="op">=</span> <span class="va">analyte</span><span class="op">)</span>, direction <span class="op">=</span> <span class="st">"y"</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1.5</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fl">1.5</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fl">1.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-101-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Great! Here is how to read the ordination plot:</p>
<ol style="list-style-type: decimal">
<li><p>When considering one analyte’s vector: the vector’s projected value on an axis shows how much its variance is aligned with that principal component.</p></li>
<li><p>When considering two analyte vectors: the angle between two vectors indicates how correlated those two variables are. If they point in the same direction, they are highly correlated. If they meet each other at 90 degrees, they are not very correlated. If they meet at ~180 degrees, they are negatively correlated. If say that one analyte is “1.9” with respect to dimension 2 and another is “-1.9” with respect to dimension 2. Let’s also say that these vectors are ~“0” with respect to dimension 1.</p></li>
</ol>
<p>With the ordination plot above, we can now see that the abundances of K, Cl, Br, and Na are the major contributors of variance to the first principal component (or the first dimension). The abundances of these elements are what make White Fish Lake and North Killeak Lake different from the other lakes. We can also see that the abundances of N, S, and Ca are the major contributors to variance in teh second dimension, whic means that these elements ar what set Wild Lake, Iniakuk Lake, Walker Lake, and several other lakes in GAAR park apart from the rest of the lakes in the data set.</p>
</div>
<div id="principal-components" class="section level2" number="4.5">
<h2>
<span class="header-section-number">4.5</span> principal components<a class="anchor" aria-label="anchor" href="#principal-components"><i class="fas fa-link"></i></a>
</h2>
<p>We also can access information about the how much of the variance in the data set is explained by each principal component, and we can plot that using ggplot:</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">AK_lakes_pca_dim</span> <span class="op">&lt;-</span> <span class="fu">runMatrixAnalysis</span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">alaska_lake_data</span>,
  analysis <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"pca_dim"</span><span class="op">)</span>,
  column_w_names_of_multiple_analytes <span class="op">=</span> <span class="st">"element"</span>,
  column_w_values_for_multiple_analytes <span class="op">=</span> <span class="st">"mg_per_L"</span>,
  columns_w_values_for_single_analyte <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"water_temp"</span>, <span class="st">"pH"</span><span class="op">)</span>,
  columns_w_additional_analyte_info <span class="op">=</span> <span class="st">"element_type"</span>,
  columns_w_sample_ID_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lake"</span>, <span class="st">"park"</span><span class="op">)</span>
<span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">AK_lakes_pca_dim</span><span class="op">)</span>
<span class="co">## # A tibble: 6 × 2</span>
<span class="co">##   principal_component percent_variance_explained</span>
<span class="co">##                 &lt;dbl&gt;                      &lt;dbl&gt;</span>
<span class="co">## 1                   1                      48.8 </span>
<span class="co">## 2                   2                      18.6 </span>
<span class="co">## 3                   3                      11.6 </span>
<span class="co">## 4                   4                       7.88</span>
<span class="co">## 5                   5                       4.68</span>
<span class="co">## 6                   6                       3.33</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">AK_lakes_pca_dim</span>, 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">principal_component</span>, y <span class="op">=</span> <span class="va">percent_variance_explained</span><span class="op">)</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-102-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Cool! We can see that the first principal component retains nearly 50% of the variance in the original dataset, while the second dimension contains only about 20%.</p>
</div>
<div id="exercises-4" class="section level2" number="4.6">
<h2>
<span class="header-section-number">4.6</span> exercises<a class="anchor" aria-label="anchor" href="#exercises-4"><i class="fas fa-link"></i></a>
</h2>
<p>In this set of exercises you will choose to complete ONE of the options below. For either option please refer to Chapter 12 for help with principal component and ordination plots. Also, when you are filling out the <code>runMatrixAnalysis()</code> template, you can use the <code><a href="https://rdrr.io/r/base/colnames.html">colnames()</a></code> function to help you specify a long list of column names rather than typing them out by hand. For example, in the periodic table data set, we can refer to a set of columns (columns 10 through 20) with the following command:</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">periodic_table_subset</span><span class="op">)</span><span class="op">[</span><span class="fl">10</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span>
<span class="co">##  [1] "electronegativity_pauling"         </span>
<span class="co">##  [2] "first_ionization_poten_eV"         </span>
<span class="co">##  [3] "second_ionization_poten_eV"        </span>
<span class="co">##  [4] "third_ionization_poten_eV"         </span>
<span class="co">##  [5] "electron_affinity_eV"              </span>
<span class="co">##  [6] "atomic_radius_ang"                 </span>
<span class="co">##  [7] "ionic_radius_ang"                  </span>
<span class="co">##  [8] "covalent_radius_ang"               </span>
<span class="co">##  [9] "atomic_volume_cm3_per_mol"         </span>
<span class="co">## [10] "electrical_conductivity_mho_per_cm"</span>
<span class="co">## [11] "specific_heat_J_per_g_K"</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">periodic_table_subset</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">18</span><span class="op">:</span><span class="fl">20</span>, <span class="fl">23</span><span class="op">:</span><span class="fl">25</span><span class="op">)</span><span class="op">]</span>
<span class="co">## [1] "atomic_volume_cm3_per_mol"         </span>
<span class="co">## [2] "electrical_conductivity_mho_per_cm"</span>
<span class="co">## [3] "specific_heat_J_per_g_K"           </span>
<span class="co">## [4] "thermal_conductivity_W_per_m_K"    </span>
<span class="co">## [5] "polarizability_A_cubed"            </span>
<span class="co">## [6] "heat_atomization_kJ_per_mol"</span></code></pre></div>
<p>We can use that command in the template, as in the example below. With the notation <code>colnames(periodic_table_subset)[c(5:7,9:25)]</code>, we can mark columns 5 - 7 and 9 - 25 as columns_w_values_for_single_analyte (note what happens when you run <code>c(5:7,9:25)</code> in the console, and what happens when you run <code>colnames(periodic_table_subset)[c(5:7,9:25)]</code> in the console). With the notation <code>colnames(periodic_table_subset)[c(1:4, 8)]</code> we can mark columns 1 - 4 and column 8 as columns_w_sample_ID_info (note what happens when you run <code>c(1:4, 8)</code> in the console, and what happens when you run <code>colnames(periodic_table_subset)[c(1:4, 8)]</code> in the console).</p>
<div id="option-1-human-metabolomics." class="section level3" number="4.6.1">
<h3>
<span class="header-section-number">4.6.1</span> option 1: human metabolomics.<a class="anchor" aria-label="anchor" href="#option-1-human-metabolomics."><i class="fas fa-link"></i></a>
</h3>
<p>This first option is to work with a dataset describing metabolomics data (i.e. abundances of &gt; 100 different biochemicals) from each of 93 human patients, some of which have Chronic Kidney Disease. If you choose this option, your task is to discover a biomarker for Chronic Kidney Disease. This means that you will need to determine a metabolite whose abundance is strongly associated with the disease. To do this you should complete the following:</p>
<ol style="list-style-type: decimal">
<li>Run a PCA analysis on <code>metabolomics_data</code> (i.e. <code>runMatrixAnalysis()</code> with <code>analysis = "pca"</code>)</li>
<li>Plot the results of the analysis to determine which principal component (i.e. dimension) separates the healthy and kidney_disease samples.</li>
<li>Obtain the ordination plot coordinates for the analytes in the PCA analysis (i.e. <code>runMatrixAnalysis()</code> with <code>analysis = "pca_ord"</code>).</li>
<li>Visualize the ordination plot and determine which of the analytes are strongly associated with the principal component (i.e. dimension) separates the healthy and kidney_disease samples.</li>
<li>Bingo! These analytes are associated with Chronic Kidney Disease and could be biomarkers for such.</li>
</ol>
</div>
<div id="option-2-grape-vine-chemistry" class="section level3" number="4.6.2">
<h3>
<span class="header-section-number">4.6.2</span> option 2: grape vine chemistry<a class="anchor" aria-label="anchor" href="#option-2-grape-vine-chemistry"><i class="fas fa-link"></i></a>
</h3>
<p>This second option is to work with a dataset describing metabolomics data (i.e. abundances of &gt; 100 different biochemicals) from 5 different wine grape varieties. If you choose this option, your task is to discover a biomarker for Chardonnay and a biomarker for Cabernet Sauvignon. This means that you will need to identify two metabolites, each of which are associated with one of those two grape varieties. To do this you should complete the following:</p>
<ol style="list-style-type: decimal">
<li>Run a PCA analysis on <code>wine_grape_data</code> (i.e. <code>runMatrixAnalysis()</code> with <code>analysis = "pca"</code>)</li>
<li>Plot the results of the analysis to determine which principal component (i.e. dimension) separates the Chardonnay samples from the other varieties and the Cabernet Sauvignon samples from the other varieties.</li>
<li>Obtain the ordination plot coordinates for the analytes in the PCA analysis (i.e. <code>runMatrixAnalysis()</code> with <code>analysis = "pca_ord"</code>).</li>
<li>Visualize the ordination plot and determine which of the analytes are strongly associated with the principal component (i.e. dimension) separates the Chardonnay samples from the other varieties and the Cabernet Sauvignon samples from the other varieties.</li>
<li>Bingo! These analytes are associated with those varieites and could be biomarkers for such.</li>
</ol>
<!-- end --><!-- start k-means -->
</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="clustering.html">clustering</a></div>
<div class="next"><a href="k-means.html">k-means</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <h2>Note: Second Edition is under construction 🏗</h2>
    <p>Now is a great time to test in-development packages and provide feedback</p>
        <ul class="list-unstyled">
<li><a href="https://forms.gle/nq9RmbxJyZXQgc948">Provide feedback (5 min)</a></li>
          <li><a href="https://geocompr.robinlovelace.net/#reproducibility">Install updated packages</a></li>
          <li><a href="https://github.com/Robinlovelace/geocompr/issues">Open an issue <i class="fas fa-question"></i></a></li>
          <li><a href="https://discord.gg/Te3gWeDwmf">Chat on Discord <i class="fab fa-discord"></i></a></li>
        </ul>
<hr>
<nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#pca">pca</a></li>
<li><a class="nav-link" href="#pca-1"><span class="header-section-number">4.3</span> pca</a></li>
<li><a class="nav-link" href="#ordination-plots"><span class="header-section-number">4.4</span> ordination plots</a></li>
<li><a class="nav-link" href="#principal-components"><span class="header-section-number">4.5</span> principal components</a></li>
<li>
<a class="nav-link" href="#exercises-4"><span class="header-section-number">4.6</span> exercises</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#option-1-human-metabolomics."><span class="header-section-number">4.6.1</span> option 1: human metabolomics.</a></li>
<li><a class="nav-link" href="#option-2-grape-vine-chemistry"><span class="header-section-number">4.6.2</span> option 2: grape vine chemistry</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/blob/main/index.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/edit/main/index.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Integrated Bioanalytics</strong>" was written by Lucas Busta. It was last built on 2022-01-07.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
