<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>phylogenies | Integrated Bioanalytics</title>
<meta name="author" content="Lucas Busta and members of the Busta lab">
<meta name="description" content="buildTree This function is a swiss army knife for tree building. It takes as input alignments or existing phylogenies from which to derive a phylogeny of interest, it can use neighbor-joining or...">
<meta name="generator" content="bookdown 0.44 with bs4_book()">
<meta property="og:title" content="phylogenies | Integrated Bioanalytics">
<meta property="og:type" content="book">
<meta property="og:description" content="buildTree This function is a swiss army knife for tree building. It takes as input alignments or existing phylogenies from which to derive a phylogeny of interest, it can use neighbor-joining or...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="phylogenies | Integrated Bioanalytics">
<meta name="twitter:description" content="buildTree This function is a swiss army knife for tree building. It takes as input alignments or existing phylogenies from which to derive a phylogeny of interest, it can use neighbor-joining or...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><link href="libs/Lato-0.4.10/font.css" rel="stylesheet">
<link href="libs/Roboto_Mono-0.4.10/font.css" rel="stylesheet">
<link href="libs/Montserrat-0.4.10/font.css" rel="stylesheet">
<script src="libs/bs3compat-0.9.0/transition.js"></script><script src="libs/bs3compat-0.9.0/tabs.js"></script><script src="libs/bs3compat-0.9.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><link rel="icon" type="image/png" sizes="32x32" href="images/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="images/favicon-16x16.png">
<script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-99618359-1', 'auto');
      ga('send', 'pageview');

    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="style/style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h2>
        <a href="index.html" title="">Integrated Bioanalytics</a>
      </h2>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">WELCOME</a></li>
<li class="book-part">GETTING STARTED</li>
<li><a class="" href="overview.html">overview</a></li>
<li><a class="" href="installation.html">installation</a></li>
<li class="book-part">DATA VISUALIZATION</li>
<li><a class="" href="data-visualization-i.html">data visualization I</a></li>
<li><a class="" href="data-visualization-ii.html">data visualization II</a></li>
<li><a class="" href="data-visualization-iii.html">data visualization III</a></li>
<li class="book-part">STATISTICAL METHODS</li>
<li><a class="" href="wrangling-and-summaries.html">wrangling and summaries</a></li>
<li><a class="" href="hierarchical-clustering.html">hierarchical clustering</a></li>
<li><a class="" href="dimensional-reduction.html">dimensional reduction</a></li>
<li><a class="" href="flat-clustering.html">flat clustering</a></li>
<li><a class="" href="comparing-means.html">comparing means</a></li>
<li class="book-part">MODELS</li>
<li><a class="" href="numerical-models.html">numerical models</a></li>
<li><a class="" href="embedding-models.html">embedding models</a></li>
<li class="book-part">GC-MS DATA</li>
<li><a class="" href="loading-analyzegcmsdata.html">loading analyzeGCMSdata</a></li>
<li><a class="" href="using-analyzegcmsdata.html">using analyzeGCMSdata</a></li>
<li><a class="" href="cdf-export.html">CDF export</a></li>
<li class="book-part">SEQUENCE ANALYSIS</li>
<li><a class="" href="homology.html">homology</a></li>
<li><a class="" href="alignments.html">alignments</a></li>
<li><a class="active" href="phylogenies.html">phylogenies</a></li>
<li><a class="" href="phylogenetic-analyses.html">phylogenetic analyses</a></li>
<li class="book-part">ANALYTICAL REPORTS</li>
<li><a class="" href="figures-captions.html">figures &amp; captions</a></li>
<li><a class="" href="question-driven-report.html">question-driven report</a></li>
<li class="book-part">APPENDIX</li>
<li><a class="" href="links.html">links</a></li>
<li><a class="" href="datasets.html">datasets</a></li>
<li><a class="" href="r-faq.html">r faq</a></li>
<li><a class="" href="templates.html">templates</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="phylogenies" class="section level1 unnumbered">
<h1>phylogenies<a class="anchor" aria-label="anchor" href="#phylogenies"><i class="fas fa-link"></i></a>
</h1>
<div id="buildtree" class="section level2 unnumbered">
<h2>buildTree<a class="anchor" aria-label="anchor" href="#buildtree"><i class="fas fa-link"></i></a>
</h2>
<p>This function is a swiss army knife for tree building. It takes as input alignments or existing phylogenies from which to derive a phylogeny of interest, it can use neighbor-joining or maximum liklihood methods (with model optimization), it can run bootstrap replicates, and it can calculate ancestral sequence states. To illustrate, let‚Äôs look at some examples:</p>
<div id="newick-input" class="section level3 unnumbered">
<h3>newick input<a class="anchor" aria-label="anchor" href="#newick-input"><i class="fas fa-link"></i></a>
</h3>
<p>Let‚Äôs use the Busta lab‚Äôs plant phylogeny [derived from Qian et al., 2016] to build a phylogeny with five species in it.</p>
<div class="sourceCode" id="cb162"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">buildTree</span><span class="op">(</span></span>
<span>  scaffold_type <span class="op">=</span> <span class="st">"newick"</span>,</span>
<span>  scaffold <span class="op">=</span> <span class="st">"https://thebustalab.github.io/data/plant_phylogeny.newick"</span>,</span>
<span>  members <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sorghum_bicolor"</span>, <span class="st">"Zea_mays"</span>, <span class="st">"Setaria_viridis"</span>, <span class="st">"Arabidopsis_thaliana"</span>, <span class="st">"Amborella_trichopoda"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Pro tip: most tree read/write functions reset node numbers. Fortify your tree and save it as a csv file to preserve node numbering.</span></span>
<span></span>
<span><span class="va">tree</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Phylogenetic tree with 5 tips and 4 internal nodes.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Tip labels:</span></span>
<span><span class="co">##   Amborella_trichopoda, Zea_mays, Sorghum_bicolor, Setaria_viridis, Arabidopsis_thaliana</span></span>
<span><span class="co">## Node labels:</span></span>
<span><span class="co">##   , , , </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Rooted; includes branch length(s).</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-526-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Cool! We got our phylogeny. What happens if we want to build a phylogeny that has a species on it that isn‚Äôt in our scaffold? For example, what if we want to build a phylogeny that includes <em>Arabidopsis neglecta</em>? We can include that name in our list of members:</p>
<div class="sourceCode" id="cb163"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">buildTree</span><span class="op">(</span></span>
<span>  scaffold_type <span class="op">=</span> <span class="st">"newick"</span>,</span>
<span>  scaffold_in_path <span class="op">=</span> <span class="st">"https://thebustalab.github.io/data/plant_phylogeny.newick"</span>,</span>
<span>  members <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sorghum_bicolor"</span>, <span class="st">"Zea_mays"</span>, <span class="st">"Setaria_viridis"</span>, <span class="st">"Arabidopsis_neglecta"</span>, <span class="st">"Amborella_trichopoda"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## IMPORTANT: Some species substitutions or removals were made as part of buildTree. Run build_tree_substitutions() to see them all.</span></span>
<span><span class="co">## Pro tip: most tree read/write functions reset node numbers. Fortify your tree and save it as a csv file to preserve node numbering.</span></span>
<span></span>
<span><span class="va">tree</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Phylogenetic tree with 5 tips and 4 internal nodes.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Tip labels:</span></span>
<span><span class="co">##   Amborella_trichopoda, Zea_mays, Sorghum_bicolor, Setaria_viridis, Arabidopsis_neglecta</span></span>
<span><span class="co">## Node labels:</span></span>
<span><span class="co">##   , , , </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Rooted; includes branch length(s).</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-527-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Note that <code>buildTree</code> informs us: ‚ÄúScaffold newick tip Arabidopsis_thaliana substituted with Arabidopsis_neglecta‚Äù. This means that <em>Arabidopsis neglecta</em> was grafted onto the tip originally occupied by <em>Arabidopsis thaliana</em>. This behaviour is useful when operating on a large phylogenetic scale (i.e.¬†where <em>exact</em> phylogeny topology is not critical below the family level). However, if a person is interested in using an existing newick tree as a scaffold for a phylogeny where genus-level topology <em>is</em> critical, then beware! Your scaffold may not be appropriate if you see that message. When operating at the genus level, you probably want to use sequence data to build your phylogeny anyway. So let‚Äôs look at how to do that:</p>
</div>
<div id="alignment-input" class="section level3 unnumbered">
<h3>alignment input<a class="anchor" aria-label="anchor" href="#alignment-input"><i class="fas fa-link"></i></a>
</h3>
<p>Arguments in this case are:</p>
<ul>
<li>‚Äúscaffold_type‚Äù: ‚Äúamin_alignment‚Äù or ‚Äúnucl_alignment‚Äù for amino acids or nucleotides.</li>
<li>‚Äúscaffold_in_path‚Äù: path to the fasta file that contains the alignment from which you want to build a tree.</li>
<li>‚Äúml‚Äù: Logical, TRUE if you want to use maximum liklihood, FALSE if not, in which case neighbor joining will ne used.</li>
<li>‚Äúmodel_test‚Äù: if you say TRUE to ‚Äúml‚Äù, should buildTree test different maximum liklihood models and then use the ‚Äúbest‚Äù one?</li>
<li>‚Äúbootstrap‚Äù: TRUE or FALSE, whether you want bootstrap values on the nodes.</li>
<li>‚Äúancestral_states‚Äù: TRUE or FALSE, should buildTree() compute the ancestral sequence at each node?</li>
<li>‚Äúroot‚Äù: NULL, or the name of an accession that should form the root of the tree.</li>
</ul>
<div class="sourceCode" id="cb164"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">buildTree</span><span class="op">(</span></span>
<span>  scaffold_type <span class="op">=</span> <span class="st">"amin_alignment"</span>,</span>
<span>  scaffold_in_path <span class="op">=</span> <span class="st">"/path_to/a_folder_for_alignments/all_amin_seqs.fa"</span>,</span>
<span>  ml <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  model_test <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  bootstrap <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ancestral_states <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  root <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div id="plotting-trees" class="section level2 unnumbered">
<h2>plotting trees<a class="anchor" aria-label="anchor" href="#plotting-trees"><i class="fas fa-link"></i></a>
</h2>
<p>There are several approaches to plotting trees. A simple one is using the base <code>plot</code> function:</p>
<div class="sourceCode" id="cb165"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test_tree_small</span> <span class="op">&lt;-</span> <span class="fu">buildTree</span><span class="op">(</span></span>
<span>  scaffold_type <span class="op">=</span> <span class="st">"newick"</span>,</span>
<span>  scaffold_in_path <span class="op">=</span> <span class="st">"https://thebustalab.github.io/data/plant_phylogeny.newick"</span>,</span>
<span>  members <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sorghum_bicolor"</span>, <span class="st">"Zea_mays"</span>, <span class="st">"Setaria_viridis"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Pro tip: most tree read/write functions reset node numbers. Fortify your tree and save it as a csv file to preserve node numbering.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">test_tree_small</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-529-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Though this can get messy when there are lots of tip labels:</p>
<div class="sourceCode" id="cb166"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">122</span><span class="op">)</span></span>
<span><span class="va">test_tree_big</span> <span class="op">&lt;-</span> <span class="fu">buildTree</span><span class="op">(</span></span>
<span>  scaffold_type <span class="op">=</span> <span class="st">"newick"</span>,</span>
<span>  scaffold_in_path <span class="op">=</span> <span class="st">"https://thebustalab.github.io/data/plant_phylogeny.newick"</span>,</span>
<span>  members <span class="op">=</span> <span class="va">plant_species</span><span class="op">$</span><span class="va">Genus_species</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html">floor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">60</span><span class="op">)</span><span class="op">*</span><span class="fl">100000</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">test_tree_big</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-530-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>One solution is to use <code>ggtree</code>, which by default doesn‚Äôt show tip labels. <code>plot</code> can do that too, but <code>ggtree</code> does a bunch of other useful things, so I recommend that:</p>
<div class="sourceCode" id="cb167"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">test_tree_big</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-531-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Another convenient fucntion is ggplot‚Äôs <code>fortify</code>. This will convert your <code>phylo</code> object into a data frame:</p>
<div class="sourceCode" id="cb168"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test_tree_big_fortified</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/fortify.html">fortify</a></span><span class="op">(</span><span class="va">test_tree_big</span><span class="op">)</span></span>
<span><span class="va">test_tree_big_fortified</span></span>
<span><span class="co">## # A tbl_tree abstraction: 101 √ó 9</span></span>
<span><span class="co">## # which can be converted to treedata or phylo </span></span>
<span><span class="co">## # via as.treedata or as.phylo</span></span>
<span><span class="co">##    parent  node branch.length label isTip     x     y branch</span></span>
<span><span class="co">##     &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt; &lt;lgl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;</span></span>
<span><span class="co">##  1     54     1          83.0 Wolf‚Ä¶ TRUE   188.     1   147.</span></span>
<span><span class="co">##  2     54     2          83.0 Spat‚Ä¶ TRUE   188.     2   147.</span></span>
<span><span class="co">##  3     55     3         138.  Dios‚Ä¶ TRUE   188.     3   120.</span></span>
<span><span class="co">##  4     58     4          42.7 Bulb‚Ä¶ TRUE   188.     5   167.</span></span>
<span><span class="co">##  5     58     5          42.7 Ober‚Ä¶ TRUE   188.     6   167.</span></span>
<span><span class="co">##  6     59     6          32.0 Poma‚Ä¶ TRUE   188.     7   172.</span></span>
<span><span class="co">##  7     59     7          32.0 Teli‚Ä¶ TRUE   188.     8   172.</span></span>
<span><span class="co">##  8     56     8         135.  Cala‚Ä¶ TRUE   188.     4   121.</span></span>
<span><span class="co">##  9     61     9         147.  Pepe‚Ä¶ TRUE   188.     9   115.</span></span>
<span><span class="co">## 10     62    10         121.  Endl‚Ä¶ TRUE   188.    10   128.</span></span>
<span><span class="co">## # ‚Ñπ 91 more rows</span></span>
<span><span class="co">## # ‚Ñπ 1 more variable: angle &lt;dbl&gt;</span></span></code></pre></div>
<p><code>ggtree</code> can still plot this dataframe, and it allows metadata to be stored in a human readable format by using mutating joins (explained below). This metadata can be plotted with standard ggplot geoms, and these dataframes can also conveniently be saved as .csv files:</p>
<div class="sourceCode" id="cb169"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## Note that "plant_species" comes with the phylochemistry source.</span></span>
<span></span>
<span><span class="va">test_tree_big_fortified_w_data</span> <span class="op">&lt;-</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">test_tree_big_fortified</span>, <span class="va">plant_species</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"label"</span> <span class="op">=</span> <span class="st">"Genus_species"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">test_tree_big_fortified_w_data</span></span>
<span><span class="co">## # A tbl_tree abstraction: 101 √ó 14</span></span>
<span><span class="co">## # which can be converted to treedata or phylo </span></span>
<span><span class="co">## # via as.treedata or as.phylo</span></span>
<span><span class="co">##    parent  node branch.length label isTip     x     y branch</span></span>
<span><span class="co">##     &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt; &lt;lgl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;</span></span>
<span><span class="co">##  1     54     1          83.0 Wolf‚Ä¶ TRUE   188.     1   147.</span></span>
<span><span class="co">##  2     54     2          83.0 Spat‚Ä¶ TRUE   188.     2   147.</span></span>
<span><span class="co">##  3     55     3         138.  Dios‚Ä¶ TRUE   188.     3   120.</span></span>
<span><span class="co">##  4     58     4          42.7 Bulb‚Ä¶ TRUE   188.     5   167.</span></span>
<span><span class="co">##  5     58     5          42.7 Ober‚Ä¶ TRUE   188.     6   167.</span></span>
<span><span class="co">##  6     59     6          32.0 Poma‚Ä¶ TRUE   188.     7   172.</span></span>
<span><span class="co">##  7     59     7          32.0 Teli‚Ä¶ TRUE   188.     8   172.</span></span>
<span><span class="co">##  8     56     8         135.  Cala‚Ä¶ TRUE   188.     4   121.</span></span>
<span><span class="co">##  9     61     9         147.  Pepe‚Ä¶ TRUE   188.     9   115.</span></span>
<span><span class="co">## 10     62    10         121.  Endl‚Ä¶ TRUE   188.    10   128.</span></span>
<span><span class="co">## # ‚Ñπ 91 more rows</span></span>
<span><span class="co">## # ‚Ñπ 6 more variables: angle &lt;dbl&gt;, Phylum &lt;chr&gt;,</span></span>
<span><span class="co">## #   Order &lt;chr&gt;, Family &lt;chr&gt;, Genus &lt;chr&gt;, species &lt;chr&gt;</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">test_tree_big_fortified_w_data</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">test_tree_big_fortified_w_data</span>, <span class="va">isTip</span> <span class="op">==</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, fill <span class="op">=</span> <span class="va">Order</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">3</span>, shape <span class="op">=</span> <span class="fl">21</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_text</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">test_tree_big_fortified_w_data</span>, <span class="va">isTip</span> <span class="op">==</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, label <span class="op">=</span> <span class="va">y</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">label</span><span class="op">)</span>, offset <span class="op">=</span> <span class="fl">10</span>, size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_void</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">discrete_palette</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">280</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span></span>
<span>    legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.15</span>, <span class="fl">0.75</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-533-1.png" width="100%" style="display: block; margin: auto;"></div>
</div>
<div id="collapsetree" class="section level2 unnumbered">
<h2>collapseTree<a class="anchor" aria-label="anchor" href="#collapsetree"><i class="fas fa-link"></i></a>
</h2>
<p>Sometimes we want to view a tree at a higher level of taxonomical organization, or some other higher level. This can be done easily using the <code>collapseTree</code> function. It takes two arguments: an un-fortified tree (<code>tree</code>), and a two-column data frame (<code>associations</code>). In the first column of the data frame are all the tip labels of the tree, and in the second column are the higher level of organization to which each tip belongs. The function will prune the tree so that only one member of the higher level of organization is included in the output. For example, let‚Äôs look at the tree from the previous section at the family level:</p>
<div class="sourceCode" id="cb170"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">collapseTree</span><span class="op">(</span></span>
<span>  tree <span class="op">=</span> <span class="va">test_tree_big</span>,</span>
<span>  associations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>    tip.label <span class="op">=</span> <span class="va">test_tree_big</span><span class="op">$</span><span class="va">tip.label</span>,</span>
<span>    family <span class="op">=</span> <span class="va">plant_species</span><span class="op">$</span><span class="va">Family</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="va">test_tree_big</span><span class="op">$</span><span class="va">tip.label</span>, <span class="va">plant_species</span><span class="op">$</span><span class="va">Genus_species</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">test_tree_big_families</span></span>
<span><span class="co">## Branch lengths have been set to one.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">test_tree_big_families</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">300</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-534-1.png" width="100%" style="display: block; margin: auto;"></div>
</div>
<div id="trees-and-traits" class="section level2 unnumbered">
<h2>trees and traits<a class="anchor" aria-label="anchor" href="#trees-and-traits"><i class="fas fa-link"></i></a>
</h2>
<p>To plot traits alongside a tree, we can use ggtree in combination with ggplot. Here is an example. First, we make the tree:</p>
<div class="sourceCode" id="cb171"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chemical_bloom_tree</span> <span class="op">&lt;-</span> <span class="fu">buildTree</span><span class="op">(</span></span>
<span>  scaffold_type <span class="op">=</span> <span class="st">"newick"</span>,</span>
<span>  scaffold_in_path <span class="op">=</span> <span class="st">"http://thebustalab.github.io/data/angiosperms.newick"</span>,</span>
<span>  members <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">chemical_blooms</span><span class="op">$</span><span class="va">label</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## IMPORTANT: Some species substitutions or removals were made as part of buildTree. Run build_tree_substitutions() to see them all.</span></span>
<span><span class="co">## Pro tip: most tree read/write functions reset node numbers. Fortify your tree and save it as a csv file to preserve node numbering.</span></span></code></pre></div>
<p>Next we join the tree with the data:</p>
<div class="sourceCode" id="cb172"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">left_join</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/fortify.html">fortify</a></span><span class="op">(</span><span class="va">chemical_bloom_tree</span><span class="op">)</span>, <span class="va">chemical_blooms</span><span class="op">)</span></span>
<span><span class="co">## Joining with `by = join_by(label)`</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 6 √ó 18</span></span>
<span><span class="co">##   parent  node branch.length label  isTip     x     y branch</span></span>
<span><span class="co">##    &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt;  &lt;lgl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;</span></span>
<span><span class="co">## 1     80     1         290.  Ginkg‚Ä¶ TRUE   352.     1   207.</span></span>
<span><span class="co">## 2     81     2         267.  Picea‚Ä¶ TRUE   352.     2   219.</span></span>
<span><span class="co">## 3     81     3         267.  Cupre‚Ä¶ TRUE   352.     3   219.</span></span>
<span><span class="co">## 4     84     4         135.  Eryth‚Ä¶ TRUE   352.     5   285.</span></span>
<span><span class="co">## 5     86     5          16.2 Iris_‚Ä¶ TRUE   352.     6   344.</span></span>
<span><span class="co">## 6     86     6          16.2 Iris_‚Ä¶ TRUE   352.     7   344.</span></span>
<span><span class="co">## # ‚Ñπ 10 more variables: angle &lt;dbl&gt;, Alkanes &lt;dbl&gt;,</span></span>
<span><span class="co">## #   Sec_Alcohols &lt;dbl&gt;, Others &lt;dbl&gt;, Fatty_acids &lt;dbl&gt;,</span></span>
<span><span class="co">## #   Alcohols &lt;dbl&gt;, Triterpenoids &lt;dbl&gt;, Ketones &lt;dbl&gt;,</span></span>
<span><span class="co">## #   Other_compounds &lt;dbl&gt;, Aldehydes &lt;dbl&gt;</span></span></code></pre></div>
<p>Now we can plot the tree:</p>
<div class="sourceCode" id="cb173"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span></span>
<span>    align <span class="op">=</span> <span class="cn">TRUE</span>, hjust <span class="op">=</span> <span class="fl">1</span>, offset <span class="op">=</span> <span class="fl">350</span>,</span>
<span>    geom <span class="op">=</span> <span class="st">"label"</span>, label.size <span class="op">=</span> <span class="fl">0</span>, size <span class="op">=</span> <span class="fl">3</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">750</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>IMPORTANT! When we plot the traits, we need to reorder whatever is on the shared axis (in this case, the y axis) so that it matches the order of the tree. In this case, we need to reorder the species names so that they match the order of the tree. We can do this by using the <code>reorder</code> function, which takes two arguments: the thing to be reordered, and the thing to be reordered by. In this case, we want to reorder the species names by their y coordinate on the tree. We can do this by using the <code>y</code> column of the data frame that we created when we fortified the tree. We can then plot the traits:</p>
<div class="sourceCode" id="cb174"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trait_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="fu">pivot_longer</span><span class="op">(</span></span>
<span>      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">data</span>, <span class="va">isTip</span> <span class="op">==</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>      cols <span class="op">=</span> <span class="fl">10</span><span class="op">:</span><span class="fl">18</span>, names_to <span class="op">=</span> <span class="st">"compound"</span>, values_to <span class="op">=</span> <span class="st">"abundance"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">compound</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html">reorder</a></span><span class="op">(</span><span class="va">label</span>, <span class="va">y</span><span class="op">)</span>, size <span class="op">=</span> <span class="va">abundance</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html">scale_y_discrete</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span></span>
<span>    plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>Finally, we can plot the two plots together using <code>plot_grid</code>. It is important to manually inspect the tree tips and the y axis text to make sure that everything lines up. We don‚Äôt want to be plotting the abundance of one species on the y axis of another species. In this case, everything looks good:</p>
<div class="sourceCode" id="cb175"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_grid</span><span class="op">(</span></span>
<span>  <span class="va">tree_plot</span>,</span>
<span>  <span class="va">trait_plot</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">1</span>, align <span class="op">=</span> <span class="st">"h"</span>, axis <span class="op">=</span> <span class="st">"tb"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-539-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Once our manual inspection is complete, we can make a new version of the plot in which the y axis text is removed from the trait plot and we can reduce the margin on the left side of the trait plot to make it look nicer:</p>
<div class="sourceCode" id="cb176"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span></span>
<span>    align <span class="op">=</span> <span class="cn">TRUE</span>, hjust <span class="op">=</span> <span class="fl">1</span>, offset <span class="op">=</span> <span class="fl">350</span>,</span>
<span>    geom <span class="op">=</span> <span class="st">"label"</span>, label.size <span class="op">=</span> <span class="fl">0</span>, size <span class="op">=</span> <span class="fl">3</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">750</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">trait_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="fu">pivot_longer</span><span class="op">(</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">data</span>, <span class="va">isTip</span> <span class="op">==</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>      cols <span class="op">=</span> <span class="fl">10</span><span class="op">:</span><span class="fl">18</span>, names_to <span class="op">=</span> <span class="st">"compound"</span>, values_to <span class="op">=</span> <span class="st">"abundance"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">compound</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html">reorder</a></span><span class="op">(</span><span class="va">label</span>, <span class="va">y</span><span class="op">)</span>, size <span class="op">=</span> <span class="va">abundance</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html">scale_y_discrete</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span></span>
<span>    axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="op">-</span><span class="fl">1.5</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="fu">plot_grid</span><span class="op">(</span></span>
<span>  <span class="va">tree_plot</span>,</span>
<span>  <span class="va">trait_plot</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">1</span>, align <span class="op">=</span> <span class="st">"h"</span>, axis <span class="op">=</span> <span class="st">"tb"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-540-1.png" width="100%" style="display: block; margin: auto;"></div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="alignments.html">alignments</a></div>
<div class="next"><a href="phylogenetic-analyses.html">phylogenetic analyses</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <h2>Note: Second Edition is under construction üèó</h2>
    <p>Now is a great time to provide feedback</p>
        <ul class="list-unstyled">
<li><a href="https://forms.gle/SZmB2Ct2exE2dBwv9">Provide feedback (5 min)</a></li>
          <!-- <li><a href="https://geocompr.robinlovelace.net/#reproducibility">Install updated packages</a></li> -->
          <!-- <li><a href="https://github.com/Robinlovelace/geocompr/issues">Open an issue <i class="fas fa-question"></i></a></li> -->
          <!-- <li><a href="https://discord.gg/Te3gWeDwmf">Chat on Discord <i class="fab fa-discord"></i></a></li> -->
        </ul>
<hr>
<nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#phylogenies">phylogenies</a></li>
<li>
<a class="nav-link" href="#buildtree">buildTree</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#newick-input">newick input</a></li>
<li><a class="nav-link" href="#alignment-input">alignment input</a></li>
</ul>
</li>
<li><a class="nav-link" href="#plotting-trees">plotting trees</a></li>
<li><a class="nav-link" href="#collapsetree">collapseTree</a></li>
<li><a class="nav-link" href="#trees-and-traits">trees and traits</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/blob/main/index.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/edit/main/index.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Integrated Bioanalytics</strong>" was written by Lucas Busta and members of the Busta lab. It was last built on 2025-09-26.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
