<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>phylogenies | Integrated Bioanalytics</title>
<meta name="author" content="Lucas Busta and members of the Busta lab">
<meta name="description" content="1.37 buildTree This function is a swiss army knife for tree building. It takes as input alignments or existing phylogenies from which to derive a phylogeny of interest, it can use neighbor-joining...">
<meta name="generator" content="bookdown 0.25 with bs4_book()">
<meta property="og:title" content="phylogenies | Integrated Bioanalytics">
<meta property="og:type" content="book">
<meta property="og:description" content="1.37 buildTree This function is a swiss army knife for tree building. It takes as input alignments or existing phylogenies from which to derive a phylogeny of interest, it can use neighbor-joining...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="phylogenies | Integrated Bioanalytics">
<meta name="twitter:description" content="1.37 buildTree This function is a swiss army knife for tree building. It takes as input alignments or existing phylogenies from which to derive a phylogeny of interest, it can use neighbor-joining...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><link href="libs/Lato-0.4.1/font.css" rel="stylesheet">
<link href="libs/Roboto%20Mono-0.4.1/font.css" rel="stylesheet">
<link href="libs/Montserrat-0.4.1/font.css" rel="stylesheet">
<script src="libs/bs3compat-0.3.1.9000/transition.js"></script><script src="libs/bs3compat-0.3.1.9000/tabs.js"></script><script src="libs/bs3compat-0.3.1.9000/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><link rel="icon" type="image/png" sizes="32x32" href="images/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="images/favicon-16x16.png">
<script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-99618359-1', 'auto');
      ga('send', 'pageview');

    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="style/style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h2>
        <a href="index.html" title="">Integrated Bioanalytics</a>
      </h2>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">WELCOME</a></li>
<li class="book-part">DATA ANALYSIS IN R</li>
<li><a class="" href="overview.html">overview</a></li>
<li><a class="" href="installation.html">installation</a></li>
<li><a class="" href="data-visualization.html">data visualization</a></li>
<li><a class="" href="data-wrangling.html">data wrangling</a></li>
<li><a class="" href="clustering.html">clustering</a></li>
<li><a class="" href="pca.html">pca</a></li>
<li><a class="" href="k-means.html">k-means</a></li>
<li><a class="" href="models.html">models</a></li>
<li><a class="" href="comparing-means.html">comparing means</a></li>
<li class="book-part">GC-MS DATA</li>
<li><a class="" href="mass-spectrometric-analysis.html">mass spectrometric analysis</a></li>
<li class="book-part">DNA AND RNA SEQUENCE DATA</li>
<li><a class="" href="data-acquisition.html">data acquisition</a></li>
<li><a class="" href="data-transfer.html">data transfer</a></li>
<li><a class="" href="sequence-assessment.html">sequence assessment</a></li>
<li><a class="" href="illumina-read-assessment.html">illumina read assessment</a></li>
<li class="book-part">TRANSCRIPTOME ANALYSIS</li>
<li><a class="" href="transcriptomic-analyses.html">transcriptomic analyses</a></li>
<li class="book-part">GENOMIC ANALYSIS</li>
<li><a class="" href="setup.html">setup</a></li>
<li><a class="" href="genome-assembly.html">genome assembly</a></li>
<li><a class="" href="comparative-genomics.html">comparative genomics</a></li>
<li class="book-part">EVOLUTIONARY ANALYSIS</li>
<li><a class="" href="blast-searches.html">blast searches</a></li>
<li><a class="" href="alignments.html">alignments</a></li>
<li><a class="active" href="phylogenies.html">phylogenies</a></li>
<li><a class="" href="phylogenetic-analyses.html">phylogenetic analyses</a></li>
<li class="book-part">SCIENTIFIC WRITING</li>
<li><a class="" href="overview-1.html">overview</a></li>
<li><a class="" href="a-mini-manuscript.html">a mini manuscript</a></li>
<li><a class="" href="figures-captions.html">figures &amp; captions</a></li>
<li><a class="" href="results-and-discussion.html">results and discussion</a></li>
<li><a class="" href="conclusion-and-introduction.html">conclusion and introduction</a></li>
<li><a class="" href="abstract-and-title.html">abstract and title</a></li>
<li><a class="" href="proposals.html">proposals</a></li>
<li class="book-part">IMAGE ANALYSIS</li>
<li><a class="" href="image-color-analysis.html">image color analysis</a></li>
<li><a class="" href="images-of-mass-spectra.html">images of mass spectra</a></li>
<li class="book-part">APPENDIX</li>
<li><a class="" href="links.html">links</a></li>
<li><a class="" href="r-faq.html">r faq</a></li>
<li><a class="" href="templates.html">templates</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="phylogenies" class="section level1 unnumbered">
<h1>phylogenies<a class="anchor" aria-label="anchor" href="#phylogenies"><i class="fas fa-link"></i></a>
</h1>
<div id="buildtree" class="section level2" number="1.37">
<h2>
<span class="header-section-number">1.37</span> buildTree<a class="anchor" aria-label="anchor" href="#buildtree"><i class="fas fa-link"></i></a>
</h2>
<p>This function is a swiss army knife for tree building. It takes as input alignments or existing phylogenies from which to derive a phylogeny of interest, it can use neighbor-joining or maximum liklihood methods (with model optimization), it can run bootstrap replicates, and it can calculate ancestral sequence states. To illustrate, let‚Äôs look at some examples:</p>
<div id="newick-input" class="section level3" number="1.37.1">
<h3>
<span class="header-section-number">1.37.1</span> newick input<a class="anchor" aria-label="anchor" href="#newick-input"><i class="fas fa-link"></i></a>
</h3>
<p>Let‚Äôs use the Busta lab‚Äôs plant phylogeny [derived from Qian et al., 2016] to build a phylogeny with five species in it.</p>
<div class="sourceCode" id="cb130"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">buildTree</span><span class="op">(</span>
  scaffold_type <span class="op">=</span> <span class="st">"newick"</span>,
  scaffold <span class="op">=</span> <span class="st">"https://thebustalab.github.io/data/plant_phylogeny.newick"</span>,
  members <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sorghum_bicolor"</span>, <span class="st">"Zea_mays"</span>, <span class="st">"Setaria_viridis"</span>, <span class="st">"Arabidopsis_thaliana"</span>, <span class="st">"Amborella_trichopoda"</span><span class="op">)</span>
<span class="op">)</span>
<span class="co">## Pro tip: most tree read/write functions reset node numbers.</span>
<span class="co">## Fortify your tree and save it as a csv file to preserve node numbering.</span>
<span class="co">## Do not save your tree as a newick or nexus file.</span>

<span class="va">tree</span>
<span class="co">## </span>
<span class="co">## Phylogenetic tree with 5 tips and 4 internal nodes.</span>
<span class="co">## </span>
<span class="co">## Tip labels:</span>
<span class="co">##   Amborella_trichopoda, Zea_mays, Sorghum_bicolor, Setaria_viridis, Arabidopsis_thaliana</span>
<span class="co">## Node labels:</span>
<span class="co">##   , , , </span>
<span class="co">## </span>
<span class="co">## Rooted; includes branch lengths.</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-179-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Cool! We got our phylogeny. What happens if we want to build a phylogeny that has a species on it that isn‚Äôt in our scaffold? For example, what if we want to build a phylogeny that includes <em>Arabidopsis neglecta</em>? We can include that name in our list of members:</p>
<div class="sourceCode" id="cb131"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">buildTree</span><span class="op">(</span>
  scaffold_type <span class="op">=</span> <span class="st">"newick"</span>,
  scaffold_in_path <span class="op">=</span> <span class="st">"https://thebustalab.github.io/data/plant_phylogeny.newick"</span>,
  members <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sorghum_bicolor"</span>, <span class="st">"Zea_mays"</span>, <span class="st">"Setaria_viridis"</span>, <span class="st">"Arabidopsis_neglecta"</span>, <span class="st">"Amborella_trichopoda"</span><span class="op">)</span>
<span class="op">)</span>
<span class="co">## Scaffold newick tip Arabidopsis_thaliana substituted with Arabidopsis_neglecta </span>
<span class="co">## Pro tip: most tree read/write functions reset node numbers.</span>
<span class="co">## Fortify your tree and save it as a csv file to preserve node numbering.</span>
<span class="co">## Do not save your tree as a newick or nexus file.</span>

<span class="va">tree</span>
<span class="co">## </span>
<span class="co">## Phylogenetic tree with 5 tips and 4 internal nodes.</span>
<span class="co">## </span>
<span class="co">## Tip labels:</span>
<span class="co">##   Amborella_trichopoda, Zea_mays, Sorghum_bicolor, Setaria_viridis, Arabidopsis_neglecta</span>
<span class="co">## Node labels:</span>
<span class="co">##   , , , </span>
<span class="co">## </span>
<span class="co">## Rooted; includes branch lengths.</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-180-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Note that <code>buildTree</code> informs us: ‚ÄúScaffold newick tip Arabidopsis_thaliana substituted with Arabidopsis_neglecta‚Äù. This means that <em>Arabidopsis neglecta</em> was grafted onto the tip originally occupied by <em>Arabidopsis thaliana</em>. This behaviour is useful when operating on a large phylogenetic scale (i.e.¬†where <em>exact</em> phylogeny topology is not critical below the family level). However, if a person is interested in using an existing newick tree as a scaffold for a phylogeny where genus-level topology <em>is</em> critical, then beware! Your scaffold may not be appropriate if you see that message. When operating at the genus level, you probably want to use sequence data to build your phylogeny anyway. So let‚Äôs look at how to do that:</p>
</div>
<div id="alignment-input" class="section level3" number="1.37.2">
<h3>
<span class="header-section-number">1.37.2</span> alignment input<a class="anchor" aria-label="anchor" href="#alignment-input"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb132"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">buildTree</span><span class="op">(</span>
  scaffold_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"amin_alignment"</span>, <span class="st">"nucl_alignment"</span>, <span class="st">"newick"</span><span class="op">)</span>,
  scaffold_in_path <span class="op">=</span> <span class="cn">NULL</span>,
  members <span class="op">=</span> <span class="cn">NULL</span>,
  ml <span class="op">=</span> <span class="cn">FALSE</span>, 
  model_test <span class="op">=</span> <span class="cn">FALSE</span>,
  bootstrap <span class="op">=</span> <span class="cn">FALSE</span>,
  ancestral_states <span class="op">=</span> <span class="cn">FALSE</span>,
  root <span class="op">=</span> <span class="cn">NULL</span>
<span class="op">)</span></code></pre></div>
</div>
</div>
<div id="plotting-trees" class="section level2" number="1.38">
<h2>
<span class="header-section-number">1.38</span> plotting trees<a class="anchor" aria-label="anchor" href="#plotting-trees"><i class="fas fa-link"></i></a>
</h2>
<p>There are several approaches to plotting trees. A simple one is using the base <code>plot</code> function:</p>
<div class="sourceCode" id="cb133"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">test_tree_small</span> <span class="op">&lt;-</span> <span class="fu">buildTree</span><span class="op">(</span>
  scaffold_type <span class="op">=</span> <span class="st">"newick"</span>,
  scaffold_in_path <span class="op">=</span> <span class="st">"https://thebustalab.github.io/data/plant_phylogeny.newick"</span>,
  members <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sorghum_bicolor"</span>, <span class="st">"Zea_mays"</span>, <span class="st">"Setaria_viridis"</span><span class="op">)</span>
<span class="op">)</span>
<span class="co">## Pro tip: most tree read/write functions reset node numbers.</span>
<span class="co">## Fortify your tree and save it as a csv file to preserve node numbering.</span>
<span class="co">## Do not save your tree as a newick or nexus file.</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">test_tree_small</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-182-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Though this can get messy when there are lots of tip labels:</p>
<div class="sourceCode" id="cb134"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">122</span><span class="op">)</span>
<span class="va">test_tree_big</span> <span class="op">&lt;-</span> <span class="fu">buildTree</span><span class="op">(</span>
  scaffold_type <span class="op">=</span> <span class="st">"newick"</span>,
  scaffold_in_path <span class="op">=</span> <span class="st">"https://thebustalab.github.io/data/plant_phylogeny.newick"</span>,
  members <span class="op">=</span> <span class="va">plant_species</span><span class="op">$</span><span class="va">Genus_species</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html">floor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">60</span><span class="op">)</span><span class="op">*</span><span class="fl">100000</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>
<span class="op">)</span>
<span class="co">## The following species belong to a genus not found in the newick scaffold and were removed: </span>
<span class="co">## Steirodiscus_schlechteri</span>
<span class="co">## Phyllagathis_marumiaetricha</span>
<span class="co">## Jungia_sordida</span>
<span class="co">## Carpacoce_heteromorpha</span>
<span class="co">## Huttonaea_woodii</span>
<span class="co">## Ferulago_macrocarpa</span>
<span class="co">## Echinospartum_lusitanicum</span>
<span class="co">## Scurrula_lepidota</span>
<span class="co">## Astronidium_sudestense</span>
<span class="co">## </span>
<span class="co">## Scaffold newick tip Peperomia_fraseri substituted with Peperomia_kimachii </span>
<span class="co">## Scaffold newick tip Telipogon_pulcher substituted with Telipogon_alberti </span>
<span class="co">## Scaffold newick tip Elaeocarpus_angustifolius substituted with Elaeocarpus_miriensis </span>
<span class="co">## Scaffold newick tip Angelica_sinensis substituted with Angelica_anomala </span>
<span class="co">## Scaffold newick tip Senecio_pterophorus substituted with Senecio_isatideus </span>
<span class="co">## Scaffold newick tip Macroscepis_hirsuta substituted with Macroscepis_pleistantha </span>
<span class="co">## Scaffold newick tip Tithonia_diversifolia substituted with Tithonia_fruticosa </span>
<span class="co">## Scaffold newick tip Spermacoce_princeae substituted with Spermacoce_latituba </span>
<span class="co">## Scaffold newick tip Bothriocline_laxa substituted with Bothriocline_amphicoma </span>
<span class="co">## Scaffold newick tip Siler_montanum substituted with Siler_zernyi </span>
<span class="co">## Scaffold newick tip Teucrium_betchei substituted with Teucrium_rotundifolium </span>
<span class="co">## Scaffold newick tip Calathea_pluriplicata substituted with Calathea_mediopicta </span>
<span class="co">## Scaffold newick tip Phyllanthus_maderaspatensis substituted with Phyllanthus_roeperianus </span>
<span class="co">## Scaffold newick tip Entada_abyssinica substituted with Entada_dolichorrhachis </span>
<span class="co">## Scaffold newick tip Pimpinella_rhodantha substituted with Pimpinella_cypria </span>
<span class="co">## Scaffold newick tip Baccharis_neglecta substituted with Baccharis_grisebachii </span>
<span class="co">## Scaffold newick tip Mikania_micrantha substituted with Mikania_longipes </span>
<span class="co">## Scaffold newick tip Raphionacme_flanaganii substituted with Raphionacme_hirsuta </span>
<span class="co">## Scaffold newick tip Senecio_polyanthemoides substituted with Senecio_conferruminatus </span>
<span class="co">## Scaffold newick tip Heracleum_austriacum substituted with Heracleum_bailletianum </span>
<span class="co">## Scaffold newick tip Bulbophyllum_hainanense substituted with Bulbophyllum_kaniense </span>
<span class="co">## Scaffold newick tip Madhuca_microphylla substituted with Madhuca_elmeri </span>
<span class="co">## Scaffold newick tip Capparis_spinosa substituted with Capparis_cantoniensis </span>
<span class="co">## Scaffold newick tip Albizia_lebbekoides substituted with Albizia_tulearensis </span>
<span class="co">## Scaffold newick tip Cephalaria_syriaca substituted with Cephalaria_dirmilensis </span>
<span class="co">## Scaffold newick tip Adesmia_exilis substituted with Adesmia_uspallatensis </span>
<span class="co">## Scaffold newick tip Artemisia_biennis substituted with Artemisia_stricta </span>
<span class="co">## Scaffold newick tip Osteospermum_ilicifolium substituted with Osteospermum_polycephalum </span>
<span class="co">## Scaffold newick tip Atriplex_californica substituted with Atriplex_lanfrancoi </span>
<span class="co">## Scaffold newick tip Oberonia_ensiformis substituted with Oberonia_anguina </span>
<span class="co">## Scaffold newick tip Ceropegia_linearis_subsp._woodii substituted with Ceropegia_andamanica </span>
<span class="co">## Scaffold newick tip Cousinia_severtzovii substituted with Cousinia_butkovii </span>
<span class="co">## Scaffold newick tip Hoya_lanceolata substituted with Hoya_loyceandrewsiana </span>
<span class="co">## Scaffold newick tip Cousinia_microcarpa substituted with Cousinia_trachylepis </span>
<span class="co">## Scaffold newick tip Bossiaea_cordigera substituted with Bossiaea_oligosperma </span>
<span class="co">## Scaffold newick tip Cayaponia_tubulosa substituted with Cayaponia_jenmanii </span>
<span class="co">## Scaffold newick tip Pomatocalpa_kunstleri substituted with Pomatocalpa_linearipetalum </span>
<span class="co">## Scaffold newick tip Endlicheria_verticillata substituted with Endlicheria_xerampela </span>
<span class="co">## Scaffold newick tip Eugenia_moschata substituted with Eugenia_azurensis </span>
<span class="co">## Scaffold newick tip Gastrolobium_punctatum substituted with Gastrolobium_spinosum </span>
<span class="co">## Scaffold newick tip Dioscorea_mexicana substituted with Dioscorea_pennellii </span>
<span class="co">## Scaffold newick tip Mitrella_kentii substituted with Mitrella_ledermannii </span>
<span class="co">## Scaffold newick tip Ageratina_adenophora substituted with Ageratina_ayerscottiana </span>
<span class="co">## Scaffold newick tip Astragalus_vogelii substituted with Astragalus_jolderensis </span>
<span class="co">## Scaffold newick tip Eremocharis_fruticosa substituted with Eremocharis_longiramea </span>
<span class="co">## Scaffold newick tip Spathiphyllum_floribundum substituted with Spathiphyllum_monachinoi </span>
<span class="co">## Scaffold newick tip Balanops_vieillardii substituted with Balanops_microstachya </span>
<span class="co">## Pro tip: most tree read/write functions reset node numbers.</span>
<span class="co">## Fortify your tree and save it as a csv file to preserve node numbering.</span>
<span class="co">## Do not save your tree as a newick or nexus file.</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">test_tree_big</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-183-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>One solution is to use <code>ggtree</code>, which by default doesn‚Äôt show tip labels. <code>plot</code> can do that too, but <code>ggtree</code> does a bunch of other useful things, so I recommend that:</p>
<div class="sourceCode" id="cb135"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">test_tree_big</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-184-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>Another convenient fucntion is ggplot‚Äôs <code>fortify</code>. This will convert your <code>phylo</code> object into a data frame:</p>
<div class="sourceCode" id="cb136"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">test_tree_big_fortified</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/fortify.html">fortify</a></span><span class="op">(</span><span class="va">test_tree_big</span><span class="op">)</span>
<span class="va">test_tree_big_fortified</span>
<span class="co">## # A tibble: 101 √ó 9</span>
<span class="co">##    parent  node branch.length label isTip     x     y branch</span>
<span class="co">##     &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt; &lt;lgl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">##  1     54     1          83.0 Wolf‚Ä¶ TRUE   188.     1   147.</span>
<span class="co">##  2     54     2          83.0 Spat‚Ä¶ TRUE   188.     2   147.</span>
<span class="co">##  3     55     3         138.  Dios‚Ä¶ TRUE   188.     3   120.</span>
<span class="co">##  4     58     4          42.7 Bulb‚Ä¶ TRUE   188.     5   167.</span>
<span class="co">##  5     58     5          42.7 Ober‚Ä¶ TRUE   188.     6   167.</span>
<span class="co">##  6     59     6          32.0 Poma‚Ä¶ TRUE   188.     7   172.</span>
<span class="co">##  7     59     7          32.0 Teli‚Ä¶ TRUE   188.     8   172.</span>
<span class="co">##  8     56     8         135.  Cala‚Ä¶ TRUE   188.     4   121.</span>
<span class="co">##  9     61     9         147.  Pepe‚Ä¶ TRUE   188.     9   115.</span>
<span class="co">## 10     62    10         121.  Endl‚Ä¶ TRUE   188.    10   128.</span>
<span class="co">## # ‚Ä¶ with 91 more rows, and 1 more variable: angle &lt;dbl&gt;</span></code></pre></div>
<p><code>ggtree</code> can still plot this dataframe, and it allows metadata to be stored in a human readable format by using mutating joins (explained below). This metadata can be plotted with standard ggplot geoms, and these dataframes can also conveniently be saved as .csv files:</p>
<div class="sourceCode" id="cb137"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co">## Note that "plant_species" comes with the phylochemistry source.</span>

<span class="va">test_tree_big_fortified_w_data</span> <span class="op">&lt;-</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">test_tree_big_fortified</span>, <span class="va">plant_species</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"label"</span> <span class="op">=</span> <span class="st">"Genus_species"</span><span class="op">)</span><span class="op">)</span>

<span class="va">test_tree_big_fortified_w_data</span>
<span class="co">## # A tibble: 101 √ó 14</span>
<span class="co">##    parent  node branch.length label isTip     x     y branch</span>
<span class="co">##     &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt; &lt;lgl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">##  1     54     1          83.0 Wolf‚Ä¶ TRUE   188.     1   147.</span>
<span class="co">##  2     54     2          83.0 Spat‚Ä¶ TRUE   188.     2   147.</span>
<span class="co">##  3     55     3         138.  Dios‚Ä¶ TRUE   188.     3   120.</span>
<span class="co">##  4     58     4          42.7 Bulb‚Ä¶ TRUE   188.     5   167.</span>
<span class="co">##  5     58     5          42.7 Ober‚Ä¶ TRUE   188.     6   167.</span>
<span class="co">##  6     59     6          32.0 Poma‚Ä¶ TRUE   188.     7   172.</span>
<span class="co">##  7     59     7          32.0 Teli‚Ä¶ TRUE   188.     8   172.</span>
<span class="co">##  8     56     8         135.  Cala‚Ä¶ TRUE   188.     4   121.</span>
<span class="co">##  9     61     9         147.  Pepe‚Ä¶ TRUE   188.     9   115.</span>
<span class="co">## 10     62    10         121.  Endl‚Ä¶ TRUE   188.    10   128.</span>
<span class="co">## # ‚Ä¶ with 91 more rows, and 6 more variables: angle &lt;dbl&gt;,</span>
<span class="co">## #   Phylum &lt;chr&gt;, Order &lt;chr&gt;, Family &lt;chr&gt;, Genus &lt;chr&gt;,</span>
<span class="co">## #   species &lt;chr&gt;</span>

<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">test_tree_big_fortified_w_data</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">test_tree_big_fortified_w_data</span>, <span class="va">isTip</span> <span class="op">==</span> <span class="cn">TRUE</span><span class="op">)</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, fill <span class="op">=</span> <span class="va">Order</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">3</span>, shape <span class="op">=</span> <span class="fl">21</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_text</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">test_tree_big_fortified_w_data</span>, <span class="va">isTip</span> <span class="op">==</span> <span class="cn">TRUE</span><span class="op">)</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, label <span class="op">=</span> <span class="va">y</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">label</span><span class="op">)</span>, offset <span class="op">=</span> <span class="fl">10</span>, size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme_void</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">discrete_palette</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">coord_cartesian</span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">280</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>
    legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.15</span>, <span class="fl">0.75</span><span class="op">)</span>
  <span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-186-1.png" width="100%" style="display: block; margin: auto;"></div>
</div>
<div id="collapsetree" class="section level2" number="1.39">
<h2>
<span class="header-section-number">1.39</span> collapseTree<a class="anchor" aria-label="anchor" href="#collapsetree"><i class="fas fa-link"></i></a>
</h2>
<p>Sometimes we want to view a tree at a higher level of taxonomical organization, or some other higher level. This can be done easily using the <code>collapseTree</code> function. It takes two arguments: an un-fortified tree (<code>tree</code>), and a two-column data frame (<code>associations</code>). In the first column of the data frame are all the tip labels of the tree, and in the second column are the higher level of organization to which each tip belongs. The function will prune the tree so that only one member of the higher level of organization is included in the output. For example, let‚Äôs look at the tree from the previous section at the family level:</p>
<div class="sourceCode" id="cb138"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">collapseTree</span><span class="op">(</span>
  tree <span class="op">=</span> <span class="va">test_tree_big</span>,
  associations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>
    tip.label <span class="op">=</span> <span class="va">test_tree_big</span><span class="op">$</span><span class="va">tip.label</span>,
    family <span class="op">=</span> <span class="va">plant_species</span><span class="op">$</span><span class="va">Family</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="va">test_tree_big</span><span class="op">$</span><span class="va">tip.label</span>, <span class="va">plant_species</span><span class="op">$</span><span class="va">Genus_species</span><span class="op">)</span><span class="op">]</span>
  <span class="op">)</span>
<span class="op">)</span> <span class="op">-&gt;</span> <span class="va">test_tree_big_families</span>

<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">test_tree_big_families</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">coord_cartesian</span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">300</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-187-1.png" width="100%" style="display: block; margin: auto;"></div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="alignments.html">alignments</a></div>
<div class="next"><a href="phylogenetic-analyses.html">phylogenetic analyses</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <h2>Note: Second Edition is under construction üèó</h2>
    <p>Now is a great time to provide feedback</p>
        <ul class="list-unstyled">
<li><a href="https://forms.gle/SZmB2Ct2exE2dBwv9">Provide feedback (5 min)</a></li>
          <!-- <li><a href="https://geocompr.robinlovelace.net/#reproducibility">Install updated packages</a></li> -->
          <!-- <li><a href="https://github.com/Robinlovelace/geocompr/issues">Open an issue <i class="fas fa-question"></i></a></li> -->
          <!-- <li><a href="https://discord.gg/Te3gWeDwmf">Chat on Discord <i class="fab fa-discord"></i></a></li> -->
        </ul>
<hr>
<nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#phylogenies">phylogenies</a></li>
<li>
<a class="nav-link" href="#buildtree"><span class="header-section-number">1.37</span> buildTree</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#newick-input"><span class="header-section-number">1.37.1</span> newick input</a></li>
<li><a class="nav-link" href="#alignment-input"><span class="header-section-number">1.37.2</span> alignment input</a></li>
</ul>
</li>
<li><a class="nav-link" href="#plotting-trees"><span class="header-section-number">1.38</span> plotting trees</a></li>
<li><a class="nav-link" href="#collapsetree"><span class="header-section-number">1.39</span> collapseTree</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/blob/main/index.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/edit/main/index.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Integrated Bioanalytics</strong>" was written by Lucas Busta and members of the Busta lab. It was last built on 2022-09-13.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
