<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>phylogenetic analyses | Integrated Bioanalytics</title>
<meta name="author" content="Lucas Busta and members of the Busta lab">
<meta name="description" content='We use a wrapper function to run phylogenetic comparative analyses. It chemical_bloom_tree &lt;- buildTree(  scaffold_type = "newick",  scaffold_in_path =...'>
<meta name="generator" content="bookdown 0.35 with bs4_book()">
<meta property="og:title" content="phylogenetic analyses | Integrated Bioanalytics">
<meta property="og:type" content="book">
<meta property="og:description" content='We use a wrapper function to run phylogenetic comparative analyses. It chemical_bloom_tree &lt;- buildTree(  scaffold_type = "newick",  scaffold_in_path =...'>
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="phylogenetic analyses | Integrated Bioanalytics">
<meta name="twitter:description" content='We use a wrapper function to run phylogenetic comparative analyses. It chemical_bloom_tree &lt;- buildTree(  scaffold_type = "newick",  scaffold_in_path =...'>
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><link href="libs/Lato-0.4.6/font.css" rel="stylesheet">
<link href="libs/Roboto_Mono-0.4.6/font.css" rel="stylesheet">
<link href="libs/Montserrat-0.4.6/font.css" rel="stylesheet">
<script src="libs/bs3compat-0.4.2/transition.js"></script><script src="libs/bs3compat-0.4.2/tabs.js"></script><script src="libs/bs3compat-0.4.2/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><link rel="icon" type="image/png" sizes="32x32" href="images/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="images/favicon-16x16.png">
<script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-99618359-1', 'auto');
      ga('send', 'pageview');

    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="style/style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h2>
        <a href="index.html" title="">Integrated Bioanalytics</a>
      </h2>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">WELCOME</a></li>
<li class="book-part">DATA ANALYSIS IN R</li>
<li><a class="" href="overview.html">overview</a></li>
<li><a class="" href="installation.html">installation</a></li>
<li><a class="" href="data-visualization-i.html">data visualization I</a></li>
<li><a class="" href="data-visualization-ii.html">data visualization II</a></li>
<li><a class="" href="data-visualization-iii.html">data visualization III</a></li>
<li><a class="" href="data-wrangling-and-summaries.html">data wrangling and summaries</a></li>
<li><a class="" href="dimensionality-reduction.html">dimensionality reduction</a></li>
<li><a class="" href="clustering.html">clustering</a></li>
<li><a class="" href="comparing-means.html">comparing means</a></li>
<li><a class="" href="models.html">models</a></li>
<li><a class="" href="special-topics.html">special topics</a></li>
<li class="book-part">GC-MS DATA</li>
<li><a class="" href="loading-analyzegcmsdata.html">loading analyzeGCMSdata</a></li>
<li><a class="" href="using-analyzegcmsdata.html">using analyzeGCMSdata</a></li>
<li><a class="" href="cdf-export.html">CDF export</a></li>
<li class="book-part">TRANSCRIPTOME ANALYSIS</li>
<li><a class="" href="transcriptome-assembly.html">transcriptome assembly</a></li>
<li class="book-part">EVOLUTIONARY ANALYSIS</li>
<li><a class="" href="similarity-searching.html">similarity searching</a></li>
<li><a class="" href="alignments.html">alignments</a></li>
<li><a class="" href="phylogenies.html">phylogenies</a></li>
<li><a class="active" href="phylogenetic-analyses.html">phylogenetic analyses</a></li>
<li><a class="" href="comparative-genomics.html">comparative genomics</a></li>
<li class="book-part">LANGUAGE MODELS</li>
<li><a class="" href="completiongpt.html">completionGPT</a></li>
<li><a class="" href="analyzeliterature.html">analyzeLiterature</a></li>
<li class="book-part">SCIENTIFIC WRITING</li>
<li><a class="" href="overview-1.html">overview</a></li>
<li><a class="" href="figures-captions.html">figures &amp; captions</a></li>
<li><a class="" href="results-and-discussion.html">results and discussion</a></li>
<li><a class="" href="conclusion-and-introduction.html">conclusion and introduction</a></li>
<li><a class="" href="abstract-and-title.html">abstract and title</a></li>
<li class="book-part">APPENDIX</li>
<li><a class="" href="links.html">links</a></li>
<li><a class="" href="r-faq.html">r faq</a></li>
<li><a class="" href="templates.html">templates</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="phylogenetic-analyses" class="section level1 unnumbered">
<h1>phylogenetic analyses<a class="anchor" aria-label="anchor" href="#phylogenetic-analyses"><i class="fas fa-link"></i></a>
</h1>
<p>We use a wrapper function to run phylogenetic comparative analyses. It</p>
<div class="sourceCode" id="cb152"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chemical_bloom_tree</span> <span class="op">&lt;-</span> <span class="fu">buildTree</span><span class="op">(</span></span>
<span>  scaffold_type <span class="op">=</span> <span class="st">"newick"</span>,</span>
<span>  scaffold_in_path <span class="op">=</span> <span class="st">"http://thebustalab.github.io/data/angiosperms.newick"</span>,</span>
<span>  members <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">chemical_blooms</span><span class="op">$</span><span class="va">label</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Scaffold newick tip Sabal_pumos substituted with Sabal_palmetto</span></span>
<span><span class="co">## Scaffold newick tip Iris_lazica substituted with Iris_sp</span></span>
<span><span class="co">## Scaffold newick tip Iris_lacustris substituted with Iris_germanica</span></span>
<span><span class="co">## Scaffold newick tip Allium_textile substituted with Allium_sp</span></span>
<span><span class="co">## Scaffold newick tip Allium_subhirsutum substituted with Allium_brevistylum</span></span>
<span><span class="co">## Scaffold newick tip Ornithogalum_saundersiae substituted with Ornithogalum_candicans</span></span>
<span><span class="co">## Scaffold newick tip Hosta_plantaginea substituted with Hosta_sp</span></span>
<span><span class="co">## Scaffold newick tip Agave_striata substituted with Agave_cerulata</span></span>
<span><span class="co">## Scaffold newick tip Agave_tequilana substituted with Agave_chrysantha</span></span>
<span><span class="co">## Scaffold newick tip Aristolochia_serpentaria substituted with Aristolochia_labiata</span></span>
<span><span class="co">## Scaffold newick tip Thalictrum_clavatum substituted with Thalictrum_rochebrunianum</span></span>
<span><span class="co">## Scaffold newick tip Delphinium_pictum substituted with Delphinium_elatum</span></span>
<span><span class="co">## Scaffold newick tip Ferocactus_recurvus substituted with Ferocactus_wislizeni</span></span>
<span><span class="co">## Scaffold newick tip Opuntia_articulata substituted with Opuntia_sp</span></span>
<span><span class="co">## Scaffold newick tip Opuntia_quimilo substituted with Opuntia_robusta</span></span>
<span><span class="co">## Scaffold newick tip Cylindropuntia_fulgida substituted with Cylindropuntia_bigelovii</span></span>
<span><span class="co">## Scaffold newick tip Cylindropuntia_prolifera substituted with Cylindropuntia_versicolor</span></span>
<span><span class="co">## Scaffold newick tip Cylindropuntia_echinocarpa substituted with Cylindropuntia_ramosissima</span></span>
<span><span class="co">## Scaffold newick tip Kirengeshoma_palmata substituted with Kirengeshoma_Palmata</span></span>
<span><span class="co">## Scaffold newick tip Lavandula_bipinnata substituted with Lavandula_sp</span></span>
<span><span class="co">## Scaffold newick tip Rudbeckia_hirta substituted with Rudbeckia_occidentalis</span></span>
<span><span class="co">## Scaffold newick tip Crassula_campestris substituted with Crassula_ovata</span></span>
<span><span class="co">## Scaffold newick tip Crassula_tillaea substituted with Crassula_deceptor</span></span>
<span><span class="co">## Scaffold newick tip Crassula_alata substituted with Crassula_arborescens</span></span>
<span><span class="co">## Scaffold newick tip Crassula_colligata substituted with Crassula_perfoliata</span></span>
<span><span class="co">## Scaffold newick tip Hylotelephium_erythrostictum substituted with Hylotelephium_sp</span></span>
<span><span class="co">## Scaffold newick tip Echeveria_setosa substituted with Echeveria_pulidonis</span></span>
<span><span class="co">## Scaffold newick tip Bryophyllum_pinnatum substituted with Bryophyllum_fedtschenkoi</span></span>
<span><span class="co">## Scaffold newick tip Kalanchoe_linearifolia substituted with Kalanchoe_marginata</span></span>
<span><span class="co">## Scaffold newick tip Kalanchoe_tomentosa substituted with Kalanchoe_luciae</span></span>
<span><span class="co">## Scaffold newick tip Kalanchoe_beharensis substituted with Kalanchoe_thrysiflora</span></span>
<span><span class="co">## Scaffold newick tip Iliamna_latibracteata substituted with Iliamna_rivularis</span></span>
<span><span class="co">## Scaffold newick tip Euphorbia_lathyris substituted with Euphorbia_resinifera</span></span>
<span><span class="co">## Scaffold newick tip Quercus_valdinervosa substituted with Quercus_muehlenbergii</span></span>
<span><span class="co">## Scaffold newick tip Rubus_repens substituted with Rubus_sp</span></span>
<span><span class="co">## Pro tip: most tree read/write functions reset node numbers.</span></span>
<span><span class="co">## Fortify your tree and save it as a csv file to preserve node numbering.</span></span>
<span><span class="co">## Do not save your tree as a newick or nexus file.</span></span>
<span></span>
<span><span class="fu">runPhylogeneticAnalyses</span><span class="op">(</span></span>
<span>    traits <span class="op">=</span> <span class="fu">pivot_longer</span><span class="op">(</span><span class="va">chemical_blooms</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"trait"</span>, values_to <span class="op">=</span> <span class="st">"value"</span><span class="op">)</span>,</span>
<span>    column_w_names_of_tiplabels <span class="op">=</span> <span class="st">"label"</span>,</span>
<span>    column_w_names_of_traits <span class="op">=</span> <span class="st">"trait"</span>,</span>
<span>    column_w_values_for_traits <span class="op">=</span> <span class="st">"value"</span>,</span>
<span>    tree <span class="op">=</span> <span class="va">chemical_bloom_tree</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Joining with `by = join_by(trait)`</span></span>
<span><span class="co">## # A tibble: 310 × 17</span></span>
<span><span class="co">##    parent  node branch.length label isTip     x     y branch</span></span>
<span><span class="co">##     &lt;int&gt; &lt;dbl&gt;         &lt;dbl&gt; &lt;chr&gt; &lt;lgl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;</span></span>
<span><span class="co">##  1     80     1         290.  Gink… TRUE   352.     1   207.</span></span>
<span><span class="co">##  2     80     1         290.  Gink… TRUE   352.     1   207.</span></span>
<span><span class="co">##  3     81     2         267.  Pice… TRUE   352.     2   219.</span></span>
<span><span class="co">##  4     81     2         267.  Pice… TRUE   352.     2   219.</span></span>
<span><span class="co">##  5     81     3         267.  Cupr… TRUE   352.     3   219.</span></span>
<span><span class="co">##  6     81     3         267.  Cupr… TRUE   352.     3   219.</span></span>
<span><span class="co">##  7     84     4         135.  Eryt… TRUE   352.     5   285.</span></span>
<span><span class="co">##  8     84     4         135.  Eryt… TRUE   352.     5   285.</span></span>
<span><span class="co">##  9     86     5          16.2 Iris… TRUE   352.     6   344.</span></span>
<span><span class="co">## 10     86     5          16.2 Iris… TRUE   352.     6   344.</span></span>
<span><span class="co">## # ℹ 300 more rows</span></span>
<span><span class="co">## # ℹ 9 more variables: angle &lt;dbl&gt;, trait &lt;chr&gt;,</span></span>
<span><span class="co">## #   value &lt;dbl&gt;, trait_type &lt;chr&gt;,</span></span>
<span><span class="co">## #   phylogenetic_signal_k_value &lt;dbl&gt;,</span></span>
<span><span class="co">## #   phylogenetic_signal_k_p_value &lt;dbl&gt;,</span></span>
<span><span class="co">## #   phylogenetic_signal_lambda_value &lt;dbl&gt;,</span></span>
<span><span class="co">## #   phylogenetic_signal_lambda_p_value &lt;dbl&gt;, pic &lt;dbl&gt;</span></span></code></pre></div>
<p>For all the below, there are some structural requirements: (i) the tree needs to be a phylo object (ii) the traits need to be a data.frame in which each row is a species and each column is a variable, and (iii) the first column in the data.frame needs to be the names of the species and they must exactly match the tip labels of the tree (though they don’t have to be in the same order), for example:</p>
<div class="sourceCode" id="cb153"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chemical_bloom_tree</span> <span class="op">&lt;-</span> <span class="fu">buildTree</span><span class="op">(</span></span>
<span>  scaffold_type <span class="op">=</span> <span class="st">"newick"</span>,</span>
<span>  scaffold_in_path <span class="op">=</span> <span class="st">"http://thebustalab.github.io/data/angiosperms.newick"</span>,</span>
<span>  members <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">chemical_blooms</span><span class="op">$</span><span class="va">label</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Scaffold newick tip Sabal_pumos substituted with Sabal_palmetto</span></span>
<span><span class="co">## Scaffold newick tip Iris_lazica substituted with Iris_sp</span></span>
<span><span class="co">## Scaffold newick tip Iris_lacustris substituted with Iris_germanica</span></span>
<span><span class="co">## Scaffold newick tip Allium_textile substituted with Allium_sp</span></span>
<span><span class="co">## Scaffold newick tip Allium_subhirsutum substituted with Allium_brevistylum</span></span>
<span><span class="co">## Scaffold newick tip Ornithogalum_saundersiae substituted with Ornithogalum_candicans</span></span>
<span><span class="co">## Scaffold newick tip Hosta_plantaginea substituted with Hosta_sp</span></span>
<span><span class="co">## Scaffold newick tip Agave_striata substituted with Agave_cerulata</span></span>
<span><span class="co">## Scaffold newick tip Agave_tequilana substituted with Agave_chrysantha</span></span>
<span><span class="co">## Scaffold newick tip Aristolochia_serpentaria substituted with Aristolochia_labiata</span></span>
<span><span class="co">## Scaffold newick tip Thalictrum_clavatum substituted with Thalictrum_rochebrunianum</span></span>
<span><span class="co">## Scaffold newick tip Delphinium_pictum substituted with Delphinium_elatum</span></span>
<span><span class="co">## Scaffold newick tip Ferocactus_recurvus substituted with Ferocactus_wislizeni</span></span>
<span><span class="co">## Scaffold newick tip Opuntia_articulata substituted with Opuntia_sp</span></span>
<span><span class="co">## Scaffold newick tip Opuntia_quimilo substituted with Opuntia_robusta</span></span>
<span><span class="co">## Scaffold newick tip Cylindropuntia_fulgida substituted with Cylindropuntia_bigelovii</span></span>
<span><span class="co">## Scaffold newick tip Cylindropuntia_prolifera substituted with Cylindropuntia_versicolor</span></span>
<span><span class="co">## Scaffold newick tip Cylindropuntia_echinocarpa substituted with Cylindropuntia_ramosissima</span></span>
<span><span class="co">## Scaffold newick tip Kirengeshoma_palmata substituted with Kirengeshoma_Palmata</span></span>
<span><span class="co">## Scaffold newick tip Lavandula_bipinnata substituted with Lavandula_sp</span></span>
<span><span class="co">## Scaffold newick tip Rudbeckia_hirta substituted with Rudbeckia_occidentalis</span></span>
<span><span class="co">## Scaffold newick tip Crassula_campestris substituted with Crassula_ovata</span></span>
<span><span class="co">## Scaffold newick tip Crassula_tillaea substituted with Crassula_deceptor</span></span>
<span><span class="co">## Scaffold newick tip Crassula_alata substituted with Crassula_arborescens</span></span>
<span><span class="co">## Scaffold newick tip Crassula_colligata substituted with Crassula_perfoliata</span></span>
<span><span class="co">## Scaffold newick tip Hylotelephium_erythrostictum substituted with Hylotelephium_sp</span></span>
<span><span class="co">## Scaffold newick tip Echeveria_setosa substituted with Echeveria_pulidonis</span></span>
<span><span class="co">## Scaffold newick tip Bryophyllum_pinnatum substituted with Bryophyllum_fedtschenkoi</span></span>
<span><span class="co">## Scaffold newick tip Kalanchoe_linearifolia substituted with Kalanchoe_marginata</span></span>
<span><span class="co">## Scaffold newick tip Kalanchoe_tomentosa substituted with Kalanchoe_luciae</span></span>
<span><span class="co">## Scaffold newick tip Kalanchoe_beharensis substituted with Kalanchoe_thrysiflora</span></span>
<span><span class="co">## Scaffold newick tip Iliamna_latibracteata substituted with Iliamna_rivularis</span></span>
<span><span class="co">## Scaffold newick tip Euphorbia_lathyris substituted with Euphorbia_resinifera</span></span>
<span><span class="co">## Scaffold newick tip Quercus_valdinervosa substituted with Quercus_muehlenbergii</span></span>
<span><span class="co">## Scaffold newick tip Rubus_repens substituted with Rubus_sp</span></span>
<span><span class="co">## Pro tip: most tree read/write functions reset node numbers.</span></span>
<span><span class="co">## Fortify your tree and save it as a csv file to preserve node numbering.</span></span>
<span><span class="co">## Do not save your tree as a newick or nexus file.</span></span></code></pre></div>
<div id="phylogeneticsignal" class="section level2 unnumbered">
<h2>phylogeneticSignal<a class="anchor" aria-label="anchor" href="#phylogeneticsignal"><i class="fas fa-link"></i></a>
</h2>
<p>Phylogenetic signal is a measure of the degree to which related species share similar trait values. It is used to determine whether a trait has evolved in a manner that is consistent with the species’ evolutionary history. <code>phylochemistry</code> provides the <code>phylogeneticSignal</code> function, which can be used to calculate phylogenetic signal for a given set of traits and a phylogenetic tree. Here is an example:</p>
<div class="sourceCode" id="cb154"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">phylogeneticSignal</span><span class="op">(</span></span>
<span>  traits <span class="op">=</span> <span class="fu">pivot_longer</span><span class="op">(</span><span class="va">chemical_blooms</span>, cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"compound"</span>, values_to <span class="op">=</span> <span class="st">"value"</span><span class="op">)</span>,</span>
<span>  column_w_names_of_tiplabels <span class="op">=</span> <span class="st">"label"</span>,</span>
<span>  column_w_names_of_traits <span class="op">=</span> <span class="st">"compound"</span>,</span>
<span>  column_w_values_for_traits <span class="op">=</span> <span class="st">"value"</span>,</span>
<span>  tree <span class="op">=</span> <span class="va">chemical_bloom_tree</span></span>
<span><span class="op">)</span></span>
<span><span class="co">##             trait trait_type n_species number_of_levels</span></span>
<span><span class="co">## 1         Alkanes continuous        78               NA</span></span>
<span><span class="co">## 2    Sec_Alcohols continuous        78               NA</span></span>
<span><span class="co">## 3          Others continuous        78               NA</span></span>
<span><span class="co">## 4     Fatty_acids continuous        78               NA</span></span>
<span><span class="co">## 5        Alcohols continuous        78               NA</span></span>
<span><span class="co">## 6   Triterpenoids continuous        78               NA</span></span>
<span><span class="co">## 7         Ketones continuous        78               NA</span></span>
<span><span class="co">## 8 Other_compounds continuous        78               NA</span></span>
<span><span class="co">## 9       Aldehydes continuous        78               NA</span></span>
<span><span class="co">##   evolutionary_transitions_observed</span></span>
<span><span class="co">## 1                                NA</span></span>
<span><span class="co">## 2                                NA</span></span>
<span><span class="co">## 3                                NA</span></span>
<span><span class="co">## 4                                NA</span></span>
<span><span class="co">## 5                                NA</span></span>
<span><span class="co">## 6                                NA</span></span>
<span><span class="co">## 7                                NA</span></span>
<span><span class="co">## 8                                NA</span></span>
<span><span class="co">## 9                                NA</span></span>
<span><span class="co">##   median_evolutionary_transitions_in_randomization</span></span>
<span><span class="co">## 1                                               NA</span></span>
<span><span class="co">## 2                                               NA</span></span>
<span><span class="co">## 3                                               NA</span></span>
<span><span class="co">## 4                                               NA</span></span>
<span><span class="co">## 5                                               NA</span></span>
<span><span class="co">## 6                                               NA</span></span>
<span><span class="co">## 7                                               NA</span></span>
<span><span class="co">## 8                                               NA</span></span>
<span><span class="co">## 9                                               NA</span></span>
<span><span class="co">##   minimum_evolutionary_transitions_in_randomization</span></span>
<span><span class="co">## 1                                                NA</span></span>
<span><span class="co">## 2                                                NA</span></span>
<span><span class="co">## 3                                                NA</span></span>
<span><span class="co">## 4                                                NA</span></span>
<span><span class="co">## 5                                                NA</span></span>
<span><span class="co">## 6                                                NA</span></span>
<span><span class="co">## 7                                                NA</span></span>
<span><span class="co">## 8                                                NA</span></span>
<span><span class="co">## 9                                                NA</span></span>
<span><span class="co">##   evolutionary_transitions_in_randomization</span></span>
<span><span class="co">## 1                                        NA</span></span>
<span><span class="co">## 2                                        NA</span></span>
<span><span class="co">## 3                                        NA</span></span>
<span><span class="co">## 4                                        NA</span></span>
<span><span class="co">## 5                                        NA</span></span>
<span><span class="co">## 6                                        NA</span></span>
<span><span class="co">## 7                                        NA</span></span>
<span><span class="co">## 8                                        NA</span></span>
<span><span class="co">## 9                                        NA</span></span>
<span><span class="co">##   phylogenetic_signal_k_value phylogenetic_signal_k_p_value</span></span>
<span><span class="co">## 1                 0.066016688                         0.167</span></span>
<span><span class="co">## 2                 2.045611108                         0.001</span></span>
<span><span class="co">## 3                 0.029719595                         0.711</span></span>
<span><span class="co">## 4                 0.069056092                         0.292</span></span>
<span><span class="co">## 5                 0.053761730                         0.354</span></span>
<span><span class="co">## 6                 0.566806846                         0.001</span></span>
<span><span class="co">## 7                 0.239488396                         0.030</span></span>
<span><span class="co">## 8                 0.018420367                         0.868</span></span>
<span><span class="co">## 9                 0.008613879                         0.928</span></span>
<span><span class="co">##   phylogenetic_signal_lambda_value</span></span>
<span><span class="co">## 1                           0.0001</span></span>
<span><span class="co">## 2                           0.9999</span></span>
<span><span class="co">## 3                           0.0001</span></span>
<span><span class="co">## 4                           0.0001</span></span>
<span><span class="co">## 5                           0.0001</span></span>
<span><span class="co">## 6                           0.9999</span></span>
<span><span class="co">## 7                           0.7874</span></span>
<span><span class="co">## 8                           0.0001</span></span>
<span><span class="co">## 9                           0.0001</span></span>
<span><span class="co">##   phylogenetic_signal_lambda_p_value</span></span>
<span><span class="co">## 1                              1.000</span></span>
<span><span class="co">## 2                              0.000</span></span>
<span><span class="co">## 3                              1.000</span></span>
<span><span class="co">## 4                              1.000</span></span>
<span><span class="co">## 5                              1.000</span></span>
<span><span class="co">## 6                              0.000</span></span>
<span><span class="co">## 7                              0.045</span></span>
<span><span class="co">## 8                              1.000</span></span>
<span><span class="co">## 9                              1.000</span></span></code></pre></div>
</div>
<div id="independentcontrasts" class="section level2 unnumbered">
<h2>independentContrasts<a class="anchor" aria-label="anchor" href="#independentcontrasts"><i class="fas fa-link"></i></a>
</h2>
<p>Phylogenetic independent contrasts are a method for analyzing the relationship between two or more traits while taking into account the evolutionary history of the species being studied. This method involves transforming the data in to “independent contrasts” to remove the effects of shared ancestry, allowing for more accurate analysis of the relationship between traits. <code>phylochemistry</code> provides the <code>independentContrasts</code> function to calculate phylogenetic independent contrasts for a given set of traits and a phylogenetic tree. Here is an example of calculating independent contrasts for an example dataset, followed by generating a linear model based on the contrasts.</p>
<div class="sourceCode" id="cb155"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">contrasts</span> <span class="op">&lt;-</span> <span class="fu">independentContrasts</span><span class="op">(</span></span>
<span>  traits <span class="op">=</span> <span class="fu">pivot_longer</span><span class="op">(</span><span class="va">chemical_blooms</span>, cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"compound"</span>, values_to <span class="op">=</span> <span class="st">"value"</span><span class="op">)</span>,</span>
<span>  column_w_names_of_tiplabels <span class="op">=</span> <span class="st">"label"</span>,</span>
<span>  column_w_names_of_traits <span class="op">=</span> <span class="st">"compound"</span>,</span>
<span>  column_w_values_for_traits <span class="op">=</span> <span class="st">"value"</span>,</span>
<span>  tree <span class="op">=</span> <span class="va">chemical_bloom_tree</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># buildLinearModel(</span></span>
<span><span class="co">#   data = contrasts,</span></span>
<span><span class="co">#   formula = "Fatty_acids = Alkanes + 0"</span></span>
<span><span class="co"># ) -&gt; model</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># ggplot(model$data) +</span></span>
<span><span class="co">#   geom_point(aes(x = input_x, y = input_y)) +</span></span>
<span><span class="co">#   geom_line(aes(x = model_x, model_y))</span></span></code></pre></div>
</div>
<div id="ancestraltraits" class="section level2 unnumbered">
<h2>ancestralTraits<a class="anchor" aria-label="anchor" href="#ancestraltraits"><i class="fas fa-link"></i></a>
</h2>
<p>Ancestral trait reconstruction is a method to infer the characteristics (or “traits”) of ancestral organisms based on the traits of their modern descendants. By examining the traits of present-day species and using phylogenetic trees, we can estimate or “reconstruct” the traits of common ancestors. This method can be applied to various types of traits, including continuously varying and discrete traits. Ancestral trait reconstruction helps us gain insights into the evolutionary processes and the historical transitions that led to current biodiversity. <code>phylochemistry</code> provides the function <code>ancestralTraits</code> to perform these operations. Note that <code>ancestralTraits</code> is different from <code>buildTree</code>s “ancestral_states”. “ancestral_states” estimates ancestral sequence states at phylogeny nodes, while <code>ancestralTraits</code> will estimate the traits of an ancestor, given the traits of extant species that are present on the leaves of a phylogeny. Here is an example. Please note that ancestralTraits accepts data in a long-style data frame.</p>
<div class="sourceCode" id="cb156"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anc_traits_tree</span> <span class="op">&lt;-</span> <span class="fu">ancestralTraits</span><span class="op">(</span></span>
<span>  traits <span class="op">=</span> <span class="fu">pivot_longer</span><span class="op">(</span><span class="va">chemical_blooms</span>, cols <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span>,</span>
<span>  column_w_names_of_tiplabels <span class="op">=</span> <span class="st">"label"</span>,</span>
<span>  column_w_names_of_traits <span class="op">=</span> <span class="st">"name"</span>,</span>
<span>  column_w_values_for_traits <span class="op">=</span> <span class="st">"value"</span>,</span>
<span>  tree <span class="op">=</span> <span class="va">chemical_bloom_tree</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">anc_traits_tree</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 6 × 11</span></span>
<span><span class="co">##   parent  node branch.length label  isTip     x     y branch</span></span>
<span><span class="co">##    &lt;int&gt; &lt;dbl&gt;         &lt;dbl&gt; &lt;chr&gt;  &lt;lgl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;</span></span>
<span><span class="co">## 1     80     1          290. Ginkg… TRUE   352.     1   207.</span></span>
<span><span class="co">## 2     80     1          290. Ginkg… TRUE   352.     1   207.</span></span>
<span><span class="co">## 3     80     1          290. Ginkg… TRUE   352.     1   207.</span></span>
<span><span class="co">## 4     80     1          290. Ginkg… TRUE   352.     1   207.</span></span>
<span><span class="co">## 5     80     1          290. Ginkg… TRUE   352.     1   207.</span></span>
<span><span class="co">## 6     80     1          290. Ginkg… TRUE   352.     1   207.</span></span>
<span><span class="co">## # ℹ 3 more variables: angle &lt;dbl&gt;, trait &lt;chr&gt;, value &lt;dbl&gt;</span></span></code></pre></div>
<p>In addition to providing ancestral state estimations, there is also a function for plotting those estimations on a phylogeny: <code>geom_ancestral_pie</code>. Here is an example. Note that <code>cols</code> is a vector of column numbers that correspond to the traits of interest. <code>pie_size</code> is the size of the pie chart that will be plotted at each node. <code>geom_ancestral_pie</code> relies on having columns in its input called <code>trait</code> and <code>value</code>, such as those output by <code>ancestralTraits</code>. Note that if you are passing an object to ggtree() that has duplicate node names, you will need to use the <code>distinct</code> function to remove the duplicates, otherwise geom_ancestral_pie will get confused about where to place the pies.</p>
<div class="sourceCode" id="cb157"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span></span>
<span>  <span class="fu">distinct</span><span class="op">(</span><span class="va">anc_traits_tree</span>, <span class="va">node</span>, .keep_all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_ancestral_pie</span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">anc_traits_tree</span>, <span class="va">isTip</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">)</span>,</span>
<span>    pie_size <span class="op">=</span> <span class="fl">0.1</span>, pie_alpha <span class="op">=</span> <span class="fl">1</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span>offset <span class="op">=</span> <span class="fl">20</span>, align <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">650</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_void</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-220-1.png" width="100%" style="display: block; margin: auto;"></div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="phylogenies.html">phylogenies</a></div>
<div class="next"><a href="comparative-genomics.html">comparative genomics</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <h2>Note: Second Edition is under construction 🏗</h2>
    <p>Now is a great time to provide feedback</p>
        <ul class="list-unstyled">
<li><a href="https://forms.gle/SZmB2Ct2exE2dBwv9">Provide feedback (5 min)</a></li>
          <!-- <li><a href="https://geocompr.robinlovelace.net/#reproducibility">Install updated packages</a></li> -->
          <!-- <li><a href="https://github.com/Robinlovelace/geocompr/issues">Open an issue <i class="fas fa-question"></i></a></li> -->
          <!-- <li><a href="https://discord.gg/Te3gWeDwmf">Chat on Discord <i class="fab fa-discord"></i></a></li> -->
        </ul>
<hr>
<nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#phylogenetic-analyses">phylogenetic analyses</a></li>
<li><a class="nav-link" href="#phylogeneticsignal">phylogeneticSignal</a></li>
<li><a class="nav-link" href="#independentcontrasts">independentContrasts</a></li>
<li><a class="nav-link" href="#ancestraltraits">ancestralTraits</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/blob/main/index.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/edit/main/index.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Integrated Bioanalytics</strong>" was written by Lucas Busta and members of the Busta lab. It was last built on 2023-10-02.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
