<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>phylogenetic analyses | Integrated Bioanalytics</title>
<meta name="author" content="Lucas Busta and members of the Busta lab">
<meta name="description" content="1.9 phylogeneticSignal Phylogenetic signal is a measure of the degree to which related species share similar trait values. It is used to determine whether a trait has evolved in a manner that is...">
<meta name="generator" content="bookdown 0.35 with bs4_book()">
<meta property="og:title" content="phylogenetic analyses | Integrated Bioanalytics">
<meta property="og:type" content="book">
<meta property="og:description" content="1.9 phylogeneticSignal Phylogenetic signal is a measure of the degree to which related species share similar trait values. It is used to determine whether a trait has evolved in a manner that is...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="phylogenetic analyses | Integrated Bioanalytics">
<meta name="twitter:description" content="1.9 phylogeneticSignal Phylogenetic signal is a measure of the degree to which related species share similar trait values. It is used to determine whether a trait has evolved in a manner that is...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><link href="libs/Lato-0.4.6/font.css" rel="stylesheet">
<link href="libs/Roboto_Mono-0.4.6/font.css" rel="stylesheet">
<link href="libs/Montserrat-0.4.6/font.css" rel="stylesheet">
<script src="libs/bs3compat-0.4.2/transition.js"></script><script src="libs/bs3compat-0.4.2/tabs.js"></script><script src="libs/bs3compat-0.4.2/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><link rel="icon" type="image/png" sizes="32x32" href="images/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="images/favicon-16x16.png">
<script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-99618359-1', 'auto');
      ga('send', 'pageview');

    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="style/style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h2>
        <a href="index.html" title="">Integrated Bioanalytics</a>
      </h2>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">WELCOME</a></li>
<li class="book-part">DATA ANALYSIS IN R</li>
<li><a class="" href="overview.html">overview</a></li>
<li><a class="" href="installation.html">installation</a></li>
<li><a class="" href="data-visualization.html">data visualization</a></li>
<li><a class="" href="data-wrangling-and-summaries.html">data wrangling and summaries</a></li>
<li><a class="" href="dimensionality-reduction.html">dimensionality reduction</a></li>
<li><a class="" href="clustering.html">clustering</a></li>
<li><a class="" href="models.html">models</a></li>
<li><a class="" href="comparing-means.html">comparing means</a></li>
<li><a class="" href="special-topics.html">special topics</a></li>
<li class="book-part">GC-MS DATA</li>
<li><a class="" href="mass-spectrometric-analysis.html">mass spectrometric analysis</a></li>
<li class="book-part">TRANSCRIPTOME ANALYSIS</li>
<li><a class="" href="transcriptome-assembly.html">transcriptome assembly</a></li>
<li class="book-part">EVOLUTIONARY ANALYSIS</li>
<li><a class="" href="blast.html">blast</a></li>
<li><a class="" href="alignments.html">alignments</a></li>
<li><a class="" href="phylogenies.html">phylogenies</a></li>
<li><a class="active" href="phylogenetic-analyses.html">phylogenetic analyses</a></li>
<li><a class="" href="comparative-genomics.html">comparative genomics</a></li>
<li class="book-part">LANGUAGE MODELS</li>
<li><a class="" href="completiongpt.html">completionGPT</a></li>
<li><a class="" href="analyzeliterature.html">analyzeLiterature</a></li>
<li class="book-part">SCIENTIFIC WRITING</li>
<li><a class="" href="overview-1.html">overview</a></li>
<li><a class="" href="figures-captions.html">figures &amp; captions</a></li>
<li><a class="" href="results-and-discussion.html">results and discussion</a></li>
<li><a class="" href="conclusion-and-introduction.html">conclusion and introduction</a></li>
<li><a class="" href="abstract-and-title.html">abstract and title</a></li>
<li class="book-part">APPENDIX</li>
<li><a class="" href="links.html">links</a></li>
<li><a class="" href="r-faq.html">r faq</a></li>
<li><a class="" href="templates.html">templates</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="phylogenetic-analyses" class="section level1 unnumbered">
<h1>phylogenetic analyses<a class="anchor" aria-label="anchor" href="#phylogenetic-analyses"><i class="fas fa-link"></i></a>
</h1>
<div id="phylogeneticsignal" class="section level2" number="1.9">
<h2>
<span class="header-section-number">1.9</span> phylogeneticSignal<a class="anchor" aria-label="anchor" href="#phylogeneticsignal"><i class="fas fa-link"></i></a>
</h2>
<p>Phylogenetic signal is a measure of the degree to which related species share similar trait values. It is used to determine whether a trait has evolved in a manner that is consistent with the species’ evolutionary history. <code>phylochemistry</code> provides the <code>phylogeneticSignal</code> function, which can be used to calculate phylogenetic signal for a given set of traits and a phylogenetic tree. Here is an example:</p>
<div class="sourceCode" id="cb151"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chemical_bloom_tree</span> <span class="op">&lt;-</span> <span class="fu">buildTree</span><span class="op">(</span></span>
<span>  scaffold_type <span class="op">=</span> <span class="st">"newick"</span>,</span>
<span>  scaffold_in_path <span class="op">=</span> <span class="st">"http://thebustalab.github.io/data/angiosperms.newick"</span>,</span>
<span>  members <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">chemical_blooms</span><span class="op">$</span><span class="va">label</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Scaffold newick tip Sabal_pumos substituted with Sabal_palmetto </span></span>
<span><span class="co">## Scaffold newick tip Iris_lazica substituted with Iris_sp </span></span>
<span><span class="co">## Scaffold newick tip Iris_lacustris substituted with Iris_germanica </span></span>
<span><span class="co">## Scaffold newick tip Allium_textile substituted with Allium_sp </span></span>
<span><span class="co">## Scaffold newick tip Allium_subhirsutum substituted with Allium_brevistylum </span></span>
<span><span class="co">## Scaffold newick tip Ornithogalum_saundersiae substituted with Ornithogalum_candicans </span></span>
<span><span class="co">## Scaffold newick tip Hosta_plantaginea substituted with Hosta_sp </span></span>
<span><span class="co">## Scaffold newick tip Agave_striata substituted with Agave_cerulata </span></span>
<span><span class="co">## Scaffold newick tip Agave_tequilana substituted with Agave_chrysantha </span></span>
<span><span class="co">## Scaffold newick tip Aristolochia_serpentaria substituted with Aristolochia_labiata </span></span>
<span><span class="co">## Scaffold newick tip Thalictrum_clavatum substituted with Thalictrum_rochebrunianum </span></span>
<span><span class="co">## Scaffold newick tip Delphinium_pictum substituted with Delphinium_elatum </span></span>
<span><span class="co">## Scaffold newick tip Ferocactus_recurvus substituted with Ferocactus_wislizeni </span></span>
<span><span class="co">## Scaffold newick tip Opuntia_articulata substituted with Opuntia_sp </span></span>
<span><span class="co">## Scaffold newick tip Opuntia_quimilo substituted with Opuntia_robusta </span></span>
<span><span class="co">## Scaffold newick tip Cylindropuntia_fulgida substituted with Cylindropuntia_bigelovii </span></span>
<span><span class="co">## Scaffold newick tip Cylindropuntia_prolifera substituted with Cylindropuntia_versicolor </span></span>
<span><span class="co">## Scaffold newick tip Cylindropuntia_echinocarpa substituted with Cylindropuntia_ramosissima </span></span>
<span><span class="co">## Scaffold newick tip Kirengeshoma_palmata substituted with Kirengeshoma_Palmata </span></span>
<span><span class="co">## Scaffold newick tip Lavandula_bipinnata substituted with Lavandula_sp </span></span>
<span><span class="co">## Scaffold newick tip Rudbeckia_hirta substituted with Rudbeckia_occidentalis </span></span>
<span><span class="co">## Scaffold newick tip Crassula_campestris substituted with Crassula_ovata </span></span>
<span><span class="co">## Scaffold newick tip Crassula_tillaea substituted with Crassula_deceptor </span></span>
<span><span class="co">## Scaffold newick tip Crassula_alata substituted with Crassula_arborescens </span></span>
<span><span class="co">## Scaffold newick tip Crassula_colligata substituted with Crassula_perfoliata </span></span>
<span><span class="co">## Scaffold newick tip Hylotelephium_erythrostictum substituted with Hylotelephium_sp </span></span>
<span><span class="co">## Scaffold newick tip Echeveria_setosa substituted with Echeveria_pulidonis </span></span>
<span><span class="co">## Scaffold newick tip Bryophyllum_pinnatum substituted with Bryophyllum_fedtschenkoi </span></span>
<span><span class="co">## Scaffold newick tip Kalanchoe_linearifolia substituted with Kalanchoe_marginata </span></span>
<span><span class="co">## Scaffold newick tip Kalanchoe_tomentosa substituted with Kalanchoe_luciae </span></span>
<span><span class="co">## Scaffold newick tip Kalanchoe_beharensis substituted with Kalanchoe_thrysiflora </span></span>
<span><span class="co">## Scaffold newick tip Iliamna_latibracteata substituted with Iliamna_rivularis </span></span>
<span><span class="co">## Scaffold newick tip Euphorbia_lathyris substituted with Euphorbia_resinifera </span></span>
<span><span class="co">## Scaffold newick tip Quercus_valdinervosa substituted with Quercus_muehlenbergii </span></span>
<span><span class="co">## Scaffold newick tip Rubus_repens substituted with Rubus_sp </span></span>
<span><span class="co">## Pro tip: most tree read/write functions reset node numbers.</span></span>
<span><span class="co">## Fortify your tree and save it as a csv file to preserve node numbering.</span></span>
<span><span class="co">## Do not save your tree as a newick or nexus file.</span></span>
<span></span>
<span><span class="fu">phylogeneticSignal</span><span class="op">(</span></span>
<span>  traits <span class="op">=</span> <span class="va">chemical_blooms</span>,</span>
<span>  tree <span class="op">=</span> <span class="va">chemical_bloom_tree</span></span>
<span><span class="op">)</span></span>
<span><span class="co">##             trait trait_type n_species number_of_levels</span></span>
<span><span class="co">## 1         Alkanes continuous        78               NA</span></span>
<span><span class="co">## 2    Sec_Alcohols continuous        78               NA</span></span>
<span><span class="co">## 3          Others continuous        78               NA</span></span>
<span><span class="co">## 4     Fatty_acids continuous        78               NA</span></span>
<span><span class="co">## 5        Alcohols continuous        78               NA</span></span>
<span><span class="co">## 6   Triterpenoids continuous        78               NA</span></span>
<span><span class="co">## 7         Ketones continuous        78               NA</span></span>
<span><span class="co">## 8 Other_compounds continuous        78               NA</span></span>
<span><span class="co">## 9       Aldehydes continuous        78               NA</span></span>
<span><span class="co">##   evolutionary_transitions_observed</span></span>
<span><span class="co">## 1                                NA</span></span>
<span><span class="co">## 2                                NA</span></span>
<span><span class="co">## 3                                NA</span></span>
<span><span class="co">## 4                                NA</span></span>
<span><span class="co">## 5                                NA</span></span>
<span><span class="co">## 6                                NA</span></span>
<span><span class="co">## 7                                NA</span></span>
<span><span class="co">## 8                                NA</span></span>
<span><span class="co">## 9                                NA</span></span>
<span><span class="co">##   median_evolutionary_transitions_in_randomization</span></span>
<span><span class="co">## 1                                               NA</span></span>
<span><span class="co">## 2                                               NA</span></span>
<span><span class="co">## 3                                               NA</span></span>
<span><span class="co">## 4                                               NA</span></span>
<span><span class="co">## 5                                               NA</span></span>
<span><span class="co">## 6                                               NA</span></span>
<span><span class="co">## 7                                               NA</span></span>
<span><span class="co">## 8                                               NA</span></span>
<span><span class="co">## 9                                               NA</span></span>
<span><span class="co">##   minimum_evolutionary_transitions_in_randomization</span></span>
<span><span class="co">## 1                                                NA</span></span>
<span><span class="co">## 2                                                NA</span></span>
<span><span class="co">## 3                                                NA</span></span>
<span><span class="co">## 4                                                NA</span></span>
<span><span class="co">## 5                                                NA</span></span>
<span><span class="co">## 6                                                NA</span></span>
<span><span class="co">## 7                                                NA</span></span>
<span><span class="co">## 8                                                NA</span></span>
<span><span class="co">## 9                                                NA</span></span>
<span><span class="co">##   evolutionary_transitions_in_randomization p_value</span></span>
<span><span class="co">## 1                                        NA   0.160</span></span>
<span><span class="co">## 2                                        NA   0.001</span></span>
<span><span class="co">## 3                                        NA   0.731</span></span>
<span><span class="co">## 4                                        NA   0.261</span></span>
<span><span class="co">## 5                                        NA   0.327</span></span>
<span><span class="co">## 6                                        NA   0.001</span></span>
<span><span class="co">## 7                                        NA   0.043</span></span>
<span><span class="co">## 8                                        NA   0.849</span></span>
<span><span class="co">## 9                                        NA   0.931</span></span></code></pre></div>
</div>
<div id="independentcontrasts" class="section level2" number="1.10">
<h2>
<span class="header-section-number">1.10</span> independentContrasts<a class="anchor" aria-label="anchor" href="#independentcontrasts"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb152"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">contrasts</span> <span class="op">&lt;-</span> <span class="fu">independentContrasts</span><span class="op">(</span></span>
<span>  traits <span class="op">=</span> <span class="va">chemical_blooms</span>,</span>
<span>  tree <span class="op">=</span> <span class="va">chemical_bloom_tree</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">buildLinearModel</span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">contrasts</span>,</span>
<span>  formula <span class="op">=</span> <span class="st">"Fatty_acids = Alcohols + 0"</span></span>
<span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">model</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">model</span><span class="op">$</span><span class="va">data</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">input_x</span>, y <span class="op">=</span> <span class="va">input_y</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">model_x</span>, <span class="va">model_y</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-213-1.png" width="100%" style="display: block; margin: auto;"></div>
</div>
<div id="ancestraltraits" class="section level2" number="1.11">
<h2>
<span class="header-section-number">1.11</span> ancestralTraits<a class="anchor" aria-label="anchor" href="#ancestraltraits"><i class="fas fa-link"></i></a>
</h2>
<p>Note that this is different from <code>buildTree</code>’s “ancestral_states”, which estimates ancestral sequence states at phylogeny nodes. Instead, <code>ancestralTraits</code> will estimate the traits of an ancestor, given the traits of extant species that are present on the leaves of a phylogeny.</p>
<p>ancestral states: <a href="https://www.phytools.org/eqg2015/asr.html" class="uri">https://www.phytools.org/eqg2015/asr.html</a></p>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="phylogenies.html">phylogenies</a></div>
<div class="next"><a href="comparative-genomics.html">comparative genomics</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <h2>Note: Second Edition is under construction 🏗</h2>
    <p>Now is a great time to provide feedback</p>
        <ul class="list-unstyled">
<li><a href="https://forms.gle/SZmB2Ct2exE2dBwv9">Provide feedback (5 min)</a></li>
          <!-- <li><a href="https://geocompr.robinlovelace.net/#reproducibility">Install updated packages</a></li> -->
          <!-- <li><a href="https://github.com/Robinlovelace/geocompr/issues">Open an issue <i class="fas fa-question"></i></a></li> -->
          <!-- <li><a href="https://discord.gg/Te3gWeDwmf">Chat on Discord <i class="fab fa-discord"></i></a></li> -->
        </ul>
<hr>
<nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#phylogenetic-analyses">phylogenetic analyses</a></li>
<li><a class="nav-link" href="#phylogeneticsignal"><span class="header-section-number">1.9</span> phylogeneticSignal</a></li>
<li><a class="nav-link" href="#independentcontrasts"><span class="header-section-number">1.10</span> independentContrasts</a></li>
<li><a class="nav-link" href="#ancestraltraits"><span class="header-section-number">1.11</span> ancestralTraits</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/blob/main/index.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/edit/main/index.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Integrated Bioanalytics</strong>" was written by Lucas Busta and members of the Busta lab. It was last built on 2023-08-27.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
