<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>phylogenetic analyses | Integrated Bioanalytics</title>
<meta name="author" content="Lucas Busta and members of the Busta lab">
<meta name="description" content='We use a wrapper function to run phylogenetic comparative analyses. It chemical_bloom_tree &lt;- buildTree(  scaffold_type = "newick",  scaffold_in_path =...'>
<meta name="generator" content="bookdown 0.44 with bs4_book()">
<meta property="og:title" content="phylogenetic analyses | Integrated Bioanalytics">
<meta property="og:type" content="book">
<meta property="og:description" content='We use a wrapper function to run phylogenetic comparative analyses. It chemical_bloom_tree &lt;- buildTree(  scaffold_type = "newick",  scaffold_in_path =...'>
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="phylogenetic analyses | Integrated Bioanalytics">
<meta name="twitter:description" content='We use a wrapper function to run phylogenetic comparative analyses. It chemical_bloom_tree &lt;- buildTree(  scaffold_type = "newick",  scaffold_in_path =...'>
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><link href="libs/Lato-0.4.10/font.css" rel="stylesheet">
<link href="libs/Roboto_Mono-0.4.10/font.css" rel="stylesheet">
<link href="libs/Montserrat-0.4.10/font.css" rel="stylesheet">
<script src="libs/bs3compat-0.9.0/transition.js"></script><script src="libs/bs3compat-0.9.0/tabs.js"></script><script src="libs/bs3compat-0.9.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><link rel="icon" type="image/png" sizes="32x32" href="images/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="images/favicon-16x16.png">
<script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-99618359-1', 'auto');
      ga('send', 'pageview');

    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="style/style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h2>
        <a href="index.html" title="">Integrated Bioanalytics</a>
      </h2>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">WELCOME</a></li>
<li class="book-part">GETTING STARTED</li>
<li><a class="" href="overview.html">overview</a></li>
<li><a class="" href="installation.html">installation</a></li>
<li class="book-part">DATA VISUALIZATION</li>
<li><a class="" href="data-visualization-i.html">data visualization I</a></li>
<li><a class="" href="data-visualization-ii.html">data visualization II</a></li>
<li><a class="" href="data-visualization-iii.html">data visualization III</a></li>
<li class="book-part">STATISTICAL METHODS</li>
<li><a class="" href="wrangling-and-summaries.html">wrangling and summaries</a></li>
<li><a class="" href="hierarchical-clustering.html">hierarchical clustering</a></li>
<li><a class="" href="dimensional-reduction.html">dimensional reduction</a></li>
<li><a class="" href="flat-clustering.html">flat clustering</a></li>
<li><a class="" href="comparing-means.html">comparing means</a></li>
<li class="book-part">MODELS</li>
<li><a class="" href="numerical-models.html">numerical models</a></li>
<li><a class="" href="embedding-models.html">embedding models</a></li>
<li class="book-part">SEQUENCE ANALYSIS</li>
<li><a class="" href="homology.html">homology</a></li>
<li><a class="" href="alignments.html">alignments</a></li>
<li><a class="" href="phylogenies.html">phylogenies</a></li>
<li><a class="active" href="phylogenetic-analyses.html">phylogenetic analyses</a></li>
<li class="book-part">ANALYTICAL REPORTS</li>
<li><a class="" href="figures-captions.html">figures &amp; captions</a></li>
<li><a class="" href="question-driven-report.html">question-driven report</a></li>
<li class="book-part">APPENDIX</li>
<li><a class="" href="links.html">links</a></li>
<li><a class="" href="datasets.html">datasets</a></li>
<li><a class="" href="r-faq.html">r faq</a></li>
<li><a class="" href="templates.html">templates</a></li>
<li><a class="" href="loading-analyzegcmsdata.html">loading analyzeGCMSdata</a></li>
<li><a class="" href="using-analyzegcmsdata.html">using analyzeGCMSdata</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="phylogenetic-analyses" class="section level1 unnumbered">
<h1>phylogenetic analyses<a class="anchor" aria-label="anchor" href="#phylogenetic-analyses"><i class="fas fa-link"></i></a>
</h1>
<p>We use a wrapper function to run phylogenetic comparative analyses. It</p>
<div class="sourceCode" id="cb173"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chemical_bloom_tree</span> <span class="op">&lt;-</span> <span class="fu">buildTree</span><span class="op">(</span></span>
<span>  scaffold_type <span class="op">=</span> <span class="st">"newick"</span>,</span>
<span>  scaffold_in_path <span class="op">=</span> <span class="st">"http://thebustalab.github.io/data/angiosperms.newick"</span>,</span>
<span>  members <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">chemical_blooms</span><span class="op">$</span><span class="va">label</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## IMPORTANT: Some species substitutions or removals were made as part of buildTree. Run build_tree_substitutions() to see them all.</span></span>
<span><span class="co">## Pro tip: most tree read/write functions reset node numbers. Fortify your tree and save it as a csv file to preserve node numbering.</span></span>
<span></span>
<span><span class="fu">runPhylogeneticAnalyses</span><span class="op">(</span></span>
<span>    traits <span class="op">=</span> <span class="fu">pivot_longer</span><span class="op">(</span><span class="va">chemical_blooms</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"trait"</span>, values_to <span class="op">=</span> <span class="st">"value"</span><span class="op">)</span>,</span>
<span>    column_w_names_of_tiplabels <span class="op">=</span> <span class="st">"label"</span>,</span>
<span>    column_w_names_of_traits <span class="op">=</span> <span class="st">"trait"</span>,</span>
<span>    column_w_values_for_traits <span class="op">=</span> <span class="st">"value"</span>,</span>
<span>    tree <span class="op">=</span> <span class="va">chemical_bloom_tree</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Joining with `by = join_by(trait)`</span></span>
<span><span class="co">## # A tibble: 310 × 17</span></span>
<span><span class="co">##    parent  node branch.length label isTip     x     y branch</span></span>
<span><span class="co">##     &lt;int&gt; &lt;dbl&gt;         &lt;dbl&gt; &lt;chr&gt; &lt;lgl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;</span></span>
<span><span class="co">##  1     80     1         290.  Gink… TRUE   352.     1   207.</span></span>
<span><span class="co">##  2     80     1         290.  Gink… TRUE   352.     1   207.</span></span>
<span><span class="co">##  3     81     2         267.  Pice… TRUE   352.     2   219.</span></span>
<span><span class="co">##  4     81     2         267.  Pice… TRUE   352.     2   219.</span></span>
<span><span class="co">##  5     81     3         267.  Cupr… TRUE   352.     3   219.</span></span>
<span><span class="co">##  6     81     3         267.  Cupr… TRUE   352.     3   219.</span></span>
<span><span class="co">##  7     84     4         135.  Eryt… TRUE   352.     5   285.</span></span>
<span><span class="co">##  8     84     4         135.  Eryt… TRUE   352.     5   285.</span></span>
<span><span class="co">##  9     86     5          16.2 Iris… TRUE   352.     6   344.</span></span>
<span><span class="co">## 10     86     5          16.2 Iris… TRUE   352.     6   344.</span></span>
<span><span class="co">## # ℹ 300 more rows</span></span>
<span><span class="co">## # ℹ 9 more variables: angle &lt;dbl&gt;, trait &lt;chr&gt;,</span></span>
<span><span class="co">## #   value &lt;dbl&gt;, trait_type &lt;chr&gt;,</span></span>
<span><span class="co">## #   phylogenetic_signal_k_value &lt;dbl&gt;,</span></span>
<span><span class="co">## #   phylogenetic_signal_k_p_value &lt;dbl&gt;,</span></span>
<span><span class="co">## #   phylogenetic_signal_lambda_value &lt;dbl&gt;,</span></span>
<span><span class="co">## #   phylogenetic_signal_lambda_p_value &lt;dbl&gt;, pic &lt;dbl&gt;</span></span></code></pre></div>
<p>For all the below, there are some structural requirements: (i) the tree needs to be a phylo object (ii) the traits need to be a data.frame in which each row is a species and each column is a variable, and (iii) the first column in the data.frame needs to be the names of the species and they must exactly match the tip labels of the tree (though they don’t have to be in the same order), for example:</p>
<div class="sourceCode" id="cb174"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chemical_bloom_tree</span> <span class="op">&lt;-</span> <span class="fu">buildTree</span><span class="op">(</span></span>
<span>  scaffold_type <span class="op">=</span> <span class="st">"newick"</span>,</span>
<span>  scaffold_in_path <span class="op">=</span> <span class="st">"http://thebustalab.github.io/data/angiosperms.newick"</span>,</span>
<span>  members <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">chemical_blooms</span><span class="op">$</span><span class="va">label</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## IMPORTANT: Some species substitutions or removals were made as part of buildTree. Run build_tree_substitutions() to see them all.</span></span>
<span><span class="co">## Pro tip: most tree read/write functions reset node numbers. Fortify your tree and save it as a csv file to preserve node numbering.</span></span></code></pre></div>
<div id="phylogeneticsignal" class="section level2 unnumbered">
<h2>phylogeneticSignal<a class="anchor" aria-label="anchor" href="#phylogeneticsignal"><i class="fas fa-link"></i></a>
</h2>
<p>Phylogenetic signal is a measure of the degree to which related species share similar trait values. It is used to determine whether a trait has evolved in a manner that is consistent with the species’ evolutionary history. <code>phylochemistry</code> provides the <code>phylogeneticSignal</code> function, which can be used to calculate phylogenetic signal for a given set of traits and a phylogenetic tree. Here is an example:</p>
<div class="sourceCode" id="cb175"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">phylogeneticSignal</span><span class="op">(</span></span>
<span>  traits <span class="op">=</span> <span class="fu">pivot_longer</span><span class="op">(</span><span class="va">chemical_blooms</span>, cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"compound"</span>, values_to <span class="op">=</span> <span class="st">"value"</span><span class="op">)</span>,</span>
<span>  column_w_names_of_tiplabels <span class="op">=</span> <span class="st">"label"</span>,</span>
<span>  column_w_names_of_traits <span class="op">=</span> <span class="st">"compound"</span>,</span>
<span>  column_w_values_for_traits <span class="op">=</span> <span class="st">"value"</span>,</span>
<span>  tree <span class="op">=</span> <span class="va">chemical_bloom_tree</span></span>
<span><span class="op">)</span></span>
<span><span class="co">##             trait trait_type n_species number_of_levels</span></span>
<span><span class="co">## 1         Alkanes continuous        78               NA</span></span>
<span><span class="co">## 2    Sec_Alcohols continuous        78               NA</span></span>
<span><span class="co">## 3          Others continuous        78               NA</span></span>
<span><span class="co">## 4     Fatty_acids continuous        78               NA</span></span>
<span><span class="co">## 5        Alcohols continuous        78               NA</span></span>
<span><span class="co">## 6   Triterpenoids continuous        78               NA</span></span>
<span><span class="co">## 7         Ketones continuous        78               NA</span></span>
<span><span class="co">## 8 Other_compounds continuous        78               NA</span></span>
<span><span class="co">## 9       Aldehydes continuous        78               NA</span></span>
<span><span class="co">##   evolutionary_transitions_observed</span></span>
<span><span class="co">## 1                                NA</span></span>
<span><span class="co">## 2                                NA</span></span>
<span><span class="co">## 3                                NA</span></span>
<span><span class="co">## 4                                NA</span></span>
<span><span class="co">## 5                                NA</span></span>
<span><span class="co">## 6                                NA</span></span>
<span><span class="co">## 7                                NA</span></span>
<span><span class="co">## 8                                NA</span></span>
<span><span class="co">## 9                                NA</span></span>
<span><span class="co">##   median_evolutionary_transitions_in_randomization</span></span>
<span><span class="co">## 1                                               NA</span></span>
<span><span class="co">## 2                                               NA</span></span>
<span><span class="co">## 3                                               NA</span></span>
<span><span class="co">## 4                                               NA</span></span>
<span><span class="co">## 5                                               NA</span></span>
<span><span class="co">## 6                                               NA</span></span>
<span><span class="co">## 7                                               NA</span></span>
<span><span class="co">## 8                                               NA</span></span>
<span><span class="co">## 9                                               NA</span></span>
<span><span class="co">##   minimum_evolutionary_transitions_in_randomization</span></span>
<span><span class="co">## 1                                                NA</span></span>
<span><span class="co">## 2                                                NA</span></span>
<span><span class="co">## 3                                                NA</span></span>
<span><span class="co">## 4                                                NA</span></span>
<span><span class="co">## 5                                                NA</span></span>
<span><span class="co">## 6                                                NA</span></span>
<span><span class="co">## 7                                                NA</span></span>
<span><span class="co">## 8                                                NA</span></span>
<span><span class="co">## 9                                                NA</span></span>
<span><span class="co">##   evolutionary_transitions_in_randomization</span></span>
<span><span class="co">## 1                                        NA</span></span>
<span><span class="co">## 2                                        NA</span></span>
<span><span class="co">## 3                                        NA</span></span>
<span><span class="co">## 4                                        NA</span></span>
<span><span class="co">## 5                                        NA</span></span>
<span><span class="co">## 6                                        NA</span></span>
<span><span class="co">## 7                                        NA</span></span>
<span><span class="co">## 8                                        NA</span></span>
<span><span class="co">## 9                                        NA</span></span>
<span><span class="co">##   phylogenetic_signal_k_value phylogenetic_signal_k_p_value</span></span>
<span><span class="co">## 1                 0.066016688                         0.167</span></span>
<span><span class="co">## 2                 2.045611108                         0.001</span></span>
<span><span class="co">## 3                 0.029719595                         0.711</span></span>
<span><span class="co">## 4                 0.069056092                         0.292</span></span>
<span><span class="co">## 5                 0.053761730                         0.354</span></span>
<span><span class="co">## 6                 0.566806846                         0.001</span></span>
<span><span class="co">## 7                 0.239488396                         0.030</span></span>
<span><span class="co">## 8                 0.018420367                         0.868</span></span>
<span><span class="co">## 9                 0.008613879                         0.928</span></span>
<span><span class="co">##   phylogenetic_signal_lambda_value</span></span>
<span><span class="co">## 1                           0.0001</span></span>
<span><span class="co">## 2                           0.9999</span></span>
<span><span class="co">## 3                           0.0001</span></span>
<span><span class="co">## 4                           0.0001</span></span>
<span><span class="co">## 5                           0.0001</span></span>
<span><span class="co">## 6                           0.9999</span></span>
<span><span class="co">## 7                           0.7874</span></span>
<span><span class="co">## 8                           0.0001</span></span>
<span><span class="co">## 9                           0.0001</span></span>
<span><span class="co">##   phylogenetic_signal_lambda_p_value</span></span>
<span><span class="co">## 1                              1.000</span></span>
<span><span class="co">## 2                              0.000</span></span>
<span><span class="co">## 3                              1.000</span></span>
<span><span class="co">## 4                              1.000</span></span>
<span><span class="co">## 5                              1.000</span></span>
<span><span class="co">## 6                              0.000</span></span>
<span><span class="co">## 7                              0.045</span></span>
<span><span class="co">## 8                              1.000</span></span>
<span><span class="co">## 9                              1.000</span></span></code></pre></div>
</div>
<div id="independentcontrasts" class="section level2 unnumbered">
<h2>independentContrasts<a class="anchor" aria-label="anchor" href="#independentcontrasts"><i class="fas fa-link"></i></a>
</h2>
<p>Phylogenetic independent contrasts are a method for analyzing the relationship between two or more traits while taking into account the evolutionary history of the species being studied. This method involves transforming the data in to “independent contrasts” to remove the effects of shared ancestry, allowing for more accurate analysis of the relationship between traits. <code>phylochemistry</code> provides the <code>independentContrasts</code> function to calculate phylogenetic independent contrasts for a given set of traits and a phylogenetic tree. Here is an example of calculating independent contrasts for an example dataset, followed by generating a linear model based on the contrasts.</p>
<div class="sourceCode" id="cb176"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">contrasts</span> <span class="op">&lt;-</span> <span class="fu">independentContrasts</span><span class="op">(</span></span>
<span>  traits <span class="op">=</span> <span class="fu">pivot_longer</span><span class="op">(</span><span class="va">chemical_blooms</span>, cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"compound"</span>, values_to <span class="op">=</span> <span class="st">"value"</span><span class="op">)</span>,</span>
<span>  column_w_names_of_tiplabels <span class="op">=</span> <span class="st">"label"</span>,</span>
<span>  column_w_names_of_traits <span class="op">=</span> <span class="st">"compound"</span>,</span>
<span>  column_w_values_for_traits <span class="op">=</span> <span class="st">"value"</span>,</span>
<span>  tree <span class="op">=</span> <span class="va">chemical_bloom_tree</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># buildLinearModel(</span></span>
<span><span class="co">#   data = contrasts,</span></span>
<span><span class="co">#   formula = "Fatty_acids = Alkanes + 0"</span></span>
<span><span class="co"># ) -&gt; model</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># ggplot(model$data) +</span></span>
<span><span class="co">#   geom_point(aes(x = input_x, y = input_y)) +</span></span>
<span><span class="co">#   geom_line(aes(x = model_x, model_y))</span></span></code></pre></div>
</div>
<div id="ancestraltraits" class="section level2 unnumbered">
<h2>ancestralTraits<a class="anchor" aria-label="anchor" href="#ancestraltraits"><i class="fas fa-link"></i></a>
</h2>
<p>Ancestral trait reconstruction is a method to infer the characteristics (or “traits”) of ancestral organisms based on the traits of their modern descendants. By examining the traits of present-day species and using phylogenetic trees, we can estimate or “reconstruct” the traits of common ancestors. This method can be applied to various types of traits, including continuously varying and discrete traits. Ancestral trait reconstruction helps us gain insights into the evolutionary processes and the historical transitions that led to current biodiversity. <code>phylochemistry</code> provides the function <code>ancestralTraits</code> to perform these operations. Note that <code>ancestralTraits</code> is different from <code>buildTree</code>s “ancestral_states”. “ancestral_states” estimates ancestral sequence states at phylogeny nodes, while <code>ancestralTraits</code> will estimate the traits of an ancestor, given the traits of extant species that are present on the leaves of a phylogeny. Here is an example. Please note that ancestralTraits accepts data in a long-style data frame.</p>
<div class="sourceCode" id="cb177"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anc_traits_tree</span> <span class="op">&lt;-</span> <span class="fu">ancestralTraits</span><span class="op">(</span></span>
<span>  traits <span class="op">=</span> <span class="fu">pivot_longer</span><span class="op">(</span><span class="va">chemical_blooms</span>, cols <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span>,</span>
<span>  column_w_names_of_tiplabels <span class="op">=</span> <span class="st">"label"</span>,</span>
<span>  column_w_names_of_traits <span class="op">=</span> <span class="st">"name"</span>,</span>
<span>  column_w_values_for_traits <span class="op">=</span> <span class="st">"value"</span>,</span>
<span>  tree <span class="op">=</span> <span class="va">chemical_bloom_tree</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">anc_traits_tree</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 6 × 11</span></span>
<span><span class="co">##   parent  node branch.length label  isTip     x     y branch</span></span>
<span><span class="co">##    &lt;int&gt; &lt;dbl&gt;         &lt;dbl&gt; &lt;chr&gt;  &lt;lgl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;</span></span>
<span><span class="co">## 1     80     1          290. Ginkg… TRUE   352.     1   207.</span></span>
<span><span class="co">## 2     80     1          290. Ginkg… TRUE   352.     1   207.</span></span>
<span><span class="co">## 3     80     1          290. Ginkg… TRUE   352.     1   207.</span></span>
<span><span class="co">## 4     80     1          290. Ginkg… TRUE   352.     1   207.</span></span>
<span><span class="co">## 5     80     1          290. Ginkg… TRUE   352.     1   207.</span></span>
<span><span class="co">## 6     80     1          290. Ginkg… TRUE   352.     1   207.</span></span>
<span><span class="co">## # ℹ 3 more variables: angle &lt;dbl&gt;, trait &lt;chr&gt;, value &lt;dbl&gt;</span></span></code></pre></div>
<p>In addition to providing ancestral state estimations, there is also a function for plotting those estimations on a phylogeny: <code>geom_ancestral_pie</code>. Here is an example. Note that <code>cols</code> is a vector of column numbers that correspond to the traits of interest. <code>pie_size</code> is the size of the pie chart that will be plotted at each node. <code>geom_ancestral_pie</code> relies on having columns in its input called <code>trait</code> and <code>value</code>, such as those output by <code>ancestralTraits</code>. Note that if you are passing an object to ggtree() that has duplicate node names, you will need to use the <code>distinct</code> function to remove the duplicates, otherwise geom_ancestral_pie will get confused about where to place the pies.</p>
<div class="sourceCode" id="cb178"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span></span>
<span>  <span class="fu">distinct</span><span class="op">(</span><span class="va">anc_traits_tree</span>, <span class="va">node</span>, .keep_all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_ancestral_pie</span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">anc_traits_tree</span>, <span class="va">isTip</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">)</span>,</span>
<span>    pie_size <span class="op">=</span> <span class="fl">0.1</span>, pie_alpha <span class="op">=</span> <span class="fl">1</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span>offset <span class="op">=</span> <span class="fl">20</span>, align <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">650</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-558-1.png" width="100%" style="display: block; margin: auto;"></div>
<hr>
<hr>
<hr>
</div>
</div>



<!-- # overview {-} -->
<!-- For your final project in this course you will use the techniques we have learned in class to analyze a large dataset, prepare high quality figures, and write a miniature manuscript describing the results: -->
<!-- ```{r fig.align='center', echo=FALSE, include=identical(knitr:::pandoc_to(), 'html'), results="markup"} -->
<!-- knitr:::include_graphics('https://thebustalab.github.io/integrated_bioanalytics/images/project_overview.png', dpi = NA) -->
<!-- ``` -->
<!-- 1. **Find a data set: Large!** >10ish variables, >5ish categories -->
<!-- + Sources: your research supervisor, CHEM5725 database spreadsheet, Google searches, Kaggle.com! -->
<!-- + Relevant to your research or interests (ideally). -->
<!-- + Requires approval from Dr. Busta. -->
<!-- 2. **Ask at least three scientific questions.** -->
<!-- + These should drive your data analyses. -->
<!-- + Requires approval from Dr. Busta. -->
<!-- 3. **Analyze your data using what you learned in class.** -->
<!-- + Refer to our book. -->
<!-- + Ask Dr. Busta for assistance. -->
<!-- 4. **Create a written overview of your analysis**. A mini-manuscript in R Markdown: -->
<!-- + *Content* similar to the articles we looked at in class, though shorter. -->
<!-- + *Layout* similar to this example ([pdf](https://github.com/thebustalab/thebustalab.github.io/blob/master/integrated_bioanalytics/final_project/final_project_example.pdf), [rmd](https://github.com/thebustalab/thebustalab.github.io/blob/master/integrated_bioanalytics/final_project/final_project_example.Rmd)). -->
<!-- ## scope {-} -->
<!-- When conducting a project of this type, it is very common for there to be mismatches in the scope of how the project was conducted and how the written report is presented (see image below). We often spend LOTS of time exploring our data and running into dead ends, conclusions that are mundane, or questions we can't answer. When we write a report on the project, we often focus the report of a specific discovery we made during our vast avenues of exploration, rather than boring the reader with all the mundane details. -->
<!-- ```{r fig.align='center', echo=FALSE, include=identical(knitr:::pandoc_to(), 'html'), results="markup"} -->
<!-- knitr:::include_graphics('https://thebustalab.github.io/integrated_bioanalytics/images/scope.jpeg', dpi = NA) -->
<!-- ``` -->
<!-- ## order and content {-} -->
<!-- The manuscript will be comprised of a title, abstract, introduction, results and discussion section, figures and captions, conclusions section, and at least five references. Please note the following when preparing your manuscript: the orders of presentation and preparation do not have to be the same (see the images below)! While in some instances a scientist may choose to write the components of a manuscript in the same order in which they appear on the page, this is not always the case. The order of preparation suggsted above is designed to minimize the amount of revision / re-writing that needs to be performed during the manuscript preparation process. Note that the suggested order of composition is in line with the class schedule for the rest of the semester. -->
<!-- ```{r fig.align='center', echo=FALSE, include=identical(knitr:::pandoc_to(), 'html'), results="markup"} -->
<!-- knitr:::include_graphics('https://thebustalab.github.io/integrated_bioanalytics/images/writing_order2.png', dpi = NA) -->
<!-- ``` -->
<!-- Here is a guide for documenting analysis in a format that is polished and comprehensible. Note that each section is written for a specific and slightly unique audience. Utilizing R Markdown, the document created will convey findings and narrate the research process. In crafting your R Markdown document, ensure that code remains concealed in the final presentation by including echo = FALSE within chunk headers. This action hides the R code blocks in the output, yet permits their execution for generating figures and results. -->
<!-- TITLE: Craft a title that’s both clear and descriptive. It should be accessible to specialists in the field as well as the wider scientific audience, avoiding overly technical language that might limit its broader appeal. -->
<!-- ABSTRACT: Summarize the study in the abstract, providing detail that will be informative to experts and also comprehensible to those outside the field. It should briefly outline the study’s aims, methods, key results, and conclusions, offering a snapshot of the entire project. -->
<!-- INTRODUCTION: In the introduction, present the context and significance of the research in a way that's understandable to researchers and those not as versed in the subject. Clearly enumerate the multiple research questions that the you aim to answer, highlighting the research's relevance and framing the inquiry. -->
<!-- METHODS: Describe where the dataset was sourced, any data processing steps taken, and provide a detailed description of the analysis procedures utilized in RStudio. This section should be detailed enough to allow for replication and validation, yet clearly written to be accessible to non-experts. The explanation of the methods should help readers understand how the research questions were addressed. -->
<!-- RESULTS AND DISCUSSION WITH FIGURES AND CAPTIONS: Integrate your findings with clear, illustrative figures and captions within this section, ensuring they can be understood independently of the text for visual learners. The written portion should concisely interpret the results in light of the research questions, discussing the significance of the findings in a way that appeals to both those interested in the analytical nuances and those seeking to understand the overall implications. -->
<!-- CONCLUSION: Summarize the primary insights and their relevance, keeping this section succinct and to the point. It should crystallize the main findings and their contribution to the field, suited for readers who seek a quick synopsis without delving into the full text. -->
<!-- REFERENCES: Include at least five references to substantiate your research, ensuring they are pertinent and formatted to facilitate easy follow-up for interested readers. The references should be organized to serve both as a trail for fellow researchers and a resource for those who are less experienced in the academic discourse. -->
<!-- ## scientific questions {-} -->
<!-- Scientific questions are pivotal in plant chemistry research, especially when examining the quantification of compounds in various plants, tissues, or environments. They orient the scope of inquiry and dictate the choice of analytical methods to draw pertinent conclusions from data. -->
<!-- DESCRIPTIVE questions might catalog the variety or concentration of phytochemicals present in a given species, often employing summary statistics to encapsulate the data: -->
<!-- - What is the average concentration of alkaloids found in the leaves of nightshade plants in temperate zones? -->
<!-- - How do the levels of flavonoids vary among different tissues of the grapevine? -->
<!-- - What is the frequency distribution of terpene profiles in pine populations across different altitudes? -->
<!-- CORRELATIVE questions investigate the relationships between environmental factors and chemical expression in plants, typically utilizing regression modeling: -->
<!-- - Does the level of UV radiation correlate with the production of protective anthocyanins in vineyard grape varieties? -->
<!-- - How is the accumulation of heavy metals in fern tissues related to soil pollution levels? -->
<!-- COMPARATIVE questions explore variations or consistencies across groups or conditions, often answered through statistical comparisons or pattern recognition methods like clustering: -->
<!-- - Which are more similar in their secondary metabolite profiles, medicinal herbs grown in greenhouse conditions or in the wild? -->
<!-- - What distinguishes the phenolic compound content in shade-grown coffee plants versus those grown in direct sunlight? -->
<!-- - Is there a significant difference in essential oil compositions between lavender plants cultivated in different soil types? -->
<!-- Unclear: How should social networking sites address the harm they cause?
Clear: What action should social networking sites like MySpace and Facebook take to protect users’ personal information and privacy?

The unclear version of this question doesn’t specify which social networking sites or suggest what kind of harm the sites might be causing. It also assumes that this “harm” is proven and/or accepted. The clearer version specifies sites (MySpace and Facebook), the type of potential harm (privacy issues), and who may be experiencing that harm (users). A strong research question should never leave room for ambiguity or interpretation.

Unfocused: What is the effect on the environment from global warming?
Focused: What is the most significant effect of glacial melting on the lives of penguins in Antarctica?

The unfocused research question is so broad that it couldn’t be adequately answered in a book-length piece, let alone a standard college-level paper. The focused version narrows down to a specific effect of global warming (glacial melting), a specific place (Antarctica), and a specific animal that is affected (penguins). It also requires the writer to take a stance on which effect has the greatest impact on the affected animal. When in doubt, make a research question as narrow and focused as possible.

Too simple: How are doctors addressing diabetes in the U.S.?
Appropriately Complex:  What main environmental, behavioral, and genetic factors predict whether Americans will develop diabetes, and how can these commonalities be used to aid the medical community in prevention of the disease? -->
<div class="chapter-nav">
<div class="prev"><a href="phylogenies.html">phylogenies</a></div>
<div class="next"><a href="figures-captions.html">figures &amp; captions</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <h2>Note: Second Edition is under construction 🏗</h2>
    <p>Now is a great time to provide feedback</p>
        <ul class="list-unstyled">
<li><a href="https://forms.gle/SZmB2Ct2exE2dBwv9">Provide feedback (5 min)</a></li>
          <!-- <li><a href="https://geocompr.robinlovelace.net/#reproducibility">Install updated packages</a></li> -->
          <!-- <li><a href="https://github.com/Robinlovelace/geocompr/issues">Open an issue <i class="fas fa-question"></i></a></li> -->
          <!-- <li><a href="https://discord.gg/Te3gWeDwmf">Chat on Discord <i class="fab fa-discord"></i></a></li> -->
        </ul>
<hr>
<nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#phylogenetic-analyses">phylogenetic analyses</a></li>
<li><a class="nav-link" href="#phylogeneticsignal">phylogeneticSignal</a></li>
<li><a class="nav-link" href="#independentcontrasts">independentContrasts</a></li>
<li><a class="nav-link" href="#ancestraltraits">ancestralTraits</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/blob/main/index.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/edit/main/index.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>
</div>
  

  

</div>
 <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Integrated Bioanalytics</strong>" was written by Lucas Busta and members of the Busta lab. It was last built on 2025-09-30.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
