<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>r faq | Integrated Bioanalytics</title>
<meta name="author" content="Lucas Busta and members of the Busta lab">
<meta name="description" content="Updating R and R Packages Close RStudio, open the plain R GUI, then run the following: On Mac: install.packages('remotes') #assuming it is not remotes installed...">
<meta name="generator" content="bookdown 0.35 with bs4_book()">
<meta property="og:title" content="r faq | Integrated Bioanalytics">
<meta property="og:type" content="book">
<meta property="og:description" content="Updating R and R Packages Close RStudio, open the plain R GUI, then run the following: On Mac: install.packages('remotes') #assuming it is not remotes installed...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="r faq | Integrated Bioanalytics">
<meta name="twitter:description" content="Updating R and R Packages Close RStudio, open the plain R GUI, then run the following: On Mac: install.packages('remotes') #assuming it is not remotes installed...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><link href="libs/Lato-0.4.6/font.css" rel="stylesheet">
<link href="libs/Roboto_Mono-0.4.6/font.css" rel="stylesheet">
<link href="libs/Montserrat-0.4.6/font.css" rel="stylesheet">
<script src="libs/bs3compat-0.4.2/transition.js"></script><script src="libs/bs3compat-0.4.2/tabs.js"></script><script src="libs/bs3compat-0.4.2/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><link rel="icon" type="image/png" sizes="32x32" href="images/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="images/favicon-16x16.png">
<script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-99618359-1', 'auto');
      ga('send', 'pageview');

    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="style/style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h2>
        <a href="index.html" title="">Integrated Bioanalytics</a>
      </h2>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">WELCOME</a></li>
<li class="book-part">DATA ANALYSIS IN R</li>
<li><a class="" href="overview.html">overview</a></li>
<li><a class="" href="installation.html">installation</a></li>
<li><a class="" href="data-visualization.html">data visualization</a></li>
<li><a class="" href="data-wrangling-and-summaries.html">data wrangling and summaries</a></li>
<li><a class="" href="dimensionality-reduction.html">dimensionality reduction</a></li>
<li><a class="" href="clustering.html">clustering</a></li>
<li><a class="" href="models.html">models</a></li>
<li><a class="" href="comparing-means.html">comparing means</a></li>
<li><a class="" href="special-topics.html">special topics</a></li>
<li class="book-part">GC-MS DATA</li>
<li><a class="" href="mass-spectrometric-analysis.html">mass spectrometric analysis</a></li>
<li class="book-part">TRANSCRIPTOME ANALYSIS</li>
<li><a class="" href="transcriptome-assembly.html">transcriptome assembly</a></li>
<li class="book-part">EVOLUTIONARY ANALYSIS</li>
<li><a class="" href="blast.html">blast</a></li>
<li><a class="" href="alignments.html">alignments</a></li>
<li><a class="" href="phylogenies.html">phylogenies</a></li>
<li><a class="" href="phylogenetic-analyses.html">phylogenetic analyses</a></li>
<li><a class="" href="comparative-genomics.html">comparative genomics</a></li>
<li class="book-part">LANGUAGE MODELS</li>
<li><a class="" href="completiongpt.html">completionGPT</a></li>
<li><a class="" href="analyzeliterature.html">analyzeLiterature</a></li>
<li class="book-part">SCIENTIFIC WRITING</li>
<li><a class="" href="overview-1.html">overview</a></li>
<li><a class="" href="figures-captions.html">figures &amp; captions</a></li>
<li><a class="" href="results-and-discussion.html">results and discussion</a></li>
<li><a class="" href="conclusion-and-introduction.html">conclusion and introduction</a></li>
<li><a class="" href="abstract-and-title.html">abstract and title</a></li>
<li class="book-part">APPENDIX</li>
<li><a class="" href="links.html">links</a></li>
<li><a class="active" href="r-faq.html">r faq</a></li>
<li><a class="" href="templates.html">templates</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="r-faq" class="section level1 unnumbered">
<h1>r faq<a class="anchor" aria-label="anchor" href="#r-faq"><i class="fas fa-link"></i></a>
</h1>
<div id="updating-r-and-r-packages" class="section level2 unnumbered">
<h2>Updating R and R Packages<a class="anchor" aria-label="anchor" href="#updating-r-and-r-packages"><i class="fas fa-link"></i></a>
</h2>
<p>Close RStudio, open the plain R GUI, then run the following:</p>
<p>On Mac:</p>
<div class="sourceCode" id="cb162"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">'remotes'</span><span class="op">)</span> <span class="co">#assuming it is not remotes installed</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">'andreacirilloac/updateR'</span><span class="op">)</span></span>
<span><span class="fu">updateR</span><span class="fu">::</span><span class="fu">updateR</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>On PC:</p>
<div class="sourceCode" id="cb163"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"installr"</span><span class="op">)</span></span>
<span><span class="fu">installr</span><span class="fu">::</span><span class="fu">updateR</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="filtering" class="section level2 unnumbered">
<h2>filtering<a class="anchor" aria-label="anchor" href="#filtering"><i class="fas fa-link"></i></a>
</h2>
<p><code>filter(&lt;data&gt;, &lt;variable&gt; &lt; 18)</code> ## less than 18</p>
<p><code>filter(&lt;data&gt;, &lt;variable&gt; &lt;= 18)</code> ## less than or equal to 18</p>
<p><code>filter(&lt;data&gt;, &lt;variable&gt; &gt; 18)</code> ## greater than 18</p>
<p><code>filter(&lt;data&gt;, &lt;variable&gt; &gt;= 18)</code> ## greater than or equal to 18</p>
<p><code>filter(&lt;data&gt;, &lt;variable&gt; == 18)</code> ## equals than 18</p>
<p><code>filter(&lt;data&gt;, &lt;variable&gt; != 18)</code> ## not equal to 18</p>
<p><code>filter(&lt;data&gt;, &lt;variable&gt; == 18 | &lt;variable&gt; == 19)</code> ## equal to 18 or 19</p>
<p><code>filter(&lt;data&gt;, &lt;variable&gt; %in% c(18, 19, 20)</code> ## equal to 18 or 19 or 20</p>
</div>
<div id="ordering-1" class="section level2 unnumbered">
<h2>ordering<a class="anchor" aria-label="anchor" href="#ordering-1"><i class="fas fa-link"></i></a>
</h2>
<p>A list of numeric element has an inherent order to it: -inf -&gt; +inf. A list of character element also has an inherent order to it: A -&gt; Z, or if it‚Äôs a mixed number and letter list (which is interpreted by R as a character list): 0 -&gt; 9 -&gt; A -&gt; Z.</p>
<p>However, there are cases where we will want a list of character elements to have some order other than A -&gt; Z. In these cases, we want to convert the list of character elements into a list of factor elements. Factors are lists of character elements that have an inherent order that is not A -&gt; Z. For example, in the plot below, the y axis is not, perhaps, in the ‚Äúcorrect‚Äù order:</p>
<div class="sourceCode" id="cb164"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">periodic_table</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">group_number</span>, x <span class="op">=</span> <span class="va">atomic_mass_rounded</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-234-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>How do we fix this? We need to convert the column <code>group_number</code> into a list of factors that have the correct order (see below). For this, we will use the command <code>factor</code>, which will accept an argument called <code>levels</code> in which we can define the order the the characters should be in:</p>
<div class="sourceCode" id="cb165"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">periodic_table</span><span class="op">$</span><span class="va">group_number</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span></span>
<span>  <span class="va">periodic_table</span><span class="op">$</span><span class="va">group_number</span>,</span>
<span>  levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"1"</span>, <span class="st">"2"</span>, <span class="st">"3"</span>, <span class="st">"4"</span>, <span class="st">"5"</span>, <span class="st">"6"</span>, <span class="st">"7"</span>, <span class="st">"8"</span>, <span class="st">"9"</span>, <span class="st">"10"</span>, <span class="st">"11"</span>, <span class="st">"12"</span>, <span class="st">"13"</span>, <span class="st">"14"</span>, <span class="st">"15"</span>, <span class="st">"16"</span>, <span class="st">"17"</span>, <span class="st">"18"</span>, <span class="st">"lanthanides"</span>, <span class="st">"actinides"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">periodic_table</span></span>
<span><span class="co">## # A tibble: 118 √ó 41</span></span>
<span><span class="co">##    atomic_number element_name atomic_symbol group_number</span></span>
<span><span class="co">##            &lt;dbl&gt; &lt;chr&gt;        &lt;chr&gt;         &lt;fct&gt;       </span></span>
<span><span class="co">##  1             1 hydrogen     H             1           </span></span>
<span><span class="co">##  2             2 helium       He            18          </span></span>
<span><span class="co">##  3             3 lithium      Li            1           </span></span>
<span><span class="co">##  4             4 beryllium    Be            2           </span></span>
<span><span class="co">##  5             5 boron        B             13          </span></span>
<span><span class="co">##  6             6 carbon       C             14          </span></span>
<span><span class="co">##  7             7 nitrogen     N             15          </span></span>
<span><span class="co">##  8             8 oxygen       O             16          </span></span>
<span><span class="co">##  9             9 fluorine     F             17          </span></span>
<span><span class="co">## 10            10 neon         Ne            18          </span></span>
<span><span class="co">## # ‚Ñπ 108 more rows</span></span>
<span><span class="co">## # ‚Ñπ 37 more variables: period &lt;dbl&gt;,</span></span>
<span><span class="co">## #   atomic_mass_rounded &lt;dbl&gt;, melting_point_C &lt;dbl&gt;,</span></span>
<span><span class="co">## #   boiling_point_C &lt;dbl&gt;, state_at_RT &lt;chr&gt;,</span></span>
<span><span class="co">## #   density_g_per_mL &lt;dbl&gt;,</span></span>
<span><span class="co">## #   electronegativity_pauling &lt;dbl&gt;,</span></span>
<span><span class="co">## #   first_ionization_poten_eV &lt;dbl&gt;, ‚Ä¶</span></span></code></pre></div>
<p>Notice that now when we look at the type of data that is contained in the column <code>group_number</code> it says ‚Äú<fct>‚Äù. This is great! It means we have converted that column into a list of factors, instead of characters. Now what happens when we make our plot?</fct></p>
<div class="sourceCode" id="cb166"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">periodic_table</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">group_number</span>, x <span class="op">=</span> <span class="va">atomic_mass_rounded</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-236-1.png" width="100%" style="display: block; margin: auto;"></div>
<p>VICTORY!</p>
</div>
<div id="column-manipulation" class="section level2 unnumbered">
<h2>column manipulation<a class="anchor" aria-label="anchor" href="#column-manipulation"><i class="fas fa-link"></i></a>
</h2>
<p>How to select specific columns:</p>
<div class="sourceCode" id="cb167"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alaska_lake_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">water_temp</span>, <span class="va">pH</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 220 √ó 2</span></span>
<span><span class="co">##    water_temp    pH</span></span>
<span><span class="co">##         &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span><span class="co">##  1       6.46  7.69</span></span>
<span><span class="co">##  2       6.46  7.69</span></span>
<span><span class="co">##  3       6.46  7.69</span></span>
<span><span class="co">##  4       6.46  7.69</span></span>
<span><span class="co">##  5       6.46  7.69</span></span>
<span><span class="co">##  6       6.46  7.69</span></span>
<span><span class="co">##  7       6.46  7.69</span></span>
<span><span class="co">##  8       6.46  7.69</span></span>
<span><span class="co">##  9       6.46  7.69</span></span>
<span><span class="co">## 10       6.46  7.69</span></span>
<span><span class="co">## # ‚Ñπ 210 more rows</span></span></code></pre></div>
<p>How to remove certain columns:</p>
<div class="sourceCode" id="cb168"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alaska_lake_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="op">!</span><span class="va">water_temp</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 220 √ó 6</span></span>
<span><span class="co">##    lake            park     pH element mg_per_L element_type</span></span>
<span><span class="co">##    &lt;chr&gt;           &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt; &lt;chr&gt;       </span></span>
<span><span class="co">##  1 Devil_Mountain‚Ä¶ BELA   7.69 C          3.4   bound       </span></span>
<span><span class="co">##  2 Devil_Mountain‚Ä¶ BELA   7.69 N          0.028 bound       </span></span>
<span><span class="co">##  3 Devil_Mountain‚Ä¶ BELA   7.69 P          0     bound       </span></span>
<span><span class="co">##  4 Devil_Mountain‚Ä¶ BELA   7.69 Cl        10.4   free        </span></span>
<span><span class="co">##  5 Devil_Mountain‚Ä¶ BELA   7.69 S          0.62  free        </span></span>
<span><span class="co">##  6 Devil_Mountain‚Ä¶ BELA   7.69 F          0.04  free        </span></span>
<span><span class="co">##  7 Devil_Mountain‚Ä¶ BELA   7.69 Br         0.02  free        </span></span>
<span><span class="co">##  8 Devil_Mountain‚Ä¶ BELA   7.69 Na         8.92  free        </span></span>
<span><span class="co">##  9 Devil_Mountain‚Ä¶ BELA   7.69 K          1.2   free        </span></span>
<span><span class="co">## 10 Devil_Mountain‚Ä¶ BELA   7.69 Ca         5.73  free        </span></span>
<span><span class="co">## # ‚Ñπ 210 more rows</span></span></code></pre></div>
</div>
<div id="user-color-palettes" class="section level2 unnumbered">
<h2>user color palettes<a class="anchor" aria-label="anchor" href="#user-color-palettes"><i class="fas fa-link"></i></a>
</h2>
<p>Suppose we want to create a specific color palette for each pack in <code>alaska_lake_data</code>. There are three unique parks:</p>
<div class="sourceCode" id="cb169"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">alaska_lake_data</span><span class="op">$</span><span class="va">park</span><span class="op">)</span></span>
<span><span class="co">## [1] "BELA" "GAAR" "NOAT"</span></span></code></pre></div>
<p>First we define the colors we want:</p>
<div class="sourceCode" id="cb170"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">custom_colors_for_lakes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#1a9850"</span>, <span class="st">"#ffffbf"</span>, <span class="st">"#d73027"</span><span class="op">)</span></span>
<span><span class="va">custom_colors_for_lakes</span></span>
<span><span class="co">## [1] "#1a9850" "#ffffbf" "#d73027"</span></span></code></pre></div>
<p>Then we name that vector according to which park we want to be which color:</p>
<div class="sourceCode" id="cb171"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">custom_colors_for_lakes</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GAAR"</span>, <span class="st">"NOAT"</span>, <span class="st">"BELA"</span><span class="op">)</span></span>
<span><span class="va">custom_colors_for_lakes</span></span>
<span><span class="co">##      GAAR      NOAT      BELA </span></span>
<span><span class="co">## "#1a9850" "#ffffbf" "#d73027"</span></span></code></pre></div>
<p>Now we feed that object to the <code>values</code> argument of scale_color_manual (or scale_fill_manual, if you want fill):</p>
<div class="sourceCode" id="cb172"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">alaska_lake_data</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">pH</span>, y <span class="op">=</span> <span class="va">water_temp</span>, fill <span class="op">=</span> <span class="va">park</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">5</span>, shape <span class="op">=</span> <span class="fl">21</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">custom_colors_for_lakes</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure">
<img src="index_files/figure-html/unnamed-chunk-242-1.png" width="100%" style="display: block; margin: auto;"><!-- end -->
</div>
<!-- start templates -->
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="links.html">links</a></div>
<div class="next"><a href="templates.html">templates</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <h2>Note: Second Edition is under construction üèó</h2>
    <p>Now is a great time to provide feedback</p>
        <ul class="list-unstyled">
<li><a href="https://forms.gle/SZmB2Ct2exE2dBwv9">Provide feedback (5 min)</a></li>
          <!-- <li><a href="https://geocompr.robinlovelace.net/#reproducibility">Install updated packages</a></li> -->
          <!-- <li><a href="https://github.com/Robinlovelace/geocompr/issues">Open an issue <i class="fas fa-question"></i></a></li> -->
          <!-- <li><a href="https://discord.gg/Te3gWeDwmf">Chat on Discord <i class="fab fa-discord"></i></a></li> -->
        </ul>
<hr>
<nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#r-faq">r faq</a></li>
<li><a class="nav-link" href="#updating-r-and-r-packages">Updating R and R Packages</a></li>
<li><a class="nav-link" href="#filtering">filtering</a></li>
<li><a class="nav-link" href="#ordering-1">ordering</a></li>
<li><a class="nav-link" href="#column-manipulation">column manipulation</a></li>
<li><a class="nav-link" href="#user-color-palettes">user color palettes</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/blob/main/index.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/thebustalab/thebustalab.github.io/tree/master/integrated_bioanalytics/edit/main/index.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Integrated Bioanalytics</strong>" was written by Lucas Busta and members of the Busta lab. It was last built on 2023-08-30.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
