<!-- # (PART) SEQUENCE DATA {-}

# data acquisition {-}

Updating MinKNOW on GridIon

`sudo apt update`
`sudo apt install ont-gridion-release`
`sudo apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1604/x86_64/3bf863cc.pub`

# data transfer {-}

Once you've completed a sequencing run, the data are can be transferred to an external hard drive, which can then be plugged into the storage computer. The next steps are:

1. Identify which files are of interest/where they are in the system. Useful commands:

Display all currently mounted filesystems (& their usage, storage space, mounting point):

`df -h`

Also:

`lsblk -f`

Display more data pertaining to the identification of disks. Can also change partitioning of hard disks:

`sudo fdisk -l` 

-Hard drives are labeled as sd’s. Organization follows as /dev/sd_ with the underscore replaced with a  letter (first hard drive starting with ‘a’ and continuing alphabetically). If partitions are present, the letter if followed by a number (starting with ‘1’ for the first partition and continuing numerically). Ex) /dev/sdb2

2. Mounting

-Use command:
`sudo mount </dev/sd_> </file_path>`
-Replace `</sd_>` with actual hard drive label and `/file_path` with the pathway for the location in which you want to mount the drive.
-Make sure the location (`</file_path>`) is preexisting location. Use the `mkdir` command to make a new directory if necessary.

3. Copy data

-Just use the `cp` command and make sure you have the right filenames and locations to transfer the data from the hard drive to the internal disk.

4. Additional Information

bootable USB: https://rufus.ie/en/#


# nanopore read assessment {-}

With your nanopore reads stored on a suitable machine, you can analyze them with several `phylochemistry` functions. Here is a quick overview:

```{r, eval = FALSE}
qc_data <- fastxQC(
    paths_to_fastxs = c(
        "/Users/bust0037/Documents/Science/Websites/thebustalab.github.io/data/example.fastq",
        "/Users/bust0037/Documents/Science/Websites/thebustalab.github.io/data/example2.fastq"
    ),
    type = "fasta",
    mode = "slow",
    max_n_seqs = 1000
)

head(qc_data)

qc_data %>%
  mutate(category = case_when(
    length > mean(qc_data$length)*5 ~ "chromosome",
    length <= mean(qc_data$length)*5 ~ "leftover_bit"
  )) %>%
  ggplot() +
    geom_treemap(aes(area = length, fill = category), color = "black", size = 1) +
    scale_fill_manual(values = c("gold", "maroon"))
```

-->

<!-- end -->
________________________________________________________________________________________________
________________________________________________________________________________________________
________________________________________________________________________________________________


<!-- # (PART) SEQUENCE DATA {-} -->

<!-- start transcriptomic analyses -->

<!-- # transcriptome assembly {-}

For nonmodel species transcriptome analysis, `transXpress` is recommended. We often use a modified version of `transXpress` we call `transXpressLite`. It uses the Trinity assembler by default, which can require at least 500GB of free disk space to run. Depending on your machine, you may also need to make some modifications to `transXpressLite` for it to run properly.

1. Download the transXpressLite code into the directory in which you wish to perform the assembly:

`git clone https://github.com/thebustalab/transXpressLite.git`

2. Rename the downloaded folder to make it unique. Perhaps:

`mv transXpressLite transXpressLite-kfed`

2. Move into that directory and set up and activate the main transXpress environment:

`cd transXpressLite-kfed`
`conda create --name transxpress`
`conda activate transxpress`

3. Create a tab-separated file called *samples.txt* with the following contents. Important! Remember that there must be an empty line on the end of the samples.tex file.
`
cond_A    cond_A_rep1    A_rep1_left.fq    A_rep1_right.fq
cond_A    cond_A_rep2    A_rep2_left.fq    A_rep2_right.fq
cond_B    cond_B_rep1    B_rep1_left.fq    B_rep1_right.fq
cond_B    cond_B_rep2    B_rep2_left.fq    B_rep2_right.fq
`

4. Start `transXpressLite`:

`./transXpress.sh`

5. Once transXpress is complete, you may wish to move its output files to a long-term storage device. You may wish to keep the following files handy for downstream analysis though:

* all the sample name folders (ex. "fed_epi_hi_rep1")
* samples.txt
* samples_trimmed.txt
* busco_report.txt
* /transdecoder/longest_orfs.pep -> is actually "transcriptome.orfs" -->

<!-- 1. It's possible that you will need to downgrade numpy for tmhmm to work (make sure you do this in the transxpress env):
`pip install "numpy<1.24.0"`
 -->
<!-- 2. You also need to add targetp to the PATH variable in the transexpress environment:
`export PATH=$PATH:/project_data/shared/general_lab_resources/targetp-2.0/bin/`
 -->

<!-- end -->
________________________________________________________________________________________________
________________________________________________________________________________________________
________________________________________________________________________________________________

<!-- # (PART) PROTEOME ANALYSIS {-} -->

<!-- start proteome analyses -->

<!-- ## structure similarity

`conda create --name foldseek`
`conda activate foldseek`
`conda install -c conda-forge -c bioconda foldseek`
 -->


<!-- end -->
________________________________________________________________________________________________
________________________________________________________________________________________________
________________________________________________________________________________________________

<!-- # (PART) GENOME ANALYSIS {-} -->

<!-- start genomic analyses -->

<!-- # setup {-}

* Get docker.

https://hub.docker.com/

* Connecting to remote host:

Make sure the remote host has openSSH installed:

`sudo apt install openssh-server`

On the client computer (usually your laptop or something), first create the key:

`ssh-keygen -t rsa`

Then copy that key to the host (usually the computer you want to connect remotely to):

`ssh-copy-id -i ~/.ssh/id_rsa.pub {username}@host.address`

Done! Log in with `ssh {username}@host.address`

* Initializing a conda environment:

`conda create --name <name> python=3.6`
`conda activate <name>`
`conda install -c bioconda flye`
`conda install -c bioconda abyss`

# genome assembly {-}

To some degree, refer to: https://github.com/dithiii/ant-pipeline/blob/main/README.md.

## assembly


### equipment

Genome assembly requires computing resources - and since not all genomes are of equal size, the computing resources required for different assemblies may differ. To run a genome assembly, start by determining what computing resources are available. Some helpful commands when investigating these resources on Linux machines:

* Assessing RAM (It is recommended to assign about 75% of available RAM to the assembly process):

`grep MeMTotal /proc/meminfo`

* Assessing CPU resources (note that "threads per CPU" can denote the availability of hyperthreading):

`lscpu`

* Assessing disk/storage space (Make sure you are running your assembly on a disk with lots of open space. Ideally > 2TB):

`df -h`

### assembly software

#### abyss

`abyss-pe k=111 name=SS1 B=10G in='SS1_1.fa SS1_2.fa'`

A rough indicator is, for 2x150bp reads and 40x coverage, the right k value is often around 70 to 90. For 2x250bp reads and 40x coverage, the right value might be around 110 to 140.

A good value for B depends on a number of factors, but primarily on the genome being assembled. A general guideline is: P. glauca (~20Gbp): B=500G; H. sapiens (~3.1Gbp): B=50G; C. elegans (~101Mbp): B=2G. Using more is fine though, 

#### canu

`docker pull staphb/canu-racon`

For assembly on the BustaLab storage box, navigate to the directory that contains your reads. Merge all reads into one file using:

`cat *.fastq > all_reads.fastq`

Then use Canu to assemble, we suggest creating a file that contains the Canu call. You can create the file using `nano`. In it, try something like:

```{r, eval = FALSE}
sudo docker run -u $(id -u) -v /data/ben_diatom2/basecalled_reads/:/canu-racon_wd staphb/canu-racon canu -p n_frust2assembly -d /canu-racon_wd/ -genomeSize=150m -nanopore /canu-racon_wd/all_reads2.fastq -minReadLength=1000 -correctedErrorRate=0.12 -minOverlapLength=500 -useGrid=false -minInputCoverage=0.5 -maxInputCoverage=100 -stopOnLowCoverage=0.5 -corMemory=48 -corThreads=4 -hapMemory=48 -hapThreads=4 -merylMemory=48 -merylthreads=4 -batMemory=48 -batThreads=4
```

Notes on Canu options:

Defaults:

* minReadLength=1000
* minOverlapLength=500bp
* correctedErrorRate=0.114
* stopOnLowCoverage <integer=10>

Essentially only speed optimization:

* For over 30X coverage:
* Nanopore flip-flop R9.4 or R10.3: try: `corMhapOptions=--threshold 0.8 –ordered-sketch-size 1000 –ordered-kmer-size 14’ correctedErrorRate=0.105`
* For over 60X coverage: 2 recommendations were made, one said to decrease slightly (~1%). Another suggested using 12%
correctedErrorRate=0.12
* Increasing minReadLength increases run time, increasing minOverlapLength improves assembly quality but increasing too much quickly degrades assemblies.

#### flye

`docker pull staphb/flye`

### kmer-based metrics

Canu will take some time to run. As it goes along, you can both check on its progress and learn about the genome you are assembling from some intermediate results. Take the k-mer data found in the .histogram files (i.e. in correction/0-mercounts/x.histogram, trimming/0-mercounts/x.histogram, unitigging/0-mercounts/x.histogram) and process them with `canuHistogramToKmerTable()`, as shown below. You can upload the output to : http://qb.cshl.edu/genomescope/genomescope2.0/. This will give you approximate genome size, ploidy, heterozygosity, repeat content, and read error rate. All good stuff to know!

```{r, eval = FALSE}
canuHistogramToKmerTable(
  file_in_path = "/Users/bust0037/Desktop/n_frust3assemblyB.ms22.histogram",
  file_out_path = "/Users/bust0037/Desktop/n_frust3assemblyB.ms22.histogram_table"
)
```

Also check on this tutorial:

[genomics tutorial](https://ucdavis-bioinformatics-training.github.io/2020-Genome_Assembly_Workshop/kmers/kmers)

#### merqury

`docker pull quay.io/chai/merqury`

`sudo docker run -u $(id -u) -v /home/bust0037/data1/Kalanchoe_DNASeq/round2_pass_reads_assembly/:/merqury/ quay.io/chai/merqury:latest quast.py -h`

References:
https://www.biorxiv.org/content/10.1101/2020.03.15.992941v1.abstract
Merqury: reference-free quality, completeness, and phasing assessment for genome assemblies

## evaluating contigs

Can these be merged into a single wrapper that can be run after each step in assembly/polishing/scaffolding etc?

### BUSCO

- Real BUSCO input will be /home/bust0037/data1/comparative_genomics/Kfedtschenkoi_382_v1.0.fa
- Note BUSCO uses current working directory for input and output

`docker pull ezlabgva/busco:v5.3.2_cv1`

`sudo docker run -u $(id -u) -v /home/bust0037/data1/comparative_genomics/:/busco_wd ezlabgva/busco:v5.3.2_cv1 busco -i k_fed.contigs.fa -o busco_our_kfed/ -m genome -l eudicots_odb10
`

### quast

Quast provides a score called ALE: alignment liklihood estimate.

`docker pull longas/quast`

`sudo docker run -u $(id -u) -v /home/bust0037/:/tmp/work/quast_results longas/quast:latest quast.py -h`

`sudo docker run -u $(id -u) -v /home/bust0037/ben_test/_ben_genomes/:/tmp/work/quast_results longas/quast:latest quast.py --fragmented /tmp/work/quast_results/ben_pre_n_frust.contigs.fasta --nanopore /tmp/work/quast_results/all_pass_reads.fastq --space-efficient --memory-efficient --fast`

## polishing contigs

### medaka

Pull the docker image:

`docker pull ontresearch/medaka`

Run medaka:

`sudo docker run -u $(id -u) -v /home/bust0037/data1/Kalanchoe_DNASeq/round2_pass_reads_assembly/:/medaka/ ontresearch/medaka:latest medaka_consensus -i medaka/all_reads.fastq -d medaka/k_fed.contigs.fasta -b 50`

Good documentation here: https://labs.epi2me.io/notebooks/Introduction_to_how_ONT's_medaka_works.html

Medaka will:
* Map all your raw reads. Look for feedback like: `[M::worker_pipeline::2924.325*0.25] mapped 75823 sequences`.
* Do something else, updates in the form of: `21.7% Done (88.2/406.1 Mbases) in 6815.1s`.

## methylation with remora

can we call methylation status on our Kalanchoe genomes? -> yes, we can use Remora -> watch for new guppy release, MinKNOW integration -> currently in bonito

```
bonito basecaller dna_r10.4_e8.1_sup@v3.4 /data/reads --modified-bases 5mC --reference ref.mmi > basecalls_with_mods.bam
bonito basecaller dna_r10.4_e8.1_sup@v3.4 --reference consensus.fasta --modified-bases 5mC
```

## scaffolding assembly

### RagTag

INSTALL PYTHON (or upgrade python) <- can this be done using docker?

INSTALL BIOCONDA
1. install miniconda:
  curl -O https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
  sh ...
  
## annotation

### augustus

see: https://hub.docker.com/r/pegi3s/autoaugustus/

```
docker pull pegi3s/autoaugustus
```

```
sudo docker run --rm -v $(pwd):$(pwd) pegi3s/autoaugustus augustus --species=help
sudo docker run --rm -v $(pwd):/augustus/ pegi3s/autoaugustus augustus --species=tomato /augustus/consensus.fasta > consensus-preductions.gff --progress=TRUE
```

## final assessment

https://www.molecularecologist.com/2017/03/29/whats-n50/

### mosdepth

`sudo docker pull quay.io/biocontainers/mosdepth:0.2.4--he527e40_0`

`sudo docker run -u $(id -u) -v /home/bust0037/data1/Kalanchoe_DNASeq/rounds_1and2_pass_assembly/:/mosdepth_wd quay.io/biocontainers/mosdepth:0.2.4--he527e40_0 mosdepth -h`

`sudo docker run -u $(id -u) -v /home/bust0037/data1/Kalanchoe_DNASeq/rounds_1and2_pass_assembly/:/mosdepth_wd quay.io/biocontainers/mosdepth:0.2.4--he527e40_0 mosdepth -n --fast-mode --by 1000 mosdepth_wd/mosdepth_out /mosdepth_wd/calls_to_draft.bam`

`sudo docker run -u $(id -u) -v /home/bust0037/data1/Kalanchoe_DNASeq/rounds_1and2_pass_assembly/:/mosdepth_wd gfanz/mosdepth -n --fast-mode --by 1000 mosdepth_wd/mosdepth_out /mosdepth_wd/calls_to_draft.bam`

`sudo docker run -u $(id -u) -v /home/bust0037/data1/Kalanchoe_DNASeq/rounds_1and2_pass_assembly/:/mosdepth_wd quay.io/biocontainers/mosdepth:0.2.4--he527e40_0 mosdepth --quantize 0:1:4:100:200: --fast-mode --by 1000 mosdepth_wd/mosdepth_out /mosdepth_wd/calls_to_draft.bam`

2. set up channels

  /home/bust0037/miniconda3/bin/conda config --add channels defaults
  /home/bust0037/miniconda3/bin/conda config --add channels bioconda
  /home/bust0037/miniconda3/bin/conda config --add channels conda-forge

3. install packages

  /home/bust0037/miniconda3/bin/conda install sibeliaz
  /home/bust0037/miniconda3/bin/conda install -c bioconda ragtag

4. run your stuff

  /home/bust0037/miniconda3/bin/conda run sibeliaz -n k_fed.contigs.scaffolded.fasta KlaxifloraFTBG2000359A_699_v3.0.fa

  /home/bust0037/miniconda3/bin/conda run ragtag

  /home/bust0037/miniconda3/bin/conda run ragtag

  ragtag.py


  /home/bust0037/meryl-1.3/bin/meryl
  /home/bust0037/merqury-1.3/merqury.sh

sh $MERQURY/best_k.sh <genome_size>


Let's look at some examples. For these example, we will use some fasta files stored in a Google Drive folder: -->

```{r}
# reads <- readFasta("https://drive.google.com/file/d/1r6E0U5LyYwjWenxy9yqh5QQ2mq1umWOW/view?usp=sharing")

# # post <- readFasta("/Users/bust0037/Desktop/ragtag.scaffold.fasta")
# n_chroms <- 18

# pb <- progress::progress_bar$new(total = n_chroms)

# out <- list()

# for (i in 1:n_chroms) {

#   pb$tick()

#   dat <- strsplit(substr(as.character(post[i]), 1, 50000000), "")[[1]]
  
#   b <- rle(dat)

#   # Create a data frame
#   dt <- data.frame(number = b$values, lengths = b$lengths, scaff = i)
#   # Get the end
#   dt$end <- cumsum(dt$lengths)
#   # Get the start
#   dt$start <- dt$end - dt$lengths + 1

#   # Select columns
#   dt <- dt[, c("number", "start", "end", "scaff")]
#   # Sort rows
#   dt <- dt[order(dt$number), ]

#   dt %>%
#     filter(number == "N") -> N_dat

#   out[[i]] <- N_dat

# }

# out <- do.call(rbind, out)


# chroms <- data.frame(
#   lengths = post@ranges@width[1:n_chroms],
#   scaff = seq(1,n_chroms,1)
# )

# ggplot() +
#   statebins:::geom_rrect(data = chroms, aes(xmin = 0, xmax = lengths, ymin = -1, ymax = 1, fill = scaff), color = "black") +
#   geom_rect(data = out, aes(xmin = start, xmax = end, ymin = -0.95, ymax = 0.95), color = "white", fill = "white", size = 0.08) +
#   facet_grid(scaff~.) +
#   scale_fill_viridis(end = 0.8) +
#   theme_classic()

# ggplot() +
#   geom_rect(data = filter(chroms, scaff == 1 | scaff == 2), aes(xmin = 0, xmax = lengths, ymin = -1, ymax = 1, fill = scaff), color = "black") +
#   geom_rect(data = filter(out, scaff == 1 | scaff == 2), aes(xmin = start, xmax = end, ymin = -0.95, ymax = 0.95), color = "white", fill = "white", size = 0.08) +
#   facet_grid(scaff~.) +
#   scale_y_continuous(limits = c(-2,2)) +
#   scale_fill_viridis(end = 0.8) +
#   theme_classic() +
#   coord_polar()
```
<!-- # annotation {-} -->

<!-- docker pull nanozoo/braker2 -->

<!-- end -->