<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 8 clustering | Integrated Bioanalytical Chemistry</title>
<meta name="author" content="Lucas Busta">
<meta name="description" content="8.1 theory “Which of my samples are most closely related?” So far we have been looking at how to plot raw data, as well as data that has been summarize across samples. This is important stuff and...">
<meta name="generator" content="bookdown 0.24.4 with bs4_book()">
<meta property="og:title" content="Chapter 8 clustering | Integrated Bioanalytical Chemistry">
<meta property="og:type" content="book">
<meta property="og:description" content="8.1 theory “Which of my samples are most closely related?” So far we have been looking at how to plot raw data, as well as data that has been summarize across samples. This is important stuff and...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 8 clustering | Integrated Bioanalytical Chemistry">
<meta name="twitter:description" content="8.1 theory “Which of my samples are most closely related?” So far we have been looking at how to plot raw data, as well as data that has been summarize across samples. This is important stuff and...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/header-attrs-2.11/header-attrs.js"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.3.1.9000/transition.js"></script><script src="libs/bs3compat-0.3.1.9000/tabs.js"></script><script src="libs/bs3compat-0.3.1.9000/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Integrated Bioanalytical Chemistry</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">integrated bioanalytical chemistry</a></li>
<li class="book-part">data analysis</li>
<li><a class="" href="installation.html"><span class="header-section-number">2</span> installation</a></li>
<li><a class="" href="R_Markdown.html"><span class="header-section-number">3</span> ggplot and markdown</a></li>
<li><a class="" href="data-visualization.html">data visualization</a></li>
<li><a class="" href="geoms_facets_scales_themes.html"><span class="header-section-number">4</span> geoms, facets, scales, themes</a></li>
<li><a class="" href="import-and-tidying.html"><span class="header-section-number">5</span> import and tidying</a></li>
<li><a class="" href="summary_statistics.html"><span class="header-section-number">6</span> the pipe and summaries</a></li>
<li class="book-part">chemometrics</li>
<li><a class="" href="overview.html">overview</a></li>
<li><a class="active" href="clustering.html"><span class="header-section-number">8</span> clustering</a></li>
<li class="book-part">instrumental analysis</li>
<li><a class="" href="chromatography-and-separations.html"><span class="header-section-number">10</span> chromatography and separations</a></li>
<li><a class="" href="mass-spectrometry.html"><span class="header-section-number">11</span> mass spectrometry</a></li>
<li><a class="" href="spectroscopy.html"><span class="header-section-number">12</span> spectroscopy</a></li>
<li><a class="" href="electrochemistry.html"><span class="header-section-number">13</span> electrochemistry</a></li>
<li class="book-part">writing</li>
<li><a class="" href="a-mini-manuscript.html">a mini manuscript</a></li>
<li><a class="" href="figures-captions.html"><span class="header-section-number">14</span> figures &amp; captions</a></li>
<li><a class="" href="results-and-discussion.html"><span class="header-section-number">15</span> results and discussion</a></li>
<li><a class="" href="conclusion-and-introduction.html"><span class="header-section-number">16</span> conclusion and introduction</a></li>
<li><a class="" href="abstract-and-title.html"><span class="header-section-number">17</span> abstract and title</a></li>
<li class="book-part">image analysis</li>
<li><a class="" href="section.html"></a></li>
<li><a class="" href="image-analysis.html">IMAGE ANALYSIS</a></li>
<li><a class="" href="image-color-analysis.html"><span class="header-section-number">18</span> image color analysis</a></li>
<li><a class="" href="images-of-mass-spectra.html"><span class="header-section-number">19</span> images of mass spectra</a></li>
<li><a class="" href="phylochemistry-1.html"><span class="header-section-number">20</span> phylochemistry</a></li>
<li class="book-part">mass spectrometric analysis</li>
<li><a class="" href="mass-spectrometric-analysis.html"><span class="header-section-number">21</span> mass spectrometric analysis</a></li>
<li class="book-part">transcriptome analysis</li>
<li><a class="" href="transcriptomic-analyses.html"><span class="header-section-number">22</span> transcriptomic analyses</a></li>
<li class="book-part">transcriptome analysis</li>
<li><a class="" href="genomic-analyses.html"><span class="header-section-number">23</span> genomic analyses</a></li>
<li class="book-part">evolutionary analysis</li>
<li><a class="" href="evolutionary-analyses.html"><span class="header-section-number">24</span> evolutionary analyses</a></li>
<li class="book-part">appendices</li>
<li><a class="" href="section-1.html"></a></li>
<li><a class="" href="appendix.html">APPENDIX</a></li>
<li><a class="" href="links.html"><span class="header-section-number">25</span> links</a></li>
<li><a class="" href="faq.html"><span class="header-section-number">26</span> faq</a></li>
<li><a class="" href="templates.html"><span class="header-section-number">27</span> templates</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/rstudio/bookdown-demo">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="clustering" class="section level1" number="8">
<h1>
<span class="header-section-number">8</span> clustering<a class="anchor" aria-label="anchor" href="#clustering"><i class="fas fa-link"></i></a>
</h1>
<div id="theory" class="section level2" number="8.1">
<h2>
<span class="header-section-number">8.1</span> theory<a class="anchor" aria-label="anchor" href="#theory"><i class="fas fa-link"></i></a>
</h2>
<p>“Which of my samples are most closely related?”</p>
<p>So far we have been looking at how to plot raw data, as well as data that has been summarize across samples. This is important stuff and very useful. However, we often have questions about how samples in our datasets relate to one another. For example: in the Alaska lakes dataset, which lake is most similar, chemically speaking, to Lake Narvakrak? Answering this requires calculating numeric distances between samples based on their chemical properties. For this, and other questions, we need to use matrix analyses. For this, we will use <code>runMatrixAnalysis()</code>, a function that is loaded into your R Session when you run the source() command.</p>
<p>Matrix analyses can be a bit difficult to set up. There are two things that we can do to help us with this: (i) we will use a template for <code>runMatrixAnalysis()</code> (see below) and (ii) it is <em>critical</em> that we think about our data in terms of <strong>samples</strong> and <strong>analytes</strong>. Let’s consider our Alaska lakes data set:</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">alaska_lake_data</span>
<span class="co">## # A tibble: 220 × 7</span>
<span class="co">##    lake                park  water_temp    pH element mg_per_L</span>
<span class="co">##    &lt;chr&gt;               &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;</span>
<span class="co">##  1 Devil_Mountain_Lake BELA        6.46  7.69 C          3.4  </span>
<span class="co">##  2 Devil_Mountain_Lake BELA        6.46  7.69 N          0.028</span>
<span class="co">##  3 Devil_Mountain_Lake BELA        6.46  7.69 P          0    </span>
<span class="co">##  4 Devil_Mountain_Lake BELA        6.46  7.69 Cl        10.4  </span>
<span class="co">##  5 Devil_Mountain_Lake BELA        6.46  7.69 S          0.62 </span>
<span class="co">##  6 Devil_Mountain_Lake BELA        6.46  7.69 F          0.04 </span>
<span class="co">##  7 Devil_Mountain_Lake BELA        6.46  7.69 Br         0.02 </span>
<span class="co">##  8 Devil_Mountain_Lake BELA        6.46  7.69 Na         8.92 </span>
<span class="co">##  9 Devil_Mountain_Lake BELA        6.46  7.69 K          1.2  </span>
<span class="co">## 10 Devil_Mountain_Lake BELA        6.46  7.69 Ca         5.73 </span>
<span class="co">## # … with 210 more rows, and 1 more variable:</span>
<span class="co">## #   element_type &lt;chr&gt;</span></code></pre></div>
<p>We can see that this dataset is comprised of measurements of various <em>analytes</em> (i.e. several chemical elements, as well as water_temp, and pH), in different <em>samples</em> (i.e. lakes). We need to tell the <code>runMatrixAnalysis()</code> function how each column relates to this samples and analytes structure. See the image below for an explanation.</p>
<div class="inline-figure"><img src="http://thebustalab.github.io/R_For_Chemists_3/images/runMatrixAnalysis1.png" width="80%" style="display: block; margin: auto;"></div>
<p>With this in mind, let’s try out our template:</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">AK_lakes_clustered</span> <span class="op">&lt;-</span> <span class="fu">runMatrixAnalysis</span><span class="op">(</span>
                                
    data <span class="op">=</span> <span class="va">alaska_lake_data</span>,

    analysis <span class="op">=</span> <span class="st">"hclust"</span>,

    column_w_names_of_multiple_analytes <span class="op">=</span> <span class="st">"element"</span>,
    column_w_values_for_multiple_analytes <span class="op">=</span> <span class="st">"mg_per_L"</span>,
    
    columns_w_values_for_single_analyte <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"water_temp"</span>, <span class="st">"pH"</span><span class="op">)</span>,
    
    columns_w_additional_analyte_info <span class="op">=</span> <span class="st">"element_type"</span>,

    columns_w_sample_ID_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lake"</span>, <span class="st">"park"</span><span class="op">)</span>

<span class="op">)</span>
<span class="va">AK_lakes_clustered</span>
<span class="co">## # A tibble: 39 × 25</span>
<span class="co">##    sample_unique_ID  lake   park  parent  node branch.length</span>
<span class="co">##    &lt;chr&gt;             &lt;chr&gt;  &lt;chr&gt;  &lt;int&gt; &lt;int&gt;         &lt;dbl&gt;</span>
<span class="co">##  1 Devil_Mountain_L… Devil… BELA      33     1          7.25</span>
<span class="co">##  2 Imuruk_Lake_BELA  Imuru… BELA      32     2          4.91</span>
<span class="co">##  3 Kuzitrin_Lake_BE… Kuzit… BELA      36     3          3.27</span>
<span class="co">##  4 Lava_Lake_BELA    Lava_… BELA      35     4          3.02</span>
<span class="co">##  5 North_Killeak_La… North… BELA      21     5        204.  </span>
<span class="co">##  6 White_Fish_Lake_… White… BELA      22     6         65.2 </span>
<span class="co">##  7 Iniakuk_Lake_GAAR Iniak… GAAR      29     7          3.60</span>
<span class="co">##  8 Kurupa_Lake_GAAR  Kurup… GAAR      31     8          8.57</span>
<span class="co">##  9 Lake_Matcharak_G… Lake_… GAAR      29     9          3.60</span>
<span class="co">## 10 Lake_Selby_GAAR   Lake_… GAAR      30    10          5.24</span>
<span class="co">## # … with 29 more rows, and 19 more variables: label &lt;chr&gt;,</span>
<span class="co">## #   isTip &lt;lgl&gt;, x &lt;dbl&gt;, y &lt;dbl&gt;, branch &lt;dbl&gt;,</span>
<span class="co">## #   angle &lt;dbl&gt;, water_temp &lt;dbl&gt;, pH &lt;dbl&gt;, C &lt;dbl&gt;,</span>
<span class="co">## #   N &lt;dbl&gt;, P &lt;dbl&gt;, Cl &lt;dbl&gt;, S &lt;dbl&gt;, F &lt;dbl&gt;, Br &lt;dbl&gt;,</span>
<span class="co">## #   Na &lt;dbl&gt;, K &lt;dbl&gt;, Ca &lt;dbl&gt;, Mg &lt;dbl&gt;</span></code></pre></div>
<p>It works! Now we can plot our cluster diagram with a ggplot add-on called ggtree. We’ve seen that ggplot takes a “data” argument (i.e. <code>ggplot(data = &lt;some_data&gt;) + geom_*()</code> etc.). In contrast, ggtree takes an argument called <code>tr</code>, though if you’re using the <code>runMatrixAnalysis()</code> function, you can treat these two (<code>data</code> and <code>tr</code>) the same, so, use: <code>ggtree(tr = &lt;output_from_runMatrixAnalysis&gt;) + geom_*()</code> etc.</p>
<p>Note that <code>ggtree</code> also comes with several great new geoms: <code><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab()</a></code> and <code><a href="https://rdrr.io/pkg/ggtree/man/geom_tippoint.html">geom_tippoint()</a></code>. Let’s try those out:</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/treedata-book/">ggtree</a></span><span class="op">)</span>
<span class="va">AK_lakes_clustered</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tippoint.html">geom_tippoint</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-92-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>Cool! Though that plot could use some tweaking… let’s try:</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">AK_lakes_clustered</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">lake</span><span class="op">)</span>, offset <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tippoint.html">geom_tippoint</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">21</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">park</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">400</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-93-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>Very nice!</p>
</div>
<div id="exercises-3" class="section level2" number="8.2">
<h2>
<span class="header-section-number">8.2</span> exercises<a class="anchor" aria-label="anchor" href="#exercises-3"><i class="fas fa-link"></i></a>
</h2>
<p>For this set of exercises, please use <code>runMatrixAnalysis()</code> to run and visualize a hierarchical cluster analysis with each of the main datasets that we have worked with so far, except for NY_trees. This means: <code>algae_data</code> (which algae strains are most similar to each other?), <code>alaska_lake_data</code> (which lakes are most similar to each other?). and <code>solvents</code> (which solvents are most similar to each other?). It also means you should use the periodic table (which elements are most similar to each other?), though please don’t use the whole periodic table, rather, use <code>periodic_table_subset</code>. For each of these, create (i) a tree diagram that shows how the “samples” in each data set are related to each other based on the numerical data associated with them, (ii) a caption for each diagram, and (iii) describe, in two or so sentences, an interesting trend you see in the diagram. You can ignore columns that contain categorical data, or you can list those columns as “additional_analyte_info.”</p>
<p>For this assignment, you may find the <code><a href="https://rdrr.io/r/base/colnames.html">colnames()</a></code> function and square bracket-subsetting useful. It will list all or a subset of the column names in a dataset for you. For example:</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">solvents</span><span class="op">)</span>
<span class="co">##  [1] "solvent"             "formula"            </span>
<span class="co">##  [3] "boiling_point"       "melting_point"      </span>
<span class="co">##  [5] "density"             "miscible_with_water"</span>
<span class="co">##  [7] "solubility_in_water" "relative_polarity"  </span>
<span class="co">##  [9] "vapor_pressure"      "CAS_number"         </span>
<span class="co">## [11] "formula_weight"      "refractive_index"   </span>
<span class="co">## [13] "specific_gravity"    "category"</span>

<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">solvents</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>
<span class="co">## [1] "solvent"       "formula"       "boiling_point"</span>

<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">solvents</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">5</span>,<span class="fl">7</span><span class="op">)</span><span class="op">]</span>
<span class="co">## [1] "solvent"             "density"            </span>
<span class="co">## [3] "solubility_in_water"</span></code></pre></div>
<!-- end -->
<!-- start principal components analysis-->
</div>
<div id="pca" class="section level2" number="8.3">
<h2>
<span class="header-section-number">8.3</span> pca<a class="anchor" aria-label="anchor" href="#pca"><i class="fas fa-link"></i></a>
</h2>
<p>“Which analytes are driving differences among my samples?”</p>
<p>In addition to heirarchical clustering, there is another way to look at our data in a cluster context - i.e. another way to identify clusters of samples that have similar properties based on the analytes in the data set. This method is called k-means, which we will look at later, because for it we first need to have a look at dimensionality reduction techniques, particularly principal components analysis (PCA).</p>
<p>explain pca as drawing a new set of axes
pca also answers questions about whether variables are related
hclust is “who is most closely related to whom?” PCA is about clustering</p>
<div id="pca-1" class="section level3" number="8.3.1">
<h3>
<span class="header-section-number">8.3.1</span> pca<a class="anchor" aria-label="anchor" href="#pca-1"><i class="fas fa-link"></i></a>
</h3>
<p>PCA looks at all the variance in a high dimensional data set and chooses new axes within that data set that align with the directions containing highest variance. These new axes are called principal components. Let’s look at an example:</p>
<div class="inline-figure"><img src="https://thebustalab.github.io/R_For_Chemists_3/images/PCA.png" width="80%" style="display: block; margin: auto;"></div>
<p>In the example above, the three dimensional space can be reduced to a two dimensional space with the principal components analysis. New axes (principal components) are selected (bold arrows on left) that become the x and y axes in the principal components space (right).</p>
<p>We can run and visualize principal components analyses using the <code>runMatrixAnalysis()</code> function as in the example below:</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">AK_lakes_pca</span> <span class="op">&lt;-</span> <span class="fu">runMatrixAnalysis</span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">alaska_lake_data</span>,
  analysis <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"pca"</span><span class="op">)</span>,
  column_w_names_of_multiple_analytes <span class="op">=</span> <span class="st">"element"</span>,
  column_w_values_for_multiple_analytes <span class="op">=</span> <span class="st">"mg_per_L"</span>,
  columns_w_values_for_single_analyte <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"water_temp"</span>, <span class="st">"pH"</span><span class="op">)</span>,
  columns_w_additional_analyte_info <span class="op">=</span> <span class="st">"element_type"</span>,
  columns_w_sample_ID_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lake"</span>, <span class="st">"park"</span><span class="op">)</span>
<span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">AK_lakes_pca</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Dim.1</span>, y <span class="op">=</span> <span class="va">Dim.2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">park</span><span class="op">)</span>, shape <span class="op">=</span> <span class="fl">21</span>, size <span class="op">=</span> <span class="fl">4</span>, alpha <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_label_repel</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">lake</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-96-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>Great! In this plot we can see that White Fish Lake and North Killeak Lake, both in BELA park, are quite different from the other parks (they are separated from the others along dimension 1, i.e. the first principal component). At the same time, Wild Lake, Iniakuk Lake, Walker Lake, and several other lakes in GAAR park are different from all the others (they are separated from the others along dimension 2, i.e. the second principal component).</p>
<p>Important question: what makes the lakes listed above different from the others? Certainly some aspect of their chemistry, since that’s the data that this analysis is built upon, but how do we determine which analyte(s) are driving the differences among the lakes that we see in the PCA plot?</p>
</div>
<div id="ordination-plots" class="section level3" number="8.3.2">
<h3>
<span class="header-section-number">8.3.2</span> ordination plots<a class="anchor" aria-label="anchor" href="#ordination-plots"><i class="fas fa-link"></i></a>
</h3>
<p>Let’s look at how to access the information about which analytes are major contributors to each principal component. This is important because it will tell you which analytes are associated with particular dimensions, and by extension, which analytes are associated with (and are markers for) particular groups in the PCA plot. This can be determined using an ordination plot. Let’s look at an example. We can obtain the ordination plot information using <code>runMatrixAnalysis()</code> with <code>analysis = "pca_ord"</code>:</p>
<pre><code>## # A tibble: 6 × 3
##   analyte      Dim.1   Dim.2
##   &lt;chr&gt;        &lt;dbl&gt;   &lt;dbl&gt;
## 1 water_temp 0.0769  -0.267 
## 2 pH         0.704    0.0190
## 3 C          0.297   -0.248 
## 4 N          0.00446  0.732 
## 5 P          0.485   -0.0817
## 6 Cl         0.978    0.0152</code></pre>
<p>We can now visualize the ordination plot using our standard ggplot plotting techniques. Note the use of <code>geom_label_repel()</code> and <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> to label certain segments in the ordination plot. You do not need to use <code>geom_label_repel()</code>, you could use the built in <code><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_label()</a></code>, but <code>geom_label_repel()</code> can make labelling your segments easier.</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">AK_lakes_pca_ord</span> <span class="op">&lt;-</span> <span class="fu">runMatrixAnalysis</span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">alaska_lake_data</span>,
  analysis <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"pca_ord"</span><span class="op">)</span>,
  column_w_names_of_multiple_analytes <span class="op">=</span> <span class="st">"element"</span>,
  column_w_values_for_multiple_analytes <span class="op">=</span> <span class="st">"mg_per_L"</span>,
  columns_w_values_for_single_analyte <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"water_temp"</span>, <span class="st">"pH"</span><span class="op">)</span>,
  columns_w_additional_analyte_info <span class="op">=</span> <span class="st">"element_type"</span>,
  columns_w_sample_ID_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lake"</span>, <span class="st">"park"</span><span class="op">)</span>
<span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">AK_lakes_pca_ord</span><span class="op">)</span>
<span class="co">## # A tibble: 6 × 3</span>
<span class="co">##   analyte      Dim.1   Dim.2</span>
<span class="co">##   &lt;chr&gt;        &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">## 1 water_temp 0.0769  -0.267 </span>
<span class="co">## 2 pH         0.704    0.0190</span>
<span class="co">## 3 C          0.297   -0.248 </span>
<span class="co">## 4 N          0.00446  0.732 </span>
<span class="co">## 5 P          0.485   -0.0817</span>
<span class="co">## 6 Cl         0.978    0.0152</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">AK_lakes_pca_ord</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html">geom_segment</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">0</span>, y <span class="op">=</span> <span class="fl">0</span>, xend <span class="op">=</span> <span class="va">Dim.1</span>, yend <span class="op">=</span> <span class="va">Dim.2</span>, color <span class="op">=</span> <span class="va">analyte</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_circle</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x0 <span class="op">=</span> <span class="fl">0</span>, y0 <span class="op">=</span> <span class="fl">0</span>, r <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_label_repel</span><span class="op">(</span>
    data <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">AK_lakes_pca_ord</span>, <span class="va">Dim.1</span> <span class="op">&gt;</span> <span class="fl">0.9</span>, <span class="va">Dim.2</span> <span class="op">&lt;</span> <span class="fl">0.1</span>, <span class="va">Dim.2</span> <span class="op">&gt;</span> <span class="op">-</span><span class="fl">0.1</span><span class="op">)</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Dim.1</span>, y <span class="op">=</span> <span class="va">Dim.2</span>, label <span class="op">=</span> <span class="va">analyte</span><span class="op">)</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1.5</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_label_repel</span><span class="op">(</span>
    data <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">AK_lakes_pca_ord</span>, <span class="va">Dim.2</span> <span class="op">&gt;</span> <span class="fl">0.5</span><span class="op">)</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Dim.1</span>, y <span class="op">=</span> <span class="va">Dim.2</span>, label <span class="op">=</span> <span class="va">analyte</span><span class="op">)</span>, direction <span class="op">=</span> <span class="st">"y"</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1.5</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fl">1.5</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fl">1.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-98-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>Great! Here is how to read the ordination plot:</p>
<ol style="list-style-type: decimal">
<li><p>When considering one analyte’s vector: the vector’s projected value on an axis shows how much its variance is aligned with that principal component.</p></li>
<li><p>When considering two analyte vectors: the angle between two vectors indicates how correlated those two variables are. If they point in the same direction, they are highly correlated. If they meet each other at 90 degrees, they are not very correlated. If they meet at ~180 degrees, they are negatively correlated. If say that one analyte is “1.9” with respect to dimension 2 and another is “-1.9” with respect to dimension 2. Let’s also say that these vectors are ~“0” with respect to dimension 1.</p></li>
</ol>
<p>With the ordination plot above, we can now see that the abundances of K, Cl, Br, and Na are the major contributors of variance to the first principal component (or the first dimension). The abundances of these elements are what make White Fish Lake and North Killeak Lake different from the other lakes. We can also see that the abundances of N, S, and Ca are the major contributors to variance in teh second dimension, whic means that these elements ar what set Wild Lake, Iniakuk Lake, Walker Lake, and several other lakes in GAAR park apart from the rest of the lakes in the data set.</p>
</div>
<div id="principal-components" class="section level3" number="8.3.3">
<h3>
<span class="header-section-number">8.3.3</span> principal components<a class="anchor" aria-label="anchor" href="#principal-components"><i class="fas fa-link"></i></a>
</h3>
<p>We also can access information about the how much of the variance in the data set is explained by each principal component, and we can plot that using ggplot:</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">AK_lakes_pca_dim</span> <span class="op">&lt;-</span> <span class="fu">runMatrixAnalysis</span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">alaska_lake_data</span>,
  analysis <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"pca_dim"</span><span class="op">)</span>,
  column_w_names_of_multiple_analytes <span class="op">=</span> <span class="st">"element"</span>,
  column_w_values_for_multiple_analytes <span class="op">=</span> <span class="st">"mg_per_L"</span>,
  columns_w_values_for_single_analyte <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"water_temp"</span>, <span class="st">"pH"</span><span class="op">)</span>,
  columns_w_additional_analyte_info <span class="op">=</span> <span class="st">"element_type"</span>,
  columns_w_sample_ID_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lake"</span>, <span class="st">"park"</span><span class="op">)</span>
<span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">AK_lakes_pca_dim</span><span class="op">)</span>
<span class="co">## # A tibble: 6 × 2</span>
<span class="co">##   principal_component percent_variance_explained</span>
<span class="co">##                 &lt;dbl&gt;                      &lt;dbl&gt;</span>
<span class="co">## 1                   1                      48.8 </span>
<span class="co">## 2                   2                      18.6 </span>
<span class="co">## 3                   3                      11.6 </span>
<span class="co">## 4                   4                       7.88</span>
<span class="co">## 5                   5                       4.68</span>
<span class="co">## 6                   6                       3.33</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">AK_lakes_pca_dim</span>, 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">principal_component</span>, y <span class="op">=</span> <span class="va">percent_variance_explained</span><span class="op">)</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-99-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>Cool! We can see that the first principal component retains nearly 50% of the variance in the original dataset, while the second dimension contains only about 20%.</p>
</div>
<div id="exercises-4" class="section level3" number="8.3.4">
<h3>
<span class="header-section-number">8.3.4</span> exercises<a class="anchor" aria-label="anchor" href="#exercises-4"><i class="fas fa-link"></i></a>
</h3>
<p>In this set of exercises you will choose to complete ONE of the options below. For either option please refer to Chapter 12 for help with principal component and ordination plots. Also, when you are filling out the <code>runMatrixAnalysis()</code> template, you can use the <code><a href="https://rdrr.io/r/base/colnames.html">colnames()</a></code> function to help you specify a long list of column names rather than typing them out by hand. For example, in the periodic table data set, we can refer to a set of columns (columns 10 through 20) with the following command:</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">periodic_table_subset</span><span class="op">)</span><span class="op">[</span><span class="fl">10</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span>
<span class="co">##  [1] "electronegativity_pauling"         </span>
<span class="co">##  [2] "first_ionization_poten_eV"         </span>
<span class="co">##  [3] "second_ionization_poten_eV"        </span>
<span class="co">##  [4] "third_ionization_poten_eV"         </span>
<span class="co">##  [5] "electron_affinity_eV"              </span>
<span class="co">##  [6] "atomic_radius_ang"                 </span>
<span class="co">##  [7] "ionic_radius_ang"                  </span>
<span class="co">##  [8] "covalent_radius_ang"               </span>
<span class="co">##  [9] "atomic_volume_cm3_per_mol"         </span>
<span class="co">## [10] "electrical_conductivity_mho_per_cm"</span>
<span class="co">## [11] "specific_heat_J_per_g_K"</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">periodic_table_subset</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">18</span><span class="op">:</span><span class="fl">20</span>, <span class="fl">23</span><span class="op">:</span><span class="fl">25</span><span class="op">)</span><span class="op">]</span>
<span class="co">## [1] "atomic_volume_cm3_per_mol"         </span>
<span class="co">## [2] "electrical_conductivity_mho_per_cm"</span>
<span class="co">## [3] "specific_heat_J_per_g_K"           </span>
<span class="co">## [4] "thermal_conductivity_W_per_m_K"    </span>
<span class="co">## [5] "polarizability_A_cubed"            </span>
<span class="co">## [6] "heat_atomization_kJ_per_mol"</span></code></pre></div>
<p>We can use that command in the template, as in the example below. With the notation <code>colnames(periodic_table_subset)[c(5:7,9:25)]</code>, we can mark columns 5 - 7 and 9 - 25 as columns_w_values_for_single_analyte (note what happens when you run <code>c(5:7,9:25)</code> in the console, and what happens when you run <code>colnames(periodic_table_subset)[c(5:7,9:25)]</code> in the console). With the notation <code>colnames(periodic_table_subset)[c(1:4, 8)]</code> we can mark columns 1 - 4 and column 8 as columns_w_sample_ID_info (note what happens when you run <code>c(1:4, 8)</code> in the console, and what happens when you run <code>colnames(periodic_table_subset)[c(1:4, 8)]</code> in the console).</p>
<div id="option-1-human-metabolomics." class="section level4" number="8.3.4.1">
<h4>
<span class="header-section-number">8.3.4.1</span> option 1: human metabolomics.<a class="anchor" aria-label="anchor" href="#option-1-human-metabolomics."><i class="fas fa-link"></i></a>
</h4>
<p>This first option is to work with a dataset describing metabolomics data (i.e. abundances of &gt; 100 different biochemicals) from each of 93 human patients, some of which have Chronic Kidney Disease. If you choose this option, your task is to discover a biomarker for Chronic Kidney Disease. This means that you will need to determine a metabolite whose abundance is strongly associated with the disease. To do this you should complete the following:</p>
<ol style="list-style-type: decimal">
<li>Run a PCA analysis on <code>metabolomics_data</code> (i.e. <code>runMatrixAnalysis()</code> with <code>analysis = "pca"</code>)</li>
<li>Plot the results of the analysis to determine which principal component (i.e. dimension) separates the healthy and kidney_disease samples.</li>
<li>Obtain the ordination plot coordinates for the analytes in the PCA analysis (i.e. <code>runMatrixAnalysis()</code> with <code>analysis = "pca_ord"</code>).</li>
<li>Visualize the ordination plot and determine which of the analytes are strongly associated with the principal component (i.e. dimension) separates the healthy and kidney_disease samples.</li>
<li>Bingo! These analytes are associated with Chronic Kidney Disease and could be biomarkers for such.</li>
</ol>
</div>
</div>
<div id="option-2-grape-vine-chemistry" class="section level3" number="8.3.5">
<h3>
<span class="header-section-number">8.3.5</span> option 2: grape vine chemistry<a class="anchor" aria-label="anchor" href="#option-2-grape-vine-chemistry"><i class="fas fa-link"></i></a>
</h3>
<p>This second option is to work with a dataset describing metabolomics data (i.e. abundances of &gt; 100 different biochemicals) from 5 different wine grape varieties. If you choose this option, your task is to discover a biomarker for Chardonnay and a biomarker for Cabernet Sauvignon. This means that you will need to identify two metabolites, each of which are associated with one of those two grape varieties. To do this you should complete the following:</p>
<ol style="list-style-type: decimal">
<li>Run a PCA analysis on <code>wine_grape_data</code> (i.e. <code>runMatrixAnalysis()</code> with <code>analysis = "pca"</code>)</li>
<li>Plot the results of the analysis to determine which principal component (i.e. dimension) separates the Chardonnay samples from the other varieties and the Cabernet Sauvignon samples from the other varieties.</li>
<li>Obtain the ordination plot coordinates for the analytes in the PCA analysis (i.e. <code>runMatrixAnalysis()</code> with <code>analysis = "pca_ord"</code>).</li>
<li>Visualize the ordination plot and determine which of the analytes are strongly associated with the principal component (i.e. dimension) separates the Chardonnay samples from the other varieties and the Cabernet Sauvignon samples from the other varieties.</li>
<li>Bingo! These analytes are associated with those varieites and could be biomarkers for such.</li>
</ol>
<!-- end --><!-- start k-means -->
</div>
</div>
<div id="k-means" class="section level2" number="8.4">
<h2>
<span class="header-section-number">8.4</span> k-means<a class="anchor" aria-label="anchor" href="#k-means"><i class="fas fa-link"></i></a>
</h2>
<p>“Do my samples fall into definable clusters?”</p>
<div id="theory-1" class="section level3" number="8.4.1">
<h3>
<span class="header-section-number">8.4.1</span> theory<a class="anchor" aria-label="anchor" href="#theory-1"><i class="fas fa-link"></i></a>
</h3>
<p>One of the questions we’ve been asking is “which of my samples are most closely related?” We’ve been answering that question using clustering. However, now that we know how to run principal components analyses, we can use another approach. This alternative approach is called k-means, and can help us decide how to assign our data into clusters. It is generally desirable to have a small number of clusters, however, this must be balanced by not having the variance within each cluster be too big. To strike this balance point, the elbow method is used. For it, we must first determine the maximum within-group variance at each possible number of clusters. An illustration of this is shown in <strong>A</strong> below:</p>
<div class="inline-figure"><img src="http://thebustalab.github.io/R_For_Chemists_3/images/kmeans.png" width="80%" style="display: block; margin: auto;"></div>
<p>One we know within-group variances, we find the “elbow” point - the point with minimum angle theta - thus picking the outcome with a good balance of cluster number and within-cluster variance (illustrated above in <strong>B</strong> and <strong>C</strong>.)</p>
<p>Let’s try k-means using <code>runMatrixAnalysis</code>. We can use it in conjunction with <code>analysis = "pca"</code> or <code>analysis = "hclust"</code>. Let’s do PCA first. To include k-means, we can just set <code>kmeans = "auto"</code>. It’s important to note that kmeans cannot handle NAs. We must set something for the <code>na_replacement</code> argument. One solution is to ignore variables that have NAs for some values, which can be done by setting <code>na_replacement = "drop"</code>.</p>
<p>With <code>kmeans = "auto"</code> and <code>na_replacement = "drop"</code>, we can now run our analyssis. The output now has an additional column called <code>kmeans_cluster</code>, which indicates what cluster each sample is in:</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">solvents_pca_kmeans</span> <span class="op">&lt;-</span> <span class="fu">runMatrixAnalysis</span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">solvents</span>,
  analysis <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"pca"</span><span class="op">)</span>,
  column_w_names_of_multiple_analytes <span class="op">=</span> <span class="cn">NULL</span>,
  column_w_values_for_multiple_analytes <span class="op">=</span> <span class="cn">NULL</span>,
  columns_w_values_for_single_analyte <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">solvents</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">7</span><span class="op">:</span><span class="fl">9</span>, <span class="fl">11</span><span class="op">:</span><span class="fl">12</span><span class="op">)</span><span class="op">]</span>,
  columns_w_additional_analyte_info <span class="op">=</span> <span class="cn">NULL</span>,
  columns_w_sample_ID_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"solvent"</span>, <span class="st">"formula"</span>, <span class="st">"miscible_with_water"</span>, <span class="st">"CAS_number"</span>, <span class="st">"category"</span><span class="op">)</span>,
  transpose <span class="op">=</span> <span class="cn">FALSE</span>,
  kmeans <span class="op">=</span> <span class="st">"auto"</span>,
  na_replacement <span class="op">=</span> <span class="st">"drop"</span>
<span class="op">)</span>
<span class="co">## Dropping any variables in your dataset that have NA as a value.</span>
<span class="co">## Variables dropped:</span>
<span class="co">## solubility_in_water vapor_pressure</span>

<span class="va">solvents_pca_kmeans</span>
<span class="co">## # A tibble: 32 × 15</span>
<span class="co">##    sample_unique_ID       solvent   formula miscible_with_w…</span>
<span class="co">##    &lt;chr&gt;                  &lt;chr&gt;     &lt;chr&gt;   &lt;lgl&gt;           </span>
<span class="co">##  1 acetic_acid_C2H4O2_TR… acetic_a… C2H4O2  TRUE            </span>
<span class="co">##  2 acetone_C3H6O_TRUE_67… acetone   C3H6O   TRUE            </span>
<span class="co">##  3 acetonitrile_C2H3N_TR… acetonit… C2H3N   TRUE            </span>
<span class="co">##  4 benzene_C6H6_FALSE_71… benzene   C6H6    FALSE           </span>
<span class="co">##  5 benzonitrile_C7H5N_FA… benzonit… C7H5N   FALSE           </span>
<span class="co">##  6 1-butanol_C4H10O_FALS… 1-butanol C4H10O  FALSE           </span>
<span class="co">##  7 2-butanone_C4H8O_FALS… 2-butano… C4H8O   FALSE           </span>
<span class="co">##  8 carbon_disulfide_CS2_… carbon_d… CS2     FALSE           </span>
<span class="co">##  9 carbon_tetrachloride_… carbon_t… CCl4    FALSE           </span>
<span class="co">## 10 chlorobenzene_C6H5Cl_… chlorobe… C6H5Cl  FALSE           </span>
<span class="co">## # … with 22 more rows, and 11 more variables:</span>
<span class="co">## #   CAS_number &lt;chr&gt;, category &lt;chr&gt;, Dim.1 &lt;dbl&gt;,</span>
<span class="co">## #   Dim.2 &lt;dbl&gt;, kmeans_cluster &lt;chr&gt;, boiling_point &lt;dbl&gt;,</span>
<span class="co">## #   melting_point &lt;dbl&gt;, density &lt;dbl&gt;,</span>
<span class="co">## #   relative_polarity &lt;dbl&gt;, formula_weight &lt;dbl&gt;,</span>
<span class="co">## #   refractive_index &lt;dbl&gt;</span></code></pre></div>
<p>We can plot the results and color them according to the group that kmeans suggested:</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">solvents_pca_kmeans</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Dim.1</span>, y <span class="op">=</span> <span class="va">Dim.2</span>, fill <span class="op">=</span> <span class="va">kmeans_cluster</span><span class="op">)</span>, shape <span class="op">=</span> <span class="fl">21</span>, size <span class="op">=</span> <span class="fl">5</span>, alpha <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-105-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>Hmmm, it looks like the elbow algorithm is suggesting lots of clusters. Why is this? Let’s look at the elbow plot itself. For this, we can just set <code>kmeans = "elbow"</code>:</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">solvents_pca_kmeans_elbow</span> <span class="op">&lt;-</span> <span class="fu">runMatrixAnalysis</span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">solvents</span>,
  analysis <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"pca"</span><span class="op">)</span>,
  column_w_names_of_multiple_analytes <span class="op">=</span> <span class="cn">NULL</span>,
  column_w_values_for_multiple_analytes <span class="op">=</span> <span class="cn">NULL</span>,
  columns_w_values_for_single_analyte <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">solvents</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">7</span><span class="op">:</span><span class="fl">9</span>, <span class="fl">11</span><span class="op">:</span><span class="fl">12</span><span class="op">)</span><span class="op">]</span>,
  columns_w_additional_analyte_info <span class="op">=</span> <span class="cn">NULL</span>,
  columns_w_sample_ID_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"solvent"</span>, <span class="st">"formula"</span>, <span class="st">"miscible_with_water"</span>, <span class="st">"CAS_number"</span>, <span class="st">"category"</span><span class="op">)</span>,
  transpose <span class="op">=</span> <span class="cn">FALSE</span>,
  kmeans <span class="op">=</span> <span class="st">"elbow"</span>,
  na_replacement <span class="op">=</span> <span class="st">"drop"</span>
<span class="op">)</span>
<span class="co">## Dropping any variables in your dataset that have NA as a value.</span>
<span class="co">## Variables dropped:</span>
<span class="co">## solubility_in_water vapor_pressure</span>

<span class="va">solvents_pca_kmeans_elbow</span>
<span class="co">## # A tibble: 31 × 2</span>
<span class="co">##    cluster_number variance_within_cluster</span>
<span class="co">##             &lt;dbl&gt;                   &lt;dbl&gt;</span>
<span class="co">##  1              1                 142804.</span>
<span class="co">##  2              2                  67355.</span>
<span class="co">##  3              3                  49545.</span>
<span class="co">##  4              4                  38964.</span>
<span class="co">##  5              5                  30702.</span>
<span class="co">##  6              6                  25212.</span>
<span class="co">##  7              7                  20188.</span>
<span class="co">##  8              8                  16508.</span>
<span class="co">##  9              9                  14428.</span>
<span class="co">## 10             10                  12265.</span>
<span class="co">## # … with 21 more rows</span></code></pre></div>
<p>This gives us the maximum variance within a cluster for each number of clusters. Let’s plot that:</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>
  <span class="va">solvents_pca_kmeans_elbow</span>,
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">cluster_number</span>, y <span class="op">=</span> <span class="va">variance_within_cluster</span><span class="op">)</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_col</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-107-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>Hmm, it looks like there aren’t any strong elbows in this plot - probably the reason that the elbow method chooses such a high number of clusters. Suppose we want to manually set the number of clusters? We can set <code>kmeans = 3</code> if we want three clusters in the output. Below, let’s do just that. Let’s also plot the results and use <code>geom_mark_ellipse</code>.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">runMatrixAnalysis</span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">solvents</span>,
  analysis <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"pca"</span><span class="op">)</span>,
  column_w_names_of_multiple_analytes <span class="op">=</span> <span class="cn">NULL</span>,
  column_w_values_for_multiple_analytes <span class="op">=</span> <span class="cn">NULL</span>,
  columns_w_values_for_single_analyte <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">solvents</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">7</span><span class="op">:</span><span class="fl">9</span>, <span class="fl">11</span><span class="op">:</span><span class="fl">12</span><span class="op">)</span><span class="op">]</span>,
  columns_w_additional_analyte_info <span class="op">=</span> <span class="cn">NULL</span>,
  columns_w_sample_ID_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"solvent"</span>, <span class="st">"formula"</span>, <span class="st">"miscible_with_water"</span>, <span class="st">"CAS_number"</span>, <span class="st">"category"</span><span class="op">)</span>,
  transpose <span class="op">=</span> <span class="cn">FALSE</span>,
  kmeans <span class="op">=</span> <span class="fl">3</span>,
  na_replacement <span class="op">=</span> <span class="st">"drop"</span>
<span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Dim.1</span>, y <span class="op">=</span> <span class="va">Dim.2</span>, fill <span class="op">=</span> <span class="va">kmeans_cluster</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">21</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_mark_ellipse</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">kmeans_cluster</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">4</span>,<span class="fl">4</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">4</span>,<span class="fl">4</span><span class="op">)</span><span class="op">)</span>
<span class="co">## Dropping any variables in your dataset that have NA as a value.</span>
<span class="co">## Variables dropped:</span>
<span class="co">## solubility_in_water vapor_pressure</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-108-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>Cool!</p>
<p>One more important point: when using kmeans, the output of runMatrixAnalysis (specifically the kmeans_cluster column) can be used to create groupings for summary statistics. For example, suppose we want two groups of solvents and we want to calculate the mean and standard deviation in boiling points for each of those groups:</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">solvents_clustered</span> <span class="op">&lt;-</span> <span class="fu">runMatrixAnalysis</span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">solvents</span>,
  analysis <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"pca"</span><span class="op">)</span>,
  column_w_names_of_multiple_analytes <span class="op">=</span> <span class="cn">NULL</span>,
  column_w_values_for_multiple_analytes <span class="op">=</span> <span class="cn">NULL</span>,
  columns_w_values_for_single_analyte <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">solvents</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">7</span><span class="op">:</span><span class="fl">9</span>, <span class="fl">11</span><span class="op">:</span><span class="fl">12</span><span class="op">)</span><span class="op">]</span>,
  columns_w_additional_analyte_info <span class="op">=</span> <span class="cn">NULL</span>,
  columns_w_sample_ID_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"solvent"</span>, <span class="st">"formula"</span>, <span class="st">"miscible_with_water"</span>, <span class="st">"CAS_number"</span>, <span class="st">"category"</span><span class="op">)</span>,
  transpose <span class="op">=</span> <span class="cn">FALSE</span>,
  kmeans <span class="op">=</span> <span class="fl">2</span>,
  na_replacement <span class="op">=</span> <span class="st">"drop"</span>
<span class="op">)</span>
<span class="co">## Dropping any variables in your dataset that have NA as a value.</span>
<span class="co">## Variables dropped:</span>
<span class="co">## solubility_in_water vapor_pressure</span>

<span class="va">solvents_clustered_summary</span> <span class="op">&lt;-</span> <span class="va">solvents_clustered</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">kmeans_cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span>mean_bp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">boiling_point</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_col</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">solvents_clustered_summary</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">kmeans_cluster</span>, y <span class="op">=</span> <span class="va">mean_bp</span><span class="op">)</span>,
    color <span class="op">=</span> <span class="st">"black"</span>, fill <span class="op">=</span> <span class="st">"white"</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">solvents_clustered</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">kmeans_cluster</span>, y <span class="op">=</span> <span class="va">boiling_point</span><span class="op">)</span>
  <span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-109-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>Very good! Since we can use the outputs of our k-means analyses to run and visualize summary statistics, it’s possible that we’ll want to see the cluster plot (dendrogram or pca plot) alongside the summary stats plot. For this we can use the <code>plot_grid</code> function from the <code>cowplot</code> package. Let’s check it out:</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">solvents_clustered</span> <span class="op">&lt;-</span> <span class="fu">runMatrixAnalysis</span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">solvents</span>,
  analysis <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"pca"</span><span class="op">)</span>,
  column_w_names_of_multiple_analytes <span class="op">=</span> <span class="cn">NULL</span>,
  column_w_values_for_multiple_analytes <span class="op">=</span> <span class="cn">NULL</span>,
  columns_w_values_for_single_analyte <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">solvents</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">7</span><span class="op">:</span><span class="fl">9</span>, <span class="fl">11</span><span class="op">:</span><span class="fl">12</span><span class="op">)</span><span class="op">]</span>,
  columns_w_additional_analyte_info <span class="op">=</span> <span class="cn">NULL</span>,
  columns_w_sample_ID_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"solvent"</span>, <span class="st">"formula"</span>, <span class="st">"miscible_with_water"</span>, <span class="st">"CAS_number"</span>, <span class="st">"category"</span><span class="op">)</span>,
  transpose <span class="op">=</span> <span class="cn">FALSE</span>,
  kmeans <span class="op">=</span> <span class="fl">4</span>,
  na_replacement <span class="op">=</span> <span class="st">"drop"</span>
<span class="op">)</span>
<span class="co">## Dropping any variables in your dataset that have NA as a value.</span>
<span class="co">## Variables dropped:</span>
<span class="co">## solubility_in_water vapor_pressure</span>

<span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"maroon"</span>, <span class="st">"gold"</span>, <span class="st">"grey"</span>, <span class="st">"white"</span><span class="op">)</span>

<span class="va">pca_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span> data <span class="op">=</span> <span class="va">solvents_clustered</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Dim.1</span>, y <span class="op">=</span> <span class="va">Dim.2</span>, fill <span class="op">=</span> <span class="va">kmeans_cluster</span><span class="op">)</span> <span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_mark_ellipse</span><span class="op">(</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">kmeans_cluster</span><span class="op">)</span>, 
    alpha <span class="op">=</span> <span class="fl">0.5</span>, label.lineheight <span class="op">=</span> <span class="fl">0.2</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">21</span>, size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"PCA dimension 1"</span>, breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="op">-</span><span class="fl">8</span>,<span class="fl">8</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"PCA dimension 2"</span>, breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="op">-</span><span class="fl">7</span>,<span class="fl">7</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">colors</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">8</span>,<span class="fl">8</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">7</span>,<span class="fl">7</span><span class="op">)</span><span class="op">)</span>

<span class="va">solvents_clustered_summary</span> <span class="op">&lt;-</span> <span class="va">solvents_clustered</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">kmeans_cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span>mean_bp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">boiling_point</span><span class="op">)</span><span class="op">)</span>

<span class="va">bar_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">geom_violin</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">solvents_clustered</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">kmeans_cluster</span>, y <span class="op">=</span> <span class="va">boiling_point</span>, fill <span class="op">=</span> <span class="va">kmeans_cluster</span><span class="op">)</span>,
    size <span class="op">=</span> <span class="fl">0.5</span>, color <span class="op">=</span> <span class="st">"black"</span>, alpha <span class="op">=</span> <span class="fl">0.6</span>, width <span class="op">=</span> <span class="fl">0.5</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html">geom_crossbar</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">solvents_clustered_summary</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">kmeans_cluster</span>, y <span class="op">=</span> <span class="va">mean_bp</span>, ymin <span class="op">=</span> <span class="va">mean_bp</span>, ymax <span class="op">=</span> <span class="va">mean_bp</span><span class="op">)</span>,
    color <span class="op">=</span> <span class="st">"black"</span>, width <span class="op">=</span> <span class="fl">0.5</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">solvents_clustered</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">kmeans_cluster</span>, y <span class="op">=</span> <span class="va">boiling_point</span><span class="op">)</span>,
    size <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="st">"black"</span>, alpha <span class="op">=</span> <span class="fl">0.6</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Boiling point"</span>, breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">250</span>,<span class="fl">20</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html">scale_x_discrete</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Cluster"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">colors</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span>

<span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">pca_plot</span>, <span class="va">bar_plot</span>, align <span class="op">=</span> <span class="st">"h"</span>, axis <span class="op">=</span> <span class="st">"b"</span>, labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-110-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>Now we are really rockin!!</p>
</div>
<div id="exercises-5" class="section level3" number="8.4.2">
<h3>
<span class="header-section-number">8.4.2</span> exercises<a class="anchor" aria-label="anchor" href="#exercises-5"><i class="fas fa-link"></i></a>
</h3>
<p>Use the wine grapes dataset (it’s stored as <code>wine_grape_data</code> after you run the <code>source(...)</code> command).</p>
<div id="question-1" class="section level4" number="8.4.2.1">
<h4>
<span class="header-section-number">8.4.2.1</span> Question 1<a class="anchor" aria-label="anchor" href="#question-1"><i class="fas fa-link"></i></a>
</h4>
<p>Run a principal components analysis on the dataset. Use <code>na_replacement = "drop"</code> (so that variables with NA values are not included in the analysis) and generate clusters automatically using kmeans by setting <code>kmeans = "auto"</code>. Make scatter plot of the results. How many clusters does kmeans recommend?</p>
</div>
<div id="question-2" class="section level4" number="8.4.2.2">
<h4>
<span class="header-section-number">8.4.2.2</span> Question 2<a class="anchor" aria-label="anchor" href="#question-2"><i class="fas fa-link"></i></a>
</h4>
<p>Modify your code from Question 1 so that only two clusters are generated. Plot the results. Use <code>geom_mark_ellipse</code> to highlight each cluster in your plot (note that the <code>fill</code> aesthetic is required to mark groups). Which varieties are put into each of the two clusters?</p>
</div>
<div id="question-3" class="section level4" number="8.4.2.3">
<h4>
<span class="header-section-number">8.4.2.3</span> Question 3<a class="anchor" aria-label="anchor" href="#question-3"><i class="fas fa-link"></i></a>
</h4>
<p>Use an ordination plot to determine what chemicals makes Chardonnay so different from the other varieties. To what class of compounds do these chemical belong?</p>
</div>
<div id="question-4" class="section level4" number="8.4.2.4">
<h4>
<span class="header-section-number">8.4.2.4</span> Question 4<a class="anchor" aria-label="anchor" href="#question-4"><i class="fas fa-link"></i></a>
</h4>
<p>Modify your code from Question 2 so that five clusters are generated. Plot the results. Use <code>geom_mark_ellipse</code> to highlight each cluster in your plot (note that the <code>fill</code> aesthetic is required to mark groups). Based on this plot, which grape variety undergoes the least amount of change, chemically speaking, between dry and well-watered conditions?</p>
</div>
<div id="question-5" class="section level4" number="8.4.2.5">
<h4>
<span class="header-section-number">8.4.2.5</span> Question 5<a class="anchor" aria-label="anchor" href="#question-5"><i class="fas fa-link"></i></a>
</h4>
<p>Run a heirarchical clustering analysis on the wine grapes data set, using kmeans to create five groups, and also continue using <code>na_replacement = "drop"</code>. Plot the results. Which grape variety undergoes the most change in terms of its chemistry between well-watered and dry conditions? (hint: remember that the x-axis shows the distances between nodes and tips, the y-axis is meaningless). Compare the method you used to compare sample shifts in question 4 (i.e. pca+kmeans) versus the method you used in this question (i.e. hclust+kmeans). Which do you is better? Would this change depending on the circumstances?</p>
</div>
<div id="question-6" class="section level4" number="8.4.2.6">
<h4>
<span class="header-section-number">8.4.2.6</span> Question 6<a class="anchor" aria-label="anchor" href="#question-6"><i class="fas fa-link"></i></a>
</h4>
<p>Google “Quercetin.” What kind of compound is it? Use the clusters created by the heirarchical clustering analysis in question 5 as groups for which to calculate summary statistics. Calculate the mean and standard deviation of the concentration of Quercetin in each group. Plot the result using <code>geom_pointrange</code> and adjust axis font sizes so that they are in good proportion with the size of the plot. Also specify a theme (for example, <code><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic()</a></code>).</p>
<p>Does one cluster have a large amount of variation in Quercetin abundance? Why do you think this might be?</p>
</div>
<div id="question-7" class="section level4" number="8.4.2.7">
<h4>
<span class="header-section-number">8.4.2.7</span> Question 7<a class="anchor" aria-label="anchor" href="#question-7"><i class="fas fa-link"></i></a>
</h4>
<p>Use <code><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">cowplot::plot_grid</a></code> to display your plots from questions 4 and 5 next to each other.</p>
</div>
<div id="challenge-optional" class="section level4" number="8.4.2.8">
<h4>
<span class="header-section-number">8.4.2.8</span> Challenge (optional)<a class="anchor" aria-label="anchor" href="#challenge-optional"><i class="fas fa-link"></i></a>
</h4>
<p>Use cowplot to display your plots from questions 4, 5, and 6 alongside each other. <strong>Make your combined plot as attractive as possible!</strong> Use each of the following:</p>
<p><code>align = TRUE</code> inside <code><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab()</a></code></p>
<p><code>nrow = 1</code> inside <code>plot_grid()</code></p>
<p><code>rel_widths = &lt;your_choice&gt;</code> inside <code>plot_grid()</code></p>
<p><code>name = &lt;your_choice&gt;</code> inside <code>scale_*_*</code></p>
<p><code>label = kmeans_cluster</code> inside <code>geom_mark_ellipse()</code></p>
<p><code>breaks = &lt;your_choice&gt;</code> inside <code><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous()</a></code> or <code><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous()</a></code> (as an example, <code>breaks = seq(0,10,1)</code>)</p>
<p>Also, consider using:</p>
<p><code>guides(fill = "none", color = "none")</code></p>
<p>Install the RColorBrewer package, and use one of its color schemes. As an example with the color scheme <code>Set1</code>:</p>
<p><code>scale_fill_brewer(palette = "Set1", na.value = "grey")</code></p>
<p><code>scale_color_brewer(palette = "Set1", na.value = "grey")</code></p>
<p>Save your plot as a png using <code><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave()</a></code>.</p>
<p>Maybe something like this:</p>
<div class="inline-figure"><img src="http://thebustalab.github.io/R_For_Chemists_3/images/Ex6Challenge.png" width="80%" style="display: block; margin: auto;"></div>
<!-- end -->
<!-- start models -->
</div>
</div>
</div>
<div id="models" class="section level2" number="8.5">
<h2>
<span class="header-section-number">8.5</span> models<a class="anchor" aria-label="anchor" href="#models"><i class="fas fa-link"></i></a>
</h2>
<div id="theory-2" class="section level3" number="8.5.1">
<h3>
<span class="header-section-number">8.5.1</span> theory<a class="anchor" aria-label="anchor" href="#theory-2"><i class="fas fa-link"></i></a>
</h3>
<p>Next on our quest to develop our abilities in analytical data exploration is modeling. We will start with some of the simplest models - linear models. There are a variety of ways to build linear models in R, but we will use a function called <code>buildLinearModel</code>. To use it, we simply give it our data, and tell it which to sets of values we want to compare. To tell it what we want to compare, we give it a formula in the form of Y = M x X + B, however, the B term and the M are implicit, so all we need to tell it is Y = X.</p>
<p>Let’s look at an example. Suppose we want to know if the abundances of ADP and AMP are related in our metabolomics dataset:</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">metabolomics_data</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">AMP</span>, y <span class="op">=</span> <span class="va">ADP</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-120-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>It looks like there might be a relationship! Let’s build a linear model for that relationship:</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu">buildLinearModel</span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">metabolomics_data</span>,
  formula <span class="op">=</span> <span class="st">"ADP = AMP"</span>
<span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">model</span>, strict.width <span class="op">=</span> <span class="st">"cut"</span><span class="op">)</span>
<span class="co">## List of 2</span>
<span class="co">##  $ metrics:'data.frame': 6 obs. of  4 variables:</span>
<span class="co">##   ..$ variable: chr [1:6] "(Intercept)" "AMP" "median_res"..</span>
<span class="co">##   ..$ value   : num [1:6] 0.7842 0.9142 0.0415 40.3224 15...</span>
<span class="co">##   ..$ type    : chr [1:6] "coefficient" "coefficient" "st"..</span>
<span class="co">##   ..$ p_value : chr [1:6] "0.4375" "0" NA NA ...</span>
<span class="co">##  $ data   :'data.frame': 92 obs. of  7 variables:</span>
<span class="co">##   ..$ input_x  : num [1:92] 13.2 13.5 14.3 13.3 12 ...</span>
<span class="co">##   ..$ input_y  : num [1:92] 12.8 13.1 13.3 13.2 11.9 ...</span>
<span class="co">##   ..$ ADP      : num [1:92] 12.8 13.1 13.3 13.2 11.9 ...</span>
<span class="co">##   ..$ AMP      : num [1:92] 13.2 13.5 14.3 13.3 12 ...</span>
<span class="co">##   ..$ residuals: num [1:92] 0.0312 -0.0217 -0.6014 0.2458 ..</span>
<span class="co">##   ..$ model_y  : num [1:92] 12.8 13.1 13.9 13 11.8 ...</span>
<span class="co">##   ..$ model_x  : num [1:92] 13.2 13.5 14.3 13.3 12 ...</span></code></pre></div>
<p>The model consists of two thigs: metrics and data. Let’s look at the metrics:</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">model</span><span class="op">$</span><span class="va">metrics</span>
<span class="co">##               variable   value        type p_value</span>
<span class="co">## 1          (Intercept)  0.7842 coefficient  0.4375</span>
<span class="co">## 2                  AMP  0.9142 coefficient       0</span>
<span class="co">## 3      median_residual  0.0415   statistic    &lt;NA&gt;</span>
<span class="co">## 4    total_sum_squares 40.3224   statistic    &lt;NA&gt;</span>
<span class="co">## 5 residual_sum_squares 15.3901   statistic    &lt;NA&gt;</span>
<span class="co">## 6            r_squared  0.6183   statistic    &lt;NA&gt;</span></code></pre></div>
<p>It shows us the intercept (b), the variable for AMP (i.e. the slope, m), as well some other things (we will talk about them in a second). The other thing the model contains is the data (below). This includes the input_x and y values. The raw values for ADP and AMP, the residuals (see below for details), and the x and y values generated by the model.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">model</span><span class="op">$</span><span class="va">data</span><span class="op">)</span>
<span class="co">##    input_x  input_y      ADP      AMP   residuals  model_y</span>
<span class="co">## 1 13.15029 12.83791 12.83791 13.15029  0.03119000 12.80672</span>
<span class="co">## 2 13.48362 13.08980 13.08980 13.48362 -0.02165141 13.11146</span>
<span class="co">## 3 14.32515 13.27943 13.27943 14.32515 -0.60138528 13.88082</span>
<span class="co">## 4 13.31191 13.20029 13.20029 13.31191  0.24581244 12.95448</span>
<span class="co">## 5 11.99764 11.93350 11.93350 11.99764  0.18057517 11.75293</span>
<span class="co">## 6 12.95966 12.83649 12.83649 12.95966  0.20405638 12.63243</span>
<span class="co">##    model_x</span>
<span class="co">## 1 13.15029</span>
<span class="co">## 2 13.48362</span>
<span class="co">## 3 14.32515</span>
<span class="co">## 4 13.31191</span>
<span class="co">## 5 11.99764</span>
<span class="co">## 6 12.95966</span></code></pre></div>
<p>Let’s plot the model!</p>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">model</span><span class="op">$</span><span class="va">data</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">input_x</span>, y <span class="op">=</span> <span class="va">input_y</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">model_x</span>, y <span class="op">=</span> <span class="va">model_y</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-124-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>Very good. Now let’s talk about evaluating the quality of our model. For this we need some means of assessing how well our line fits our data. We will use residuals - the distance between each of our points and our line.</p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">model</span><span class="op">$</span><span class="va">data</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">input_x</span>, y <span class="op">=</span> <span class="va">input_y</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">model_x</span>, y <span class="op">=</span> <span class="va">model_y</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html">geom_segment</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">input_x</span>, y <span class="op">=</span> <span class="va">input_y</span>, xend <span class="op">=</span> <span class="va">input_x</span>, yend <span class="op">=</span> <span class="va">model_y</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-125-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>We can calculate the sum of the squared residuals:</p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span>
  <span class="op">(</span><span class="va">model</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">input_y</span> <span class="op">-</span> <span class="va">model</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">model_y</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span>
, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">## [1] 15.39014</span></code></pre></div>
<p>15.39! Let’s call that the “residual sum of the squares.” So. 15.39.. does that mean our model is good? I don’t know. We have to compare that number to something. Let’s compare it to a super simple model that is just defined by the mean y value of the input data.</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">metabolomics_data</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">AMP</span>, y <span class="op">=</span> <span class="va">ADP</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">ADP</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-127-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>A pretty bad model, I agree. How much better is our linear model that the flat line model? Let’s create a measure of the distance between each point and the point predicted for that same x value on the model:</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span>
  <span class="op">(</span><span class="va">metabolomics_data</span><span class="op">$</span><span class="va">ADP</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">metabolomics_data</span><span class="op">$</span><span class="va">ADP</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span>
, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">## [1] 40.32239</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">metabolomics_data</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">AMP</span>, y <span class="op">=</span> <span class="va">ADP</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">ADP</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html">geom_segment</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">AMP</span>, y <span class="op">=</span> <span class="va">ADP</span>, xend <span class="op">=</span> <span class="va">AMP</span>, yend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">ADP</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-128-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>40.32! Wow. Let’s call that the “total sum of the squares,” and now we can compare that to our “residual sum of the squares”:</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fl">1</span><span class="op">-</span><span class="op">(</span><span class="fl">15.39</span><span class="op">/</span><span class="fl">40.32</span><span class="op">)</span>
<span class="co">## [1] 0.6183036</span></code></pre></div>
<p>0.68! Alright. That is our R squared value. It is equal to 1 minus the ratio of the “residual sum of the squares” to the “total sum of the squares.” Now, let’s put it all together and make it pretty:</p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">top</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">model</span><span class="op">$</span><span class="va">data</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">input_x</span>, y <span class="op">=</span> <span class="va">input_y</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">model_x</span>, y <span class="op">=</span> <span class="va">model_y</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotate.html">annotate</a></span><span class="op">(</span>geom <span class="op">=</span> <span class="st">"table"</span>,
    x <span class="op">=</span> <span class="fl">11.4</span>,
    y <span class="op">=</span> <span class="fl">16</span>,
    label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">model</span><span class="op">$</span><span class="va">metrics</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span><span class="op">(</span>ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">16</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">bottom</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">model</span><span class="op">$</span><span class="va">data</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_col</a></span><span class="op">(</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">input_x</span>, y <span class="op">=</span> <span class="va">residuals</span><span class="op">)</span>,
    width <span class="op">=</span> <span class="fl">0.03</span>, color <span class="op">=</span> <span class="st">"black"</span>, position <span class="op">=</span> <span class="st">"dodge"</span>, alpha <span class="op">=</span> <span class="fl">0.5</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span>

<span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">top</span>, <span class="va">bottom</span>, ncol <span class="op">=</span> <span class="fl">1</span>, labels <span class="op">=</span> <span class="st">"AUTO"</span>, rel_heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-130-1.png" width="80%" style="display: block; margin: auto;"></div>
</div>
<div id="exercises-6" class="section level3" number="8.5.2">
<h3>
<span class="header-section-number">8.5.2</span> exercises<a class="anchor" aria-label="anchor" href="#exercises-6"><i class="fas fa-link"></i></a>
</h3>
<p>To practice creating linear models, try the following:</p>
<ol style="list-style-type: decimal">
<li><p>Choose one of the datasets we have used so far, and run a principal components analysis on it. Note that the output of the analysis when you run “pca_ord” contains the Dimension 1 coordinate “Dim.1” for each sample, as well as the abundance of each analyte in that sample.</p></li>
<li><p>Using the information from the ordination plot, identify two analytes: one that has a variance that is strongly and positively correlated with the first principal component (i.e. dimension 1), and one that has a variance that is slightly less strongly, but still positively correlated with the first principal component. Using <code>buildLinearModel</code>, create and plot two linear models, one that regresses each of those analytes against dimension 1. Which has the greater r-squared value? Based on what you know about PCA, does that make sense?</p></li>
<li><p>Choose two analytes: one should be one of the analytes from question 2 above, the other should be an analyte that, according to your PCA ordination analysis, is negatively correlated with the first principal component. Using <code>buildLinearModel</code> create plots showing how those two analytes are correlated with dimension 1. One should be positively correlated, and the other negatively correlated. Enhance the plots by including in them a visual represetation of the residuals.</p></li>
</ol>
<!-- end --><!-- start comparing means -->
</div>
</div>
<div id="comparing-means" class="section level2" number="8.6">
<h2>
<span class="header-section-number">8.6</span> comparing means<a class="anchor" aria-label="anchor" href="#comparing-means"><i class="fas fa-link"></i></a>
</h2>
<div class="inline-figure"><img src="http://thebustalab.github.io/R_For_Chemists_3/images/hawaii_aquifers.jpeg" width="80%" style="display: block; margin: auto;"></div>
<p>shapiroTest
leveneTest
tTest
wilcoxTest
anovaTest
tukeyTest
kruskalTest
dunnTest</p>
<p><strong>“Are these two things the same?”</strong></p>
<p>Often, we want to know if our study subjects contain different amounts of certain analytes. For example, “Does this lake over here contain more potassium than that lake over there?” For this, we need statistical tests. Here, we will have a look at comparing mean values for analyte abundance in situations with two samples and in situations with more than two samples.</p>
<p>I find many of the concepts discussed in this chapter easier to think about with an example in mind. For that, suppose that you are an analytical chemist on Hawaii that is studying the chemistry of the island’s aquifers. you have the data set <code>hawaii_aquifers</code>. You can see in the output below the structure of the data set - we have 990 measurements of a 9 different analytes in multiple wells that draw on a set of 10 aquifers.</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hawaii_aquifers</span>
<span class="co">## # A tibble: 990 × 6</span>
<span class="co">##    aquifer_code well_name      longitude latitude analyte   </span>
<span class="co">##    &lt;chr&gt;        &lt;chr&gt;              &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;     </span>
<span class="co">##  1 aquifer_1    Alewa_Heights…        NA       NA SiO2      </span>
<span class="co">##  2 aquifer_1    Alewa_Heights…        NA       NA Cl        </span>
<span class="co">##  3 aquifer_1    Alewa_Heights…        NA       NA Mg        </span>
<span class="co">##  4 aquifer_1    Alewa_Heights…        NA       NA Na        </span>
<span class="co">##  5 aquifer_1    Alewa_Heights…        NA       NA K         </span>
<span class="co">##  6 aquifer_1    Alewa_Heights…        NA       NA SO4       </span>
<span class="co">##  7 aquifer_1    Alewa_Heights…        NA       NA HCO3      </span>
<span class="co">##  8 aquifer_1    Alewa_Heights…        NA       NA dissolved…</span>
<span class="co">##  9 aquifer_1    Alewa_Heights…        NA       NA Ca        </span>
<span class="co">## 10 aquifer_1    Beretania_Hig…        NA       NA SiO2      </span>
<span class="co">## # … with 980 more rows, and 1 more variable:</span>
<span class="co">## #   abundance &lt;dbl&gt;</span>
<span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">hawaii_aquifers</span><span class="op">$</span><span class="va">aquifer_code</span><span class="op">)</span>
<span class="co">##  [1] "aquifer_1"  "aquifer_2"  "aquifer_3"  "aquifer_4" </span>
<span class="co">##  [5] "aquifer_5"  "aquifer_6"  "aquifer_7"  "aquifer_8" </span>
<span class="co">##  [9] "aquifer_9"  "aquifer_10"</span></code></pre></div>
<p>Importantly, there are many wells that draw on each aquifer, as shown in the graph below.</p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hawaii_aquifers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">aquifer_code</span>, <span class="va">well_name</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">aquifer_code</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span>n_wells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">well_name</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">aquifers_summarized</span>

<span class="va">aquifers_summarized</span>
<span class="co">## # A tibble: 10 × 2</span>
<span class="co">##    aquifer_code n_wells</span>
<span class="co">##    &lt;chr&gt;          &lt;int&gt;</span>
<span class="co">##  1 aquifer_1         12</span>
<span class="co">##  2 aquifer_10         7</span>
<span class="co">##  3 aquifer_2          5</span>
<span class="co">##  4 aquifer_3          3</span>
<span class="co">##  5 aquifer_4         16</span>
<span class="co">##  6 aquifer_5          4</span>
<span class="co">##  7 aquifer_6         12</span>
<span class="co">##  8 aquifer_7          9</span>
<span class="co">##  9 aquifer_8          3</span>
<span class="co">## 10 aquifer_9         30</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">aquifers_summarized</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_col</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">n_wells</span>, y <span class="op">=</span> <span class="va">aquifer_code</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-134-1.png" width="80%" style="display: block; margin: auto;"></div>
<!-- To run these statistical analyses, we will need several new R packages: `rstatix`, `agricolae`, and `multcompView`. Please install these with `install.packages("rstatix")`, `install.packages("agricolae")`, and `install.packages("multcompView")`. Load them into your R session using `library(rstatix)`, `library(agricolae)`, and `library(multcompView)`.
 -->
<!-- ```{r, echo = FALSE, message = FALSE}
source("https://thebustalab.github.io/phylochemistry/phylochemistry.R")
library(rstatix)
library(agricolae)
library(multcompView)
``` -->
<div id="definitions" class="section level3" number="8.6.1">
<h3>
<span class="header-section-number">8.6.1</span> definitions<a class="anchor" aria-label="anchor" href="#definitions"><i class="fas fa-link"></i></a>
</h3>
<ol style="list-style-type: decimal">
<li><p><strong>populations and independent measurements</strong>: When we are comparing means, we are comparing two <em>sets</em> of values. It is important to consider where these values came from in the first place. In particular, it is usually useful to think of these values as representatives of larger populations. In the example of our aquifer data set, we can think of the measurements from different wells drawing on the same aquifer as independent measurements of the “population” (i.e. the aquifer).</p></li>
<li><p><strong>the null hypothesis</strong>: When we conduct a statistical test, we are testing the null hypothesis. The null (think “default”) hypothesis is that there is no difference bewteen the means (hence the name “null”). In the example of our aquifers, let’s say that we’re interested in whether two aquifers have different abundances of potassium - in this case the null hypothesis is that they do not differ, in other words, that they have the same amount of potassium.</p></li>
<li><p><strong>the <em>p</em> value</strong>: The <em>p</em> value represents the probability of getting data as extreme as our results if the null hypothesis is true. In other words - the <em>p</em> value is the probability that we would observe the differences we did, if in fact there were no differences in the means at all. To continue with our example: suppose we measure potassium levels in 10% of the wells that access each aquifer and find that aquifer_1 has potassium levels of 14 +/- 2 and aquifer_2 has potassium levels of 12 +/- 1. Suppose that we then conduct a statistical test and get a p value of 0.04. This means that, assuming the aquifers have the same magneisum levels (i.e. assuming the null hypothesis is true), there is a 4% chance that we would get the measured values that we did. In other words, IF the aquifers have the same potassium abundance, it is pretty unlikely that we would have obtained the measurements that we did.</p></li>
</ol>
<p>Please note that the the <em>p</em> value is not the probability of a detected difference being a false positive. The probability of a false positive requires additional information in order to be calculated. For further discussion please see the end of this chapter.</p>
</div>
<div id="test-selection" class="section level3" number="8.6.2">
<h3>
<span class="header-section-number">8.6.2</span> test selection<a class="anchor" aria-label="anchor" href="#test-selection"><i class="fas fa-link"></i></a>
</h3>
<p>There are many different types of statistical tests. Below is a flow chart illustrating how it is recommended that statistical tests be used in this course. You can see that there are three regimes of tests: variance and normality tests (blue), parametric tests (green), and non-parametric tests (orange):</p>
<div class="inline-figure"><img src="http://thebustalab.github.io/R_For_Chemists_3/images/stats.png" width="80%" style="display: block; margin: auto;"></div>
<p>When we are comparing means, we need to first determine what kind of statistical tests we can use with our data. If (i) our data can be reasonably modelled by a normal distribution and (ii) the variances about the two means are similar, then we can use the more powerful “parametric” tests (i.e. tests that will be more likely to detect a difference in means, assuming one exists). If one of these criteria are not met, then we need to use less powerful “non-parametric” tests.</p>
<p>We can check our data for normality and similar variances using the Shapiro test and the Levene test. Let’s use the <code>hawaii_aquifers</code> data as an example, and let’s consider only the element potassium:</p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">K_data</span> <span class="op">&lt;-</span> <span class="va">hawaii_aquifers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">analyte</span> <span class="op">==</span> <span class="st">"K"</span><span class="op">)</span>
  <span class="va">K_data</span>
<span class="co">## # A tibble: 110 × 6</span>
<span class="co">##    aquifer_code well_name         longitude latitude analyte</span>
<span class="co">##    &lt;chr&gt;        &lt;chr&gt;                 &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;  </span>
<span class="co">##  1 aquifer_1    Alewa_Heights_Sp…       NA      NA   K      </span>
<span class="co">##  2 aquifer_1    Beretania_High_S…       NA      NA   K      </span>
<span class="co">##  3 aquifer_1    Beretania_Low_Se…       NA      NA   K      </span>
<span class="co">##  4 aquifer_1    Kuliouou_Well         -158.     21.3 K      </span>
<span class="co">##  5 aquifer_1    Manoa_Well_II         -158.     21.3 K      </span>
<span class="co">##  6 aquifer_1    Moanalua_Wells_P…     -158.     21.4 K      </span>
<span class="co">##  7 aquifer_1    Moanalua_Wells_P…     -158.     21.4 K      </span>
<span class="co">##  8 aquifer_1    Moanalua_Wells_P…     -158.     21.4 K      </span>
<span class="co">##  9 aquifer_1    Nuuanu_Aerator_W…     -158.     21.4 K      </span>
<span class="co">## 10 aquifer_1    Palolo_Tunnel         -158.     21.3 K      </span>
<span class="co">## # … with 100 more rows, and 1 more variable:</span>
<span class="co">## #   abundance &lt;dbl&gt;</span></code></pre></div>
<p>To work with two means, let’s just look at aquifers 1 and 6:</p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">K_data_1_2</span> <span class="op">&lt;-</span> <span class="va">K_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">aquifer_code</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"aquifer_1"</span>, <span class="st">"aquifer_6"</span><span class="op">)</span><span class="op">)</span>

<span class="va">K_data_1_2</span>
<span class="co">## # A tibble: 24 × 6</span>
<span class="co">##    aquifer_code well_name         longitude latitude analyte</span>
<span class="co">##    &lt;chr&gt;        &lt;chr&gt;                 &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;  </span>
<span class="co">##  1 aquifer_1    Alewa_Heights_Sp…       NA      NA   K      </span>
<span class="co">##  2 aquifer_1    Beretania_High_S…       NA      NA   K      </span>
<span class="co">##  3 aquifer_1    Beretania_Low_Se…       NA      NA   K      </span>
<span class="co">##  4 aquifer_1    Kuliouou_Well         -158.     21.3 K      </span>
<span class="co">##  5 aquifer_1    Manoa_Well_II         -158.     21.3 K      </span>
<span class="co">##  6 aquifer_1    Moanalua_Wells_P…     -158.     21.4 K      </span>
<span class="co">##  7 aquifer_1    Moanalua_Wells_P…     -158.     21.4 K      </span>
<span class="co">##  8 aquifer_1    Moanalua_Wells_P…     -158.     21.4 K      </span>
<span class="co">##  9 aquifer_1    Nuuanu_Aerator_W…     -158.     21.4 K      </span>
<span class="co">## 10 aquifer_1    Palolo_Tunnel         -158.     21.3 K      </span>
<span class="co">## # … with 14 more rows, and 1 more variable: abundance &lt;dbl&gt;</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">K_data_1_2</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">aquifer_code</span>, y <span class="op">=</span> <span class="va">abundance</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-137-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>Are these data normally distributed? Do they have similar variance? Let’s get a first approximation by looking at a plot:</p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">K_data_1_2</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">abundance</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">30</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">aquifer_code</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">..density..</span><span class="op">*</span><span class="fl">10</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-138-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>Based on this graphic, it’s hard to say! Let’s use a statistical test to help. When we want to run the Shaprio test, we are looking to see if each group has normally distributed here (here group is “aquifer_code,” i.e. aquifer_1 and aquifer_6). This means we need to <code>group_by(aquifer_code)</code> before we run the test:</p>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">K_data_1_2</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">aquifer_code</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> 
  <span class="fu">shapiro_test</span><span class="op">(</span><span class="va">abundance</span><span class="op">)</span>
<span class="co">## # A tibble: 2 × 4</span>
<span class="co">##   aquifer_code variable  statistic     p</span>
<span class="co">##   &lt;chr&gt;        &lt;chr&gt;         &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">## 1 aquifer_1    abundance     0.885 0.102</span>
<span class="co">## 2 aquifer_6    abundance     0.914 0.239</span></code></pre></div>
<p>Both p-values are above 0.05! This means that the distributions are not significantly different from a normal distribution. What about the variances about the two means? Are they similar? For this we need a Levene test. With that test, we are not looking within each group, but rather across groups - this means we do NOT need to <code>group_by(aquifer_code)</code> and should specify a <code>y ~ x</code> formula instead:</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">K_data_1_2</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu">levene_test</span><span class="op">(</span><span class="va">abundance</span> <span class="op">~</span> <span class="va">aquifer_code</span><span class="op">)</span>
<span class="co">## # A tibble: 1 × 4</span>
<span class="co">##     df1   df2 statistic     p</span>
<span class="co">##   &lt;int&gt; &lt;int&gt;     &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">## 1     1    22     0.289 0.596</span></code></pre></div>
<p>The p-value from this test is 0.596! This means that their variances are not significantly different.</p>
</div>
<div id="two-means" class="section level3" number="8.6.3">
<h3>
<span class="header-section-number">8.6.3</span> two means<a class="anchor" aria-label="anchor" href="#two-means"><i class="fas fa-link"></i></a>
</h3>
<p>Now, since our data passed both test, this means we can use a normal t-test. A t-test is a parametric test. This means that it relies on modelling the data using a normal distribution in order to make comparisons. It is also a powerful test. This means that it is likely to detect a difference in means, assuming one is present. Let’s try it out:</p>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">K_data_1_2</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu">t_test</span><span class="op">(</span><span class="va">abundance</span> <span class="op">~</span> <span class="va">aquifer_code</span><span class="op">)</span>
<span class="co">## # A tibble: 1 × 8</span>
<span class="co">##   .y.       group1 group2    n1    n2 statistic    df      p</span>
<span class="co">## * &lt;chr&gt;     &lt;chr&gt;  &lt;chr&gt;  &lt;int&gt; &lt;int&gt;     &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">## 1 abundance aquif… aquif…    12    12     -2.75  20.5 0.0121</span></code></pre></div>
<p>A p-value of 0.012! This is below 0.05, meaning that there is a 95% chance that the two means are different. Suppose that our data had not passed the Shapiro and/or Levene tests. We would then need to use a Wilcox test. The Wilcox test is a non-parametric test, which means that it does not use a normal distribution to model the data in order to make comparisons. This means that is a less powerful test than the t-test, which means that it is less likely to detect a difference in the means, assuming there is one. For fun, let’s try that one out and compare the p-values from the two methods:</p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">K_data_1_2</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu">wilcox_test</span><span class="op">(</span><span class="va">abundance</span> <span class="op">~</span> <span class="va">aquifer_code</span><span class="op">)</span>
<span class="co">## # A tibble: 1 × 7</span>
<span class="co">##   .y.       group1    group2       n1    n2 statistic      p</span>
<span class="co">## * &lt;chr&gt;     &lt;chr&gt;     &lt;chr&gt;     &lt;int&gt; &lt;int&gt;     &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">## 1 abundance aquifer_1 aquifer_6    12    12      33.5 0.0282</span></code></pre></div>
<p>A p-value of 0.028! This is higher than the value given by the t-test (0.012). That is because the Wilcox test is a less powerful test: it is less likely to detect differences in means, assuming they exist.</p>
</div>
<div id="more-than-two-means" class="section level3" number="8.6.4">
<h3>
<span class="header-section-number">8.6.4</span> more than two means<a class="anchor" aria-label="anchor" href="#more-than-two-means"><i class="fas fa-link"></i></a>
</h3>
<p>In the previous section we compared two means. What if we want to compare means from more than two study subjects? The first step is again to determine which tests to use. Let’s consider our hawaii aquifer data again, though this time let’s use all the aquifers, not just two:</p>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">K_data</span> <span class="op">&lt;-</span> <span class="va">hawaii_aquifers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">analyte</span> <span class="op">==</span> <span class="st">"K"</span><span class="op">)</span>

<span class="va">K_data</span>
<span class="co">## # A tibble: 110 × 6</span>
<span class="co">##    aquifer_code well_name         longitude latitude analyte</span>
<span class="co">##    &lt;chr&gt;        &lt;chr&gt;                 &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;  </span>
<span class="co">##  1 aquifer_1    Alewa_Heights_Sp…       NA      NA   K      </span>
<span class="co">##  2 aquifer_1    Beretania_High_S…       NA      NA   K      </span>
<span class="co">##  3 aquifer_1    Beretania_Low_Se…       NA      NA   K      </span>
<span class="co">##  4 aquifer_1    Kuliouou_Well         -158.     21.3 K      </span>
<span class="co">##  5 aquifer_1    Manoa_Well_II         -158.     21.3 K      </span>
<span class="co">##  6 aquifer_1    Moanalua_Wells_P…     -158.     21.4 K      </span>
<span class="co">##  7 aquifer_1    Moanalua_Wells_P…     -158.     21.4 K      </span>
<span class="co">##  8 aquifer_1    Moanalua_Wells_P…     -158.     21.4 K      </span>
<span class="co">##  9 aquifer_1    Nuuanu_Aerator_W…     -158.     21.4 K      </span>
<span class="co">## 10 aquifer_1    Palolo_Tunnel         -158.     21.3 K      </span>
<span class="co">## # … with 100 more rows, and 1 more variable:</span>
<span class="co">## #   abundance &lt;dbl&gt;</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">K_data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">aquifer_code</span>, x <span class="op">=</span> <span class="va">abundance</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"maroon"</span>, alpha <span class="op">=</span> <span class="fl">0.6</span>, size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-143-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>Let’s check visually to see if each group is normally distributed and to see if they have roughly equal variance:</p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">K_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">aquifer_code</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">abundance</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">30</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">aquifer_code</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">..density..</span><span class="op">*</span><span class="fl">10</span><span class="op">)</span>, colour <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-144-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>Again, it is somewhat hard to tell visually if these data are normally distributed. It seems pretty likely that they have different variances about the means, but let’s check using the Shapiro and Levene tests. Don’t forget: with the Shaprio test, we are looking within each group and so need to <code><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code>, with the Levene test, we are looking across groups, and so need to provide a <code>y~x</code> formula:</p>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">K_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">aquifer_code</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> 
  <span class="fu">shapiro_test</span><span class="op">(</span><span class="va">abundance</span><span class="op">)</span>
<span class="co">## # A tibble: 10 × 4</span>
<span class="co">##    aquifer_code variable  statistic          p</span>
<span class="co">##    &lt;chr&gt;        &lt;chr&gt;         &lt;dbl&gt;      &lt;dbl&gt;</span>
<span class="co">##  1 aquifer_1    abundance     0.885 0.102     </span>
<span class="co">##  2 aquifer_10   abundance     0.864 0.163     </span>
<span class="co">##  3 aquifer_2    abundance     0.913 0.459     </span>
<span class="co">##  4 aquifer_3    abundance     0.893 0.363     </span>
<span class="co">##  5 aquifer_4    abundance     0.948 0.421     </span>
<span class="co">##  6 aquifer_5    abundance     0.902 0.421     </span>
<span class="co">##  7 aquifer_6    abundance     0.914 0.239     </span>
<span class="co">##  8 aquifer_7    abundance     0.915 0.355     </span>
<span class="co">##  9 aquifer_8    abundance     0.842 0.220     </span>
<span class="co">## 10 aquifer_9    abundance     0.786 0.00000866</span></code></pre></div>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">K_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu">levene_test</span><span class="op">(</span><span class="va">abundance</span> <span class="op">~</span> <span class="va">aquifer_code</span><span class="op">)</span>
<span class="co">## # A tibble: 1 × 4</span>
<span class="co">##     df1   df2 statistic       p</span>
<span class="co">##   &lt;int&gt; &lt;int&gt;     &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">## 1     9   100      3.12 0.00239</span></code></pre></div>
<p>Based on these tests, it looks like the data for aquifer 9 is significantly different from a normal distribution (Shaprio test p = 0.000008), and the variances are certainly different from one another (Levene test p = 0.002).</p>
<p>Let’s assume for a second that our data passed these tests. This means that we could reasonably model our data with normal distributions and use a parametric test to compare means. This means that we can use an ANOVA to test for differences in means.</p>
<div id="anova-tukey-tests" class="section level4" number="8.6.4.1">
<h4>
<span class="header-section-number">8.6.4.1</span> ANOVA, Tukey tests<a class="anchor" aria-label="anchor" href="#anova-tukey-tests"><i class="fas fa-link"></i></a>
</h4>
<p>We will use the <code>anova_test</code> function from the package <code>rstatix</code>. It will tell us if any of the means in the data are statistically different from one another. However, if there are differences between the means, it will not tell us which of them are different.</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">K_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu">anova_test</span><span class="op">(</span><span class="va">abundance</span> <span class="op">~</span> <span class="va">aquifer_code</span><span class="op">)</span>
<span class="co">## Coefficient covariances computed by hccm()</span>
<span class="co">## ANOVA Table (type II tests)</span>
<span class="co">## </span>
<span class="co">##         Effect DFn DFd      F        p p&lt;.05   ges</span>
<span class="co">## 1 aquifer_code   9 100 10.021 7.72e-11     * 0.474</span></code></pre></div>
<p>A p-value of 7.7e-11! There are definitely some significant differences among this group. But, WHICH are different from one another though? For this, we need to run Tukey’s Honest Significant Difference test (implemented using <code>tukey_hsd</code>). This will essentially run t-test on all the pairs of study subjects that we can derive from our data set (in this example, aquifer_1 vs. aquifer_2, aquifer_1 vs. aquifer_3, etc.). After that, it will correct the p-values according to the number of comparisons that it performed. This controls the rate of type I error that we can expect from the test. These corrected values are provided to us in the <code>p.adj</code> column.</p>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">K_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu">tukey_hsd</span><span class="op">(</span><span class="va">abundance</span> <span class="op">~</span> <span class="va">aquifer_code</span><span class="op">)</span>
<span class="co">## # A tibble: 45 × 9</span>
<span class="co">##    term         group1     group2 null.value estimate conf.low</span>
<span class="co">##  * &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;       &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;</span>
<span class="co">##  1 aquifer_code aquifer_1  aquif…          0  0.00357   -2.00 </span>
<span class="co">##  2 aquifer_code aquifer_1  aquif…          0  1.44      -0.668</span>
<span class="co">##  3 aquifer_code aquifer_1  aquif…          0  0.375     -2.35 </span>
<span class="co">##  4 aquifer_code aquifer_1  aquif…          0 -1.15      -2.75 </span>
<span class="co">##  5 aquifer_code aquifer_1  aquif…          0 -0.845     -3.09 </span>
<span class="co">##  6 aquifer_code aquifer_1  aquif…          0  1.98       0.261</span>
<span class="co">##  7 aquifer_code aquifer_1  aquif…          0  2.70       0.837</span>
<span class="co">##  8 aquifer_code aquifer_1  aquif…          0 -0.125     -2.85 </span>
<span class="co">##  9 aquifer_code aquifer_1  aquif…          0 -0.378     -1.78 </span>
<span class="co">## 10 aquifer_code aquifer_10 aquif…          0  1.44      -0.910</span>
<span class="co">## # … with 35 more rows, and 3 more variables:</span>
<span class="co">## #   conf.high &lt;dbl&gt;, p.adj &lt;dbl&gt;, p.adj.signif &lt;chr&gt;</span></code></pre></div>
<p>Using the output from our tukey test, we can determine which means are similar. We can do this using the p_groups function:</p>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">groups_based_on_tukey</span> <span class="op">&lt;-</span> <span class="va">K_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu">tukey_hsd</span><span class="op">(</span><span class="va">abundance</span> <span class="op">~</span> <span class="va">aquifer_code</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu">p_groups</span><span class="op">(</span><span class="op">)</span>
<span class="va">groups_based_on_tukey</span>
<span class="co">##             treatment group spaced_group</span>
<span class="co">## aquifer_1   aquifer_1    ab         ab  </span>
<span class="co">## aquifer_10 aquifer_10   abc         abc </span>
<span class="co">## aquifer_2   aquifer_2   acd         a cd</span>
<span class="co">## aquifer_3   aquifer_3  abcd         abcd</span>
<span class="co">## aquifer_4   aquifer_4     b          b  </span>
<span class="co">## aquifer_5   aquifer_5    ab         ab  </span>
<span class="co">## aquifer_6   aquifer_6    cd           cd</span>
<span class="co">## aquifer_7   aquifer_7     d            d</span>
<span class="co">## aquifer_8   aquifer_8   abc         abc </span>
<span class="co">## aquifer_9   aquifer_9    ab         ab</span></code></pre></div>
<p>We can use the output from <code>p_groups</code> to annotate our plot:</p>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">K_data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">aquifer_code</span>, x <span class="op">=</span> <span class="va">abundance</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"maroon"</span>, alpha <span class="op">=</span> <span class="fl">0.6</span>, size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_text</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">groups_based_on_tukey</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">treatment</span>, x <span class="op">=</span> <span class="fl">9</span>, label <span class="op">=</span> <span class="va">group</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-150-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>Excellent! This plot shows us, using the letters on the same line with each aquifer, which means are the same and which are different. If a letter is shared among the labels in line with two aquifers, it means that their means do not differ significantly. For example, aquifer 2 and aquifer 6 both have “b” in their labels, so their means are not different - and are the same as those of aquifers 3 and 10.</p>
</div>
<div id="kruskal-dunn-tests" class="section level4" number="8.6.4.2">
<h4>
<span class="header-section-number">8.6.4.2</span> Kruskal, Dunn tests<a class="anchor" aria-label="anchor" href="#kruskal-dunn-tests"><i class="fas fa-link"></i></a>
</h4>
<p>The above ANOVA example is great, but remember - our data did not pass the Shapiro or Levene tests. This means not all our data can be modelled by a normal distribution and taht we need to use a non-parametric test. The non-parametric alternative to the ANOVA is called the Kruskal test. Like the Wilcox test, it is less powerful that its parametric relative, meaning that it is less likely to detected differences, should they exist. However, since our data do not pass the Shapiro/Levene tests, we have to resort to the Kruskal test. Let’s try it out:</p>
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">K_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu">kruskal_test</span><span class="op">(</span><span class="va">abundance</span> <span class="op">~</span> <span class="va">aquifer_code</span><span class="op">)</span>
<span class="co">## # A tibble: 1 × 6</span>
<span class="co">##   .y.           n statistic    df            p method       </span>
<span class="co">## * &lt;chr&gt;     &lt;int&gt;     &lt;dbl&gt; &lt;int&gt;        &lt;dbl&gt; &lt;chr&gt;        </span>
<span class="co">## 1 abundance   110      57.7     9 0.0000000037 Kruskal-Wall…</span></code></pre></div>
<p>A p-value of 3.9e-9! This is higher than the p-value from running ANOVA on the same data (remember, the Kruskal test is less powerful). Never the less, the value is still well below 0.05, meaning that some of the means are different. So, how do we determine WHICH are different from one another? When we ran ANOVA the follow-up test (the post hoc test) was Tukey’s HSD. After the Kruskal test, the post hoc test we use is the Dunn test. Let’s try:</p>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">K_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu">dunn_test</span><span class="op">(</span><span class="va">abundance</span> <span class="op">~</span> <span class="va">aquifer_code</span><span class="op">)</span>
<span class="co">## # A tibble: 45 × 9</span>
<span class="co">##    .y.    group1 group2    n1    n2 statistic       p  p.adj</span>
<span class="co">##  * &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;  &lt;int&gt; &lt;int&gt;     &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">##  1 abund… aquif… aquif…    12     7    -0.205 0.838   1     </span>
<span class="co">##  2 abund… aquif… aquif…    12     6     2.25  0.0242  0.702 </span>
<span class="co">##  3 abund… aquif… aquif…    12     3     0.911 0.362   1     </span>
<span class="co">##  4 abund… aquif… aquif…    12    17    -2.70  0.00702 0.232 </span>
<span class="co">##  5 abund… aquif… aquif…    12     5    -1.15  0.252   1     </span>
<span class="co">##  6 abund… aquif… aquif…    12    12     2.53  0.0113  0.351 </span>
<span class="co">##  7 abund… aquif… aquif…    12     9     3.02  0.00254 0.0967</span>
<span class="co">##  8 abund… aquif… aquif…    12     3     0.182 0.855   1     </span>
<span class="co">##  9 abund… aquif… aquif…    12    36    -0.518 0.605   1     </span>
<span class="co">## 10 abund… aquif… aquif…     7     6     2.20  0.0278  0.777 </span>
<span class="co">## # … with 35 more rows, and 1 more variable:</span>
<span class="co">## #   p.adj.signif &lt;chr&gt;</span></code></pre></div>
<p>This gives us adjusted p-values for all pairwise comparisons. Once again, we can use <code>p_groups()</code> to give us a compact letter display for each group, which can then be used to annotate the plot:</p>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">groups_based_on_dunn</span> <span class="op">&lt;-</span> <span class="va">K_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu">dunn_test</span><span class="op">(</span><span class="va">abundance</span> <span class="op">~</span> <span class="va">aquifer_code</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu">p_groups</span><span class="op">(</span><span class="op">)</span>
<span class="va">groups_based_on_dunn</span>
<span class="co">##             treatment group spaced_group</span>
<span class="co">## aquifer_1   aquifer_1  abcd         abcd</span>
<span class="co">## aquifer_10 aquifer_10  abcd         abcd</span>
<span class="co">## aquifer_2   aquifer_2   abc         abc </span>
<span class="co">## aquifer_3   aquifer_3  abcd         abcd</span>
<span class="co">## aquifer_4   aquifer_4     d            d</span>
<span class="co">## aquifer_5   aquifer_5   acd         a cd</span>
<span class="co">## aquifer_6   aquifer_6    ab         ab  </span>
<span class="co">## aquifer_7   aquifer_7     b          b  </span>
<span class="co">## aquifer_8   aquifer_8  abcd         abcd</span>
<span class="co">## aquifer_9   aquifer_9    cd           cd</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">K_data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">aquifer_code</span>, x <span class="op">=</span> <span class="va">abundance</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"black"</span>, alpha <span class="op">=</span> <span class="fl">0.4</span>, size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Potassium abundance"</span>, breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">10</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html">scale_y_discrete</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Aquifer code"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_text</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">groups_based_on_dunn</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">treatment</span>, x <span class="op">=</span> <span class="fl">9</span>, label <span class="op">=</span> <span class="va">group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-153-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>Note that these groupings are different from those generated by ANOVA/Tukey.</p>
</div>
</div>
<div id="pairs-of-means" class="section level3" number="8.6.5">
<h3>
<span class="header-section-number">8.6.5</span> pairs of means<a class="anchor" aria-label="anchor" href="#pairs-of-means"><i class="fas fa-link"></i></a>
</h3>
<p>Oftentimes we have more than two means to compare, but rather than wanting to compare all means at once, we want to compare them in a pairwise fashion. For example, suppose we want to know if any of the aquifers contain different amounts of Na and Cl. We are not interested in testing for differences among <em>all</em> values of Na and Cl, rather, we want to test all <em>pairs</em> of Na and Cl values arising from each aquifer. That is to say, we want to compare the means in each facet of the plot below:</p>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hawaii_aquifers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">analyte</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Na"</span>, <span class="st">"Cl"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">analyte</span>, y <span class="op">=</span> <span class="va">abundance</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">geom_violin</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span><span class="op">(</span><span class="va">.</span><span class="op">~</span><span class="va">aquifer_code</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-154-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>Fortunately, we can use an approach that is very similar to the what we’ve learned in the earlier portions of this chapter, just with minor modifications. Let’s have a look! We start with the Shapiro and Levene tests, as usual (note that we group using two variables when using the Shapiro test so that each analyte within each aquifer is considered as an individual distribution):</p>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hawaii_aquifers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">analyte</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Na"</span>, <span class="st">"Cl"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">analyte</span>, <span class="va">aquifer_code</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu">shapiro_test</span><span class="op">(</span><span class="va">abundance</span><span class="op">)</span>
<span class="co">## # A tibble: 20 × 5</span>
<span class="co">##    aquifer_code analyte variable  statistic        p</span>
<span class="co">##    &lt;chr&gt;        &lt;chr&gt;   &lt;chr&gt;         &lt;dbl&gt;    &lt;dbl&gt;</span>
<span class="co">##  1 aquifer_1    Cl      abundance     0.900 1.59e- 1</span>
<span class="co">##  2 aquifer_10   Cl      abundance     0.486 1.09e- 5</span>
<span class="co">##  3 aquifer_2    Cl      abundance     0.869 2.24e- 1</span>
<span class="co">##  4 aquifer_3    Cl      abundance     0.75  0       </span>
<span class="co">##  5 aquifer_4    Cl      abundance     0.903 7.49e- 2</span>
<span class="co">##  6 aquifer_5    Cl      abundance     0.767 4.22e- 2</span>
<span class="co">##  7 aquifer_6    Cl      abundance     0.741 2.15e- 3</span>
<span class="co">##  8 aquifer_7    Cl      abundance     0.893 2.12e- 1</span>
<span class="co">##  9 aquifer_8    Cl      abundance     0.878 3.17e- 1</span>
<span class="co">## 10 aquifer_9    Cl      abundance     0.414 7.58e-11</span>
<span class="co">## 11 aquifer_1    Na      abundance     0.886 1.06e- 1</span>
<span class="co">## 12 aquifer_10   Na      abundance     0.593 2.26e- 4</span>
<span class="co">## 13 aquifer_2    Na      abundance     0.884 2.88e- 1</span>
<span class="co">## 14 aquifer_3    Na      abundance     0.822 1.69e- 1</span>
<span class="co">## 15 aquifer_4    Na      abundance     0.933 2.41e- 1</span>
<span class="co">## 16 aquifer_5    Na      abundance     0.782 5.71e- 2</span>
<span class="co">## 17 aquifer_6    Na      abundance     0.764 3.80e- 3</span>
<span class="co">## 18 aquifer_7    Na      abundance     0.915 3.51e- 1</span>
<span class="co">## 19 aquifer_8    Na      abundance     0.855 2.53e- 1</span>
<span class="co">## 20 aquifer_9    Na      abundance     0.544 2.09e- 9</span></code></pre></div>
<p>Looks like some of those distributions are significantly different from normal! Let’s run the levene test anyway. Note that for this particular case of the Levene test, we are interested in testing whether each pair of distributions has similar variances. For that we need to feed the Levene test data that is grouped by aquifer_code (so that it tests each pair as a group), then we need to specify the y ~ x formula (which in this case is abundance ~ analyte):</p>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hawaii_aquifers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">analyte</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Na"</span>, <span class="st">"Cl"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">aquifer_code</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu">levene_test</span><span class="op">(</span><span class="va">abundance</span> <span class="op">~</span> <span class="va">analyte</span><span class="op">)</span>
<span class="co">## # A tibble: 10 × 5</span>
<span class="co">##    aquifer_code   df1   df2 statistic       p</span>
<span class="co">##    &lt;chr&gt;        &lt;int&gt; &lt;int&gt;     &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">##  1 aquifer_1        1    22   10.5    0.00375</span>
<span class="co">##  2 aquifer_10       1    12    0.0535 0.821  </span>
<span class="co">##  3 aquifer_2        1    10    0.0243 0.879  </span>
<span class="co">##  4 aquifer_3        1     4    0.320  0.602  </span>
<span class="co">##  5 aquifer_4        1    32    1.57   0.219  </span>
<span class="co">##  6 aquifer_5        1     8    0.474  0.511  </span>
<span class="co">##  7 aquifer_6        1    22    1.03   0.322  </span>
<span class="co">##  8 aquifer_7        1    16    1.54   0.232  </span>
<span class="co">##  9 aquifer_8        1     4    0.515  0.512  </span>
<span class="co">## 10 aquifer_9        1    70    1.07   0.304</span></code></pre></div>
<p>It looks like the variances of the pair in aquifer 1 have significantly different variances. So - we for sure need to be using non-parametric testing. If this were a simple case of two means we would use the <code>wilcox_test</code>, but we have may pairs, so we will use <code>pairwise_wilcox_test</code> (note that with this test there are options for various styles of controlling for multiple comparisons, see: <code>?pairwise_wilcox_test</code>):</p>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hawaii_aquifers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">analyte</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Na"</span>, <span class="st">"Cl"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">aquifer_code</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu">pairwise_wilcox_test</span><span class="op">(</span><span class="va">abundance</span><span class="op">~</span><span class="va">analyte</span><span class="op">)</span>
<span class="co">## # A tibble: 10 × 10</span>
<span class="co">##    aquifer_code .y.       group1 group2    n1    n2 statistic</span>
<span class="co">##  * &lt;chr&gt;        &lt;chr&gt;     &lt;chr&gt;  &lt;chr&gt;  &lt;int&gt; &lt;int&gt;     &lt;dbl&gt;</span>
<span class="co">##  1 aquifer_1    abundance Cl     Na        12    12      99.5</span>
<span class="co">##  2 aquifer_10   abundance Cl     Na         7     7      14  </span>
<span class="co">##  3 aquifer_2    abundance Cl     Na         6     6      36  </span>
<span class="co">##  4 aquifer_3    abundance Cl     Na         3     3       3  </span>
<span class="co">##  5 aquifer_4    abundance Cl     Na        17    17     189  </span>
<span class="co">##  6 aquifer_5    abundance Cl     Na         5     5      17.5</span>
<span class="co">##  7 aquifer_6    abundance Cl     Na        12    12      53  </span>
<span class="co">##  8 aquifer_7    abundance Cl     Na         9     9      42  </span>
<span class="co">##  9 aquifer_8    abundance Cl     Na         3     3       6  </span>
<span class="co">## 10 aquifer_9    abundance Cl     Na        36    36     248. </span>
<span class="co">## # … with 3 more variables: p &lt;dbl&gt;, p.adj &lt;dbl&gt;,</span>
<span class="co">## #   p.adj.signif &lt;chr&gt;</span></code></pre></div>
<p>Excellent! It looks like there is a statistically significant difference between the means of the abundances of Cl and Na in aquifer_2 and (surprisingly?) in aquifer_9 (perhaps due to the large number of observations?).</p>
<p>What would we have done if our Shaprio and Levene tests had revealed no significant differences? Well, a <code>pairwise_t_test</code> of course!</p>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hawaii_aquifers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">analyte</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Na"</span>, <span class="st">"Cl"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">aquifer_code</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu">pairwise_t_test</span><span class="op">(</span><span class="va">abundance</span><span class="op">~</span><span class="va">analyte</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">test_output</span>
  <span class="va">test_output</span>
<span class="co">## # A tibble: 10 × 10</span>
<span class="co">##    aquifer_code .y.       group1 group2    n1    n2        p</span>
<span class="co">##  * &lt;chr&gt;        &lt;chr&gt;     &lt;chr&gt;  &lt;chr&gt;  &lt;int&gt; &lt;int&gt;    &lt;dbl&gt;</span>
<span class="co">##  1 aquifer_1    abundance Cl     Na        12    12  4.69e-2</span>
<span class="co">##  2 aquifer_10   abundance Cl     Na         7     7  8.82e-1</span>
<span class="co">##  3 aquifer_2    abundance Cl     Na         6     6  3.75e-5</span>
<span class="co">##  4 aquifer_3    abundance Cl     Na         3     3  6.83e-1</span>
<span class="co">##  5 aquifer_4    abundance Cl     Na        17    17  1.03e-1</span>
<span class="co">##  6 aquifer_5    abundance Cl     Na         5     5  1.45e-1</span>
<span class="co">##  7 aquifer_6    abundance Cl     Na        12    12  5.66e-1</span>
<span class="co">##  8 aquifer_7    abundance Cl     Na         9     9  5.21e-1</span>
<span class="co">##  9 aquifer_8    abundance Cl     Na         3     3  4.28e-1</span>
<span class="co">## 10 aquifer_9    abundance Cl     Na        36    36  9.48e-1</span>
<span class="co">## # … with 3 more variables: p.signif &lt;chr&gt;, p.adj &lt;dbl&gt;,</span>
<span class="co">## #   p.adj.signif &lt;chr&gt;</span></code></pre></div>
<p>Excellent, now we see how to run parametric and non-parametric pairwise comparisons. How do we annotate plots with the output of these tests? Here is an example:</p>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>
  xmin <span class="op">=</span> <span class="va">test_output</span><span class="op">$</span><span class="va">group1</span>,
  xmax <span class="op">=</span> <span class="va">test_output</span><span class="op">$</span><span class="va">group2</span>,
  y_position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">150</span>, <span class="fl">150</span>, <span class="fl">150</span>, <span class="fl">175</span>, <span class="fl">80</span>, <span class="fl">50</span>, <span class="fl">300</span>, <span class="fl">150</span>, <span class="fl">50</span>, <span class="fl">125</span><span class="op">)</span>,
  text <span class="op">=</span> <span class="va">test_output</span><span class="op">$</span><span class="va">p.signif</span>,
  text_size <span class="op">=</span> <span class="fl">10</span>,
  text_vert_offset <span class="op">=</span> <span class="fl">10</span>,
  text_horiz_offset <span class="op">=</span> <span class="fl">1.5</span>,
  tip_length_xmin <span class="op">=</span> <span class="fl">5</span>,
  tip_length_xmax <span class="op">=</span> <span class="fl">5</span>,
  aquifer_code <span class="op">=</span> <span class="va">test_output</span><span class="op">$</span><span class="va">aquifer_code</span>
<span class="op">)</span>

<span class="va">hawaii_aquifers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">analyte</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Na"</span>, <span class="st">"Cl"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">analyte</span>, y <span class="op">=</span> <span class="va">abundance</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">geom_violin</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"gold"</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">21</span>, fill <span class="op">=</span> <span class="st">"maroon"</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span><span class="op">(</span><span class="va">.</span><span class="op">~</span><span class="va">aquifer_code</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geomSignif</span><span class="op">(</span>data <span class="op">=</span> <span class="va">anno</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html">scale_x_discrete</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Analyte"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Abundance"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>
    text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">16</span><span class="op">)</span>
    <span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="index_files/figure-html/unnamed-chunk-159-1.png" width="80%" style="display: block; margin: auto;"></div>
</div>
<div id="further-reading" class="section level3" number="8.6.6">
<h3>
<span class="header-section-number">8.6.6</span> further reading<a class="anchor" aria-label="anchor" href="#further-reading"><i class="fas fa-link"></i></a>
</h3>
<p>For more on comparing multiple means in R: <a href="https://www.datanovia.com/en/courses/comparing-multiple-means-in-r/">www.datanovia.com</a></p>
<p>For more on parametric versus non-parametric tests: <a href="https://statisticsbyjim.com/hypothesis-testing/nonparametric-parametric-tests/">Statistics by Jim</a></p>
<p>For more on interpreting <em>p</em> values: [<a href="https://link.springer.com/article/10.1007/s00259-019-04467-5">The p value wars (again) by Ulrich Dirnagl</a>]</p>
</div>
<div id="exercises-7" class="section level3" number="8.6.7">
<h3>
<span class="header-section-number">8.6.7</span> exercises<a class="anchor" aria-label="anchor" href="#exercises-7"><i class="fas fa-link"></i></a>
</h3>
<p>Using the <code>hawaii_aquifers</code> data set, please complete the following:</p>
<ol style="list-style-type: decimal">
<li><p>Choose one analyte and filter the data so only the rows for that analyte are shown.</p></li>
<li><p>Choose two of the aquifers. Are the mean abundances for your chosen analyte different in these two aquifers? Don’t forget to test your data for normality and homogeneity of variance before selecting a statistical test. Use a plot to illustrate whether the means are similar or different.</p></li>
<li><p>Choose a second analyte, different from the first one you chose. Considering all the aquifers in the dataset, do any of them have the same abundance of this analyte? Again, don’t forget about normality and homogeneity of variance tests. Use a plot to illustrate your answer.</p></li>
<li><p>Repeat #3 above, but switch the type of test used (i.e. use non-parametric if you used parametric for #3 and vice-versa). Compare the <em>p</em> values and <em>p</em> groups obtained by the two methods. Use a graphic to illustrate this. Why are they different?</p></li>
</ol>
<!-- end --><!-- start Mid Term Exam --><!-- # {-}
# MIDTERM EXAM {-}

For this examination, imagine that you are an analytical chemist that was just hired by Fireside Brewing, a large beer brewing company in Montana. At Fireside Brewing, beer is brewed using recipes that can include barley, corn, hops, hop oil, and/or hop extract. With these six ingredients the company can make all the beer varieties discussed below.

The person that previously worked in your position, Mr. Porter, recently completed a chemical analysis of the major ingredients that Fireside Brewing uses. However, Mr. Porter didn't have time to finish analyzing the data that was collected.

At Fireside Brewing, your boss is Mrs. Pilsner, a woman who has worked in the brewing industry for decades and is a certified cicerone (a professional beer taster). As an expert taster, she is familiar with all the classes of compounds in beer ingredients and their corresponding flavors, but she is not an expert in chemistry or data analysis.

For your first task in this new position, Mrs. Pilsner wants you to finish analyzing Mr. Porter's data. Mrs. Pilsner has not given you specific instructions on how to analyze the data, rather, she has a set of specific questions that she wants answered so that she can improve Fireside Brewing. Mr. Porter left some notes about how to perform the analyses, but, since he was close to retirement and spent too much time riding his bicycle and drinking beer, the notes are not complete. Use Mr. Porter's incomplete notes to help you answer the questions below.

**Answer each question with both (i) at least one high quality plot that provides a visual answer and (ii) a caption that provides a written answer. If a question necessitates an answer that is difficult to show with a plot (for example, a single *p* value), include that answer in the caption.**

Write your answers in an R Markdown document, compile the document so that everything is shown: the code, your writing, and the plots. Submit the compiled pdf on Canvas.

*note: a high quality plot is one in which, for example, axes tick labels do not overlap but also fill the space available to them, colors are used, raw data is plotted (if possible), axes labels are customized, an appropriate theme is chosen, and geoms are chosen carefully. The plots should be visually attractive and professional. You are a new employee - impress people!*

<img src="http://thebustalab.github.io/R_For_Chemists_3/images/plot_quality.jpg" width="80%" style="display: block; margin: auto;" />


## Data set 1: beer_components

Load this data set by running our source() command. It will then be present in your environment as `beer_components`.

### Question 1 [7 points]:

Mrs. Pilsner: *We are having problems with our Pale Ale. Its flavor is not consistent from batch to batch. We need to identify the ingredient that is causing this problem. What is the variability of each analyte class (ketone, sesquiterpene, etc.) in each of the ingredients we use (barley, corn, hop extract, hop oil, and hops), and particuarly, which ingredient and which class are most highly variable?*



### Question 2 [12 points]:

Mrs. Pilsner: *We are experimenting with a new beer that uses a 1:1:1 mixture of all the hop products that we use (hops, hop oil, and hop extract). I want to try to predict what type of flavor profile this beer might have. Using Mr. Porter's data for those three ingredients (hops, hop oil, and hop extract), can you predict what the relative abundance of each analyte class will be in the mixture?*



### Question 3 [15 points]:

Mrs. Pilsner: *I recently heard a rumor from other cicerones that, compared to other types of barley, barley containing significantly more 3-methylbutanal than all other aldehydes leads to a maltier flavor. Statistically speaking, does our barley currently contain significantly more 3-methylbutanal than all other aldehydes?*

note from Mr. Porter: *Reminder to self: before conducting statistical tests, filter out analytes whose abundance == 0 in all five replicates, otherwise the tests will be biased.*



### Question 4 [15 points]:

Mrs. Pilsner: *When I taste our beers made with hop extract, they don't taste very hoppy compared to our beers made with hop oil or hops. Why is this - how do ALL our different ingredients (including corn and barley) cluster according to their average chemical composition?*



### Question 5 [*extra credit*, 4 points]:

Mrs. Pilsner: *We are trying to reduce the price of our India Session Ale by replacing its hops with either hop extract or hop oil. Chemically speaking, which is more similar to hops - hop extract, or hop oil?*



## Data set 2: hops_components

After you answer all Mrs. Pilsner's questions about Mr. Porter's data, Mrs. Pilsner tells you about another data set that Fireside Brewing has. This data set was collected by Mr. Porter's intern, an undergraduate named Ms. Amber Stout. Amber measured the chemical composition of many different hop varieties. Please answer Mrs. Pilsner's questions below about the hops data in the same way you did for the questions above.

To load this second data set, the hops data set, ensure you have run our source() command. It will then be present in your environment as `hops_components`.

### Question 1 [12 points]:

Mrs. Pilsner: *There are three major types of hops - aroma hops, dry hopping hops, and bittering hops. We had an emergency scenario last month where we ran out of bittering hops due to covid-related supply line problems. We needed to substitute either aroma hops or dry hopping hops into the recipe in place of bittering hops. At the time, we used aroma hops. Was that a good decision? If that emergency were to arise again, should we substitute aroma hops again, or should we use dry hopping hops instead? As much as possible, we would want the product from the emergency-substituted batch to be as similar to the normal batch made with bittering hops.*



### Question 2 [15 points]:

Mrs. Pilsner: *We buy our hops from many different countries of origin. I've always wondered which country has the most unique hop varieties. Can you make a plot that shows me which country has the most hop varieties that have no overlap with the chemical profiles of other countries' hops (as a proxy for flavor profiles)? What chemical(s) largely drive this difference?*

note from Mr. Porter: *note to self... by reducing the dimensionality of the chemical dataset to two dimensions, it should be possible to draw ellipses that encapsulate the chemical profile (as a proxy for "flavor profile") of each country's hops.*



### Question 3 [12 points]:

Mrs. Pilsner: *If we wanted to prepare our own hop oil, which type of hops should we use - aroma hops, bittering hops, or dry hopping hops? In other words, is there a statistical difference in the amount of total oil that each type of hops has? We want to pick the type of hops that would maximize our oil yield.*

Mr. Porter: *Note to self: here we are **not** interested in comparing `total_oil` from different `hop_varieties`, but instead in comparing `total_oil` from hops with different `hop_brewing_usage`.*



### Question 4 [12 points]:

Mrs. Pilsner: *Cascade hops are a very popular variety, and also a variety we use in producing our flagship IPA. However, due to COVID-19, cascade hops are on back order for six months. What style of hop is most similar to cascade? I would like to order it to use in place of cascade hops for the batch of IPA we will brew two weeks from now.*



### Question 5 [*extra credit*, 4 points]:

Mrs. Pilsner: *We are interested in producing a new type of beer with the lowest amount of acids possible. Which variety of hops has the lowest total amount of acids?*


--><!-- end --><hr>
<hr>
<hr>
</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="overview.html">overview</a></div>
<div class="next"><a href="chromatography-and-separations.html"><span class="header-section-number">10</span> chromatography and separations</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#clustering"><span class="header-section-number">8</span> clustering</a></li>
<li><a class="nav-link" href="#theory"><span class="header-section-number">8.1</span> theory</a></li>
<li><a class="nav-link" href="#exercises-3"><span class="header-section-number">8.2</span> exercises</a></li>
<li>
<a class="nav-link" href="#pca"><span class="header-section-number">8.3</span> pca</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#pca-1"><span class="header-section-number">8.3.1</span> pca</a></li>
<li><a class="nav-link" href="#ordination-plots"><span class="header-section-number">8.3.2</span> ordination plots</a></li>
<li><a class="nav-link" href="#principal-components"><span class="header-section-number">8.3.3</span> principal components</a></li>
<li><a class="nav-link" href="#exercises-4"><span class="header-section-number">8.3.4</span> exercises</a></li>
<li><a class="nav-link" href="#option-2-grape-vine-chemistry"><span class="header-section-number">8.3.5</span> option 2: grape vine chemistry</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#k-means"><span class="header-section-number">8.4</span> k-means</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#theory-1"><span class="header-section-number">8.4.1</span> theory</a></li>
<li><a class="nav-link" href="#exercises-5"><span class="header-section-number">8.4.2</span> exercises</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#models"><span class="header-section-number">8.5</span> models</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#theory-2"><span class="header-section-number">8.5.1</span> theory</a></li>
<li><a class="nav-link" href="#exercises-6"><span class="header-section-number">8.5.2</span> exercises</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#comparing-means"><span class="header-section-number">8.6</span> comparing means</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#definitions"><span class="header-section-number">8.6.1</span> definitions</a></li>
<li><a class="nav-link" href="#test-selection"><span class="header-section-number">8.6.2</span> test selection</a></li>
<li><a class="nav-link" href="#two-means"><span class="header-section-number">8.6.3</span> two means</a></li>
<li><a class="nav-link" href="#more-than-two-means"><span class="header-section-number">8.6.4</span> more than two means</a></li>
<li><a class="nav-link" href="#pairs-of-means"><span class="header-section-number">8.6.5</span> pairs of means</a></li>
<li><a class="nav-link" href="#further-reading"><span class="header-section-number">8.6.6</span> further reading</a></li>
<li><a class="nav-link" href="#exercises-7"><span class="header-section-number">8.6.7</span> exercises</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/rstudio/bookdown-demo/blob/master/index.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/rstudio/bookdown-demo/edit/master/index.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Integrated Bioanalytical Chemistry</strong>" was written by Lucas Busta. It was last built on 2022-01-04.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
