--- 
title: "Phylochemistry"
author: "Lucas Busta"
date: "`r Sys.Date()`"
site: bookdown::bookdown_site
always_allow_html: yes
output:
  bookdown::gitbook:
    highlight: tango
    df_print: tibble
    css: style.css
    config:
      toc:
        collapse: section
      fontsettings:
        theme: white
        family: sans
        size: 2 
      toolbar:
        position: fixed     
documentclass: book
bibliography: [bio723-refs.bib]
biblio-style: apalike
link-citations: yes
description: "Vingette for the R Package Phylochemistry."
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, prompt = FALSE, eval = TRUE, 
                      warning = FALSE, comment="##", cache = TRUE,
                      fig.width = 6, fig.height = 4, #results = "hide",
                      collapse=TRUE, results='markup', max.print=6)

options(pillar.sigfig = 3)
```

# Introduction 

```{r fig.align='center', echo=FALSE, include=identical(knitr:::pandoc_to(), 'html'), results="markup"}
knitr::include_graphics('http://thebustalab.github.io/phylochemistry/figures/phylochemistry_logo.jpg', dpi = NA)
```

`phylochemistry` enables analyses of naturally occurring chemicals and genomic data in the context of phylogeny.

## Importing data

Let's import some data and get down to business! First we need some data. Let's use this cool algae chemistry dataset. First, remember that RSudio operates as a session. This means that we need to make sure any libraries/packages we need are available to the current session. To import data, let's use the `tidyverse` function called `read_csv`. This means we need to first make sure the `tidyverse` is loaded into the current session, then we can run the `read_csv` command.

```{r}
library(tidyverse)
```

Next, we need to find the path to our data. If you are on a PC, right-click the file of interest (in this case algae_data.csv), go to preferences, then copy the ... If you are on a Mac, find the file right-click on it, hold down the option key, and then click "Copy <file> as pathname". One you have the path, insert it IN QUOTES into the read_csv command.

```{r}
read_csv()
```


<!--chapter:end:index.Rmd-->


# Setting up a `phylochemistry` analysis

Placeholder


## Installing `phylochemistry` directly from github and view documentation contents
## Conventions in `phylochemsitry`
### File paths and directory paths
### Monolists
### Recommendations
#### Storing your data as a list of lists
#### in that case...

<!--chapter:end:01-setting_up_phylochemistry.Rmd-->


# Analyzing chromatography data with `phylochemistry`

Placeholder


## Preparing the raw data and metadata
## Using the `phylochemistry` integrationApp

<!--chapter:end:02-analyzing_GC_MS_data.Rmd-->


# Gene mining

Placeholder


## BLAST
### Using the blastTranscriptomes command

<!--chapter:end:03-gene_mining.Rmd-->

