--- 
title: "Phylochemistry"
author: "Lucas Busta"
date: "`r Sys.Date()`"
output:
  bookdown::gitbook:
    css: fix.css
    config:
      toc:
        collapse: section
      fontsettings:
        theme: night
        family: sans
        size: 2 
      toolbar:
        position: fixed     
description: "Vingette for the R Package Phylochemistry."
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, prompt = FALSE, eval = TRUE, 
                      warning = FALSE, comment="##", cache = TRUE,
                      fig.width = 6, fig.height = 4, #results = "hide",
                      collapse=TRUE, results='markup', max.print=6)

options(pillar.sigfig = 3)
```

<!-- start phylochemistry --> 

# phylochemistry

![](http://thebustalab.github.io/phylochemistry/figures/phylochemistry_logo.jpg)

`phylochemistry` is a set of functions for chemical, transcriptomic, and genomic analysis. These tools are provided though a combination of new functions and wrapped features of previously developed packages. A number of new functions to streamline analyses in this interdisciplinary space are also provided. This repository contains the latest version of the `phylochemistry` R package.

## install

`phylochemistry` can be installed by following these four steps:

1. Install `msa` and `phangorn` by running the following code in R
```{r, eval = FALSE, message = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
BiocManager::install("msa")
BiocManager::install("ggtree")
BiocManager::install("phangorn")
BiocManager::install("seqinr")
BiocManager::install("Biostrings")
BiocManager::install("GenomicRanges")
```
2. Install prequisite R packages by running the following code in R
```{r, eval = FALSE, message = FALSE}
install.packages("tidyverse")
install.packages("DT")
install.packages("RColorBrewer")
install.packages("xcms")
install.packages("data.table")
install.packages("rhandsontable")
install.packages("ape")
install.packages("ips")
```

3. (for Windows only) [install X.Org](https://www.x.org/wiki/)

4. (for Mac only) [install XQuartz](https://www.xquartz.org/)

5. Load `phylochemistry` directly into your R session using the following:
```{r, message = FALSE}
source("http://thebustalab.github.io/phylochemistry/phylochemistry.R")
```

## new features

1. A Shiny app for GC-FID and GC-MS data analysis.
2. Open reading frame extraction from multiple fasta files.
3. Store BLAST results in a .csv file.
4. Minor ticks for ggplot2 axes.
5. Phylogenetic signal for discrete traits.
6. Analyze multiple sequence alignments for sites associated with user-defined function
7. Multiple column name, multiple row name data structures (aka "polylists").
8. Draw annotated multiple sequence alignments.
9. Use image analysis to automatically get the csv of a mass spectrum from a published image.
10. Draw chemical structures in R from a csv of molecular coordinates.

## wrapped features

1. BLAST transcriptomes, via [NCBI BLAST+](https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&DOC_TYPE=Download).
2. Multiple sequence alignments and codon alignments of amino acid and nucleotide sequences, via [msa](https://bioconductor.org/packages/release/bioc/html/msa.html) and [orthologr](https://github.com/HajkD/orthologr).
3. Phylogenetic tree construction (including g-blocks trimming, pruning, ancestral states reconstruction), via [phangorn](https://cran.r-project.org/web/packages/phangorn/index.html).
4. Systematic read/write functions (csv, newick, wide tables, fasta, summary statistic tables, GFFs, chromatograms, mass spectra).
5. Phylogenetic signal for continuous traits.

<!-- end --> 