--- 
title: "Phylochemistry"
author: "Lucas Busta"
date: "`r Sys.Date()`"
output:
  bookdown::gitbook:
    css: fix.css
    config:
      toc:
        collapse: section
      fontsettings:
        theme: sepia
        family: sans
        size: 2 
      toolbar:
        position: fixed     
description: "Vingette for the R Package Phylochemistry."
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, prompt = FALSE, eval = TRUE, 
                      warning = FALSE, comment="##", cache = TRUE,
                      fig.width = 6, fig.height = 4, #results = "hide",
                      collapse=TRUE, results='markup', max.print=6)

options(pillar.sigfig = 3)
```

<!-- start phylochemistry --> 

# phylochemistry

![](http://thebustalab.github.io/images/phylochemistry_logo.jpg)

`phylochemistry` is a set of functions for chemical, transcriptomic, and genomic analysis. These tools are provided though a combination of new computational functions and wrapped features of previously developed packages. A number of new organizational and data handling functions to streamline analyses in this interdisciplinary space are also provided. This page provides access to the latest version of `phylochemistry`.

## requirements

* To run `phylochemistry`, you need to have R and RStudio installed. For instructions on how to install those, please see [this page](https://thebustalab.github.io/R_For_Chemists/installation.html).

## load phylochemistry

Phylochemistry is not an R package, but rather a set of components that you can add to your R environment by running an R script hosted on this site. `phylochemistry` requires a number of existing R packages in order to run, but don't worry,`phylochemistry` will help you install these packages if they are not installed already. 

1. Load `phylochemistry` directly into your R session by running the following command in RStudio:
```{r, message = FALSE, eval = FALSE}
source("http://thebustalab.github.io/phylochemistry/phylochemistry.R")
```
 \
Sometimes running the command above generates the message "You need to install the following packages before proceeding [...] Run: installPhylochemistry() to automatically install the required packages.". This means that some of the prerequisite packages that phylochemistry needs are not installed. If this happens, run the following:\
```{r, message = FALSE, eval = FALSE}
installPhylochemistry()
```
 \
Once that is complete, then try the source() command again:
```{r, message = FALSE, eval = FALSE}
source("http://thebustalab.github.io/phylochemistry/phylochemistry.R")
```

<!-- 1. (for Windows only) [install X.Org](https://www.x.org/wiki/)

1. (for Mac only) [install XQuartz](https://www.xquartz.org/)
 -->

## R scripts on Google Drive

Sometimes we want to save our R scripts on Google Drive. If you have an R script on Google Drive and want to open it in RStudio, get the share link for the file and use the following command:
```{r, message = FALSE, eval = FALSE}
openRGD("file_share_link_here")
```
 \
When you do this, "IN_USE___" will appear in front of the file name in Google Drive, so that others will know that you are using it. When you are done using the file, you can save and close it using:
```{r, message = FALSE, eval = FALSE}
closeRGD("file_share_link_here")
```

## new features

1. A Shiny app for GC-FID and GC-MS data analysis, including a large MS library.
2. Open reading frame extraction from multiple fasta files.
3. BLAST searches that export .fasta files of hits and store results in a .csv file.
4. Minor ticks for ggplot2 axes.
5. Phylogenetic signal for discrete traits.
6. Analyze multiple sequence alignments for sites associated with user-defined function
7. Multiple column name, multiple row name data structures (aka "polylists").
8. Draw annotated multiple sequence alignments.
9. Use image analysis to automatically get the csv of a mass spectrum from a published image.
10. Draw chemical structures in R from a csv of molecular coordinates.

## wrapped features

1. BLAST transcriptomes, via [NCBI BLAST+](https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&DOC_TYPE=Download).
2. Multiple sequence alignments and codon alignments of amino acid and nucleotide sequences, via [msa](https://bioconductor.org/packages/release/bioc/html/msa.html) and [orthologr](https://github.com/HajkD/orthologr).
3. Phylogenetic tree construction (including g-blocks trimming, pruning, ancestral states reconstruction), via [phangorn](https://cran.r-project.org/web/packages/phangorn/index.html).
4. Systematic read/write functions (csv, newick, wide tables, fasta, summary statistic tables, GFFs, chromatograms, mass spectra).
5. Phylogenetic signal for continuous traits, via [picante](https://cran.r-project.org/web/packages/picante/index.html).

<!-- end -->

<!-- start chemical analyses -->
# chemical analyses

## integrationAppLite

`phylochemistry` provides a simple application for integrating and analyzing GC-MS data. With it, you can analyze .CDF files, which contain essentially all the data from a GC-MS run, and can be exported from most GC-MS systems using the software provided by the manufacturer. Instructions for this are provided at the end of this chapter. To run the lite version of the integration app, use the following guidelines:

1. Create a new folder on your hard drive and place your CDF file into that folder. It doesn't matter what the name of that folder is, but it must not contain special characters (including a space ` ` in the name). For example, if my CDF file is called "sorghum_bicolor.CDF", then I might create a folder called `gc_data` on my hard drive, and place the "sorghum_bicolor.CDF" file in that folder.

2. In RStudio, run the source command to load `phylochemistry`:

```{r, message = FALSE, eval = FALSE}
source("http://thebustalab.github.io/phylochemistry/phylochemistry.R")
```
 \

3. In RStudio, run the `integrationAppLite` command on the *folder* that contains your CDF file. 
 \

If you are on a Mac, *use single forward slashes*. For example:
```{r, message = FALSE, eval = FALSE}
integrationAppLite("/Volumes/My_Drive/gc_data")
```
 \

If you are on a PC, *use double back slashes*. For example:
```{r, message = FALSE, eval = FALSE}
integrationAppLite("C:\\Users\\My Profile\\gc_data")
```
 \

The first time you open your datafile, it may take a while to load. This is normal - the program is analyzing all the data points in your data file. For a typical exploratory GC-MS run of around 60 minutes, this is more than 2.5 million data points! So please be patient. After you open your data file once, subsequent openings will not take so long.

Please watch this overview video for a demonstration of how to use the integration app.

As a reference, below are the key commands used to operate the integration app. This is the information that is covered in the overview video.

To control the chromatogram window:

* shift + q = update
* shift + a = add selected peak
* shift + r = remove selected peak
* shift + f = forward
* shift + d = backward
* shift + c = zoom in
* shift + v = zoom out
* shift + z = save table

To control the mass spectrum window:

* shift+1 = extract mass spectra from highlighted chromatogram region, plot average mass spectrum in panel 1.
* shift+2 = refresh mass spectrum in panel 1. This is used for zooming in on a region of the mass spectrum that you have highlighted. A spectrum needs to first be extracted for this to be possible.
* shift+3 = extract mass spectra from highlighted chromatogram region, subtract their average from the mass spectrum in panel 1.
* shift+4 = search current spectrum in panel 1 against library of mass spectra.


## CDF export

1. On the GC-MS computer, open Enhanced Data Analysis
2. File > Export Data To .AIA Format, Create New Directory (“OK”) > Desktop (create a folder with a name you will remember)
3. Select all the datafiles you wish to analyze and process them, saving the output into the folder you just created
4. Copy the .D files for the samples you wish to analyze to the same folder
5. Move this folder to your personal computer
6. Create one folder for each sample, and put the corresponding .CDF file into that folder. 
<!-- end -->

<!-- start transcriptomic analyses -->
# transcriptomic analyses

## BLAST
<!-- end -->

<!-- start genomic analyses -->
# genomic analyses

## loading GFF files
<!-- end -->

<!-- start evolutionary analyses -->
# evolutionary analyses

## buildTree

### Simple template

```{r, eval = FALSE}
buildTree(
  scaffold_type = "newick",
  scaffold_in_path = NULL,
  members = NULL
)
```

### Full template

```{r}
buildTree(
  scaffold_type = c("amin_alignment", "nucl_alignment", "newick"),
  scaffold_in_path = NULL,
  members = NULL,
  gblocks = FALSE, 
  gblocks_path = NULL,
  ml = FALSE, 
  model_test = FALSE,
  bootstrap = FALSE,
  rois = FALSE, 
  rois_data = NULL,
  ancestral_states = FALSE,
  root = NULL
)
```

## collapseTree
<!-- end -->