<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Phylochemistry</title>
  <meta name="description" content="Vingette for the R Package Phylochemistry." />
  <meta name="generator" content="bookdown 0.23 and GitBook 2.6.7" />

  <meta property="og:title" content="Phylochemistry" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Vingette for the R Package Phylochemistry." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Phylochemistry" />
  
  <meta name="twitter:description" content="Vingette for the R Package Phylochemistry." />
  

<meta name="author" content="Lucas Busta" />


<meta name="date" content="2021-08-26" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  

<link rel="next" href="chemical-analyses.html"/>
<script src="libs/header-attrs-2.10/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="fix.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="phylochemistry.html"><a href="phylochemistry.html"><i class="fa fa-check"></i><b>1</b> phylochemistry</a>
<ul>
<li class="chapter" data-level="1.1" data-path="phylochemistry.html"><a href="phylochemistry.html#requirements"><i class="fa fa-check"></i><b>1.1</b> requirements</a></li>
<li class="chapter" data-level="1.2" data-path="phylochemistry.html"><a href="phylochemistry.html#load-phylochemistry"><i class="fa fa-check"></i><b>1.2</b> load phylochemistry</a></li>
<li class="chapter" data-level="1.3" data-path="phylochemistry.html"><a href="phylochemistry.html#r-scripts-on-google-drive"><i class="fa fa-check"></i><b>1.3</b> R scripts on Google Drive</a></li>
<li class="chapter" data-level="1.4" data-path="phylochemistry.html"><a href="phylochemistry.html#new-features"><i class="fa fa-check"></i><b>1.4</b> new features</a></li>
<li class="chapter" data-level="1.5" data-path="phylochemistry.html"><a href="phylochemistry.html#wrapped-features"><i class="fa fa-check"></i><b>1.5</b> wrapped features</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="chemical-analyses.html"><a href="chemical-analyses.html"><i class="fa fa-check"></i><b>2</b> chemical analyses</a>
<ul>
<li class="chapter" data-level="2.1" data-path="chemical-analyses.html"><a href="chemical-analyses.html#integrationapplite"><i class="fa fa-check"></i><b>2.1</b> integrationAppLite</a></li>
<li class="chapter" data-level="2.2" data-path="chemical-analyses.html"><a href="chemical-analyses.html#cdf-export"><i class="fa fa-check"></i><b>2.2</b> CDF export</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="transcriptomic-analyses.html"><a href="transcriptomic-analyses.html"><i class="fa fa-check"></i><b>3</b> transcriptomic analyses</a>
<ul>
<li class="chapter" data-level="3.1" data-path="transcriptomic-analyses.html"><a href="transcriptomic-analyses.html#blast"><i class="fa fa-check"></i><b>3.1</b> BLAST</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="genomic-analyses.html"><a href="genomic-analyses.html"><i class="fa fa-check"></i><b>4</b> genomic analyses</a>
<ul>
<li class="chapter" data-level="4.1" data-path="genomic-analyses.html"><a href="genomic-analyses.html#loading-gff-files"><i class="fa fa-check"></i><b>4.1</b> loading GFF files</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Phylochemistry</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">Phylochemistry</h1>
<p class="author"><em>Lucas Busta</em></p>
<p class="date"><em>2021-08-26</em></p>
</div>
<div id="phylochemistry" class="section level1" number="1">
<h1><span class="header-section-number">1</span> phylochemistry</h1>
<p><img src="http://thebustalab.github.io/images/phylochemistry_logo.jpg" /></p>
<p><code>phylochemistry</code> is a set of functions for chemical, transcriptomic, and genomic analysis. These tools are provided though a combination of new computational functions and wrapped features of previously developed packages. A number of new organizational and data handling functions to streamline analyses in this interdisciplinary space are also provided. This page provides access to the latest version of <code>phylochemistry</code>.</p>
<div id="requirements" class="section level2" number="1.1">
<h2><span class="header-section-number">1.1</span> requirements</h2>
<ul>
<li>To run <code>phylochemistry</code>, you need to have R and RStudio installed. For instructions on how to install those, please see <a href="https://thebustalab.github.io/R_For_Chemists/installation.html">this page</a>.</li>
</ul>
</div>
<div id="load-phylochemistry" class="section level2" number="1.2">
<h2><span class="header-section-number">1.2</span> load phylochemistry</h2>
<p>Phylochemistry is not an R package, but rather a set of components that you can add to your R environment by running an R script hosted on this site. <code>phylochemistry</code> requires a number of existing R packages in order to run, but don’t worry,<code>phylochemistry</code> will help you install these packages if they are not installed already.</p>
<ol style="list-style-type: decimal">
<li>Load <code>phylochemistry</code> directly into your R session by running the following command in RStudio:</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="phylochemistry.html#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;http://thebustalab.github.io/phylochemistry/phylochemistry.R&quot;</span>)</span></code></pre></div>
<p><br />
Sometimes running the command above generates the message “You need to install the following packages before proceeding […] Run: installPhylochemistry() to automatically install the required packages.”. This means that some of the prerequisite packages that phylochemistry needs are not installed. If this happens, run the following:<br />
</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="phylochemistry.html#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">installPhylochemistry</span>()</span></code></pre></div>
<p><br />
Once that is complete, then try the source() command again:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="phylochemistry.html#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;http://thebustalab.github.io/phylochemistry/phylochemistry.R&quot;</span>)</span></code></pre></div>
<!-- 1. (for Windows only) [install X.Org](https://www.x.org/wiki/)

1. (for Mac only) [install XQuartz](https://www.xquartz.org/)
 -->
</div>
<div id="r-scripts-on-google-drive" class="section level2" number="1.3">
<h2><span class="header-section-number">1.3</span> R scripts on Google Drive</h2>
<p>Sometimes we want to save our R scripts on Google Drive. If you have an R script on Google Drive and want to open it in RStudio, get the share link for the file and use the following command:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="phylochemistry.html#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">openRGD</span>(<span class="st">&quot;file_share_link_here&quot;</span>)</span></code></pre></div>
<p><br />
When you do this, "IN_USE___" will appear in front of the file name in Google Drive, so that others will know that you are using it. When you are done using the file, you can save and close it using:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="phylochemistry.html#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">closeRGD</span>(<span class="st">&quot;file_share_link_here&quot;</span>)</span></code></pre></div>
</div>
<div id="new-features" class="section level2" number="1.4">
<h2><span class="header-section-number">1.4</span> new features</h2>
<ol style="list-style-type: decimal">
<li>A Shiny app for GC-FID and GC-MS data analysis, including a large MS library.</li>
<li>Open reading frame extraction from multiple fasta files.</li>
<li>BLAST searches that export .fasta files of hits and store results in a .csv file.</li>
<li>Minor ticks for ggplot2 axes.</li>
<li>Phylogenetic signal for discrete traits.</li>
<li>Analyze multiple sequence alignments for sites associated with user-defined function</li>
<li>Multiple column name, multiple row name data structures (aka “polylists”).</li>
<li>Draw annotated multiple sequence alignments.</li>
<li>Use image analysis to automatically get the csv of a mass spectrum from a published image.</li>
<li>Draw chemical structures in R from a csv of molecular coordinates.</li>
</ol>
</div>
<div id="wrapped-features" class="section level2" number="1.5">
<h2><span class="header-section-number">1.5</span> wrapped features</h2>
<ol style="list-style-type: decimal">
<li>BLAST transcriptomes, via <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">NCBI BLAST+</a>.</li>
<li>Multiple sequence alignments and codon alignments of amino acid and nucleotide sequences, via <a href="https://bioconductor.org/packages/release/bioc/html/msa.html">msa</a> and <a href="https://github.com/HajkD/orthologr">orthologr</a>.</li>
<li>Phylogenetic tree construction (including g-blocks trimming, pruning, ancestral states reconstruction), via <a href="https://cran.r-project.org/web/packages/phangorn/index.html">phangorn</a>.</li>
<li>Systematic read/write functions (csv, newick, wide tables, fasta, summary statistic tables, GFFs, chromatograms, mass spectra).</li>
<li>Phylogenetic signal for continuous traits, via <a href="https://cran.r-project.org/web/packages/picante/index.html">picante</a>.</li>
</ol>
<!-- end -->
</div>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="chemical-analyses.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "sepia",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": false,
"toc": {
"collapse": "section"
},
"toolbar": {
"position": "fixed"
}
});
});
</script>

</body>

</html>
