<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 4 Gene mining | Phylochemistry</title>
  <meta name="description" content="Vingette for the R Package Phylochemistry." />
  <meta name="generator" content="bookdown 0.18 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 4 Gene mining | Phylochemistry" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Vingette for the R Package Phylochemistry." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 4 Gene mining | Phylochemistry" />
  
  <meta name="twitter:description" content="Vingette for the R Package Phylochemistry." />
  

<meta name="author" content="Lucas Busta" />


<meta name="date" content="2020-03-20" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="analyzing-chromatography-data-with-phylochemistry.html"/>

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { background-color: #f8f8f8; }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.0.1" data-path="index.html"><a href="index.html#install-phylochemistry-directly-from-github-and-view-documentation-contents"><i class="fa fa-check"></i><b>1.0.1</b> Install <code>phylochemistry</code> directly from github and view documentation contents</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="setting-up-a-phylochemistry-analysis.html"><a href="setting-up-a-phylochemistry-analysis.html"><i class="fa fa-check"></i><b>2</b> Setting up a <code>phylochemistry</code> analysis</a><ul>
<li class="chapter" data-level="2.1" data-path="setting-up-a-phylochemistry-analysis.html"><a href="setting-up-a-phylochemistry-analysis.html#conventions-in-phylochemsitry"><i class="fa fa-check"></i><b>2.1</b> Conventions in <code>phylochemsitry</code></a><ul>
<li class="chapter" data-level="2.1.1" data-path="setting-up-a-phylochemistry-analysis.html"><a href="setting-up-a-phylochemistry-analysis.html#monolists"><i class="fa fa-check"></i><b>2.1.1</b> Monolists</a></li>
<li class="chapter" data-level="2.1.2" data-path="setting-up-a-phylochemistry-analysis.html"><a href="setting-up-a-phylochemistry-analysis.html#storing-your-data-as-a-list-of-lists"><i class="fa fa-check"></i><b>2.1.2</b> Storing your data as a list of lists</a></li>
<li class="chapter" data-level="2.1.3" data-path="setting-up-a-phylochemistry-analysis.html"><a href="setting-up-a-phylochemistry-analysis.html#nomenclature"><i class="fa fa-check"></i><b>2.1.3</b> Nomenclature</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="setting-up-a-phylochemistry-analysis.html"><a href="setting-up-a-phylochemistry-analysis.html#setting-up-a-new-phylochemistry-analysis"><i class="fa fa-check"></i><b>2.2</b> Setting up a new <code>phylochemistry</code> analysis</a><ul>
<li class="chapter" data-level="2.2.1" data-path="setting-up-a-phylochemistry-analysis.html"><a href="setting-up-a-phylochemistry-analysis.html#install-phylochemistry-directly-from-github-and-view-documentation-contents-1"><i class="fa fa-check"></i><b>2.2.1</b> Install <code>phylochemistry</code> directly from github and view documentation contents</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="setting-up-a-phylochemistry-analysis.html"><a href="setting-up-a-phylochemistry-analysis.html#visualizing-data-types-separately"><i class="fa fa-check"></i><b>2.3</b> Visualizing data types separately</a></li>
<li class="chapter" data-level="2.4" data-path="setting-up-a-phylochemistry-analysis.html"><a href="setting-up-a-phylochemistry-analysis.html#building-a-phylochem-object-from-scratch"><i class="fa fa-check"></i><b>2.4</b> Building a phylochem object from scratch</a><ul>
<li class="chapter" data-level="2.4.1" data-path="setting-up-a-phylochemistry-analysis.html"><a href="setting-up-a-phylochemistry-analysis.html#create-a-phylochem-object-called-fads-and-bind-taxonomic-data-to-it"><i class="fa fa-check"></i><b>2.4.1</b> Create a phylochem object called “fads” and bind taxonomic data to it</a></li>
<li class="chapter" data-level="2.4.2" data-path="setting-up-a-phylochemistry-analysis.html"><a href="setting-up-a-phylochemistry-analysis.html#create-a-phylochem-object-called-fads-and-bind-taxonomic-data-to-it-1"><i class="fa fa-check"></i><b>2.4.2</b> Create a phylochem object called “fads” and bind taxonomic data to it</a></li>
<li class="chapter" data-level="2.4.3" data-path="setting-up-a-phylochemistry-analysis.html"><a href="setting-up-a-phylochemistry-analysis.html#bind-chemical-data-to-the-fads-phylochem-object"><i class="fa fa-check"></i><b>2.4.3</b> Bind chemical data to the “fads” phylochem object</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="analyzing-chromatography-data-with-phylochemistry.html"><a href="analyzing-chromatography-data-with-phylochemistry.html"><i class="fa fa-check"></i><b>3</b> Analyzing chromatography data with <code>phylochemistry</code></a><ul>
<li class="chapter" data-level="3.1" data-path="analyzing-chromatography-data-with-phylochemistry.html"><a href="analyzing-chromatography-data-with-phylochemistry.html#preparing-the-raw-data-and-metadata"><i class="fa fa-check"></i><b>3.1</b> Preparing the raw data and metadata</a></li>
<li class="chapter" data-level="3.2" data-path="analyzing-chromatography-data-with-phylochemistry.html"><a href="analyzing-chromatography-data-with-phylochemistry.html#using-the-phylochemistry-integrationapp"><i class="fa fa-check"></i><b>3.2</b> Using the <code>phylochemistry</code> integrationApp</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="gene-mining.html"><a href="gene-mining.html"><i class="fa fa-check"></i><b>4</b> Gene mining</a><ul>
<li class="chapter" data-level="4.1" data-path="gene-mining.html"><a href="gene-mining.html#blast"><i class="fa fa-check"></i><b>4.1</b> BLAST</a><ul>
<li class="chapter" data-level="4.1.1" data-path="gene-mining.html"><a href="gene-mining.html#using-the-blasttranscriptomes-command"><i class="fa fa-check"></i><b>4.1.1</b> Using the blastTranscriptomes command</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Phylochemistry</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="gene-mining" class="section level1">
<h1><span class="header-section-number">Chapter 4</span> Gene mining</h1>
<p>Very often, we are interested in learning about genes in a particular species, or in a set of species. This section will walk you through the functions offered by <code>phylochemistry</code> that help solve problems in this area.</p>
<div id="blast" class="section level2">
<h2><span class="header-section-number">4.1</span> BLAST</h2>
<p>One of the most common tasks we want to perform are BLAST searches. For this, we need a query (what you are looking to find - “the needle”) and a subject (“the haystack”). Often, these two take the forms of the sequence of a known gene (the query), and a transcriptome (the haystack). Let’s look at an example.</p>
<p>Note: this example requires BLAST+ to be installed. See: <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download" class="uri">https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download</a>, or google “BLAST+”.</p>
<div id="using-the-blasttranscriptomes-command" class="section level3">
<h3><span class="header-section-number">4.1.1</span> Using the blastTranscriptomes command</h3>
<p>Let’s load the phylochemistry library and set up our phylochemistry directory</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">library</span>(phylochemistry)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2">phylochemistry_example_dir &lt;-<span class="st"> &quot;/Users/lucasbusta/Desktop/phylochemistry_example_data/&quot;</span></a>
<a class="sourceLine" id="cb1-3" data-line-number="3">phylochemistry_example_dir</a></code></pre></div>
<pre><code>## [1] &quot;/Users/lucasbusta/Desktop/phylochemistry_example_data/&quot;</code></pre>
<p>Next, we’ll need to make a named object that contans the path to the subject transcriptome</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">transcriptome_path &lt;-<span class="st"> </span><span class="kw">paste0</span>(phylochemistry_example_dir, <span class="st">&quot;sequences_by_species/Sorghum_bicolor/Sbicolor_454_v3.1.1.cds_primaryTranscriptOnly.fa&quot;</span>)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="kw">names</span>(transcriptome_path) &lt;-<span class="st"> &quot;Sorghum_bicolor&quot;</span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3">transcriptome_path</a></code></pre></div>
<pre><code>##                                                                                                                               Sorghum_bicolor 
## &quot;/Users/lucasbusta/Desktop/phylochemistry_example_data/sequences_by_species/Sorghum_bicolor/Sbicolor_454_v3.1.1.cds_primaryTranscriptOnly.fa&quot;</code></pre>
<p>Finally, we can run the blast. The blastTranscriptomes() command needs a variety of information to run:</p>
<ul>
<li>transcriptomes - the named list of paths to your transcriptomes (in this example, the <code>transcriptome_path</code> object we created).</li>
<li>query_in_path - the path to the query. Note that this can be a single or multifasta file.</li>
<li>blast_module_directory_path - the path to the directory of the BLAST+ module on your computer.</li>
<li>monolist_out_path - the path to the monolist that you want the results written to.</li>
<li>sequences_of_interest_directory_path - the path to the directory where you want blast hits to be stored.</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">blastTranscriptomes</span>(</a>
<a class="sourceLine" id="cb5-2" data-line-number="2">  <span class="dt">transcriptomes =</span> transcriptome_path,</a>
<a class="sourceLine" id="cb5-3" data-line-number="3">  <span class="dt">query_in_path =</span> <span class="kw">paste0</span>(phylochemistry_example_dir, <span class="st">&quot;sequences_by_species/Arabidopsis_thaliana/AtThalaniolCyclase.fa&quot;</span>),</a>
<a class="sourceLine" id="cb5-4" data-line-number="4">  <span class="dt">blast_module_directory_path =</span> <span class="st">&quot;/usr/local/ncbi/blast/bin/&quot;</span>,</a>
<a class="sourceLine" id="cb5-5" data-line-number="5">  <span class="dt">blast_type =</span> <span class="st">&quot;dc-megablast&quot;</span>,</a>
<a class="sourceLine" id="cb5-6" data-line-number="6">  <span class="dt">monolist_out_path =</span> <span class="kw">paste0</span>(phylochemistry_example_dir, <span class="st">&quot;monolists/blast_hits.csv&quot;</span>),</a>
<a class="sourceLine" id="cb5-7" data-line-number="7">  <span class="dt">sequences_of_interest_directory_path =</span> <span class="kw">paste0</span>(phylochemistry_example_dir, <span class="st">&quot;sequences_of_interest/&quot;</span>)</a>
<a class="sourceLine" id="cb5-8" data-line-number="8">)</a></code></pre></div>
<pre><code>## Found 15 BLAST hits found for AtThalaniolCyclase in Sorghum_bicolor
## Done!</code></pre>
<p>Now that this has run, let’s look at the results. There are two aspects to the results. One is the monolist output:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">sequences &lt;-<span class="st"> </span><span class="kw">readMonolist</span>(<span class="dt">monolist_in_path =</span> <span class="kw">paste0</span>(phylochemistry_example_dir, <span class="st">&quot;monolists/blast_hits.csv&quot;</span>))</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">sequences</a></code></pre></div>
<pre><code>## # A tibble: 14 x 8
##    accession  Genus  species annotation length subset_all query  longestORF
##    &lt;chr&gt;      &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;       &lt;int&gt; &lt;lgl&gt;      &lt;chr&gt;       &lt;int&gt;
##  1 Sobic.010… Sorgh… bicolor Sobic.010…   2205 TRUE       AtTha…       2064
##  2 Sobic.005… Sorgh… bicolor Sobic.005…    828 TRUE       AtTha…        828
##  3 Sobic.005… Sorgh… bicolor Sobic.005…   2301 TRUE       AtTha…       2301
##  4 Sobic.008… Sorgh… bicolor Sobic.008…   2277 TRUE       AtTha…       2277
##  5 Sobic.008… Sorgh… bicolor Sobic.008…   2274 TRUE       AtTha…       2274
##  6 Sobic.004… Sorgh… bicolor Sobic.004…    249 TRUE       AtTha…        249
##  7 Sobic.004… Sorgh… bicolor Sobic.004…   2286 TRUE       AtTha…       2286
##  8 Sobic.004… Sorgh… bicolor Sobic.004…   2274 TRUE       AtTha…       2274
##  9 Sobic.007… Sorgh… bicolor Sobic.007…   2187 TRUE       AtTha…       2187
## 10 Sobic.007… Sorgh… bicolor Sobic.007…   2277 TRUE       AtTha…       2277
## 11 Sobic.007… Sorgh… bicolor Sobic.007…   2289 TRUE       AtTha…       2289
## 12 Sobic.007… Sorgh… bicolor Sobic.007…   2382 TRUE       AtTha…       2382
## 13 Sobic.001… Sorgh… bicolor Sobic.001…   2385 TRUE       AtTha…       2385
## 14 Sobic.002… Sorgh… bicolor Sobic.002…   1896 TRUE       AtTha…       1896</code></pre>
<p>In the above we can see the results of the BLAST. There were 14 hits, and it this dataframe now contains a variety of information about them.</p>
<p>The second aspect to the results is of course the sequences of the hits themselves. These are stored in the <code>sequences_of_interest_directory_path</code>.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="kw">dir</span>(<span class="kw">paste0</span>(phylochemistry_example_dir, <span class="st">&quot;sequences_of_interest/&quot;</span>))</a></code></pre></div>
<pre><code>##  [1] &quot;Sobic.001G179740.1.fa&quot; &quot;Sobic.002G142300.1.fa&quot;
##  [3] &quot;Sobic.004G014400.1.fa&quot; &quot;Sobic.004G037100.1.fa&quot;
##  [5] &quot;Sobic.004G037200.1.fa&quot; &quot;Sobic.005G092500.3.fa&quot;
##  [7] &quot;Sobic.005G109500.1.fa&quot; &quot;Sobic.007G043100.1.fa&quot;
##  [9] &quot;Sobic.007G078200.1.fa&quot; &quot;Sobic.007G085800.1.fa&quot;
## [11] &quot;Sobic.007G090469.1.fa&quot; &quot;Sobic.008G142400.1.fa&quot;
## [13] &quot;Sobic.008G167100.1.fa&quot; &quot;Sobic.010G254000.1.fa&quot;</code></pre>
<p>Above, you can see that this directory now contains a file that corresponds to each BLAST hit. We can look at one of these files:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">Biostrings<span class="op">::</span><span class="kw">readDNAStringSet</span>(<span class="dt">filepath =</span> <span class="kw">paste0</span>(phylochemistry_example_dir, <span class="st">&quot;sequences_of_interest/&quot;</span>, <span class="st">&quot;Sobic.001G179740.1.fa&quot;</span>))</a></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 1
##     width seq                                          names               
## [1]  2385 ATGTGGCGGCTAAAGATCGGC...TTCTTGCAAAAAAGAGTTAA Sobic.001G179740.1</code></pre>
<p>So, now we have our BLAST hits all organized! One way in which we may want to view them is using a phylogenetic tree. See the next chapter for more information on this.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="analyzing-chromatography-data-with-phylochemistry.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "section"
},
"toolbar": {
"position": "fixed"
}
});
});
</script>

</body>

</html>
