<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Phylochemistry</title>
  <meta name="description" content="Vingette for the R Package Phylochemistry." />
  <meta name="generator" content="bookdown 0.18 and GitBook 2.6.7" />

  <meta property="og:title" content="Phylochemistry" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Vingette for the R Package Phylochemistry." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Phylochemistry" />
  
  <meta name="twitter:description" content="Vingette for the R Package Phylochemistry." />
  

<meta name="author" content="Lucas Busta" />


<meta name="date" content="2020-08-31" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  

<link rel="next" href="setting-up-a-phylochemistry-analysis.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { background-color: #f8f8f8; }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#installing-r"><i class="fa fa-check"></i><b>1.1</b> Installing <code>R</code></a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#installing-rstudio"><i class="fa fa-check"></i><b>1.2</b> Installing <code>RStudio</code></a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#check-your-installations-of-r-and-rstudio"><i class="fa fa-check"></i><b>1.3</b> Check your installations of R and RStudio</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#install-phylochemistry-directly-from-github-and-view-documentation-contents"><i class="fa fa-check"></i><b>1.4</b> Install <code>phylochemistry</code> directly from github and view documentation contents</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="setting-up-a-phylochemistry-analysis.html"><a href="setting-up-a-phylochemistry-analysis.html"><i class="fa fa-check"></i><b>2</b> Setting up a <code>phylochemistry</code> analysis</a><ul>
<li class="chapter" data-level="2.1" data-path="setting-up-a-phylochemistry-analysis.html"><a href="setting-up-a-phylochemistry-analysis.html#conventions-in-phylochemsitry"><i class="fa fa-check"></i><b>2.1</b> Conventions in <code>phylochemsitry</code></a><ul>
<li class="chapter" data-level="2.1.1" data-path="setting-up-a-phylochemistry-analysis.html"><a href="setting-up-a-phylochemistry-analysis.html#file-paths-and-directory-paths"><i class="fa fa-check"></i><b>2.1.1</b> File paths and directory paths</a></li>
<li class="chapter" data-level="2.1.2" data-path="setting-up-a-phylochemistry-analysis.html"><a href="setting-up-a-phylochemistry-analysis.html#monolists"><i class="fa fa-check"></i><b>2.1.2</b> Monolists</a></li>
<li class="chapter" data-level="2.1.3" data-path="setting-up-a-phylochemistry-analysis.html"><a href="setting-up-a-phylochemistry-analysis.html#recommendations"><i class="fa fa-check"></i><b>2.1.3</b> Recommendations</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="analyzing-chromatography-data-with-phylochemistry.html"><a href="analyzing-chromatography-data-with-phylochemistry.html"><i class="fa fa-check"></i><b>3</b> Analyzing chromatography data with <code>phylochemistry</code></a><ul>
<li class="chapter" data-level="3.1" data-path="analyzing-chromatography-data-with-phylochemistry.html"><a href="analyzing-chromatography-data-with-phylochemistry.html#preparing-the-raw-data-and-metadata"><i class="fa fa-check"></i><b>3.1</b> Preparing the raw data and metadata</a></li>
<li class="chapter" data-level="3.2" data-path="analyzing-chromatography-data-with-phylochemistry.html"><a href="analyzing-chromatography-data-with-phylochemistry.html#using-the-phylochemistry-integrationapp"><i class="fa fa-check"></i><b>3.2</b> Using the <code>phylochemistry</code> integrationApp</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="gene-mining.html"><a href="gene-mining.html"><i class="fa fa-check"></i><b>4</b> Gene mining</a><ul>
<li class="chapter" data-level="4.1" data-path="gene-mining.html"><a href="gene-mining.html#blast"><i class="fa fa-check"></i><b>4.1</b> BLAST</a><ul>
<li class="chapter" data-level="4.1.1" data-path="gene-mining.html"><a href="gene-mining.html#using-the-blasttranscriptomes-command"><i class="fa fa-check"></i><b>4.1.1</b> Using the blastTranscriptomes command</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Phylochemistry</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">Phylochemistry</h1>
<p class="author"><em>Lucas Busta</em></p>
<p class="date"><em>2020-08-31</em></p>
</div>
<div id="introduction" class="section level1">
<h1><span class="header-section-number">Chapter 1</span> Introduction</h1>
<p><img src="http://lucasbusta.github.io/phylochemistry/figures/phylochemistry_logo.jpg" style="display: block; margin: auto;" /></p>
<p><code>phylochemistry</code> enables analyses of natural chemicals and gene sequences in the context of phylogeny.</p>
<div id="installing-r" class="section level2">
<h2><span class="header-section-number">1.1</span> Installing <code>R</code></h2>
<p>R is the computing language we will use to run our chemometric analyses and produce high quality plots. If you already have R installed, you can go straight to installing RStudio. If not, follow these steps to install R:</p>
<ol style="list-style-type: decimal">
<li>Go to <a href="https://cran.r-project.org/" class="uri">https://cran.r-project.org/</a></li>
<li>Click on “Download R for &lt;your operating system&gt;” (see footnote), depending on your operating system you will select “Download R for Linux”, “Download R for (Mac) OS X”, or “Download R for Windows”.</li>
</ol>
<p>footnote: We will use &lt;this notation&gt; quite a bit. It indicates a place where you should insert information, data, or something similar that corresponds to your particular situation. In this example it means insert “your operating system”, i.e. Linux, (Mac) OS X, or Windows.</p>
<ol start="3" style="list-style-type: decimal">
<li><p>For Mac: download the .pkg file for the latest release. As of 8/31/2020, this is R-4.0.2.pkg. For PC: click “install R for the first time”, then click “Download R 4.0.2 for Windows”.</p></li>
<li><p>After the executable finishes downloading (in Windows, it is a file with .exe extension; for Mac, it is a .dmg file or a .dmg inside a .pkg file), open the file as an administrator, and follow the installation instructions. R should install without any problems. You can click OK for all of the windows that pop-up during installation, and choose a “regular” installation (if given the choice).</p></li>
</ol>
<p>If you have trouble installing R please google “Install R Mac” or “Install R PC” and following one the many video tutorials out there. If you have tried this and are still having trouble, please contact me.</p>
</div>
<div id="installing-rstudio" class="section level2">
<h2><span class="header-section-number">1.2</span> Installing <code>RStudio</code></h2>
<p>Once we install R, we can install RStudio, which is essentially a convenient way of interacting with R. Some people do not like RStudio and prefer to interact with R directly. This is fine, but many beginning R users find RStudio helpful, so I recommend it. Follow these steps to install RStudio:</p>
<ol style="list-style-type: decimal">
<li>Go to <a href="https://rstudio.com/" class="uri">https://rstudio.com/</a></li>
<li>Click “DOWNLOAD” at the top of the page.</li>
<li>Click the “DOWNLOAD” button that corresponds to RStudio Desktop with the free Open Source License.</li>
<li>The page may automatically detect which operating system you are using and recommend a version for you. If it does, download that file (.exe for PC or .dmg for Mac). If not, scroll down to the “All Installers” section and download the file that is right for you. Open the file as an administrator, and follow the installation instructions. RStudio should install without any problems. You can click OK for all of the windows that pop-up during installation, and choose a “regular” installation (if given the choice).</li>
</ol>
<p>If you have trouble installing RStudio please google “Install RStudio Mac” or “Install RStudio PC” and following one the many video tutorials out there. If you have tried this and are still having trouble, please contact me.</p>
</div>
<div id="check-your-installations-of-r-and-rstudio" class="section level2">
<h2><span class="header-section-number">1.3</span> Check your installations of R and RStudio</h2>
<p>Open RStudio by clicking on the appropriate file in your applications folder, or wherever it is saved on your computer. You will see several windows. One is the Code Editor, one is the R Console, one is the Workspace and History, and one is the Plots and Files window.</p>
<p><img src="http://lucasbusta.github.io/phylochemistry/figures/rstudio_components.png" style="display: block; margin: auto;" /></p>
<p>The R Console window should have a carat in it. Type <code>head(Indomet)</code>. This should display the first six lines of a data set describing the pharmacokinets of indomethacin. This is one of the built in datasets in R.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">head</span>(Indometh)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="co">## # A tibble: 6 x 3</span></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="co">##   Subject  time  conc</span></a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="co">##   &lt;ord&gt;   &lt;dbl&gt; &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="co">## 1 1        0.25  1.5 </span></a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="co">## 2 1        0.5   0.94</span></a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="co">## 3 1        0.75  0.78</span></a>
<a class="sourceLine" id="cb1-8" data-line-number="8"><span class="co">## 4 1        1     0.48</span></a>
<a class="sourceLine" id="cb1-9" data-line-number="9"><span class="co">## 5 1        1.25  0.37</span></a>
<a class="sourceLine" id="cb1-10" data-line-number="10"><span class="co">## 6 1        2     0.19</span></a></code></pre></div>
<p>Next, type <code>plot(Indometh)</code> into the R Console. This will plot the indomethacin dataset in a basic way.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw">plot</span>(Indometh)</a></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-4-1.png" width="576" /></p>
<p>If both the above commands (<code>head(Indometh)</code> and <code>plot(Indometh)</code>) worked and there were no error messages during installation, then you should be ready to proceed.</p>
</div>
<div id="install-phylochemistry-directly-from-github-and-view-documentation-contents" class="section level2">
<h2><span class="header-section-number">1.4</span> Install <code>phylochemistry</code> directly from github and view documentation contents</h2>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;LucasBusta/phylochemistry&quot;</span>, <span class="dt">quiet =</span> <span class="ot">FALSE</span>, <span class="dt">force =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="co">## Downloading GitHub repo LucasBusta/phylochemistry@master</span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="co">##   </span></a>
<a class="sourceLine" id="cb3-4" data-line-number="4">   checking <span class="cf">for</span> file ‘<span class="op">/</span>private<span class="op">/</span>var<span class="op">/</span>folders<span class="op">/</span>d4<span class="op">/</span>m9ll00q10hdd6kz0vk9q0nx80000gn<span class="op">/</span>T<span class="op">/</span>Rtmp04nFW7<span class="op">/</span>remotes127f43ffe8406<span class="op">/</span>LucasBusta<span class="op">-</span>phylochemistry<span class="op">-</span>75de128<span class="op">/</span>DESCRIPTION’ ...</a>
<a class="sourceLine" id="cb3-5" data-line-number="5">  </a>
<a class="sourceLine" id="cb3-6" data-line-number="6">✓  checking <span class="cf">for</span> file ‘<span class="op">/</span>private<span class="op">/</span>var<span class="op">/</span>folders<span class="op">/</span>d4<span class="op">/</span>m9ll00q10hdd6kz0vk9q0nx80000gn<span class="op">/</span>T<span class="op">/</span>Rtmp04nFW7<span class="op">/</span>remotes127f43ffe8406<span class="op">/</span>LucasBusta<span class="op">-</span>phylochemistry<span class="op">-</span>75de128<span class="op">/</span>DESCRIPTION’</a>
<a class="sourceLine" id="cb3-7" data-line-number="7"><span class="co">## </span></a>
<a class="sourceLine" id="cb3-8" data-line-number="8">  </a>
<a class="sourceLine" id="cb3-9" data-line-number="9">─  preparing ‘phylochemistry’<span class="op">:</span></a>
<a class="sourceLine" id="cb3-10" data-line-number="10"><span class="co">## </span></a>
<a class="sourceLine" id="cb3-11" data-line-number="11"><span class="st">  </span></a>
<a class="sourceLine" id="cb3-12" data-line-number="12"><span class="st">   </span>checking DESCRIPTION meta<span class="op">-</span>information ...</a>
<a class="sourceLine" id="cb3-13" data-line-number="13">  </a>
<a class="sourceLine" id="cb3-14" data-line-number="14">✓  checking DESCRIPTION meta<span class="op">-</span>information</a>
<a class="sourceLine" id="cb3-15" data-line-number="15"><span class="co">## </span></a>
<a class="sourceLine" id="cb3-16" data-line-number="16">  </a>
<a class="sourceLine" id="cb3-17" data-line-number="17">─  checking <span class="cf">for</span> LF line<span class="op">-</span>endings <span class="cf">in</span> source and make files and shell scripts</a>
<a class="sourceLine" id="cb3-18" data-line-number="18"><span class="co">## </span></a>
<a class="sourceLine" id="cb3-19" data-line-number="19">  </a>
<a class="sourceLine" id="cb3-20" data-line-number="20">─  checking <span class="cf">for</span> empty or unneeded directories</a>
<a class="sourceLine" id="cb3-21" data-line-number="21"><span class="co">## </span></a>
<a class="sourceLine" id="cb3-22" data-line-number="22">  </a>
<a class="sourceLine" id="cb3-23" data-line-number="23">─  building ‘phylochemistry_<span class="dv">0</span>.<span class="dv">0</span>.<span class="dv">0</span>.<span class="fl">9000.</span>tar.gz’</a>
<a class="sourceLine" id="cb3-24" data-line-number="24"><span class="co">## </span></a>
<a class="sourceLine" id="cb3-25" data-line-number="25">  </a>
<a class="sourceLine" id="cb3-26" data-line-number="26">   </a>
<a class="sourceLine" id="cb3-27" data-line-number="27"><span class="co">## </span></a>
<a class="sourceLine" id="cb3-28" data-line-number="28">packages &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;phylochemistry&quot;</span>)</a>
<a class="sourceLine" id="cb3-29" data-line-number="29"><span class="kw">lapply</span>(packages, require, <span class="dt">character.only =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb3-30" data-line-number="30"><span class="co">## [[1]]</span></a>
<a class="sourceLine" id="cb3-31" data-line-number="31"><span class="co">## [1] TRUE</span></a>
<a class="sourceLine" id="cb3-32" data-line-number="32"><span class="kw">help</span>(<span class="dt">package =</span> <span class="st">&quot;phylochemistry&quot;</span>)</a></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="setting-up-a-phylochemistry-analysis.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "section"
},
"toolbar": {
"position": "fixed"
}
});
});
</script>

</body>

</html>
